---
title: "F1 ASE Analysis with MBASED"
author: Lynn Ly
output: html_document
---

Author comment
File description comment, including purpose of program, inputs, and outputs
source() and library() statements
Function definitions
Executed statements, if applicable (e.g., print, plot)

Use with pre-filtered VCF files

```{r setup, include=FALSE}
library(MBASED)
library(tidyverse)
```

```{r AnnotateSNPs}
AnnotateSNPs <- function(SNPdata, gff.mRNA){
  # Combines SNP/SNV loci with gene names
  #
  # Args:
  #   SNP.data: SNP data containing positions to be matched with genomic features
  #   gff: A gff file containing only CHROM, START, END, GeneID
  #
  # Returns:
  #   SNP.data with a new column, GeneID
  
  colnames(gff.mRNA) <- c("CHROM", "start", "end", "name") 
  
  genes <- GRanges(seqnames = Rle(gff.mRNA$CHROM),
                   ranges = IRanges(start = gff.mRNA$start, end = gff.mRNA$end), 
                   names = gff.mRNA$name)
  
  SNPs <- GRanges(seqnames = Rle(SNPdata$CHROM), 
                 ranges = IRanges(start = SNPdata$POS, SNPdata$POS), 
                 CHROM = SNPdata$CHROM,
                 POS = SNPdata$POS)
  
  # Overlap SNP position with gene range 
  overlappedGenes <- mergeByOverlaps(SNPs, genes)
  overlappedGenes <- overlappedGenes[, c(2, 3, 5)]
  colnames(overlappedGenes) <- c("CHROM", "POS", "GeneID")
  
  annotatedSNPdata <- SNPdata %>% 
    left_join(as.data.frame(overlappedGenes), by=c("CHROM", "POS")) 
  
  return(annotatedSNPdata)  
}
```

```{r Functions for analyzing results}
SummarizeASEResults_1s <- function(MBASEDOutput) {
  geneOutputDF <- data.frame(
    majorAlleleFrequency = assays(MBASEDOutput)$majorAlleleFrequency[,1],
    pValueASE = assays(MBASEDOutput)$pValueASE[,1],
    pValueHeterogeneity = assays(MBASEDOutput)$pValueHeterogeneity[,1])
  
  lociOutputGR <- rowRanges(metadata(MBASEDOutput)$locusSpecificResults)
  lociOutputGR$allele1IsMajor <-  assays(metadata(MBASEDOutput)$locusSpecificResults)$allele1IsMajor[,1]
  lociOutputGR$MAF <- assays(metadata(MBASEDOutput)$locusSpecificResults)$MAF[,1]
  lociOutputList <- split(lociOutputGR, factor(lociOutputGR$aseID, levels = unique(lociOutputGR$aseID)))
  return(list(geneOutput=geneOutputDF, locusOutput=lociOutputList))
}

ExtractASE <- function(MBASEDOutput) {
  # Extract only desired genes
  # Modify ASEindexes to vary the strictness of selection.
  results <- SummarizeASEResults_1s(MBASEDOutput)

  ASEindexes <- results$geneOutput$pValueASE * 46577 < 0.05 & 
    results$geneOutput$majorAlleleFrequency > 0.7
  
  significantResults <- list(results$geneOutput[ASEindexes, ], 
                             results$locusOutput[ASEindexes, ])
  return(significantResults)
}
```

```{r From input to MBASED Output}
# Please change numSim to at least 1,000,000 for final analysis

  annotatedData <- AnnotateSNPs(SNPdata = F1.young.GQ.filtered, gff.mRNA = gff.mRNA)
  
  # Remove SNVs with no associated genes
  annotatedData <- filter(annotatedData, !is.na(GeneID))
  
  mySNVs <- GRanges(
    seqnames = annotatedData$CHROM,
    ranges = IRanges(start = annotatedData$POS, width = 1),
    aseID = as.vector(annotatedData$GeneID),
    allele1 = annotatedData$REF,
    allele2 = annotatedData$ALT)

SingleSample <- function(annotatedData, mySNVs, genotype){
  # create RangedSummarizedExperiment object as input for runMBASED
  # MBASEDprep[1] is the annotated data. MBASEDprep[2] is mySNVs
  RO <- paste(genotype, "RO", sep = "_")
  AO <- paste(genotype, "AO", sep = "_")
  
  mySample <- SummarizedExperiment(
    assays = list(lociAllele1Counts = matrix(annotatedData[, RO]),
                lociAllele2Counts = matrix(annotatedData[, AO])),
    rowRanges=mySNVs)
  
  MBASEDOutput <- runMBASED(
    ASESummarizedExperiment = mySample,
    numSim = 1000000,
    isPhased = FALSE) # We do have phasing information though, right?

  return(MBASEDOutput)
}

MBASED.F1.414 <- SingleSample(annotatedData, mySNVs, genotype = "F1_414")
save(MBASED.F1.414, "MBASED.F1.414.Rdata")
MBASED.F1.415 <- SingleSample(annotatedData, mySNVs, genotype = "F1_415")
save(MBASED.F1.414, "MBASED.F1.415.Rdata")
MBASED.Ae <- SingleSample(annotatedData, mySNVs, genotype = "Ae")
save(MBASED.F1.414, "MBASED.F1.Ae.Rdata")
MBASED.Ol <- SingleSample(annotatedData, mySNVs, genotype = "Ol")
save(MBASED.F1.414, "MBASED.F1.Ol.Rdata")
```

```{r Data to use}
# Change the following line if not using B. napus
gff.mRNA <- read.table("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/Reference/B.napus/gff.mRNA")

# Data to use
load("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/mizukikadowaki/project/output/F1.young.GQ.filtered.Rdata")
```

In Progress: TwoSample analysis

```{r}
TwoSample <- function(annotatedData, mySNVs, genotype1, genotype2){
  RO1 <- paste(genotype1, "RO", sep = "_")
  AO1 <- paste(genotype1, "AO", sep = "_")
  RO2 <- paste(genotype2, "RO", sep = "_")
  AO2 <- paste(genotype2, "AO", sep = "_")
  
  mySample <- SummarizedExperiment(
    assays = list(lociAllele1Counts = matrix(c(annotatedData[, RO1], annotatedData[, RO2]),
                                             ncol = 2, 
                                             dimnames = list(names(mySNVs), c(genotype1, genotype2))),
                lociAllele2Counts = matrix(c(annotatedData[, RO1], annotatedData[, RO2]),
                                             ncol = 2, 
                                             dimnames = list(names(mySNVs), c(genotype1, genotype2))),
    rowRanges=mySNVs))
  
  MBASEDOutput <- runMBASED(
    ASESummarizedExperiment = mySample,
    numSim = 1000,
    isPhased = FALSE) # We do have phasing information though, right?

  return(MBASEDOutput)
}
```


```{r}
dim(MBASED.F1.414)
sig.F1.414 <- ExtractASE(MBASED.F1.414)
dim(sigF1.414) #2147 genes found to be in ASE

```

