---
title: '505_late_silique'
author: "Ruijuan Li"
date: "4/7/2017"
output: 
  html_document: 
    keep_md: yes 
---

### data description 
```{r}
# RNA seq data from late silique, batch C: 47 lines; batch D: 41 lines, more lines will come... 
```

# preprocess 
```{r}
# 1) download 

# 2) trimming 
# batch C: https://github.com/leejimmy93/KIAT_cabernet/blob/master/505/trimming.slurm
# batch D: https://github.com/leejimmy93/KIAT_whitney/blob/master/505/trimming_batchD.sh

# 3) mapping using STAR 
# batch C: https://github.com/leejimmy93/KIAT_cabernet/blob/master/505/mapping_array.slurm (sometimes there is error: could not load STAR module, for this case, load STAR module in the environment before sbatch) 
# mapping for sample88 cannot be finished on whitney, change strategy, "--outSAMtype BAM Unsorted" 
```

# analyze mapping result (QC & mapping rate)
```{r}
# star_extract_stats.sh get star mapping stats 
```

# mapping result analysis 
```{r, results='hide'}
source("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/analysis/function_BnRNAseq.R")

# sample description 
sample_des_c <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/data/batch_c.csv", header = T) 
sample_des_c$batch <- rep("C", nrow(sample_des_c))
sample_des_c_sub <- sample_des_c[,c("Sample.ID","TotalReads","batch")]
sample_des_d <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/data/batch_d.csv", header = T)
sample_des_d$batch <- rep("D", nrow(sample_des_d))
sample_des_d_sub <- sample_des_d[,c("Sample.ID","TotalReads","batch")]
head(sample_des_d_sub)

# mapping result
mapping_result <- read.table("~/Desktop/Brassica_project/KIAT_RNA_seq/505/data/Star_Stats.tab", header = T)
head(mapping_result)
mapping_result$Sample.ID <- gsub("(Sample_)([[:print:]])", "\\2", mapping_result$Sample) 

# merge sample description & mapping result for batch A and batch B 
batch.c.stats <- merge(sample_des_c_sub, mapping_result, by="Sample.ID")
batch.d.stats <- merge(sample_des_d_sub, mapping_result, by="Sample.ID")
mapping_result_505 <- rbind(batch.c.stats, batch.d.stats)

mapping_result_505$batch
mapping_result_505$Number_Input_Reads/(mapping_result_505$TotalReads/2)

### calculate average 
mean(mapping_result_505[mapping_result_505$batch=="C",]$TotalReads/2) # 29598998
mean(mapping_result_505[mapping_result_505$batch=="D",]$TotalReads/2) # 27072598 

# number of total reads 
library(ggplot2) 
p.total.reads <- ggplot(data=mapping_result_505)
p.total.reads <- p.total.reads + geom_bar(aes(x=reorder(Sample.ID, TotalReads/2), y=TotalReads/2, fill=batch), stat = "identity")
p.total.reads <- p.total.reads + labs(list(title = "", x = "", y = "Number of raw reads"))
p.total.reads <- p.total.reads + theme(axis.text.x = element_text(angle = 90, size = 8)) 
p.total.reads   
# ggsave(p.total.reads, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/total_reads.png", height = 8, width = 11) 

# look at HQ, mapped, multimapped in detail for each sample (memory too low, cannot view histogram of total reads...)
colnames(mapping_result_505)
mapping_result_505_sub <- data.frame(Sample.ID = mapping_result_505$Sample.ID, 
                                     batch=mapping_result_505$batch,
                                    trimmed.off = mapping_result_505$TotalReads/2 - mapping_result_505$Number_Input_Reads,
                                    unmapped = mapping_result_505$Number_Input_Reads * (as.numeric(mapping_result_505$Percent_Unmapped_Mismatches + mapping_result_505$Percent_Unmapped_Other + mapping_result_505$Percent_Unmapped_Too_Short)/100), 
                                    multi_too_many_mapped = mapping_result_505$Number_Multi_Mapped + mapping_result_505$Number_Too_Many_Multi_Mapped, 
                                    unqiuemapped = mapping_result_505$Number_Unique_Mapped
) 

mapping_result_505_sub 
dim(mapping_result_505_sub) # 88 6 

mapping_result_505_sub.long <- melt(mapping_result_505_sub, id.vars = c("Sample.ID", "batch"))
mapping_result_505_sub.long.c <- mapping_result_505_sub.long[mapping_result_505_sub.long$batch=="C",]
mapping_result_505_sub.long.d <- mapping_result_505_sub.long[mapping_result_505_sub.long$batch=="D",]

pl.mapping.indi.c <- ggplot(data = mapping_result_505_sub.long.c)
pl.mapping.indi.c <- pl.mapping.indi.c + geom_bar(aes(x=Sample.ID, y = value, fill=variable), stat = "identity")
pl.mapping.indi.c <- pl.mapping.indi.c + labs(list(title = "", x = "individual sample", y = "reads number")) + theme(axis.text.x=element_text(angle = 90, size = 8)) + ylim(0, 4.5e+7) 
# p.total.reads <- p.total.reads + theme(axis.text.x = element_text(angle = 90, size = 8)) 
pl.mapping.indi.c

pl.mapping.indi.d <- ggplot(data = mapping_result_505_sub.long.d)
pl.mapping.indi.d <- pl.mapping.indi.d + geom_bar(aes(x=Sample.ID, y = value, fill=variable), stat = "identity")
pl.mapping.indi.d <- pl.mapping.indi.d + labs(list(title = "", x = "individual sample", y = "reads number")) + theme(axis.text.x=element_text(angle = 90, size = 8)) + ylim(0, 4.5e+7)
pl.mapping.indi.d 
 
library(cowplot)
pl.mapping.indi <- plot_grid(
  pl.mapping.indi.c +labs(title="mapping result of batch c")+theme(legend.position = "none"),
  pl.mapping.indi.d +labs(title="mapping result of batch d") +theme(legend.position = "none"),
  align = 'vh', ncol=2, nrow = 1, labels=c("",""))

get_legend<-function(a.gplot){
  tmp <- ggplot_gtable(ggplot_build(a.gplot))
  leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box")
  legend <- tmp$grobs[[leg]]
  return(legend)}

legend <- get_legend(pl.mapping.indi.c)

# pl.mapping.indi <- plot_grid(arrangeGrob(pl.mapping.indi.a + theme(legend.position="none") + labs(title="mapping result of batch a"),
#                          pl.mapping.indi.b + theme(legend.position="none") + labs(title="mapping result of batch b"),
#                          nrow=1),
#              legend, nrow=2,heights=c(10, 1))

pl.mapping.indi.final <- plot_grid(pl.mapping.indi, legend, rel_widths = c(3, 1))

pl.mapping.indi.final 
save_plot("~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/pl_mapping_indi.png", pl.mapping.indi.final, base_aspect_ratio = 1, base_width = 12)  


# look at histogram of HQ reads 
colnames(mapping_result_505)
p.HQ.perc <- ggplot(data=mapping_result_505)
p.HQ.perc <- p.HQ.perc + geom_histogram(aes(x=mapping_result_505$Number_Input_Reads/(mapping_result_505$TotalReads/2), fill=batch), binwidth = 0.01) 
p.HQ.perc <- p.HQ.perc + labs(list(title = "histogram of HQ reads percentage", x = "Percent of high qulity reads", y = "number of samples"))
p.HQ.perc 
# ggsave(p.HQ.perc, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/HQ_hist.png", height = 8, width = 11) 

# histogram of unqiue mapped reads 
p.uniq.mapping <- ggplot(data=mapping_result_505)
p.uniq.mapping <- p.uniq.mapping + geom_histogram(aes(x=Percent_Unique_Mapped, fill=batch), binwidth = 2.5) 
p.uniq.mapping <- p.uniq.mapping + labs(list(title = "histogram of uniquely mapped reads percentage", x = "Percent of uniquely mapped reads", y = "number of samples"))
p.uniq.mapping  
# ggsave(p.uniq.mapping, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/unique_mapped.png", height = 8, width = 11) 

# hisogram of multimapped & too many reads 
p.multi.too.many.mapping <- ggplot(data=mapping_result_505)
p.multi.too.many.mapping <- p.multi.too.many.mapping + geom_histogram(aes(x=mapping_result_505$Percent_Multi_Mapped+mapping_result_505$Percent_Too_Many_Multi_Mapped, fill=batch), binwidth = 2.5) 
p.multi.too.many.mapping <- p.multi.too.many.mapping + labs(list(title = "", x = "Percent of multi & too many mapped reads", y = "number of samples"))

# histogram of multi mapped 
p.multi.mapping <- ggplot(data=mapping_result_505)
p.multi.mapping <- p.multi.mapping + geom_histogram(aes(x=mapping_result_505$Percent_Multi_Mapped, fill=batch), binwidth = 2.5) 
p.multi.mapping <- p.multi.mapping + labs(list(title = "histogram of multi mapped reads percentage", x = "Percent of multi mapped reads", y = "number of samples"))
p.multi.mapping
# ggsave(p.multi.mapping, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/multi_mapped.png", height = 8, width = 11) 

# histogram of too many mapped 
p.toomany.mapping <- ggplot(data=mapping_result_505)
p.toomany.mapping <- p.toomany.mapping + geom_histogram(aes(x=mapping_result_505$Percent_Too_Many_Multi_Mapped, fill=batch), binwidth = 2.5) 
p.toomany.mapping <- p.toomany.mapping + labs(list(title = "histogram of too many mapped reads percentage", x = "Percent of toomany mapped reads", y = "number of samples"))
p.toomany.mapping
# ggsave(p.toomany.mapping, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/toomany_mapped.png", height = 8, width = 11) 

# histogram of unmapped 
p.un.mapping <- ggplot(data=mapping_result_505)
p.un.mapping <- p.un.mapping + geom_histogram(aes(x=mapping_result_505$Percent_Unmapped_Mismatches+mapping_result_505$Percent_Unmapped_Other+mapping_result_505$Percent_Unmapped_Too_Short, fill=batch), binwidth = 2.5) 
p.un.mapping <- p.un.mapping + labs(list(title = "histogram of unmapped reads percentage", x = "Percent of unmapped reads", y = "number of samples"))
p.un.mapping
# ggsave(p.un.mapping, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/un_mapped.png", height = 8, width = 11)

# mapping stats 
colnames(mapping_result_505)
# average percentage of HQ, unique, multi, toomany, unmapped 
average.mapping <- aggregate(mapping_result_505[,c("TotalReads","Number_Input_Reads", "Percent_Unique_Mapped", "Percent_Multi_Mapped", "Percent_Too_Many_Multi_Mapped", "Percent_Unmapped_Too_Short")], list(mapping_result_505$batch), mean)

rownames(average.mapping) <- average.mapping$Group.1
average.mapping$Percent_HQ <- average.mapping$Number_Input_Reads/(average.mapping$TotalReads/2)*100
average.mapping

average.mapping.long <- melt(average.mapping[,c("Percent_Unique_Mapped", "Percent_Multi_Mapped", "Percent_Too_Many_Multi_Mapped", "Percent_Unmapped_Too_Short","Percent_HQ","Group.1")], id.vars = "Group.1")

average.mapping.long

p.average.mapping <- ggplot(data = average.mapping.long)
p.average.mapping <- p.average.mapping + geom_bar(aes(x=variable, y=value, fill=variable), stat = "identity")
p.average.mapping <- p.average.mapping + facet_wrap(~Group.1, ncol = 2)
p.average.mapping <- p.average.mapping + labs(list(title = "mapping stats of 505", x = "", y = "percentage"))
p.average.mapping <- p.average.mapping + theme(axis.text.x = element_text(angle = 90), legend.position = "none")
p.average.mapping <- p.average.mapping + geom_text(data = average.mapping.long, aes(x=variable, y=value, label=factor(round(value,0)))) 
p.average.mapping
# ggsave(p.average.mapping, filename =  "~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/average_mapping.png", height = 6, width = 9) 

### split this figure to HQ ratio & mapping ratio (two plots)
average.mapping.long.HQ <- 
  average.mapping.long[average.mapping.long$variable=="Percent_HQ",]

average.mapping.long.HQ

p.average.HQ <- ggplot(data = average.mapping.long.HQ)
p.average.HQ <- p.average.HQ + geom_bar(aes(x=Group.1, y=value, fill=variable), stat = "identity")
# p.average.mapping <- p.average.mapping + facet_wrap(~Group.1, ncol = 2)
p.average.HQ <- p.average.HQ + labs(list(title = "", x = "", y = "percentage of total raw reads"))
p.average.HQ <- p.average.HQ + theme(legend.position = "none")
p.average.HQ <- p.average.HQ + geom_text(data = average.mapping.long.HQ, aes(x=Group.1, y=value, label=factor(round(value,0)))) 
p.average.HQ

ggsave(p.average.HQ, filename =  "~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/average_HQ_late_silique.png", height = 4, width = 2) 

average.mapping.long.mapping <- 
  average.mapping.long[average.mapping.long$variable!="Percent_HQ",]

p.average.mapping <- ggplot(data = average.mapping.long.mapping)
p.average.mapping <- p.average.mapping + geom_bar(aes(x=variable, y=value, fill=variable), stat = "identity")
p.average.mapping <- p.average.mapping + facet_wrap(~Group.1, ncol = 2)
p.average.mapping <- p.average.mapping + labs(list(title = "", x = "", y = "percentage of HQ reads"))
p.average.mapping <- p.average.mapping + theme(axis.text.x = element_text(angle = 90), legend.position = "none")
p.average.mapping <- p.average.mapping + geom_text(data = average.mapping.long.mapping, aes(x=variable, y=value, label=factor(round(value,0)))) 
p.average.mapping 

ggsave(p.average.mapping, filename =  "~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/average_mapping_late_silique.png", height = 6, width = 5) 

```
 

# SNP calling (wait untill I get QC and mapping rate)
```{r} 
# 1) prepare data for SNP calling 
# Prep4Freebayes.sh (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/Prep4Freebayes.sh)

# 2) split bam file by chromosome 
# Bam_Split_By_Chrom.sh (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/Bam_Split_By_Chrom.sh) 

# 3) SNP calling 
# Freebayes_Array_no_pop_late_silique.slurm (https://github.com/leejimmy93/KIAT_cabernet/blob/master/505/Freebayes_Array_no_pop_late_silique.slurm) 

# scp to whitney 
# scp * ruijuanli@whitney.plb.ucdavis.edu:/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/vcf_late_silique_no_pop 

# 4) compress & index 
# compress_index_vcf.sh 
# (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/compress_index_vcf.sh)

# 5) combine into one big vcf file 
# concat.sh
# (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/concat.sh)
# output: 505.vcf.gz  

# 6) filter 
# filter.sh 
# (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/filter.sh)
# output: 505.filtered.vcf.gz 

# 7) calculate a bunch of stats (haven't done this step yet...)
# ./calc.sh 505_filtered.vcf.gz 505_filtered 
# https://github.com/leejimmy93/KIAT_whitney/blob/master/505/calc.sh 
# output: many stats start with 505_filtered
```

# stats, wait for late 
```{r}
# output: many stats start with 505_filtered  
```

```{r}
# chromosome distribution 
geno_505_hmp.late_silique <- read.table("~/505/vcf_late_silique_no_pop/combined/505_filtered_sorted.hmp.txt", head=T)
head(geno_505_hmp.late_silique)
colnames(geno_505_hmp.late_silique)
nrow(geno_505_hmp.late_silique) # 242314  

distri <- geno_505_hmp.late_silique[,c("chrom", "pos")] 
head(distri)
colnames(distri) <- c("CHROM", "POS")

class(distri)
data <- distri
head(data)

# below the block of code should go into function... 
chr_ID <- unique(data$CHROM)
number.unique <- list() 

for (chr in chr_ID) {
  number.unique[chr] <- sum(sum(data$CHROM==chr))
  print(number.unique[chr])
}

SNP_chr.unique <- as.data.frame(t(as.data.frame(number.unique)))
SNP_chr.unique$ID <- rownames(SNP_chr.unique)
head(SNP_chr.unique)
# split to main & random chromosomes 
# main 
SNP_chr_main.unique <- SNP_chr.unique[!grepl("RANDOM", SNP_chr.unique$ID),]  
head(SNP_chr_main.unique)
SNP_chr_main.unique$subgenome <- gsub("(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\1", SNP_chr_main.unique$ID)   
SNP_chr_main.unique$chr_ID <- gsub("(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\2", SNP_chr_main.unique$ID) 

# do a little calculation: how many SNPs on main & random chromosomes? 
sum(SNP_chr_main.unique$V1) # 200897        
head(SNP_chr_main.unique)     

sum(SNP_chr_main.unique[SNP_chr_main.unique$subgenome=="A",]$V1)/sum(SNP_chr_main.unique$V1) # 56% 

# make a plot to see how many SNPs on each chromosome 
library(ggplot2)
pl.SNP.main.1.unique <- ggplot(data = SNP_chr_main.unique) 
pl.SNP.main.1.unique <- pl.SNP.main.1.unique + geom_bar(aes(x=factor(chr_ID), y = V1, fill=subgenome), stat = "identity")
pl.SNP.main.1.unique <- pl.SNP.main.1.unique + facet_wrap(~subgenome, ncol = 1)
# pl.SNP.main.1 <- pl.SNP.main.1 + geom_text(aes(x=factor(chr_ID), y = number, label=factor(number), size=1)) 
pl.SNP.main.1.unique <- pl.SNP.main.1.unique + labs(list(title = "", x = "chromosome ID", y = "number of SNPs"))
pl.SNP.main.1.unique <- pl.SNP.main.1.unique + theme(legend.position = "none")
pl.SNP.main.1.unique 
ggsave("~/505/output/figure/late_silique/SNP.main.png", width = 9, height = 7) 

########## Per Mb ... 
head(data)
data <- data[!grepl("RANDOM", data$CHROM),]  
data$subgenome <-  gsub("(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\1", data$CHROM) 
data$Chr_ID <- gsub("(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\2", data$CHROM) 
head(data) 

# write.csv(vcf.Ae.Ol.intersect.df.2, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/parent_SNP/output/vcf.Ae.Ol.intersect.df.2.csv") # sort in excel manually 

# cf.Ae.Ol.intersect.df.2.sorted <-  read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/parent_SNP/output/vcf.Ae.Ol.intersect.df.2.csv")
# head(vcf.Ae.Ol.intersect.df.2.sorted) 

library(ggplot2)
pl.SNP.main.2.unique <- ggplot(data = data)
pl.SNP.main.2.unique <- pl.SNP.main.2.unique + geom_histogram(aes(x=POS, fill=subgenome), binwidth = 1000000) 
pl.SNP.main.2.unique <- pl.SNP.main.2.unique + facet_grid(Chr_ID ~subgenome)
pl.SNP.main.2.unique <- pl.SNP.main.2.unique + labs(list(title = "", x = "chromosome ID", y = "number of SNPs"))
pl.SNP.main.2.unique <- pl.SNP.main.2.unique + theme(legend.position = "none")
pl.SNP.main.2.unique 

ggsave("~/505/output/figure/late_silique/SNP.main.PerMb.png", width = 9, height = 7) 
 
# calculate SNPs per Mb number 
chrom.length <- read.table("~/Reference/B.napus/Brassica_napus_v4.1.length", header=F)
chrom.length
nrow(geno_505_hmp.late_silique)/sum(chrom.length$V2)*1000000 

```

# cluster analysis   
```{r} 
# reform_505_late_silique.R to reform vcf file for GWAS 
# output: late_silique_505.2 

# cluster analysis 
load("~/505/output/late_silique/late_silique_505.2.Rdata")
class(late_silique_505.2.t)
dim(late_silique_505.2.t) #  88 347282 
late_silique_505.2.t[1:10, 1:10]
# late_silique_505.2.t <- as.data.frame(late_silique_505.2.t)
# geno.numeric <- data.matrix(late_silique_505.2.t)
# geno.numeric[1:10, 1:10]

geno.numeric <- apply(late_silique_505.2.t, 2, function(x) as.integer(x))
rnames <- rownames(late_silique_505.2.t)
rownames(geno.numeric) <- rnames

geno.numeric.df <- as.data.frame(geno.numeric)

geno.numeric.df.2 <- geno.numeric.df[which(!rownames(geno.numeric.df) %in% "X505.K88"),]
### 
geno.numeric.2 <- as.matrix(geno.numeric.df.2)

genDist <- as.matrix(dist(geno.numeric.2))

#perform the multi-dimensional scaling
geno.mds <- as.data.frame(cmdscale(genDist))
head(geno.mds) #now we have 2 dimensions
dim(geno.mds)
plot(geno.mds)  

############  
save(geno.numeric, file="~/505/output/late_silique_505_geno_numeric.Rdata") 

## add country of origin, winter or spring type to the figure ... 
sample_des_c_revised <- read.csv("~/505/data/phenotype/batch_c_revised.csv", header=T)
sample_des_c_sub <- sample_des_c_revised[,c("Sample.ID", "Name","Sample.Description.Origin", "Sample.Description.Type")]
sample_des_d_revised <- read.csv("~/505/data/phenotype/batch_d_revised.csv", header=T)
sample_des_d_sub <- sample_des_d_revised[,c("Sample.ID", "Name", "Sample.Description.Origin", "Sample.Description.Type")]
head(sample_des_c_sub)
sample_des_c_sub$Sample.ID
sample_des_d_sub$Sample.ID

sample_des_revised <- rbind(sample_des_c_sub, sample_des_d_sub)
sample_des_revised <- sample_des_revised[-c(48, 90:93),]
sample_des_revised$Sample.ID
save(sample_des_revised, file="~/505/output/late_silique/sample_des_revised.Rdata")

# MDS result 
# reformat MDS result sample ID (X remove & . to -)
geno.mds$Sample.ID <- gsub("\\.","\\-", rownames(geno.mds)) 
geno.mds$Sample.ID <- gsub("X", "", geno.mds$Sample.ID)
class(geno.mds$Sample.ID)

geno.mds$Sample.ID
sample_des_revised$Sample.ID

# merge 
geno.mds.all.des <- merge(geno.mds, sample_des_revised, by="Sample.ID")
head(geno.mds.all.des)


library(ggrepel)
set.seed(111) 

p.mds.country <- ggplot(data=geno.mds.all.des, aes(V1, V2, color=factor(Sample.Description.Origin)))
p.mds.country <- p.mds.country + geom_point(size=1) 
p.mds.country <- p.mds.country + scale_color_brewer(type="qual",palette="Set1")
p.mds.country <- p.mds.country + geom_text_repel(aes(x=V1, y=V2, label=factor(Name)))
p.mds.country <- p.mds.country + facet_wrap(~Sample.Description.Type)
p.mds.country 
ggsave(p.mds.country, filename="~/505/output/figure/late_silique/p.mds.country.png", height=20, width=20)

p.mds.country.2 <- ggplot(data=geno.mds.all.des, aes(V1, V2, color=factor(Sample.Description.Origin)))
p.mds.country.2 <- p.mds.country.2 + geom_point(size=1) 
p.mds.country.2 <- p.mds.country.2 + scale_color_brewer(type="qual",palette="Set1")
p.mds.country.2 <- p.mds.country.2 + geom_text_repel(aes(x=V1, y=V2, label=factor(Name)))
# p.mds.country <- p.mds.country + facet_wrap(~Sample.Description.Type)
p.mds.country.2
ggsave(p.mds.country.2, filename="~/505/output/figure/leaf/p.mds.country.2.png", height=20, width=20)

p.mds.country.2 <- ggplot(data=geno.mds.all.des, aes(V1, V2, color=factor(Sample.Description.Origin)))
p.mds.country.2 <- p.mds.country.2 + geom_point(size=1) 
p.mds.country.2 <- p.mds.country.2 + scale_color_brewer(type="qual",palette="Set1")
p.mds.country.2 <- p.mds.country.2 + geom_text_repel(aes(x=V1, y=V2, label=factor(Name)))
# p.mds.country <- p.mds.country + facet_wrap(~Sample.Description.Type)
p.mds.country.2
ggsave(p.mds.country.2, filename="~/505/output/figure/late_silique/p.mds.country.2.png", height=20, width=20)
 
# cluster for only the common lines 
load("~/505/output/common.lines.Rdata")
common.lines$Taxa.y

rownames(geno.numeric.df.2) 
# change rownames 
rownames(geno.numeric.df.2) <- gsub("\\.","\\-", rownames(geno.numeric.df.2))
rownames(geno.numeric.df.2) <- gsub("X", "", rownames(geno.numeric.df.2))

common.lines$Taxa.y <- as.character(common.lines$Taxa.y)
rownames(geno.numeric.df.2) 
common.lines$Taxa.y <- gsub("\\_","\\-", common.lines$Taxa.y)

geno.numeric.df.common <- geno.numeric.df.2[rownames(geno.numeric.df.2) %in% common.lines$Taxa.y,]
dim(geno.numeric.df.common)  
# check to make sure I get the right subset 

identical(rownames(geno.numeric.df.common)[order(rownames(geno.numeric.df.common))], common.lines$Taxa.y[order(common.lines$Taxa.y)])
geno.numeric.common <- as.matrix(geno.numeric.df.common)

genDist <- as.matrix(dist(geno.numeric.common))

#perform the multi-dimensional scaling
geno.mds <- as.data.frame(cmdscale(genDist))
head(geno.mds) #now we have 2 dimensions
dim(geno.mds)
plot(geno.mds)  

## add country of origin, winter or spring type to the figure ... 
sample_des_c_revised <- read.csv("~/505/data/phenotype/batch_c_revised.csv", header=T)
sample_des_c_sub <- sample_des_c_revised[,c("Sample.ID", "Name","Sample.Description.Origin", "Sample.Description.Type")]
sample_des_d_revised <- read.csv("~/505/data/phenotype/batch_d_revised.csv", header=T)
sample_des_d_sub <- sample_des_d_revised[,c("Sample.ID", "Name", "Sample.Description.Origin", "Sample.Description.Type")]
head(sample_des_c_sub)
sample_des_c_sub$Sample.ID
sample_des_d_sub$Sample.ID

sample_des_revised <- rbind(sample_des_c_sub, sample_des_d_sub)
sample_des_revised <- sample_des_revised[-c(48, 90:93),]
sample_des_revised$Sample.ID
save(sample_des_revised, file="~/505/output/late_silique/sample_des_revised.Rdata")

# MDS result 
# reformat MDS result sample ID (X remove & . to -)
# geno.mds$Sample.ID <- gsub("\\.","\\-", rownames(geno.mds)) 
# geno.mds$Sample.ID <- gsub("X", "", geno.mds$Sample.ID)
class(geno.mds$Sample.ID) 
geno.mds

geno.mds$Sample.ID
sample_des_revised$Sample.ID <- gsub("\\_", "\\-", sample_des_revised$Sample.ID)

# merge 
geno.mds.all.des <- merge(geno.mds, sample_des_revised, by="Sample.ID")
head(geno.mds.all.des)  
dim(geno.mds.all.des)

library(ggrepel)
set.seed(111) 

# p.mds.country <- ggplot(data=geno.mds.all.des, aes(V1, V2, color=factor(Sample.Description.Origin)))
# p.mds.country <- p.mds.country + geom_point(size=1) 
# p.mds.country <- p.mds.country + scale_color_brewer(type="qual",palette="Set1")
# p.mds.country <- p.mds.country + geom_text_repel(aes(x=V1, y=V2, label=factor(Name)))
# p.mds.country <- p.mds.country + facet_wrap(~Sample.Description.Type)
# p.mds.country 
# ggsave(p.mds.country, filename="~/505/output/figure/late_silique/p.mds.country.png", height=20, width=20)

p.mds.country.2 <- ggplot(data=geno.mds.all.des, aes(V1, V2, color=factor(Sample.Description.Origin)))
p.mds.country.2 <- p.mds.country.2 + geom_point(size=1) 
p.mds.country.2 <- p.mds.country.2 + scale_color_brewer(type="qual",palette="Set1")
p.mds.country.2 <- p.mds.country.2 + geom_text_repel(aes(x=V1, y=V2, label=factor(Name)))
# p.mds.country <- p.mds.country + facet_wrap(~Sample.Description.Type)
p.mds.country.2
ggsave(p.mds.country.2, filename="~/505/output/figure/late_silique/p.mds.country.2.common.png", height=20, width=20)

p.mds.country.2 <- ggplot(data=geno.mds.all.des, aes(V1, V2, color=factor(Sample.Description.Origin)))
p.mds.country.2 <- p.mds.country.2 + geom_point(size=1) 
p.mds.country.2 <- p.mds.country.2 + scale_color_brewer(type="qual",palette="Set1")
p.mds.country.2 <- p.mds.country.2 + geom_text_repel(aes(x=V1, y=V2, label=factor(Name)))
# p.mds.country <- p.mds.country + facet_wrap(~Sample.Description.Type)
p.mds.country.2
ggsave(p.mds.country.2, filename="~/505/output/figure/late_silique/p.mds.country.2.png", height=20, width=20)


``` 

<<<<<<< HEAD
=======

>>>>>>> 85a7272415f2dc9d23046be2677e4fb5387f8baa
# GWAS
```{r}
# reformat for GWAS 
# get hapmap file using tassel 

######## use tassel to get hapmap format ########### 
# run_pipeline.pl -SortGenotypeFilePlugin -inputFile 505_filtered_GQ.vcf.gz -outputFile 505_filtered_GQ_sorted -fileType VCF
# run_pipeline.pl -Xmx5g -fork1 -vcf 505_filtered_GQ_sorted.vcf -export -exportType Hapmap -runfork1 
# output: 505_filtered_GQ_sorted.hmp.txt  
# edit output file to remove "#" after rs & assembly 
#################################################### 
```

# GWAS run
```{r}
# GWAS_505_late_silique.R  
```

# GWAS result plot 
```{r} 
library("qqman")
Erucic_acid <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/myGAPIT/late_silique/GAPIT..Erucic_acid.GWAS.Results.csv", header = T)

Oil_content <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/myGAPIT/late_silique/GAPIT..Oil_content.GWAS.Results.csv", header = T)

Oleic_acid <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/myGAPIT/late_silique/GAPIT..Oleic_acid.GWAS.Results.csv", header = T)

colnames(Erucic_acid)[1:4] <- c("SNP", "CHR", "BP", "P")
colnames(Oil_content)[1:4] <- c("SNP", "CHR", "BP", "P")
colnames(Oleic_acid)[1:4] <- c("SNP", "CHR", "BP", "P")

png("~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/figure/leaf/Erucic_acid.png", width=12, height=10, units="in", res=300)
par(mfrow=c(3,1)) 
manhattan(Erucic_acid, main = "Erucic acid", ylim = c(0, 8), cex = 0.6, 
    cex.axis = 0.9, col = c("blue4", "orange3"), supggestiveline = 5, genomewideline = F 
    ) 

manhattan(Oleic_acid, main = "Oleic acid", ylim = c(0, 8), cex = 0.6, 
    cex.axis = 0.9, col = c("blue4", "orange3"), suggestiveline = 5, genomewideline = F 
    ) 

manhattan(Oil_content, main = "Oil content", ylim = c(0, 8), cex = 0.6, 
    cex.axis = 0.9, col = c("blue4", "orange3"), suggestiveline = 5, genomewideline = F
    ) 
dev.off() 
```

# histogram of phenotype data 
```{r}
load("~/505/output/pheno_data_505_late_silique.R")
pheno_data_505

pheno_data_505_melt <- melt(pheno_data_505[,2:4])

# plot 
library(ggplot2)

P.oil.late.silique <- ggplot(data=pheno_data_505_melt)
P.oil.late.silique <- P.oil.late.silique + geom_histogram(aes(value, fill=variable))
P.oil.late.silique <- P.oil.late.silique + facet_wrap(~variable, ncol=1)
P.oil.late.silique <- P.oil.late.silique + labs(list(x="value", y="Number of varieties"))
P.oil.late.silique <- P.oil.late.silique + theme(legend.position = "none") 

P.oil.late.silique 
ggsave(P.oil, filename="~/505/output/figure/late_silique/P.oil.png", width=6, height=10)

# do seperate graph for unique lines 
common.lines
pheno_data_505[pheno_data_505$Taxa %in% common.lines$Taxa.y,]

pheno.unique.late.silique <- pheno_data_505[which(!(pheno_data_505$Taxa %in% common.lines$Taxa.y)),]

pheno_data_505_melt_unique <- melt(pheno.unique.late.silique[,2:4])

P.oil <- ggplot(data=pheno_data_505_melt_unique)
P.oil <- P.oil + geom_histogram(aes(value, fill=variable))
P.oil <- P.oil + facet_wrap(~variable, ncol=1)
P.oil <- P.oil + labs(list(x="value", y="Number of varieties"))
P.oil <- P.oil + theme(legend.position = "none") 
P.oil

ggsave(P.oil, filename="~/505/output/figure/leaf/P.oil.late.silique.unique.png", width=6, height=10)

```
>>>>>>> 85a7272415f2dc9d23046be2677e4fb5387f8baa 

# 
```{r}
geno_505_hmp <- read.table("~/505/vcf_late_silique_no_pop/combined/505_filtered_sorted.hmp.txt", head=FALSE)
# get chrom name to numerics
geno_505_hmp$V3 <- as.numeric(geno_505_hmp$V3)
geno_505_hmp[1,3] <- "chrom"
geno_505_hmp <- geno_505_hmp[,-91]
geno_505_hmp[1,]


```


