---
title: "check_Ae_Ol_heterzygosity"
author: "Ruijuan Li"
date: "5/8/2017"
output: html_document
---

Goal: Shinje mentioned that Da-Ae went through 6 generations of selfing but she is afraid that some genes are still segregating. So we need to check that. For that purpose, we can call SNPs between Da-Ae and reference genome using freebayes, with biological replicates combined or seperate. For result with rep sepereated, the expectation is that there should not be heterzygosity at all. For result with reps combined, if there is heterzygosity (especially unexpcted hetezygosity and segregation), it can be an indication of segregation and contamination... I will also do this for Da-Ol. 

# data 
I use flowering data for this anaysis 
```{r}
# https://github.com/leejimmy93/KIAT/blob/master/SNP_calling_Ae_flowering_combined_withGQ.sh
# https://github.com/leejimmy93/KIAT/blob/master/SNP_calling_Ae_flowering_seperate_withGQ.sh 
``` 

# SNP filtering 
```{r}
# https://github.com/leejimmy93/KIAT/blob/master/filter_Ae_flowering.sh 
```

# analysis 
import data 
```{r}
# import 
Ae_flowering_combined <- read.table("~/Desktop/Brassica_project/KIAT_RNA_seq/Da-Ae_Da_Ol-1_heterzygosity_check/Ae_combined_filtered.recode.vcf", as.is=T,na.strings = ".")

Ae_flowering_rep1 <- read.table("~/Desktop/Brassica_project/KIAT_RNA_seq/Da-Ae_Da_Ol-1_heterzygosity_check/6_filtered.recode.vcf", as.is = T, na.strings = ".")

Ae_flowering_rep2 <- read.table("~/Desktop/Brassica_project/KIAT_RNA_seq/Da-Ae_Da_Ol-1_heterzygosity_check/Ae_Gae_2_filtered.recode.vcf", as.is = T, na.strings = ".")

Ae_flowering_rep3 <- read.table("~/Desktop/Brassica_project/KIAT_RNA_seq/Da-Ae_Da_Ol-1_heterzygosity_check/Ae_Gae_3_filtered.recode.vcf", as.is = T, na.strings = ".")

dim(Ae_flowering_combined); dim(Ae_flowering_rep1); dim(Ae_flowering_rep2); dim(Ae_flowering_rep3)
# [1] 145556     10
# [1] 82117    10
# [1] 62509    10
# [1] 52531    10

head(Ae_flowering_combined)
```

reformat data 
```{r}
####### function ######################################
SNP.freebayes.reformat.Ae <- function(vcf, vcf.header){ 
  colnames(vcf) <- vcf.header
  head(vcf)
  colnames(vcf)[10] <- "Ae"
  
  vcf$Ae[is.na(vcf$Ae)] <- "NA:NA:NA:NA:NA:NA:NA:NA"
  
  Ae.tmp.unique <- matrix(
    unlist(strsplit(vcf$Ae,split = ":")),
    nrow=nrow(vcf),  
    byrow=TRUE
  )
  
  colnames(Ae.tmp.unique) <- paste("Ae",c("gt", "gt.qual", "tot.depth","ref.depth","ref.qual","alt.depth","alt.qual","gen.lik"),sep="_")
  
  vcf.reform <- cbind(vcf,Ae.tmp.unique,stringsAsFactors=FALSE)
  
  vcf.reform[,c("Ae_tot.depth","Ae_gt.qual","Ae_ref.depth","Ae_ref.qual","Ae_alt.depth","Ae_alt.qual")] <- 
    apply(vcf.reform[,c("Ae_tot.depth","Ae_gt.qual","Ae_ref.depth","Ae_ref.qual","Ae_alt.depth","Ae_alt.qual")],
          2,
          as.numeric) 
  
return(vcf.reform)
  }

###################################################

# combined 
# get header 
header.combined <- system("grep '#C' ~/Desktop/Brassica_project/KIAT_RNA_seq/Da-Ae_Da_Ol-1_heterzygosity_check/Ae_combined_filtered.recode.vcf",intern = TRUE) 
header.combined <- sub("#","",header.combined) #get rid of the pound sign
header.combined <- unlist(strsplit(header.combined,split="\t"))
header.combined

# use function to reformat vcf file 
Ae_flowering_combined.reform <- SNP.freebayes.reformat.Ae(vcf = Ae_flowering_combined, vcf.header = header.combined)
dim(Ae_flowering_combined.reform) # 145556     18    
table(Ae_flowering_combined.reform$Ae_gt)

# rep1 
# get header 
header.rep1 <- system("grep '#C' ~/Desktop/Brassica_project/KIAT_RNA_seq/Da-Ae_Da_Ol-1_heterzygosity_check/6_filtered.recode.vcf",intern = TRUE) 
header.rep1 <- sub("#","",header.rep1) #get rid of the pound sign
header.rep1 <- unlist(strsplit(header.rep1,split="\t"))
header.rep1

# use function to reformat vcf file 
Ae_flowering_rep1.reform <- SNP.freebayes.reformat.Ae(vcf = Ae_flowering_rep1, vcf.header = header.rep1)
dim(Ae_flowering_rep1.reform) # 82117     
table(Ae_flowering_rep1.reform$Ae_gt)

# rep2 
# get header 
header.rep2 <- system("grep '#C'  ~/Desktop/Brassica_project/KIAT_RNA_seq/Da-Ae_Da_Ol-1_heterzygosity_check/Ae_Gae_2_filtered.recode.vcf",intern = TRUE) 
header.rep2 <- sub("#","",header.rep2) #get rid of the pound sign
header.rep2 <- unlist(strsplit(header.rep2,split="\t"))
header.rep2

# use function to reformat vcf file 
Ae_flowering_rep2.reform <- SNP.freebayes.reformat.Ae(vcf = Ae_flowering_rep2, vcf.header = header.rep2)
dim(Ae_flowering_rep2.reform) # 145556     18    
table(Ae_flowering_rep2.reform$Ae_gt)

# rep3 
# get header 
header.rep3 <- system("grep '#C' ~/Desktop/Brassica_project/KIAT_RNA_seq/Da-Ae_Da_Ol-1_heterzygosity_check/Ae_Gae_3_filtered.recode.vcf",intern = TRUE) 
header.rep3 <- sub("#","",header.rep3) #get rid of the pound sign
header.rep3 <- unlist(strsplit(header.rep3,split="\t"))
header.rep3

# use function to reformat vcf file 
Ae_flowering_rep3.reform <- SNP.freebayes.reformat.Ae(vcf = Ae_flowering_rep3, vcf.header = header.rep3)
dim(Ae_flowering_rep3.reform) # 145556     18    
table(Ae_flowering_rep3.reform$Ae_gt)  
```

If most of the heterozygous problem was caused by homeologous exchange, heterozygosity should appear in clusters 
```{r}
Ae_flowering_combined.reform$Ae_het <- ifelse(Ae_flowering_combined.reform$Ae_gt=="0/1", 1, 0)
Ae_flowering_combined.reform$Ae_het

# calculate het ratio for a window size of 1Mb 
binsize <- 1000000

Ae_flowering_combined.reform <- Ae_flowering_combined.reform[grep("random", Ae_flowering_combined.reform$CHROM, invert = T),]
dim(Ae_flowering_combined.reform) # 117569     21

Ae_flowering_combined.reform$subgenome <- gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\2", Ae_flowering_combined.reform$CHROM)
Ae_flowering_combined.reform$chr_ID <- gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\3", Ae_flowering_combined.reform$CHROM) 

# number of SNPs on A & C subgenome
sum(Ae_flowering_combined.reform$subgenome=="A")/nrow(Ae_flowering_combined.reform) # 54.4%
sum(Ae_flowering_combined.reform$subgenome=="C") # 53618 

## make plot for main & random chromosome seperately (distribution of SNPs per Mb)
library(ggplot2)
pl.SNP.Ae <- ggplot(data = Ae_flowering_combined.reform)
pl.SNP.Ae <- pl.SNP.Ae + geom_histogram(aes(x=POS,y=..count.., fill=subgenome), binwidth = binsize, alpha=0.5)  
pl.SNP.Ae <- pl.SNP.Ae + facet_grid(chr_ID ~subgenome)
pl.SNP.Ae <- pl.SNP.Ae + labs(list(title = "", x = "chromosome ID", y = "number of SNPs"))
pl.SNP.Ae <- pl.SNP.Ae + theme(legend.position = "none")
pl.SNP.Ae 

Ae_flowering_combined.reform.Ae.het <- 
Ae_flowering_combined.reform %>%
  filter(Ae_het == 1)
dim(Ae_flowering_combined.reform) # 117569     21 
dim(Ae_flowering_combined.reform.Ae.het) # 54966    21 

# number of Ae heterozygous SNPs 
pl.SNP.Ae.het <- ggplot(data = Ae_flowering_combined.reform.Ae.het)
pl.SNP.Ae.het <- pl.SNP.Ae.het + geom_histogram(aes(x=POS, fill=subgenome), binwidth = binsize)
pl.SNP.Ae.het <- pl.SNP.Ae.het + facet_grid(chr_ID ~subgenome)
pl.SNP.main.2.unique <- pl.SNP.Ae.het + labs(list(title = "", x = "chromosome ID", y = "number of SNPs"))
pl.SNP.Ae.het <- pl.SNP.Ae.het + theme(legend.position = "none")
pl.SNP.Ae.het

# overlay the two 
library(ggplot2)
pl.SNP.Ae.ratio <- ggplot(data = Ae_flowering_combined.reform)
pl.SNP.Ae.ratio <- pl.SNP.Ae.ratio + geom_histogram(aes(x=POS,y=..count..), binwidth = binsize, alpha=0.3) 
pl.SNP.Ae.ratio <- pl.SNP.Ae.ratio + facet_grid(chr_ID ~subgenome)
pl.SNP.Ae.ratio <- pl.SNP.Ae.ratio + labs(list(title = "", x = "physical position on chromosome", y = "number of SNPs"))
pl.SNP.Ae.ratio <- pl.SNP.Ae.ratio + theme(legend.position = "none")
pl.SNP.Ae.ratio <- pl.SNP.Ae.ratio + geom_histogram(data = Ae_flowering_combined.reform.Ae.het, aes(x=POS,y=..count.., fill=subgenome), binwidth = binsize, alpha=0.8)
pl.SNP.Ae.ratio 
ggsave(pl.SNP.Ae.ratio, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/parent_SNP/output/figure/pl.SNP.Ae.ratio.png", width = 10, height = 15)

# ratio of heterozygous SNPs (het SNP/ total number of SNPs in a window size of 1Mb)
het.data <- ggplot_build(pl.SNP.Ae.het)
het.data.final <- het.data$data[[1]]

SNP.data <- ggplot_build(pl.SNP.Ae)
SNP.data.final <- SNP.data$data[[1]]

het.ratio <- het.data.final$count/SNP.data.final$count

het.ratio
length(het.ratio) # 1178

bin.number <- length(unique(SNP.data.final$xmax)) # 1178, so every existing chromosome need to have 62 bins 

CHROM=c(paste("chrA", rep("01", 62), sep = ""), paste("chrC", rep("01", 62), sep = ""), paste("chrA", rep("02", 62), sep = ""), paste("chrC", rep("02", 62), sep = ""), paste("chrA", rep("03", 62), sep = ""), paste("chrC", rep("03", 62), sep = ""), paste("chrA", rep("04", 62), sep = ""), paste("chrC", rep("04", 62), sep = ""), paste("chrA", rep("05", 62), sep = ""), paste("chrC", rep("05", 62), sep = ""), paste("chrA", rep("06", 62), sep = ""), paste("chrC", rep("06", 62), sep = ""), paste("chrA", rep("07", 62), sep = ""), paste("chrC", rep("07", 62), sep = ""), paste("chrA", rep("08", 62), sep = ""), paste("chrC", rep("08", 62), sep = ""), paste("chrA", rep("09", 62), sep = ""), paste("chrC", rep("09", 62), sep = ""),
paste("chrA", rep("10", 62), sep = ""))

het.ratio.final <- data.frame(CHROM=CHROM,
                    bin=rep(1:bin.number, 19),
                    het_ratio = het.ratio
)

# reform & plot 
het.ratio.final$subgenome <- gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\2", het.ratio.final$CHROM)
het.ratio.final$chr_ID <- gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\3", het.ratio.final$CHROM) 

het.ratio.final[het.ratio.final$CHROM=="chrC03",]
het.ratio.final

## make plot for main & random chromosome seperately (distribution of SNPs per Mb)
library(ggplot2)
pl.SNP.Ae.het.ratio.final <- ggplot(data = het.ratio.final)
pl.SNP.Ae.het.ratio.final <- pl.SNP.Ae.het.ratio.final + geom_col(aes(x=bin,y=het_ratio, fill=subgenome)) 
pl.SNP.Ae.het.ratio.final <- pl.SNP.Ae.het.ratio.final + facet_grid(chr_ID ~subgenome)
pl.SNP.Ae.het.ratio.final <- pl.SNP.Ae.het.ratio.final + labs(list(title = "", x = "bin in 1Mb", y = "ratio of heterozygosity"))
pl.SNP.Ae.het.ratio.final <- pl.SNP.Ae.het.ratio.final + theme(legend.position = "none")
pl.SNP.Ae.het.ratio.final 

ggsave(pl.SNP.Ae.het.ratio.final, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/parent_SNP/output/figure/pl.SNP.Ae.ratio.final.png", width = 10, height = 15)

```

By checking the result from last code chunk, subgenome C has higher het ratio although its SNP density is lower, probably because it has more repeated regions, which cause reads from A subgeome more falsefully mapped to different/similar regions of C subgenome. So many of detected SNPs on C sub genome are then appeared as het. Reads from C also map to A but they are more clustered, because A subgeome is more unqiue compared to C subgenome. 

A subgenome size: 314.2 Mb; C subgenome size: 525.8 Mb  
A subgenome gene number: 44,452
C subgenome gene number: 56,055 
Reads uniquely mapped to A subgenome:
34,174,744   
Reads uniquely mapped to C subgenome: 
36,333,626

# For SNPs that are heterzygous in rep combined sample, what are their genotypes in the rep seperate samples. 
```{r}
# keep only common SNPs for the 4 dataset (1 rep combined & 3 rep seperate samples)
# https://github.com/leejimmy93/KIAT/blob/master/get_common_SNP.sh

# analyze vcf file 



```
























