---
title: "F2_map_construction"
author: "Ruijuan Li"
date: "4/18/2017"
output: html_document
---

# Goal
The goal of this analysis is to build genetic map using F2 genotyping information, F2 genotyping was done based on the SNPs we identified from its parents 

# install & load package 
```{r}
# install.packages("onemap")
# install.packages("tkrplot", type="source") 
library(tcltk)
library(tkrplot)
library(onemap)  
library(ggplot2)
library(reshape) 
# library(tidyverse) 
```

# import & format data (MAC version)
```{r}
F2_geno_data <- read.table("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/F2_Final_SNP_Calls", header = T)
head(F2_geno_data) 
dim(F2_geno_data) # 18226   172 
class(F2_geno_data)

rownames(F2_geno_data) <- paste(F2_geno_data$CHROM, F2_geno_data$POS, sep = "_")
F2_geno_data <- F2_geno_data[, -c(1:6)]
dim(F2_geno_data) # 18226   166 
colnames(F2_geno_data) <- gsub("([[:print:]]+)(\\.)([[:digit:]])", "\\3", colnames(F2_geno_data))
colnames(F2_geno_data)  
```

# check missing rate & remove data with high missing rate above 0.85 
```{r}
F2_geno_data_with_missing_rate <- mutate(F2_geno_data, 
       missing_rate = round(apply(F2_geno_data, 1, function(x) sum(is.na(x)))/94, 2))

rownames(F2_geno_data_with_missing_rate) <- rownames(F2_geno_data) 
hist(F2_geno_data_with_missing_rate$missing_rate)
max(F2_geno_data_with_missing_rate$missing_rate) # 0.83, the highest missing rate is 0.83 

# remove markers with missing rate greater than 0.85 
F2_geno_data_with_missing_rate_filtered <- F2_geno_data_with_missing_rate[F2_geno_data_with_missing_rate$missing_rate < 0.15,]
rownames(F2_geno_data_with_missing_rate_filtered)

dim(F2_geno_data_with_missing_rate_filtered) # 3077 167
hist(F2_geno_data_with_missing_rate_filtered$missing_rate)

F2_geno_data_with_missing_rate_filtered.final <- subset(F2_geno_data_with_missing_rate_filtered, select=-missing_rate)
rownames(F2_geno_data_with_missing_rate_filtered.final)

dim(F2_geno_data_with_missing_rate_filtered.final) # 3077 166 
rownames(F2_geno_data_with_missing_rate_filtered.final) 
```


# calculate pairwise corrlelation and remove SNPs with corrlelation greater or equal to 0.9
some SNPs should have exactly the same genotype across all individuals, they are redundant in terms of genetic map construction, so they should be removed. find those SNPs by doing correlation test. 
```{r}
F2_geno_data_t <- as.data.frame(t(F2_geno_data_with_missing_rate_filtered.final))
F2_geno_data_t.numeric <- data.matrix(F2_geno_data_t)

# delete markers with all same genotypes across individuals 
test <- apply(F2_geno_data_t.numeric, 2, function(x) length(unique(x[!is.na(x)])))
filter.polymorphsm <- test != 1 

F2_geno_data_t.numeric.2 <- F2_geno_data_t.numeric[,filter.polymorphsm]
dim(F2_geno_data_t.numeric.2) # 166 3075  

# also output save markers with same genotype across individuals
polymorphism <- test == 1
SNP.not.poly <- colnames(F2_geno_data_t.numeric)[polymorphism]
length(SNP.not.poly) # two markers are not polymorphism across individuals 
save(SNP.not.poly, file="SNP.not.poly.Rdata")

# correlation test
options(warn=-1) # suppress warning message
F2_SNP_correlation <- cor(F2_geno_data_t.numeric.2, use = "pairwise.complete.obs")
save(F2_SNP_correlation, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_SNP_correlation.Rdata")
options(warn=0) # unsuppress warning message
dim(F2_SNP_correlation) # 3075 3075 

# calcualte number of SNPs with correlation of 1
nrow(which(F2_SNP_correlation == 1 & lower.tri(F2_SNP_correlation), arr.ind = T, useNames = T)) # 150 pairs of SNPs have identical genotype data 

# find SNPs with correlation of 1 and remove them 
dup.cordinate <- which(F2_SNP_correlation == 1 & lower.tri(F2_SNP_correlation), arr.ind = T, useNames = F)
length(dup.cordinate) # 300 
dup.cordinate.df <- as.data.frame(dup.cordinate)
sample.ID <- colnames(F2_SNP_correlation)

# extract duplicate pair information based on their coordicate
dup.pair <- data.frame(matrix(nrow = nrow(dup.cordinate), ncol = 2))
for (i in 1:nrow(dup.cordinate)){
 dup.pair[i,1] <- sample.ID[dup.cordinate[i,1]]
 dup.pair[i,2] <- sample.ID[dup.cordinate[i,2]]
}

length(dup.pair[,2]) # 150 
rem <- unique(dup.pair[,2])
length(rem) # 143 markers are removed due to pairwise correlation of 1 

# save data
save(dup.pair, file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/dup.pair.Rdata")
save(rem, file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/rem.Rdata") 

### 
F2_geno_data_1 <- F2_geno_data_with_missing_rate_filtered.final[!(rownames(F2_geno_data_with_missing_rate_filtered.final) %in% rem),]  

# remove non polymorphism SNPs 
F2_geno_data_2 <- F2_geno_data_1[!(rownames(F2_geno_data_1) %in% SNP.not.poly),]

dim(F2_geno_data_2) # 2932  166, end up with 2932 SNPs for map construction 

save(F2_geno_data_2, file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_data_2.Rdata") # this is non-redundant SNPs with missing data less than 0.85   
```

# format data for onemap  
```{r}
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_data_2.Rdata")
dim(F2_geno_data_2) # 2932 166 

# formating data for onemap 
F2_geno_data_2 <- as.matrix(F2_geno_data_2)
F2_geno_data_2[is.na(F2_geno_data_2)] <- "-"
F2_geno_data_2 <- gsub("0/0", "A", F2_geno_data_2) 
F2_geno_data_2 <- gsub("0/1", "H", F2_geno_data_2)
F2_geno_data_2 <- gsub("1/1", "B", F2_geno_data_2)

F2_geno_data_2 <- as.data.frame(F2_geno_data_2)
dim(F2_geno_data_2) # 2932 166  
write.table(F2_geno_data_2, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_data.txt")
# change file format in linux 
# cat F2_geno_data.txt | sed 's/"//g' | awk '{first = $1; $1 = ""; print $0}' | sed 's/ //g' > tmp
# tail -2932 tmp > tmp.1 

write.table(rownames(F2_geno_data_2), file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/marker_info.txt" )
# cat marker_info.txt | awk '{print "*"$2}' | tail -2932 | sed 's/"//g' > marker_info_reform.txt

# paste marker_info_reform.txt tmp.1 | awk '{print $1 " " $2}' > F2_geno_for_one_map.txt 
# cat header_one_map_input F2_geno_for_one_map.txt > F2_geno_for_one_map_final.txt 
```

# calculate two point rf & assign to linkage groups (on whitney) 
```{r}
# F2_map_construction_LOD3_rf0.5_missing_0.15.R 
# scp rf.Rdata to my MAC
# after checking the group result, only one linkage group was formed... 
```

# assign markers to different linkage groups based on their chromsome assignment 
```{r}
# get the number of markers for each chromosome 
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_data_2.Rdata")
dim(F2_geno_data_2) # 2932 166

F2_geno_data_2$Chr <- gsub("([[:print:]]+)(_)([[:digit:]]+)", "\\1", rownames(F2_geno_data_2)) 
F2_geno_data_2$Pos <- gsub("([[:print:]]+)(_)([[:digit:]]+)", "\\3", rownames(F2_geno_data_2)) 
F2_geno_data_2$index <- seq(1: nrow(F2_geno_data_2))

head(F2_geno_data_2[F2_geno_data_2$Chr=="chrA04",],1)
tail(F2_geno_data_2[F2_geno_data_2$Chr=="chrC09",],1)

sum(F2_geno_data_2$Chr=="chrA01") # 1-192  
sum(F2_geno_data_2$Chr=="chrA02") # 78 193-270
sum(F2_geno_data_2$Chr=="chrA03") # 267 271-537
sum(F2_geno_data_2$Chr=="chrA04") # 136 538-673
sum(F2_geno_data_2$Chr=="chrA05") # 129 674-802
sum(F2_geno_data_2$Chr=="chrA06") # 379 803-1181
sum(F2_geno_data_2$Chr=="chrA07") # 242 1182-1423
sum(F2_geno_data_2$Chr=="chrA08") # 151 1424-1574
sum(F2_geno_data_2$Chr=="chrA09") # 307 1575-1881
sum(F2_geno_data_2$Chr=="chrA10") # 253 1882-2134
sum(F2_geno_data_2$Chr=="chrC01") # 106 2135-2240
sum(F2_geno_data_2$Chr=="chrC02") # 59 2241-2299
sum(F2_geno_data_2$Chr=="chrC03") # 158 2300-2457
sum(F2_geno_data_2$Chr=="chrC04") # 143 2458-2600
sum(F2_geno_data_2$Chr=="chrC05") # 43 2601-2643
sum(F2_geno_data_2$Chr=="chrC06") # 85 2644-2728
sum(F2_geno_data_2$Chr=="chrC07") # 100 2729-2828
sum(F2_geno_data_2$Chr=="chrC08") # 60 2829-2888
sum(F2_geno_data_2$Chr=="chrC09") # 44 2889-2932  

# https://github.com/leejimmy93/KIAT/map_A01.R .... mapC09.R 
```

# get marker index within each chromosome 
```{r}
###################### below can be useful when I only need marker order within each chromosome ## 
# figure out the index positon within each chromosome for all chromosomes 
F2_geno_data_info <- F2_geno_data_2[,c("Chr", "Pos", "index")]
dim(F2_geno_data_info) # 2932 3 
head(F2_geno_data_info, 1000)

out <- split( F2_geno_data_info, f = F2_geno_data_info$Chr)
length(out) # 19 
out

order <- list()

for (i in 1:length(out)){
  out[[i]]$Pos <- as.numeric(out[[i]]$Pos)
  order[[i]] <- out[[i]][order(out[[i]]$Pos),]$index 
}

order 
save(order, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/order.Rdata")
```

# after getting good number of linkage group, work on each LG seperately... 
A01 
```{r}
#################### A01 ##############################
# load data 
F2.data <- read.mapmaker(file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_for_one_map_final.txt") 
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/twopts.f2.05.12.Rdata")

# mapping order 
LG.f2.arbi.A01 <- make.seq(twopts.f2.05.12, order[[1]])

LG.f2.arbi.A01.map <- map(LG.f2.arbi.A01)
rf.graph.table(LG.f2.arbi.A01.map) 

draw.map(LG.f2.arbi.A01.map, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T) 

# safe mode order (safe order should be recommended because we have enough marker here)
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.A01.Rdata")
LG1.f2.safe <- make.seq(LG.f2.ord.A01, "safe")

LG1.f2.safe 

# use touchdown
LG1.f2.ord <- order.seq(input.seq = LG1.f2, n.init = 5,
                        subset.search = "twopt",
                        twopt.alg = "rcd", THRES = 6,
                        draw.try = T, wait = 1,
                        touchdown = T)

LG1.f2.ord

# safe mode order (safe order should be recommended because we have enough marker here)
LG1.f2.safe <- make.seq(LG1.f2.ord, "safe")
LG1.f2.safe 
# all order
LG1.f2.final <- make.seq(LG1.f2.ord, "force") 
LG1.f2.final

# ripple
ripple.seq(LG1.f2.final, ws = 5, LOD = 6) 

#########
###### don't consider mapping postion at all
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/LG.f2.ord.A01.Rdata")
LG.f2.ord.A01 

# set map function 
set.map.fun(type="kosambi")

# safe mode order (only marker with unique positions)
LG.f2.safe.A01 <- make.seq(LG.f2.ord.A01, "safe")
LG.f2.safe.A01
# all order (all markers)
LG.f2.final.A01 <- make.seq(LG.f2.ord.A01, "force")
LG.f2.final.A01

# print the recombination fraction mix 
rf.graph.table(LG.f2.safe.A01) 
par(mfrow=c(1,1))
draw.map(LG.f2.safe.A01, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

# map estimation for arbitrary order 
LG.f2.arbi.A01 <- make.seq(twopts.f2.LOD3_rf0.5, order[[1]])
LG.f2.arbi.A01.map <- map(LG.f2.arbi.A01)
rf.graph.table(LG.f2.arbi.A01.map) 

draw.map(LG.f2.arbi.A01.map, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)  

####### new stuff
F2.data <- read.mapmaker(file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_for_one_map_final.txt")
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/twopts.LOD3_rf0.5.Rdata")

mark.f2.A01 <- make.seq(twopts.f2.LOD3_rf0.5, 1:192)

LG.f2.ord.A01 <- order.seq(input.seq = mark.f2.A01, n.init = 5,
                        subset.search = "twopt",
                        twopt.alg = "rcd", THRES = 6,
                        draw.try = T, wait = 1,
                        touchdown = T)


```

# A02
```{r}
################# A02 ##################################
######## import data #######
# marker data 
F2.data <- read.mapmaker(file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_for_one_map_final.txt") 
F2.data <- read.mapmaker(file="~/F2/output/missing_rate_0.15/F2_geno_for_one_map_final.txt")
F2.data 
# This is an object of class 'f2.onemap'
#     No. individuals:     166 
#     No. markers:         2932 
#     Percent genotyped:   96 
# 
#     Number of markers per type:
#        AA : AB : BB -->  2932
# 
# This data contains no phenotypic information

# two-point rf data  
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/twopts.f2.05.12.Rdata")
# load("~/F2/output/missing_rate_0.15/twopts.f2.05.12.Rdata")
########## order maker within chromosome #############
###### don't consider mapping postion at all
# https://github.com/leejimmy93/KIAT/blob/master/map_construction_missing_0.15/map_A02_2.R_final 
# mark.f2.A02 <- make.seq(twopts.f2.05.12, 193:270)
# LG.f2.ord.A02 <- order.seq(input.seq = mark.f2.A02, n.init = 5,
#                         subset.search = "twopt",
#                         twopt.alg = "rcd", THRES = 6, 
#                         draw.try = T, wait = 1,
#                         touchdown = T)

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.A02.Rdata")
load("~/F2/output/missing_rate_0.15/LG.f2.ord.A02.Rdata")
LG.f2.A02 <- make.seq(LG.f2.ord.A02) 
LG.f2.A02
rf.graph.table(LG.f2.A02) 
par(mfrow=c(1,1)) 
draw.map(LG.f2.A02, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T) 

# all order
LG.f2.A02.final <- make.seq(LG.f2.ord.A02, "force")
LG.f2.A02.final
rf.graph.table(LG.f2.A02.final) 
draw.map(LG.f2.A02.final, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T) 

# ripple
ripple.seq(LG.f2.A02.final, ws = 5, LOD = 6) 

# set map function 
set.map.fun(type="kosambi")

# map estimation for arbitrary order 
LG.f2.arbi.A02 <- make.seq(twopts.f2.05.12, order[[2]])
LG.f2.arbi.A02 
LG.f2.arbi.A02.map <- map(LG.f2.arbi.A02)  
rf.graph.table(LG.f2.arbi.A02.map) 

draw.map(LG.f2.arbi.A02.map, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T) 

#####
mark.f2.A02 <- make.seq(twopts.f2.05.12, 193:270)
LG.f2.ord.A02.1 <- order.seq(input.seq = mark.f2.A02, n.init = 5,
                        subset.search = "twopt",
                        twopt.alg = "rcd", THRES = 7,
                        draw.try = T, wait = 1
                        )

LG.f2.A02.2 <- make.seq(LG.f2.ord.A02.1, "safe")
LG.f2.A02.2
rf.graph.table(LG.f2.A02)  
draw.map(LG.f2.A02.2, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T) 

```

# C01 
```{r}
################# C01 ##############################

######## import data #######
# marker data 
F2.data <- read.mapmaker(file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_for_one_map_final.txt") 
F2.data 
# This is an object of class 'f2.onemap'
#     No. individuals:     166  
#     No. markers:         2932 
#     Percent genotyped:   96 
# 
#     Number of markers per type:
#        AA : AB : BB -->  2932
# 
# This data contains no phenotypic information

# two-point rf data  
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/twopts.f2.05.12.Rdata")

########## order maker within chromosome #############
###### don't consider mapping postion at all
# https://github.com/leejimmy93/KIAT/blob/master/map_construction_missing_0.15/map_C01_2.R_final 
# mark.f2.C01 <- make.seq(twopts.f2.05.12, 2135:2240)
# LG.f2.ord.C01 <- order.seq(input.seq = mark.f2.C01, n.init = 5,
#                         subset.search = "twopt",
#                         twopt.alg = "rcd", THRES = 6, # notice I used 6 here, may need to change later... 
#                         draw.try = T, wait = 1,
#                         touchdown = T) 

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.C01.Rdata")
LG.f2.ord.C01 

# set map function 
set.map.fun(type="kosambi")  

# order marker 
LG.f2.C01 <- make.seq(LG.f2.ord.C01) 
LG.f2.C01
rf.graph.table(LG.f2.C01) 
par(mfrow=c(1,1))
draw.map(LG.f2.C01, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)




LG1.f2.C01.rcd <- order.seq(input.seq = LG1.f2.C01, n.init = 5, 
                            subset.search = "twopt",
                            twopt.alg = "rcd", THRES = 3,
                            draw.try = T, wait = 1
                            )

LG1.f2.C01.rcd

LG1.f2.C01.rcd <- order.seq(input.seq = LG1.f2.C01, n.init = 5, 
                            subset.search = "twopt",
                            twopt.alg = "rcd", THRES = 3,
                            draw.try = T, wait = 1,
                            touchdown = T
                            )

ripple.seq(LG1.f2.C01, ws=5, LOD=3)
LG1.f2.C01

# safe mode order (only marker with unique positions)
LG1.f2.safe.C01 <- make.seq(LG1.f2.ord.C01, "safe")
LG1.f2.safe.C01
# all order (all markers)
LG.f2.force.C01 <- make.seq(LG.f2.ord.C01, "force")
LG.f2.force.C01
rf.graph.table(LG.f2.force.C01) 
par(mfrow=c(1,1))
draw.map(LG.f2.force.C01, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

# print the recombination fraction mix 
rf.graph.table(LG1.f2.safe.C01) 
par(mfrow=c(1,1))
draw.map(LG1.f2.safe.C01, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

# map estimation for arbitrary order 
LG.f2.arbi.C01 <- make.seq(twopts.f2.05.12, order[[11]])
LG.f2.arbi.C01
LG.f2.arbi.C01.map <- map(LG.f2.arbi.C01) 
rf.graph.table(LG.f2.arbi.C01.map) 

draw.map(LG.f2.arbi.C01.map, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T) 

# set map function 
set.map.fun(type = "haldane") 
```

# C05 (a total of 43 markers)
```{r}
######## import data #######
# marker data 
F2.data <- read.mapmaker(file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_for_one_map_final.txt") 
F2.data 

# two-point rf data  
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/twopts.f2.05.12.Rdata")

########## order maker within chromosome #############
###### don't consider mapping postion at all
# https://github.com/leejimmy93/KIAT/blob/master/map_construction_missing_0.15/map_C05_2.R_final 
# mark.f2.C05 <- make.seq(twopts.f2.05.12, 2601:2643)
# LG.f2.ord.C05 <- order.seq(input.seq = mark.f2.C05, n.init = 5,
#                         subset.search = "twopt",
#                         twopt.alg = "rcd", THRES = 6,
#                         draw.try = T, wait = 1,
#                         touchdown = T) 

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.C05.Rdata")
LG.f2.ord.C05

# set map function 
set.map.fun(type="kosambi")  

# order marker 
LG.f2.C05 <- make.seq(LG.f2.ord.C05) 
LG.f2.C05
rf.graph.table(LG.f2.C05) 
par(mfrow=c(1,1))
draw.map(LG.f2.C05, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)
# looks off 

# use higher LOD score to gather markers with "right" order to start from, LOD score of 6 no touchdown 
mark.f2.C05 <- make.seq(twopts.f2.05.12, 2601:2643) 
LG.f2.ord.C05.2 <- order.seq(input.seq = mark.f2.C05, n.init = 5,
                        subset.search = "twopt",
                        twopt.alg = "rcd", THRES = 6,
                        draw.try = T, wait = 1
                        )

LG.f2.C05.2 <- make.seq(LG.f2.ord.C05.2) 
LG.f2.C05.2
rf.graph.table(LG.f2.C05.2) 
par(mfrow=c(1,1))
draw.map(LG.f2.C05.2, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T) 
# this method dosn't seem work... order looks off still 

# seems like a lower LOD can make the map look better, reduce LOD threshold to 3 and use touchdown 
LG.f2.ord.C05.3 <- order.seq(input.seq = mark.f2.C05, n.init = 5,
                        subset.search = "twopt",
                        twopt.alg = "rcd", THRES = 3,
                        draw.try = T, wait = 1,
                        touchdown = T
                        )

LG.f2.C05.3 <- make.seq(LG.f2.ord.C05.3) 
LG.f2.C05.3
rf.graph.table(LG.f2.C05.3) 
par(mfrow=c(1,1))
draw.map(LG.f2.C05.3, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T) 
####### actually no big difference... just use the 1st one, flip one of the marker chuncks 
LG.f2.C05.force <- make.seq(LG.f2.ord.C05, "force") 
LG.f2.C05.force
rf.graph.table(LG.f2.C05.force) 
par(mfrow=c(1,1))
draw.map(LG.f2.C05.force, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)
# looks off
# move the last two markers to the begining of the map chrC05_163897 (2609) & chrC05_323355 (2625)
LG.f2.C05.force$seq.num[1:41]
C01.order.1 <- c(2609, 2625, 2602, 2603, 2608, 2627, 2635, 2636, 2637, 2641, 2634, 2638, LG.f2.C05.force$seq.num[1:31])

LG.f2.arbi.C05 <- make.seq(twopts.f2.05.12, C01.order.1)

LG.f2.arbi.C05
LG.f2.arbi.C05.map <- map(LG.f2.arbi.C05) 
rf.graph.table(LG.f2.arbi.C05.map) 

draw.map(LG.f2.arbi.C05.map, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T) 

LG.f2.arbi.C05$seq.num
order.2 <- c(LG.f2.arbi.C05$seq.num[6:12], LG.f2.arbi.C05$seq.num[1:5], LG.f2.arbi.C05$seq.num[13:41])
LG.f2.arbi.C05.2 <- make.seq(twopts.f2.05.12, order.2)

LG.f2.arbi.C05.map.2 <- map(LG.f2.arbi.C05.2) 
rf.graph.table(LG.f2.arbi.C05.map.2) 

LG.f2.arbi.C05.2$seq.num
LG.f2.arbi.C05.3 <- drop.marker(LG.f2.arbi.C05.map.2, c(2604, 2605, 2626, 2631, 2632, 2630))
LG.f2.arbi.C05.map.3 <- map(LG.f2.arbi.C05.3) 
rf.graph.table(LG.f2.arbi.C05.map.3) 

LG.f2.arbi.C05.3$seq.num
```


# just export the orignal order 
```{r}
F2.data <- read.mapmaker(file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_for_one_map_final.txt") 
F2.data 
# This is an object of class 'f2.onemap'
#     No. individuals:     166 
#     No. markers:         2932 
#     Percent genotyped:   96 
# 
#     Number of markers per type:
#        AA : AB : BB -->  2932
# 
# This data contains no phenotypic information

# two-point rf data  
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/twopts.f2.05.12.Rdata") 

####### A subgenome ############### 
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.A01.Rdata")
LG.f2.A01.force <- make.seq(LG.f2.ord.A01, "force") 
LG.f2.A01.force
rf.graph.table(LG.f2.A01.force) 
par(mfrow=c(1,1))
draw.map(LG.f2.A01.force, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.A02.Rdata")
LG.f2.A02.force <- make.seq(LG.f2.ord.A02, "force") 
rf.graph.table(LG.f2.A02.force) 
par(mfrow=c(1,1))
draw.map(LG.f2.A02.force, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.A03.Rdata")
LG.f2.A03.force <- make.seq(LG.f2.ord.A03, "force") 
rf.graph.table(LG.f2.A03.force) 
par(mfrow=c(1,1))
draw.map(LG.f2.A02.force, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.A04.Rdata")
LG.f2.A04.force <- make.seq(LG.f2.ord.A04, "force") 
rf.graph.table(LG.f2.A04.force) 
par(mfrow=c(1,1))
draw.map(LG.f2.A04.force, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.A05.Rdata")
LG.f2.A05.force <- make.seq(LG.f2.ord.A05, "force") 
rf.graph.table(LG.f2.A05.force) 
par(mfrow=c(1,1))
draw.map(LG.f2.A05.force, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.A06.Rdata")
LG.f2.A06.force <- make.seq(LG.f2.ord.A06, "force") 
LG.f2.A06.force
rf.graph.table(LG.f2.A06.force) 
par(mfrow=c(1,1))
draw.map(LG.f2.A06.force, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.A07.Rdata")
LG.f2.A07.force <- make.seq(LG.f2.ord.A07, "force") 
rf.graph.table(LG.f2.A07.force) 
par(mfrow=c(1,1))
draw.map(LG.f2.A07.force, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.A08.Rdata")
LG.f2.A08.force <- make.seq(LG.f2.ord.A08, "force") 
rf.graph.table(LG.f2.A08.force) 
par(mfrow=c(1,1))
draw.map(LG.f2.A08.force, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.A09.Rdata")
LG.f2.A09.force <- make.seq(LG.f2.ord.A09, "force") 
rf.graph.table(LG.f2.A09.force) 
par(mfrow=c(1,1))
draw.map(LG.f2.A09.force, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.A10.Rdata")
LG.f2.A10.force <- make.seq(LG.f2.ord.A10, "force") 
rf.graph.table(LG.f2.A10.force) 
par(mfrow=c(1,1))
draw.map(LG.f2.A10.force, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T) 

map.Asub <- list(LG.f2.A01.force, LG.f2.A02.force, LG.f2.A03.force, LG.f2.A04.force, LG.f2.A05.force, LG.f2.A06.force, LG.f2.A07.force, LG.f2.A08.force, LG.f2.A09.force, LG.f2.A10.force)
draw.map(map.Asub, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal = T) 

write.map(map.Asub, "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/map.Asub.map")

######## C subgenome #################
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.C01.Rdata")
LG.f2.C01.force <- make.seq(LG.f2.ord.C01, "force") 
LG.f2.C01.force
rf.graph.table(LG.f2.C01.force) 
par(mfrow=c(1,1))
draw.map(LG.f2.C01.force, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.C02.Rdata")
LG.f2.C02.force <- make.seq(LG.f2.ord.C02, "force") 
rf.graph.table(LG.f2.C02.force) 
par(mfrow=c(1,1))
draw.map(LG.f2.C02.force, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.C03.Rdata")
LG.f2.C03.force <- make.seq(LG.f2.ord.C03, "force") 
rf.graph.table(LG.f2.C03.force) 
par(mfrow=c(1,1))
draw.map(LG.f2.C03.force, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.C04.Rdata")
LG.f2.C04.force <- make.seq(LG.f2.ord.C04, "force") 
rf.graph.table(LG.f2.C04.force) 
par(mfrow=c(1,1))
draw.map(LG.f2.C04.force, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.C05.Rdata")
LG.f2.C05.force <- make.seq(LG.f2.ord.C05, "force") 
rf.graph.table(LG.f2.C05.force) 
par(mfrow=c(1,1))
draw.map(LG.f2.C05.force, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.C06.Rdata")
LG.f2.C06.force <- make.seq(LG.f2.ord.C06, "force") 
LG.f2.C06.force
rf.graph.table(LG.f2.C06.force) 
par(mfrow=c(1,1))
draw.map(LG.f2.C06.force, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.C07.Rdata")
LG.f2.C07.force <- make.seq(LG.f2.ord.C07, "force") 
rf.graph.table(LG.f2.C07.force) 
par(mfrow=c(1,1))
draw.map(LG.f2.C07.force, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.C08.Rdata")
LG.f2.C08.force <- make.seq(LG.f2.ord.C08, "force") 
rf.graph.table(LG.f2.C08.force) 
par(mfrow=c(1,1))
draw.map(LG.f2.C08.force, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.C09.Rdata")
LG.f2.C09.force <- make.seq(LG.f2.ord.C09, "force") 
rf.graph.table(LG.f2.C09.force) 
par(mfrow=c(1,1))
draw.map(LG.f2.C09.force, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

map.Csub <- list(LG.f2.C01.force, LG.f2.C02.force, LG.f2.C03.force, LG.f2.C04.force, LG.f2.C05.force, LG.f2.C06.force, LG.f2.C07.force, LG.f2.C08.force, LG.f2.C09.force)
draw.map(map.Csub, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal = T) 

write.map(map.Csub, "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/map.Csub.map") 

map.AC <- list(LG.f2.A01.force, LG.f2.A02.force, LG.f2.A03.force, LG.f2.A04.force, LG.f2.A05.force, LG.f2.A06.force, LG.f2.A07.force, LG.f2.A08.force, LG.f2.A09.force, LG.f2.A10.force, LG.f2.C01.force, LG.f2.C02.force, LG.f2.C03.force, LG.f2.C04.force, LG.f2.C05.force, LG.f2.C06.force, LG.f2.C07.force, LG.f2.C08.force, LG.f2.C09.force)
draw.map(map.AC, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal = T) 

write.map(map.AC, "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/map.AC.map") 
```

########### R/qtl ######################## 
Because I didn't see much potential in using one map to build a reasonable map, I decide to use R/qtl to build the map 

# data import 
```{r}
library("qtl")

# import data 
map.F2 <- read.cross("mm", file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/F2_geno_for_one_map_final.txt", mapfile = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/map.AC.map")

# newmap <- est.map(map.F2, tol=1e-6, map.function="kosambi") 
# newmap doesn't look better at all 
```

# a bunch of checks in R/qtl 
```{r}
summary(map.F2)
plotMap(map.F2)
# plot.rf(map.F2)

totmar(map.F2)

# 1) segregation distortion 
test <- geno.table(map.F2)
seg_dist <- test[test$P.value < 0.01,]
seg_dist
dim(seg_dist) # 340 markers 

##### drop markers with segregation distortion 
# map.F2.after.drop <- drop.marker(map.F2, rownames(seg_dist)) # doesn't work... 
map.F2

# rffull.Asub <- pull.rf(map.F2)

# work with each chromosome seperately 
# plot.rf(map.F2, chr = "chrA02")
rfA02 <- pull.rf(map.F2, chr = 'chrA02')
chrA02 <- markernames(map.F2, chr = 'chrA02')
chrA02

# 2) compare individual's genotype similarity. Assumption: no two individuals with very similary genotypes
cg <- comparegeno(map.F2) 
dim(cg) 
hist(cg, breaks = 200,
     xlab="Proportion of identical genotypes")
rug(cg)
which(cg >0.9, arr.ind = TRUE) # no individuals with high genotype similarities

# No apparent problem, make a plot to see the data
plot.rf(map.F2, col.scheme = "redblue", alternate.chrid = T)

# looks like they are on the right chromosomes but wrong order within each chromosome

# estimate the genetic map
nm <- est.map(map.F2)
plot(nm, alternate.chrid = T) 
pull.map(cross_m, chr = c("A03","A05")) # longer length was plot than what was indicated in the data? 
pull.map(cross_m) 
```

# check double cross over 
```{r}
cross.drop.marker <- map.F2

for (chr in names(cross.drop.marker$geno)) { # for each chromosome in cross genotype data
  my.chr <- get(chr,cross.drop.marker$geno) # return the genotype data, including data & map
  print(paste(chr,"NA before",sum(is.na(my.chr$data)))) 
  if(ncol(my.chr$data) > 3) { 
    my.chr$data[,2:(ncol(my.chr$data)-1)] <- sapply(2:(ncol(my.chr$data)-1),function(i) {
      apply(my.chr$data[,(i-1):(i+1)],1,function(gt) {
        if (any(is.na(gt))) return(gt[2]) #technically should be looking at the next genotyped marker.
        if ( (length(unique(gt)) == 2) & (gt[1] == gt[3])) return(NA)
        if ( length(unique(gt))  == 3) return(NA)
        return(gt[2])
      })
    })
  }
  cross.drop.marker$geno <- within(cross.drop.marker$geno,assign(chr,my.chr))
  print(paste(chr,"NA after",sum(is.na(get(chr,cross.drop.marker$geno)$data))))
} 

write.csv(pull.geno(cross.drop.marker),file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/cross.drop.marker.csv")

newmap.drop.marker <- est.map(cross.drop.marker,verbose=T,error.prob=.01) # why going through 
# the last section of code doesn't assign new map 
plot.map(map.F2, alternate.chrid = T) # the old genetic map
plot.map(map.F2,newmap.drop.marker, alternate.chrid = T) # genetic map comparison

geno.image(cross.drop.marker, alternate.chrid = T) # grid with color pixels for different genotypes

save(newmap.drop.marker, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/newmap.drop.marker.Rdata")
```

#################### below is taking a different strategy ###################################
```{r}
# talking to Maria, she suggested to use madmapper to assign linkage group before order marker within each chromosome. 


```

















