---
title: "F2_map_construction"
author: "Ruijuan Li"
date: "4/18/2017"
output: html_document
---

# Goal
The goal of this analysis is to build genetic map using F2 genotyping information, F2 genotyping was done based on the SNPs we identified from its parents 

# install & load package 
```{r}
# install.packages("onemap")
# install.packages("tkrplot", type="source") 
library(tcltk)
library(tkrplot)
library(onemap)  
library(ggplot2)
library(reshape)
library(dplyr)
```

# import & foramt data (MAC version)
```{r}
F2_geno_data <- read.table("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/F2_Final_SNP_Calls", header = T)
head(F2_geno_data) 
dim(F2_geno_data) 
class(F2_geno_data)

rownames(F2_geno_data) <- paste(F2_geno_data$CHROM, F2_geno_data$POS, sep = "_")
F2_geno_data <- F2_geno_data[, -c(1:6)]
dim(F2_geno_data) # 18226   166 
colnames(F2_geno_data) <- gsub("([[:print:]]+)(\\.)([[:digit:]])", "\\3", colnames(F2_geno_data))
head(F2_geno_data)  

## sample 200 markers to try this method ... 
# test <- sample_n(F2_geno_data, size = 200, replace = F) 
# dim(test)
```

# import & foramt data (whitney version)
```{r}
F2_geno_data <- read.table("~/F2/data/F2_Final_SNP_Calls", header = T)
head(F2_geno_data) 
dim(F2_geno_data) 
class(F2_geno_data)

rownames(F2_geno_data) <- paste(F2_geno_data$CHROM, F2_geno_data$POS, sep = "_")
F2_geno_data <- F2_geno_data[, -c(1:6)]
dim(F2_geno_data) # 18226   166 
colnames(F2_geno_data) <- gsub("([[:print:]]+)(\\.)([[:digit:]])", "\\3", colnames(F2_geno_data))
head(F2_geno_data)  

```

# remove redundant SNPs

some SNPs should have exactly the same genotype across all individuals, they are redundant in terms of genetic map construction, so they should be removed. find those SNPs by doing correlation test. 

```{r}
# problem: 
# 1)  12 markers are not polymorphism across individuals 
# 2)  because there are missing data present in the matrix, when computing for correlation, only non missing data are accepted, so the pairwise correlation resulted in many markers that end up with same genotypes for the shared non-missing data, this cause error in cor() functon, I suppressed the "0 as standard deviation" warning message. finally found out there are 43422 pairs of SNPs that have this kind of problem 

# script "SNP_pairwise_correlatition.R" deal with problem 1 & 2 
# output file /Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/F2_SNP_correlation.Rdata is the correlation matrix for individuals in F2 population... 
# output file SNP.not.poly.Rdata has the 12 SNPs that are not polymorphism across F2 pops 

# 3) 579 pairs of SNPs have correlation of 1 
# script "identify_identical_SNP.R" identifys SNPs with pairwise correlation of 1 
# output dup.pair.Rdata (identical pairs); rem.Rdata (SNPs that should be removed) 

load("~/F2/output/F2_SNP_correlation.Rdata")
# code to remove duplicated SNPs 
length(which(F2_SNP_correlation >= 0.9 & lower.tri(F2_SNP_correlation))) # 579 SNPs with exactly the same genotype across F2 individuals

dup.cordinate <- which(F2_SNP_correlation ==1 & lower.tri(F2_SNP_correlation), arr.ind = T, useNames = F)
dup.cordinate.df <- as.data.frame(dup.cordinate)
sample.ID <- colnames(F2_SNP_correlation)
length(sample.ID) # 18214

dup.pair <- data.frame(matrix(nrow = nrow(dup.cordinate), ncol = 2))
for (i in 1:nrow(dup.cordinate)){
 dup.pair[i,1] <- sample.ID[dup.cordinate[i,1]]
 dup.pair[i,2] <- sample.ID[dup.cordinate[i,2]]
}

rem <- unique(dup.pair[,2])
### remove the redundant SNPs

load("~/F2/output/rem.Rdata")
load("~/F2/output/SNP.not.poly.Rdata")
# remove identical SNPs 
F2_geno_data_1 <- F2_geno_data[!(rownames(F2_geno_data) %in% rem),]  

# remove non polymorphism SNPs 
F2_geno_data_2 <- F2_geno_data_1[!(rownames(F2_geno_data_1) %in% SNP.not.poly),]

dim(F2_geno_data_2) # 5765   166
save(F2_geno_data_2, file="/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/F2_geno_data_2.Rdata") # this is non-redundant SNPs 
```

# format data for onemap 
```{r}
install.packages("tidyverse")
library("tidyverse")

load("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/F2_geno_data_2.Rdata")
dim(F2_geno_data_2) # 5765 166 

F2_geno_data_2_with_missing_rate <- mutate(F2_geno_data_2, 
       missing_rate = round(apply(F2_geno_data_2, 1, function(x) sum(is.na(x)))/94, 2))

hist(F2_geno_data_2_with_missing_rate$missing_rate)
max(F2_geno_data_2_with_missing_rate$missing_rate) 

dim(filter(F2_geno_data_2_with_missing_rate, missing_rate < 0.2))



# formating data for onemap 
F2_geno_data_2 <- as.matrix(F2_geno_data_2)
F2_geno_data_2[is.na(F2_geno_data_2)] <- "-"
F2_geno_data_2 <- gsub("0/0", "A", F2_geno_data_2) 
F2_geno_data_2 <- gsub("0/1", "H", F2_geno_data_2)
F2_geno_data_2 <- gsub("1/1", "B", F2_geno_data_2)
# F2_geno_data <- t(F2_geno_data)

F2_geno_data_2 <- as.data.frame(F2_geno_data_2)
dim(F2_geno_data_2) #  
write.table(F2_geno_data_2, file = "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/F2_geno_data.txt")
# change file format in linux 
# cat F2_geno_data.txt | sed 's/"//g' | awk '{first = $1; $1 = ""; print $0}' | sed 's/ //g' > tmp
# tail -18226 tmp > tmp.1 

write.table(rownames(F2_geno_data_2), file = "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/marker_info.txt" )
# cat marker_info.txt | awk '{print "*"$2}' | tail -18226 | sed 's/"//g' > marker_info_reform.txt

# paste marker_info_reform.txt tmp.1 | awk '{print $1 " " $2}' > F2_geno_for_one_map.txt 
# cat header_one_map_input F2_geno_for_one_map.txt > F2_geno_for_one_map_final.txt 


########### start to build linkage group ################################
data <- read.mapmaker(file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/F2_geno_for_one_map_final.txt")
# data
# save(data, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/data.Rdata") # the data saved here can be used later for map construction 

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/data.Rdata")
# estimate two-point rf
twopts.f2 <- rf.2pts(data, LOD = 6, max.rf = 0.25) ##### run this on whitney....  
?rf.2pts
twopts.f2

# assign marker to linkage groups
mark.all.f2 <- make.seq(twopts.f2, "all")
mark.all.f2
?make.seq 
LGs.f2 <- group(mark.all.f2, LOD = 6, max.rf = 0.25)
LGs.f2
# assign marker to linkage groups # this is not a good way, maybe should make.seq with a subset of markers, say, markers on specific chromosomes. then order marker within each chromosome. 

# so get the number of markers for each chromosome 
load("~/F2/output/F2_geno_data_2.Rdata")
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_data_2.Rdata")
dim(F2_geno_data_2) 
head(F2_geno_data)
F2_geno_data_2$Chr <- gsub("([[:print:]]+)(_)([[:digit:]]+)", "\\1", rownames(F2_geno_data_2))
F2_geno_data_2$Pos <- gsub("([[:print:]]+)(_)([[:digit:]]+)", "\\3", rownames(F2_geno_data_2)) 
F2_geno_data_2$index <- seq(1: nrow(F2_geno_data_2))
sum(F2_geno_data_2$Chr=="chrA01") # 403 
sum(F2_geno_data_2$Chr=="chrA02") # 248 404-651
sum(F2_geno_data_2$Chr=="chrA03") # 504 652-1155
sum(F2_geno_data_2$Chr=="chrA04") # 287 1156-1442
sum(F2_geno_data_2$Chr=="chrA05") # 282 1443-1724
sum(F2_geno_data_2$Chr=="chrA06") # 446 1725-2170
sum(F2_geno_data_2$Chr=="chrA07") # 429 2171-2599
sum(F2_geno_data_2$Chr=="chrA08") # 239 2600-2838
sum(F2_geno_data_2$Chr=="chrA09") # 554 2839-3392
sum(F2_geno_data_2$Chr=="chrA10") # 367 3393-3759
sum(F2_geno_data_2$Chr=="chrC01") # 227 3760-3986
sum(F2_geno_data_2$Chr=="chrC02") # 237 3987-4223
sum(F2_geno_data_2$Chr=="chrC03") # 444 4224-4667
sum(F2_geno_data_2$Chr=="chrC04") # 202 4668-4869
sum(F2_geno_data_2$Chr=="chrC05") # 174 4870-5043
sum(F2_geno_data_2$Chr=="chrC06") # 197 5044-5240
sum(F2_geno_data_2$Chr=="chrC07") # 209 5241-5449
sum(F2_geno_data_2$Chr=="chrC08") # 185 5450-5634
sum(F2_geno_data_2$Chr=="chrC09") # 131 5635-5765 

# figure out the index positon within each chromosome for all chromosomes 
F2_geno_data_info <- F2_geno_data_2[,c("Chr", "Pos", "index")]

out <- split( F2_geno_data_info , f = F2_geno_data_info$Chr)
length(out)

order <- list()

for (i in 1:length(out)){
  out[[i]]$Pos <- as.numeric(out[[i]]$Pos)
  order[[i]] <- out[[i]][order(out[[i]]$Pos),]$index 
}

order 
```

```{r} 
?make.seq
# LGs.f2 <- group(mark.all.f2, LOD = 6, max.rf = 0.25)
# LGs.f2
# save(LGs.f2, file="/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/LGs.f2.Rdata") 
``` 


# after getting good number of linkage group, work on each LG seperately... 
```{r}
#################### A01 ##############################
F2.data <- read.mapmaker(file="~/F2/output/F2_geno_for_one_map_final.txt")
load("~/F2/output/twopts.f2.04.21.Rdata")
load("~/F2/output/LG1.f2.ord.A01.Rdata")
# map_A01.R 

LG.f2.arbi.A01 <- make.seq(twopts.f2.04.24, order[[1]])

LG.f2.arbi.A01.map <- map(LG.f2.arbi.A01)
rf.graph.table(LG.f2.arbi.A01.map) 

draw.map(LG.f2.arbi.A01.map, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T) 


######## wait for next step analysis ############ 
LG1.f2.ord

# safe mode order (safe order should be recommended because we have enough marker here)
LG1.f2.safe <- make.seq(LG1.f2.ord, "safe")
LG1.f2.safe 


LG1.f2 <- make.seq(LGs.f2, 1) 
LG1.f2

# use touchdown
LG1.f2.ord <- order.seq(input.seq = LG1.f2, n.init = 5,
                        subset.search = "twopt",
                        twopt.alg = "rcd", THRES = 6,
                        draw.try = T, wait = 1,
                        touchdown = T)

LG1.f2.ord

# safe mode order (safe order should be recommended because we have enough marker here)
LG1.f2.safe <- make.seq(LG1.f2.ord, "safe")
LG1.f2.safe
# all order
LG1.f2.final <- make.seq(LG1.f2.ord, "force") 
LG1.f2.final

# ripple
ripple.seq(LG1.f2.final, ws = 5, LOD = 6) 

################# A02 ##################################
# map_A02.R 


# safe mode order
LG2.f2.safe <- make.seq(LG2.f2.ord, "safe")
LG2.f2.safe
# all order
LG2.f2.final <- make.seq(LG2.f2.ord, "force")
LG2.f2.final

# ripple
ripple.seq(LG2.f2.final, ws = 5, LOD = 6) 

######################## A03 ########################

# map_A03.R 


# safe mode order
LG3.f2.safe <- make.seq(LG3.f2.ord, "safe")
LG3.f2.safe
# all order
LG3.f2.final <- make.seq(LG3.f2.ord, "force")
LG3.f2.final

# ripple
ripple.seq(LG3.f2.final, ws = 3, LOD = 6) 

######################## A04 ######################
########### use Kosambi map function ################
# map_A01.R 

LG4.f2.ord

# safe mode order
LG4.f2.safe <- make.seq(LG4.f2.ord, "safe")
LG4.f2.safe
# all order
LG4.f2.final <- make.seq(LG4.f2.ord, "force")
LG4.f2.final

######################################
#Markers               Position           
#
#60 FPBPN1720              0.00           
#59 FPBPN1183             21.79           
#58 FPBPN1466             23.96           
#22 FPBRN0403             99.48           
#23 FPBPN0589            123.34           
#24 FPBPN0595            124.96           
#25 FPBPN0006            132.46           
#26 FPBPN0294            138.91           
#27 FPBPN0205            151.91           
#28 FPBRN0426            171.45           
#29 FPBPN1534            177.44           
#30 FPBPN1503            178.58           
#31 FPBPN1712            187.25           
#32 FPBRN0434            192.02           
#33 FPBRN0876            207.88           
#40 FPBPN1533            270.61           
#41 FPBPN0296            272.67           
#42 FPBPN0399            284.70           
#43 FPBPN1863            304.96           
#44 FPBRN0153            311.49           
#45 FPBPN0005            324.22           
#46 FPBRN0154            325.08           
#47 FPBPN0291            330.16           
#48 FPBPN0283            331.39           
#53 FPBPN1083            335.94           
#49 FPBPN0790            339.05           
#50 FPBPN1637            341.32           
#51 FPBPN1563            345.21           
#52 FPBPN1651            346.72           
#54 FPBPN1982            357.41           
#55 FPBPN1702            360.15           
#56 FPBPN0045            365.75           
#34 FPBPN1414            380.72           
#35 FPBPN0559            390.62           
#36 FPBPN1369            395.71           
#37 FPBPN0197            400.49           
#38 FPBPN0164            409.48           
#39 FPBPN1107            419.04           
#57 FPBPN0082            440.07           
#
#39 markers            log-likelihood: -2303.75 
###############################################

# ripple
ripple.seq(LG4.f2.final, ws = 5, LOD = 6)

################### use haldane map function ##########
set.map.fun(type = "haldane")

LG4h.f2 <- make.seq(LGs.f2, 4)
LG4h.f2

# use touchdown
LG4h.f2.ord <- order.seq(input.seq = LG4h.f2, n.init = 5,
                        subset.search = "twopt",
                        twopt.alg = "rcd", THRES = 6,
                        draw.try = T, wait = 1,
                        touchdown = T)

LG4h.f2.ord

# safe mode order
LG4h.f2.safe <- make.seq(LG4h.f2.ord, "safe")
LG4h.f2.safe
# all order
LG4h.f2.final <- make.seq(LG4h.f2.ord, "force")
LG4h.f2.final
### almost no difference between haldane & kosambi fxn, except that 
# kosambi gives shorter map distance... the marker order is a little 
# different (not much) from FnP's order got from mapdisto kosambi funcion. 
######################################################

###################### A05 ########################

# map_A05.R 

# safe mode order
LG5.f2.safe <- make.seq(LG5.f2.ord, "safe")
LG5.f2.safe
# all order
LG5.f2.final <- make.seq(LG5.f2.ord, "force")
LG5.f2.final

################# A06 ##############################
# map_A06.R 

# safe mode order
LG6.f2.safe <- make.seq(LG6.f2.ord, "safe")
LG6.f2.safe
# all order
LG6.f2.final <- make.seq(LG6.f2.ord, "force")
LG6.f2.final

################# A07 ##############################
# map_A07.R  


# safe mode order
LG7.f2.safe <- make.seq(LG7.f2.ord, "safe")
LG7.f2.safe
# all order
LG7.f2.final <- make.seq(LG7.f2.ord, "force")
LG7.f2.final

################# A08 ##############################
# map_A08.R 


# safe mode order
LG8.f2.safe <- make.seq(LG8.f2.ord, "safe")
LG8.f2.safe
# all order
LG8.f2.final <- make.seq(LG8.f2.ord, "force")
LG8.f2.final

################# A09 ##############################
# map_A09.R 

# safe mode order
LG9.f2.safe <- make.seq(LG9.f2.ord, "safe")
LG9.f2.safe
# all order
LG9.f2.final <- make.seq(LG9.f2.ord, "force")
LG9.f2.final

################# A10 ##############################
# map_A10.R 

# safe mode order
LG10.f2.safe <- make.seq(LG10.f2.ord, "safe")
LG10.f2.safe
# all order
LG10.f2.final <- make.seq(LG10.f2.ord, "force")
LG10.f2.final

################# C01 ##############################
# map_C01.R 
F2.data <- read.mapmaker(file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_for_one_map_final.txt")

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/twopts.f2.04.21.Rdata")

###### don't consider mapping postion at all
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG1.f2.ord.C01.Rdata")
LG1.f2.ord.C01

# set map function 
set.map.fun(type="kosambi")

# safe mode order (only marker with unique positions)
LG1.f2.safe.C01 <- make.seq(LG1.f2.ord.C01, "safe")
LG1.f2.safe.C01
# all order (all markers)
LG1.f2.final.C01 <- make.seq(LG1.f2.ord.C01, "force")
LG1.f2.final.C01

# print the recombination fraction mix 
rf.graph.table(LG1.f2.safe.C01) 
par(mfrow=c(1,1))
draw.map(LG1.f2.final.C01, names= F, grid=F, cex.mrk=0.5)
draw.map(LG1.f2.safe.C01, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

# map estimation for arbitrary order 
LG1.f2.arbi.C01 <- make.seq(twopts.f2.04.24, order[[11]])
LG1.f2.arbi.C01.map <- map(LG1.f2.arbi.C01)
rf.graph.table(LG1.f2.arbi.C01.map) 

draw.map(LG1.f2.arbi.C01.map, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T) 

################# LG12/A10 ##############################
LG12.f2 <- make.seq(LGs.f2, 12)
LG12.f2

# use touchdown
LG12.f2.ord <- order.seq(input.seq = LG12.f2, n.init = 5,
                        subset.search = "twopt",
                        twopt.alg = "rcd", THRES = 6,
                        draw.try = T, wait = 1,
                        touchdown = T)

LG12.f2.ord

# safe mode order
LG12.f2.safe <- make.seq(LG12.f2.ord, "safe")
LG12.f2.safe
# all order
LG12.f2.final <- make.seq(LG12.f2.ord, "force")
LG12.f2.final

################# LG13/C01 ##############################
LG13.f2 <- make.seq(LGs.f2, 13)
LG13.f2

# use touchdown
LG13.f2.ord <- order.seq(input.seq = LG13.f2, n.init = 5,
                        subset.search = "twopt",
                        twopt.alg = "rcd", THRES = 6,
                        draw.try = T, wait = 1,
                        touchdown = T)

LG13.f2.ord

# safe mode order
LG13.f2.safe <- make.seq(LG13.f2.ord, "safe")
LG13.f2.safe
# all order
LG13.f2.final <- make.seq(LG13.f2.ord, "force")
LG13.f2.final

################# LG14/C01-1 ##############################
LG14.f2 <- make.seq(LGs.f2, 14)
LG14.f2

# use touchdown
LG14.f2.ord <- order.seq(input.seq = LG14.f2, n.init = 5,
                        subset.search = "twopt",
                        twopt.alg = "rcd", THRES = 6,
                        draw.try = T, wait = 1,
                        touchdown = T)

LG14.f2.ord

# safe mode order
LG14.f2.safe <- make.seq(LG14.f2.ord, "safe")
LG14.f2.safe
# all order
LG14.f2.final <- make.seq(LG14.f2.ord, "force")
LG14.f2.final

################# LG15/C02 ##############################
LG15.f2 <- make.seq(LGs.f2, 15)
LG15.f2

# use touchdown
LG15.f2.ord <- order.seq(input.seq = LG15.f2, n.init = 5,
                        subset.search = "twopt",
                        twopt.alg = "rcd", THRES = 6,
                        draw.try = T, wait = 1,
                        touchdown = T)

LG15.f2.ord

# safe mode order
LG15.f2.safe <- make.seq(LG15.f2.ord, "safe")
LG15.f2.safe
# all order
LG15.f2.final <- make.seq(LG15.f2.ord, "force")
LG15.f2.final

################# LG16/C02-2 ##############################
LG16.f2 <- make.seq(LGs.f2, 16)
LG16.f2

# use touchdown
LG16.f2.ord <- order.seq(input.seq = LG16.f2, n.init = 5,
                         subset.search = "twopt",
                         twopt.alg = "rcd", THRES = 6,
                         draw.try = T, wait = 1,
                         touchdown = T)

LG16.f2.ord

# safe mode order
LG16.f2.safe <- make.seq(LG16.f2.ord, "safe")
LG16.f2.safe
# all order
LG16.f2.final <- make.seq(LG16.f2.ord, "force")
LG16.f2.final

################# LG17/C02-3 ##############################
LG17.f2 <- make.seq(LGs.f2, 17)
LG17.f2

# use touchdown
LG17.f2.ord <- order.seq(input.seq = LG17.f2, n.init = 2,
                         subset.search = "twopt",
                         twopt.alg = "rcd", THRES = 6,
                         draw.try = T, wait = 1,
                         touchdown = T)

LG17.f2.ord

# safe mode order
LG17.f2.safe <- make.seq(LG17.f2.ord, "safe")
LG17.f2.safe
# all order
LG17.f2.final <- make.seq(LG17.f2.ord, "force")
LG17.f2.final

################# LG18/C003 ##############################
LG18.f2 <- make.seq(LGs.f2, 18)
LG18.f2

# use touchdown
LG18.f2.ord <- order.seq(input.seq = LG18.f2, n.init = 5,
                         subset.search = "twopt",
                         twopt.alg = "rcd", THRES = 6,
                         draw.try = T, wait = 1,
                         touchdown = T)

LG18.f2.ord

# safe mode order
LG18.f2.safe <- make.seq(LG18.f2.ord, "safe")
LG18.f2.safe
# all order
LG18.f2.final <- make.seq(LG18.f2.ord, "force")
LG18.f2.final

################# LG19/C04 ##############################
LG19.f2 <- make.seq(LGs.f2, 19)
LG19.f2

# use touchdown
LG19.f2.ord <- order.seq(input.seq = LG19.f2, n.init = 5,
                         subset.search = "twopt",
                         twopt.alg = "rcd", THRES = 6,
                         draw.try = T, wait = 1,
                         touchdown = T)

LG19.f2.ord

# safe mode order
LG19.f2.safe <- make.seq(LG19.f2.ord, "safe")
LG19.f2.safe
# all order
LG19.f2.final <- make.seq(LG19.f2.ord, "force")
LG19.f2.final

################# LG20/C05 ##############################
LG20.f2 <- make.seq(LGs.f2, 20)
LG20.f2

# use touchdown
LG20.f2.ord <- order.seq(input.seq = LG20.f2, n.init = 5,
                         subset.search = "twopt",
                         twopt.alg = "rcd", THRES = 6,
                         draw.try = T, wait = 1,
                         touchdown = T)

LG20.f2.ord

# safe mode order
LG20.f2.safe <- make.seq(LG20.f2.ord, "safe")
LG20.f2.safe
# all order
LG20.f2.final <- make.seq(LG20.f2.ord, "force")
LG20.f2.final

################# LG21/C06 ##############################
LG21.f2 <- make.seq(LGs.f2, 21)
LG21.f2

# use touchdown
LG21.f2.ord <- order.seq(input.seq = LG21.f2, n.init = 5,
                         subset.search = "twopt",
                         twopt.alg = "rcd", THRES = 6,
                         draw.try = T, wait = 1,
                         touchdown = T)

LG21.f2.ord

# safe mode order
LG21.f2.safe <- make.seq(LG21.f2.ord, "safe")
LG21.f2.safe
# all order
LG21.f2.final <- make.seq(LG21.f2.ord, "force")
LG21.f2.final

################# LG22/C07 ##############################
LG22.f2 <- make.seq(LGs.f2, 22)
LG22.f2

# use touchdown
LG22.f2.ord <- order.seq(input.seq = LG22.f2, n.init = 5,
                         subset.search = "twopt",
                         twopt.alg = "rcd", THRES = 6,
                         draw.try = T, wait = 1,
                         touchdown = T)

LG22.f2.ord

# safe mode order
LG22.f2.safe <- make.seq(LG22.f2.ord, "safe")
LG22.f2.safe
# all order
LG22.f2.final <- make.seq(LG22.f2.ord, "force")
LG22.f2.final

################# LG23/C08 ##############################
LG23.f2 <- make.seq(LGs.f2, 23)
LG23.f2

# use touchdown
LG23.f2.ord <- order.seq(input.seq = LG23.f2, n.init = 5,
                         subset.search = "twopt",
                         twopt.alg = "rcd", THRES = 6,
                         draw.try = T, wait = 1,
                         touchdown = T)

LG23.f2.ord

# safe mode order
LG23.f2.safe <- make.seq(LG23.f2.ord, "safe")
LG23.f2.safe
# all order
LG23.f2.final <- make.seq(LG23.f2.ord, "force")
LG23.f2.final

################# LG24/C09 ##############################
LG24.f2 <- make.seq(LGs.f2, 24)
LG24.f2

# use touchdown
LG24.f2.ord <- order.seq(input.seq = LG24.f2, n.init = 5,
                         subset.search = "twopt",
                         twopt.alg = "rcd", THRES = 6,
                         draw.try = T, wait = 1,
                         touchdown = T)

LG24.f2.ord

# safe mode order
LG24.f2.safe <- make.seq(LG24.f2.ord, "safe")
LG24.f2.safe
# all order
LG24.f2.final <- make.seq(LG24.f2.ord, "force")
LG24.f2.final

################# LG25/C09-1 ##############################
LG25.f2 <- make.seq(LGs.f2, 25)
LG25.f2

# use touchdown
LG25.f2.ord <- order.seq(input.seq = LG25.f2, n.init = 5,
                         subset.search = "twopt",
                         twopt.alg = "rcd", THRES = 6,
                         draw.try = T, wait = 1,
                         touchdown = T)

LG25.f2.ord

# safe mode order
LG25.f2.safe <- make.seq(LG25.f2.ord, "safe")
LG25.f2.safe
# all order
LG25.f2.final <- make.seq(LG25.f2.ord, "force")
LG25.f2.final




###################################################
########### drop marker with double crossover ###########
data <- read.mapmaker(file = "F2_415_dropp_geno_Ruijuan.txt")
data

# estimate two-point rf
twopts.f2 <- rf.2pts(data, LOD = 3, max.rf = 0.5)
?rf.2pts

# assign marker to linkage groups
mark.all.f2 <- make.seq(twopts.f2, "all")
?make.seq
LGs.f2 <- group(mark.all.f2, LOD = 3, max.rf = 0.5)
LGs.f2

#################### LG1 ##############################
LG1.f2 <- make.seq(LGs.f2, 1)
LG1.f2

# use touchdown
LG1.f2.ord <- order.seq(input.seq = LG1.f2, n.init = 5,
                        subset.search = "twopt",
                        twopt.alg = "rcd", THRES = 6,
                        draw.try = T, wait = 1,
                        touchdown = T)

LG1.f2.ord

# safe mode order
LG1.f2.safe <- make.seq(LG1.f2.ord, "safe")
LG1.f2.safe
# all order
LG1.f2.final <- make.seq(LG1.f2.ord, "force")
LG1.f2.final

# ripple
ripple.seq(LG1.f2.final, ws = 5, LOD = 3) 
``` 
