---
title: "F2_map_construction"
author: "Ruijuan Li"
date: "4/18/2017"
output: html_document
---

# Goal
The goal of this analysis is to build genetic map using F2 genotyping information, F2 genotyping was done based on the SNPs we identified from its parents 

# install & load package 
```{r}
# install.packages("onemap")
# install.packages("tkrplot", type="source") 
library(tcltk)
library(tkrplot)
library(onemap)  
library(ggplot2)
library(reshape) 
```

# import & format data (MAC version)
```{r}
F2_geno_data <- read.table("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/F2_Final_SNP_Calls", header = T)
head(F2_geno_data) 
dim(F2_geno_data) # 18226   172 
class(F2_geno_data)

rownames(F2_geno_data) <- paste(F2_geno_data$CHROM, F2_geno_data$POS, sep = "_")
F2_geno_data <- F2_geno_data[, -c(1:6)]
dim(F2_geno_data) # 18226   166 
colnames(F2_geno_data) <- gsub("([[:print:]]+)(\\.)([[:digit:]])", "\\3", colnames(F2_geno_data))
head(F2_geno_data)  
```

# check missing rate & remove data with high missing rate above 0.85 
```{r}
F2_geno_data_with_missing_rate <- mutate(F2_geno_data, 
       missing_rate = round(apply(F2_geno_data, 1, function(x) sum(is.na(x)))/94, 2))

rownames(F2_geno_data_with_missing_rate) <- rownames(F2_geno_data)
hist(F2_geno_data_with_missing_rate$missing_rate)
max(F2_geno_data_with_missing_rate$missing_rate) # 0.83, the highest missing rate is 0.83 

# remove markers with missing rate greater than 0.85 
F2_geno_data_with_missing_rate_filtered <- F2_geno_data_with_missing_rate[F2_geno_data_with_missing_rate$missing_rate < 0.15,]
rownames(F2_geno_data_with_missing_rate_filtered)

dim(F2_geno_data_with_missing_rate_filtered) # 3077 167
hist(F2_geno_data_with_missing_rate_filtered$missing_rate)

F2_geno_data_with_missing_rate_filtered.final <- subset(F2_geno_data_with_missing_rate_filtered, select=-missing_rate)
rownames(F2_geno_data_with_missing_rate_filtered.final)

dim(F2_geno_data_with_missing_rate_filtered.final) # 3077 166 
rownames(F2_geno_data_with_missing_rate_filtered.final) 
```


# calculate pairwise corrlelation and remove SNPs with corrlelation greater or equal to 0.9
some SNPs should have exactly the same genotype across all individuals, they are redundant in terms of genetic map construction, so they should be removed. find those SNPs by doing correlation test. 
```{r}
F2_geno_data_t <- as.data.frame(t(F2_geno_data_with_missing_rate_filtered.final))
F2_geno_data_t.numeric <- data.matrix(F2_geno_data_t)

# delete markers with all same genotypes across individuals 
test <- apply(F2_geno_data_t.numeric, 2, function(x) length(unique(x[!is.na(x)])))
filter.polymorphsm <- test != 1 

F2_geno_data_t.numeric.2 <- F2_geno_data_t.numeric[,filter.polymorphsm]
dim(F2_geno_data_t.numeric.2) # 166 3075  

# also output save markers with same genotype across individuals
polymorphism <- test == 1
SNP.not.poly <- colnames(F2_geno_data_t.numeric)[polymorphism]
length(SNP.not.poly) # two markers are not polymorphism across individuals 
save(SNP.not.poly, file="SNP.not.poly.Rdata")

# correlation test
options(warn=-1) # suppress warning message
F2_SNP_correlation <- cor(F2_geno_data_t.numeric.2, use = "pairwise.complete.obs")
save(F2_SNP_correlation, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_SNP_correlation.Rdata")
options(warn=0) # unsuppress warning message
dim(F2_SNP_correlation) # 3075 3075 

# calcualte number of SNPs with correlation of 1
nrow(which(F2_SNP_correlation == 1 & lower.tri(F2_SNP_correlation), arr.ind = T, useNames = T)) # 150 pairs of SNPs have identical genotype data 

# find SNPs with correlation of 1 and remove them 
dup.cordinate <- which(F2_SNP_correlation == 1 & lower.tri(F2_SNP_correlation), arr.ind = T, useNames = F)
length(dup.cordinate) # 300 
dup.cordinate.df <- as.data.frame(dup.cordinate)
sample.ID <- colnames(F2_SNP_correlation)

# extract duplicate pair information based on their coordicate
dup.pair <- data.frame(matrix(nrow = nrow(dup.cordinate), ncol = 2))
for (i in 1:nrow(dup.cordinate)){
 dup.pair[i,1] <- sample.ID[dup.cordinate[i,1]]
 dup.pair[i,2] <- sample.ID[dup.cordinate[i,2]]
}

length(dup.pair[,2]) # 150 
rem <- unique(dup.pair[,2])
length(rem) # 143 markers are removed due to pairwise correlation of 1 

# save data
save(dup.pair, file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/dup.pair.Rdata")
save(rem, file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/rem.Rdata") 

### 
F2_geno_data_1 <- F2_geno_data_with_missing_rate_filtered.final[!(rownames(F2_geno_data_with_missing_rate_filtered.final) %in% rem),]  

# remove non polymorphism SNPs 
F2_geno_data_2 <- F2_geno_data_1[!(rownames(F2_geno_data_1) %in% SNP.not.poly),]

dim(F2_geno_data_2) # 2932  166, end up with 2932 SNPs for map construction 

save(F2_geno_data_2, file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_data_2.Rdata") # this is non-redundant SNPs with missing data less than 0.85   
```

# format data for onemap  
```{r}
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_data_2.Rdata")
dim(F2_geno_data_2) # 2932 166 

# formating data for onemap 
F2_geno_data_2 <- as.matrix(F2_geno_data_2)
F2_geno_data_2[is.na(F2_geno_data_2)] <- "-"
F2_geno_data_2 <- gsub("0/0", "A", F2_geno_data_2) 
F2_geno_data_2 <- gsub("0/1", "H", F2_geno_data_2)
F2_geno_data_2 <- gsub("1/1", "B", F2_geno_data_2)

F2_geno_data_2 <- as.data.frame(F2_geno_data_2)
dim(F2_geno_data_2) # 2932 166  
write.table(F2_geno_data_2, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_data.txt")
# change file format in linux 
# cat F2_geno_data.txt | sed 's/"//g' | awk '{first = $1; $1 = ""; print $0}' | sed 's/ //g' > tmp
# tail -2932 tmp > tmp.1 

write.table(rownames(F2_geno_data_2), file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/marker_info.txt" )
# cat marker_info.txt | awk '{print "*"$2}' | tail -2932 | sed 's/"//g' > marker_info_reform.txt

# paste marker_info_reform.txt tmp.1 | awk '{print $1 " " $2}' > F2_geno_for_one_map.txt 
# cat header_one_map_input F2_geno_for_one_map.txt > F2_geno_for_one_map_final.txt 
```

# calculate two point rf & assign to linkage groups (on whitney)
```{r}
# F2_map_construction_LOD3_rf0.5_missing_0.15.R 
# scp rf.Rdata to my MAC
# after checking the group result, only one linkage group was formed... 
```

# assign markers to different linkage groups based on their chromsome assignment 
```{r}
# get the number of markers for each chromosome 
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_data_2.Rdata")
dim(F2_geno_data_2) # 2932 166
head(F2_geno_data_2) 

F2_geno_data_2$Chr <- gsub("([[:print:]]+)(_)([[:digit:]]+)", "\\1", rownames(F2_geno_data_2))
F2_geno_data_2$Pos <- gsub("([[:print:]]+)(_)([[:digit:]]+)", "\\3", rownames(F2_geno_data_2)) 
F2_geno_data_2$index <- seq(1: nrow(F2_geno_data_2))

head(F2_geno_data_2[F2_geno_data_2$Chr=="chrA04",],1)
tail(F2_geno_data_2[F2_geno_data_2$Chr=="chrC09",],1)

sum(F2_geno_data_2$Chr=="chrA01") # 1-192  
sum(F2_geno_data_2$Chr=="chrA02") # 78 193-270
sum(F2_geno_data_2$Chr=="chrA03") # 267 271-537
sum(F2_geno_data_2$Chr=="chrA04") # 136 538-673
sum(F2_geno_data_2$Chr=="chrA05") # 129 674-802
sum(F2_geno_data_2$Chr=="chrA06") # 379 803-1181
sum(F2_geno_data_2$Chr=="chrA07") # 242 1182-1423
sum(F2_geno_data_2$Chr=="chrA08") # 151 1424-1574
sum(F2_geno_data_2$Chr=="chrA09") # 307 1575-1881
sum(F2_geno_data_2$Chr=="chrA10") # 253 1882-2134
sum(F2_geno_data_2$Chr=="chrC01") # 106 2135-2240
sum(F2_geno_data_2$Chr=="chrC02") # 59 2241-2299
sum(F2_geno_data_2$Chr=="chrC03") # 158 2300-2457
sum(F2_geno_data_2$Chr=="chrC04") # 143 2458-2600
sum(F2_geno_data_2$Chr=="chrC05") # 43 2601-2643
sum(F2_geno_data_2$Chr=="chrC06") # 85 2644-2728
sum(F2_geno_data_2$Chr=="chrC07") # 100 2729-2828
sum(F2_geno_data_2$Chr=="chrC08") # 60 2829-2888
sum(F2_geno_data_2$Chr=="chrC09") # 44 2889-2932  

# https://github.com/leejimmy93/KIAT/map_A01.R .... mapC09.R 
```

# get marker index within each chromosome 
```{r}
###################### below can be useful when I only need marker order within each chromosome ## 
# figure out the index positon within each chromosome for all chromosomes 
F2_geno_data_info <- F2_geno_data_2[,c("Chr", "Pos", "index")]
dim(F2_geno_data_info) # 2932 3 
head(F2_geno_data_info, 1000)

out <- split( F2_geno_data_info, f = F2_geno_data_info$Chr)
length(out) # 19 
out

order <- list()

for (i in 1:length(out)){
  out[[i]]$Pos <- as.numeric(out[[i]]$Pos)
  order[[i]] <- out[[i]][order(out[[i]]$Pos),]$index 
}

order 
save(order, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/order.Rdata")
```

# after getting good number of linkage group, work on each LG seperately... 
A01 
```{r}
#################### A01 ##############################
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/twopts.f2.05.12.Rdata")
LG.f2.arbi.A01 <- make.seq(twopts.f2.05.12, order[[1]])

LG.f2.arbi.A01.map <- map(LG.f2.arbi.A01)
rf.graph.table(LG.f2.arbi.A01.map) 

draw.map(LG.f2.arbi.A01.map, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T) 


######## wait for next step analysis ############ 
LG1.f2.ord

# safe mode order (safe order should be recommended because we have enough marker here)
LG1.f2.safe <- make.seq(LG1.f2.ord, "safe")
LG1.f2.safe 


LG1.f2 <- make.seq(LGs.f2, 1) 
LG1.f2

# use touchdown
LG1.f2.ord <- order.seq(input.seq = LG1.f2, n.init = 5,
                        subset.search = "twopt",
                        twopt.alg = "rcd", THRES = 6,
                        draw.try = T, wait = 1,
                        touchdown = T)

LG1.f2.ord

# safe mode order (safe order should be recommended because we have enough marker here)
LG1.f2.safe <- make.seq(LG1.f2.ord, "safe")
LG1.f2.safe
# all order
LG1.f2.final <- make.seq(LG1.f2.ord, "force") 
LG1.f2.final

# ripple
ripple.seq(LG1.f2.final, ws = 5, LOD = 6) 

#########
###### don't consider mapping postion at all
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/LG.f2.ord.A01.Rdata")
LG.f2.ord.A01

# set map function 
set.map.fun(type="kosambi")

# safe mode order (only marker with unique positions)
LG.f2.safe.A01 <- make.seq(LG.f2.ord.A01, "safe")
LG.f2.safe.A01
# all order (all markers)
LG.f2.final.A01 <- make.seq(LG.f2.ord.A01, "force")
LG.f2.final.A01

# print the recombination fraction mix 
rf.graph.table(LG.f2.safe.A01) 
par(mfrow=c(1,1))
draw.map(LG.f2.safe.A01, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

# map estimation for arbitrary order 
LG.f2.arbi.A01 <- make.seq(twopts.f2.LOD3_rf0.5, order[[1]])
LG.f2.arbi.A01.map <- map(LG.f2.arbi.A01)
rf.graph.table(LG.f2.arbi.A01.map) 

draw.map(LG.f2.arbi.A01.map, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)  

####### new stuff
F2.data <- read.mapmaker(file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_for_one_map_final.txt")
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/twopts.LOD3_rf0.5.Rdata")

mark.f2.A01 <- make.seq(twopts.f2.LOD3_rf0.5, 1:192)

LG.f2.ord.A01 <- order.seq(input.seq = mark.f2.A01, n.init = 5,
                        subset.search = "twopt",
                        twopt.alg = "rcd", THRES = 6,
                        draw.try = T, wait = 1,
                        touchdown = T)


```

# A02
```{r}
################# A02 ##################################
######## import data #######
# marker data 
F2.data <- read.mapmaker(file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_for_one_map_final.txt") 
F2.data <- read.mapmaker(file="~/F2/output/missing_rate_0.15/F2_geno_for_one_map_final.txt")
F2.data 
# This is an object of class 'f2.onemap'
#     No. individuals:     166 
#     No. markers:         2932 
#     Percent genotyped:   96 
# 
#     Number of markers per type:
#        AA : AB : BB -->  2932
# 
# This data contains no phenotypic information

# two-point rf data  
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/twopts.f2.05.12.Rdata")
load("~/F2/output/missing_rate_0.15/twopts.f2.05.12.Rdata")
########## order maker within chromosome #############
###### don't consider mapping postion at all
# https://github.com/leejimmy93/KIAT/blob/master/map_construction_missing_0.15/map_A02_2.R_final 
# mark.f2.A02 <- make.seq(twopts.f2.05.12, 193:270)
# LG.f2.ord.A02 <- order.seq(input.seq = mark.f2.A02, n.init = 5,
#                         subset.search = "twopt",
#                         twopt.alg = "rcd", THRES = 6, 
#                         draw.try = T, wait = 1,
#                         touchdown = T)

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.A02.Rdata")
load("~/F2/output/missing_rate_0.15/LG.f2.ord.A02.Rdata")
LG.f2.A02 <- make.seq(LG.f2.ord.A02) 
LG.f2.A02
rf.graph.table(LG.f2.A02) 
par(mfrow=c(1,1)) 
draw.map(LG.f2.A02, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T) 

# all order
LG.f2.A02.final <- make.seq(LG.f2.ord.A02, "force")
LG.f2.A02.final
rf.graph.table(LG.f2.A02.final) 
draw.map(LG.f2.A02.final, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T) 

### remove markers with double crossover 
LG.f2.A02.final$seq.num







# ripple
ripple.seq(LG.f2.A02.final, ws = 5, LOD = 6) 

# set map function 
set.map.fun(type="kosambi")

# map estimation for arbitrary order 
LG.f2.arbi.A02 <- make.seq(twopts.f2.05.12, order[[2]])
LG.f2.arbi.A02 
LG.f2.arbi.A02.map <- map(LG.f2.arbi.A02)  
rf.graph.table(LG.f2.arbi.A02.map) 

draw.map(LG.f2.arbi.A02.map, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)   
```

```{r}
######################## A03 ########################

# map_A03.R 


# safe mode order
LG3.f2.safe <- make.seq(LG3.f2.ord, "safe")
LG3.f2.safe
# all order
LG3.f2.final <- make.seq(LG3.f2.ord, "force")
LG3.f2.final

# ripple
ripple.seq(LG3.f2.final, ws = 3, LOD = 6) 

######################## A04 ######################
########### use Kosambi map function ################
# map_A01.R 

LG4.f2.ord

# safe mode order
LG4.f2.safe <- make.seq(LG4.f2.ord, "safe")
LG4.f2.safe
# all order
LG4.f2.final <- make.seq(LG4.f2.ord, "force")
LG4.f2.final 

# ripple
ripple.seq(LG4.f2.final, ws = 5, LOD = 6)

################### use haldane map function ##########
set.map.fun(type = "haldane")

LG4h.f2 <- make.seq(LGs.f2, 4)
LG4h.f2

# use touchdown
LG4h.f2.ord <- order.seq(input.seq = LG4h.f2, n.init = 5,
                        subset.search = "twopt",
                        twopt.alg = "rcd", THRES = 6,
                        draw.try = T, wait = 1,
                        touchdown = T)

LG4h.f2.ord

# safe mode order
LG4h.f2.safe <- make.seq(LG4h.f2.ord, "safe")
LG4h.f2.safe
# all order
LG4h.f2.final <- make.seq(LG4h.f2.ord, "force")
LG4h.f2.final
### almost no difference between haldane & kosambi fxn, except that 
# kosambi gives shorter map distance... the marker order is a little 
# different (not much) from FnP's order got from mapdisto kosambi funcion. 
######################################################

###################### A05 ########################

# map_A05.R 

# safe mode order
LG5.f2.safe <- make.seq(LG5.f2.ord, "safe")
LG5.f2.safe
# all order
LG5.f2.final <- make.seq(LG5.f2.ord, "force")
LG5.f2.final

################# A06 ##############################
# map_A06.R 

# safe mode order
LG6.f2.safe <- make.seq(LG6.f2.ord, "safe")
LG6.f2.safe
# all order
LG6.f2.final <- make.seq(LG6.f2.ord, "force")
LG6.f2.final

################# A07 ##############################
# map_A07.R  


# safe mode order
LG7.f2.safe <- make.seq(LG7.f2.ord, "safe")
LG7.f2.safe
# all order
LG7.f2.final <- make.seq(LG7.f2.ord, "force")
LG7.f2.final

################# A08 ##############################
# map_A08.R 


# safe mode order
LG8.f2.safe <- make.seq(LG8.f2.ord, "safe")
LG8.f2.safe
# all order
LG8.f2.final <- make.seq(LG8.f2.ord, "force")
LG8.f2.final

################# A09 ##############################
# map_A09.R 

# safe mode order
LG9.f2.safe <- make.seq(LG9.f2.ord, "safe")
LG9.f2.safe
# all order
LG9.f2.final <- make.seq(LG9.f2.ord, "force")
LG9.f2.final

################# A10 ##############################
# map_A10.R 

# safe mode order
LG10.f2.safe <- make.seq(LG10.f2.ord, "safe")
LG10.f2.safe
# all order
LG10.f2.final <- make.seq(LG10.f2.ord, "force")
LG10.f2.final
```

# C01 
```{r}
################# C01 ##############################

######## import data #######
# marker data 
F2.data <- read.mapmaker(file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_for_one_map_final.txt") 
F2.data 
# This is an object of class 'f2.onemap'
#     No. individuals:     166 
#     No. markers:         2932 
#     Percent genotyped:   96 
# 
#     Number of markers per type:
#        AA : AB : BB -->  2932
# 
# This data contains no phenotypic information

# two-point rf data  
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/twopts.f2.05.12.Rdata")

########## order maker within chromosome #############
###### don't consider mapping postion at all
# https://github.com/leejimmy93/KIAT/blob/master/map_construction_missing_0.15/map_C01_2.R_final 
# mark.f2.C01 <- make.seq(twopts.f2.05.12, 2135:2240)
# LG.f2.ord.C01 <- order.seq(input.seq = mark.f2.C01, n.init = 5,
#                         subset.search = "twopt",
#                         twopt.alg = "rcd", THRES = 6, # notice I used 6 here, may need to change later... 
#                         draw.try = T, wait = 1,
#                         touchdown = T) 

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.C01.Rdata")
LG.f2.ord.C01 

# set map function 
set.map.fun(type="kosambi")  

# order marker 
LG.f2.C01 <- make.seq(LG.f2.ord.C01) 
LG.f2.C01
rf.graph.table(LG.f2.C01) 
par(mfrow=c(1,1))
draw.map(LG.f2.C01, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)




LG1.f2.C01.rcd <- order.seq(input.seq = LG1.f2.C01, n.init = 5, 
                            subset.search = "twopt",
                            twopt.alg = "rcd", THRES = 3,
                            draw.try = T, wait = 1
                            )

LG1.f2.C01.rcd

LG1.f2.C01.rcd <- order.seq(input.seq = LG1.f2.C01, n.init = 5, 
                            subset.search = "twopt",
                            twopt.alg = "rcd", THRES = 3,
                            draw.try = T, wait = 1,
                            touchdown = T
                            )

ripple.seq(LG1.f2.C01, ws=5, LOD=3)
LG1.f2.C01

# safe mode order (only marker with unique positions)
LG1.f2.safe.C01 <- make.seq(LG1.f2.ord.C01, "safe")
LG1.f2.safe.C01
# all order (all markers)
LG.f2.force.C01 <- make.seq(LG.f2.ord.C01, "force")
LG.f2.force.C01
rf.graph.table(LG.f2.force.C01) 
par(mfrow=c(1,1))
draw.map(LG.f2.force.C01, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

# print the recombination fraction mix 
rf.graph.table(LG1.f2.safe.C01) 
par(mfrow=c(1,1))
draw.map(LG1.f2.safe.C01, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

# map estimation for arbitrary order 
LG.f2.arbi.C01 <- make.seq(twopts.f2.05.12, order[[11]])
LG.f2.arbi.C01
LG.f2.arbi.C01.map <- map(LG.f2.arbi.C01) 
rf.graph.table(LG.f2.arbi.C01.map) 

draw.map(LG.f2.arbi.C01.map, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T) 

# set map function 
set.map.fun(type = "haldane") 

########## the previous one 
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_before_filter/twopts.f2.04.21.Rdata")
F2.data <- read.mapmaker(file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_before_filter/F2_geno_for_one_map_final.txt")

max(order[[11]])

LG.f2.arbi.C01 <- make.seq(twopts.f2.LOD3_rf0.5, order[[11]])
LG.f2.arbi.C01 
LG.f2.arbi.C01.map <- map(LG.f2.arbi.C01)
rf.graph.table(LG1.f2.arbi.C01.map) 

```

```{r}
################# LG12/A10 ##############################
LG12.f2 <- make.seq(LGs.f2, 12)
LG12.f2

# use touchdown
LG12.f2.ord <- order.seq(input.seq = LG12.f2, n.init = 5,
                        subset.search = "twopt",
                        twopt.alg = "rcd", THRES = 6,
                        draw.try = T, wait = 1,
                        touchdown = T)

LG12.f2.ord

# safe mode order
LG12.f2.safe <- make.seq(LG12.f2.ord, "safe")
LG12.f2.safe
# all order
LG12.f2.final <- make.seq(LG12.f2.ord, "force")
LG12.f2.final

################# LG13/C01 ##############################
LG13.f2 <- make.seq(LGs.f2, 13)
LG13.f2

# use touchdown
LG13.f2.ord <- order.seq(input.seq = LG13.f2, n.init = 5,
                        subset.search = "twopt",
                        twopt.alg = "rcd", THRES = 6,
                        draw.try = T, wait = 1,
                        touchdown = T)

LG13.f2.ord

# safe mode order
LG13.f2.safe <- make.seq(LG13.f2.ord, "safe")
LG13.f2.safe
# all order
LG13.f2.final <- make.seq(LG13.f2.ord, "force")
LG13.f2.final

################# LG14/C01-1 ##############################
LG14.f2 <- make.seq(LGs.f2, 14)
LG14.f2

# use touchdown
LG14.f2.ord <- order.seq(input.seq = LG14.f2, n.init = 5,
                        subset.search = "twopt",
                        twopt.alg = "rcd", THRES = 6,
                        draw.try = T, wait = 1,
                        touchdown = T)

LG14.f2.ord

# safe mode order
LG14.f2.safe <- make.seq(LG14.f2.ord, "safe")
LG14.f2.safe
# all order
LG14.f2.final <- make.seq(LG14.f2.ord, "force")
LG14.f2.final

################# LG15/C02 ##############################
LG15.f2 <- make.seq(LGs.f2, 15)
LG15.f2

# use touchdown
LG15.f2.ord <- order.seq(input.seq = LG15.f2, n.init = 5,
                        subset.search = "twopt",
                        twopt.alg = "rcd", THRES = 6,
                        draw.try = T, wait = 1,
                        touchdown = T)

LG15.f2.ord

# safe mode order
LG15.f2.safe <- make.seq(LG15.f2.ord, "safe")
LG15.f2.safe
# all order
LG15.f2.final <- make.seq(LG15.f2.ord, "force")
LG15.f2.final

################# LG16/C02-2 ##############################
LG16.f2 <- make.seq(LGs.f2, 16)
LG16.f2

# use touchdown
LG16.f2.ord <- order.seq(input.seq = LG16.f2, n.init = 5,
                         subset.search = "twopt",
                         twopt.alg = "rcd", THRES = 6,
                         draw.try = T, wait = 1,
                         touchdown = T)

LG16.f2.ord

# safe mode order
LG16.f2.safe <- make.seq(LG16.f2.ord, "safe")
LG16.f2.safe
# all order
LG16.f2.final <- make.seq(LG16.f2.ord, "force")
LG16.f2.final

################# LG17/C02-3 ##############################
LG17.f2 <- make.seq(LGs.f2, 17)
LG17.f2

# use touchdown
LG17.f2.ord <- order.seq(input.seq = LG17.f2, n.init = 2,
                         subset.search = "twopt",
                         twopt.alg = "rcd", THRES = 6,
                         draw.try = T, wait = 1,
                         touchdown = T)

LG17.f2.ord

# safe mode order
LG17.f2.safe <- make.seq(LG17.f2.ord, "safe")
LG17.f2.safe
# all order
LG17.f2.final <- make.seq(LG17.f2.ord, "force")
LG17.f2.final

################# LG18/C003 ##############################
LG18.f2 <- make.seq(LGs.f2, 18)
LG18.f2

# use touchdown
LG18.f2.ord <- order.seq(input.seq = LG18.f2, n.init = 5,
                         subset.search = "twopt",
                         twopt.alg = "rcd", THRES = 6,
                         draw.try = T, wait = 1,
                         touchdown = T)

LG18.f2.ord

# safe mode order
LG18.f2.safe <- make.seq(LG18.f2.ord, "safe")
LG18.f2.safe
# all order
LG18.f2.final <- make.seq(LG18.f2.ord, "force")
LG18.f2.final

################# LG19/C04 ##############################
LG19.f2 <- make.seq(LGs.f2, 19)
LG19.f2

# use touchdown
LG19.f2.ord <- order.seq(input.seq = LG19.f2, n.init = 5,
                         subset.search = "twopt",
                         twopt.alg = "rcd", THRES = 6,
                         draw.try = T, wait = 1,
                         touchdown = T)

LG19.f2.ord

# safe mode order
LG19.f2.safe <- make.seq(LG19.f2.ord, "safe")
LG19.f2.safe
# all order
LG19.f2.final <- make.seq(LG19.f2.ord, "force")
LG19.f2.final

################# LG20/C05 ##############################
LG20.f2 <- make.seq(LGs.f2, 20)
LG20.f2

# use touchdown
LG20.f2.ord <- order.seq(input.seq = LG20.f2, n.init = 5,
                         subset.search = "twopt",
                         twopt.alg = "rcd", THRES = 6,
                         draw.try = T, wait = 1,
                         touchdown = T)

LG20.f2.ord

# safe mode order
LG20.f2.safe <- make.seq(LG20.f2.ord, "safe")
LG20.f2.safe
# all order
LG20.f2.final <- make.seq(LG20.f2.ord, "force")
LG20.f2.final

################# LG21/C06 ##############################
LG21.f2 <- make.seq(LGs.f2, 21)
LG21.f2

# use touchdown
LG21.f2.ord <- order.seq(input.seq = LG21.f2, n.init = 5,
                         subset.search = "twopt",
                         twopt.alg = "rcd", THRES = 6,
                         draw.try = T, wait = 1,
                         touchdown = T)

LG21.f2.ord

# safe mode order
LG21.f2.safe <- make.seq(LG21.f2.ord, "safe")
LG21.f2.safe
# all order
LG21.f2.final <- make.seq(LG21.f2.ord, "force")
LG21.f2.final

################# LG22/C07 ##############################
LG22.f2 <- make.seq(LGs.f2, 22)
LG22.f2

# use touchdown
LG22.f2.ord <- order.seq(input.seq = LG22.f2, n.init = 5,
                         subset.search = "twopt",
                         twopt.alg = "rcd", THRES = 6,
                         draw.try = T, wait = 1,
                         touchdown = T)

LG22.f2.ord

# safe mode order
LG22.f2.safe <- make.seq(LG22.f2.ord, "safe")
LG22.f2.safe
# all order
LG22.f2.final <- make.seq(LG22.f2.ord, "force")
LG22.f2.final

################# LG23/C08 ##############################
LG23.f2 <- make.seq(LGs.f2, 23)
LG23.f2

# use touchdown
LG23.f2.ord <- order.seq(input.seq = LG23.f2, n.init = 5,
                         subset.search = "twopt",
                         twopt.alg = "rcd", THRES = 6,
                         draw.try = T, wait = 1,
                         touchdown = T)

LG23.f2.ord

# safe mode order
LG23.f2.safe <- make.seq(LG23.f2.ord, "safe")
LG23.f2.safe
# all order
LG23.f2.final <- make.seq(LG23.f2.ord, "force")
LG23.f2.final

################# LG24/C09 ##############################
LG24.f2 <- make.seq(LGs.f2, 24)
LG24.f2

# use touchdown
LG24.f2.ord <- order.seq(input.seq = LG24.f2, n.init = 5,
                         subset.search = "twopt",
                         twopt.alg = "rcd", THRES = 6,
                         draw.try = T, wait = 1,
                         touchdown = T)

LG24.f2.ord

# safe mode order
LG24.f2.safe <- make.seq(LG24.f2.ord, "safe")
LG24.f2.safe
# all order
LG24.f2.final <- make.seq(LG24.f2.ord, "force")
LG24.f2.final

################# LG25/C09-1 ##############################
LG25.f2 <- make.seq(LGs.f2, 25)
LG25.f2

# use touchdown
LG25.f2.ord <- order.seq(input.seq = LG25.f2, n.init = 5,
                         subset.search = "twopt",
                         twopt.alg = "rcd", THRES = 6,
                         draw.try = T, wait = 1,
                         touchdown = T)

LG25.f2.ord

# safe mode order
LG25.f2.safe <- make.seq(LG25.f2.ord, "safe")
LG25.f2.safe
# all order
LG25.f2.final <- make.seq(LG25.f2.ord, "force")
LG25.f2.final




###################################################
########### drop marker with double crossover ###########
data <- read.mapmaker(file = "F2_415_dropp_geno_Ruijuan.txt")
data

# estimate two-point rf
twopts.f2 <- rf.2pts(data, LOD = 3, max.rf = 0.5)
?rf.2pts

# assign marker to linkage groups
mark.all.f2 <- make.seq(twopts.f2, "all")
?make.seq
LGs.f2 <- group(mark.all.f2, LOD = 3, max.rf = 0.5)
LGs.f2

#################### LG1 ##############################
LG1.f2 <- make.seq(LGs.f2, 1)
LG1.f2

# use touchdown
LG1.f2.ord <- order.seq(input.seq = LG1.f2, n.init = 5,
                        subset.search = "twopt",
                        twopt.alg = "rcd", THRES = 6,
                        draw.try = T, wait = 1,
                        touchdown = T)

LG1.f2.ord

# safe mode order
LG1.f2.safe <- make.seq(LG1.f2.ord, "safe")
LG1.f2.safe
# all order
LG1.f2.final <- make.seq(LG1.f2.ord, "force")
LG1.f2.final

# ripple
ripple.seq(LG1.f2.final, ws = 5, LOD = 3) 
``` 


```{r}
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.A01.Rdata")
LG.f2.A01 <- make.seq(LG.f2.ord.A01) 
LG.f2.A01
rf.graph.table(LG.f2.A01) 
par(mfrow=c(1,1))
draw.map(LG.f2.A01, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)


load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.A03.Rdata")
LG.f2.A03 <- make.seq(LG.f2.ord.A03) 
LG.f2.A03
rf.graph.table(LG.f2.A03) 
par(mfrow=c(1,1))
draw.map(LG.f2.A02, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.A04.Rdata")
LG.f2.A04 <- make.seq(LG.f2.ord.A04) 
LG.f2.A04
rf.graph.table(LG.f2.A04) 
par(mfrow=c(1,1))
draw.map(LG.f2.A04, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.A05.Rdata")
LG.f2.A05 <- make.seq(LG.f2.ord.A05) 
LG.f2.A05
rf.graph.table(LG.f2.A05) 
par(mfrow=c(1,1))
draw.map(LG.f2.A05, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.A06.Rdata")
LG.f2.A06 <- make.seq(LG.f2.ord.A06) 
LG.f2.A06
rf.graph.table(LG.f2.A06) 
par(mfrow=c(1,1))
draw.map(LG.f2.A06, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.A07.Rdata")
LG.f2.A07 <- make.seq(LG.f2.ord.A07) 
LG.f2.A07
rf.graph.table(LG.f2.A07) 
par(mfrow=c(1,1))
draw.map(LG.f2.A07, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.A08.Rdata")
LG.f2.A08 <- make.seq(LG.f2.ord.A08) 
LG.f2.A08
rf.graph.table(LG.f2.A08) 
par(mfrow=c(1,1))
draw.map(LG.f2.A08, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.A09.Rdata")
LG.f2.A09 <- make.seq(LG.f2.ord.A09) 
LG.f2.A09
rf.graph.table(LG.f2.A09) 
par(mfrow=c(1,1))
draw.map(LG.f2.A09, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.A10.Rdata")
LG.f2.A10 <- make.seq(LG.f2.ord.A10) 
LG.f2.A10
rf.graph.table(LG.f2.A10) 
par(mfrow=c(1,1))
draw.map(LG.f2.A10, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)

load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.15/right/LG.f2.ord.C01.Rdata")
LG.f2.C01 <- make.seq(LG.f2.ord.C01) 
LG.f2.C01
rf.graph.table(LG.f2.C01) 
par(mfrow=c(1,1))
draw.map(LG.f2.C01, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T) 

maps <- list(LG.f2.A07, LG.f2.A08, LG.f2.A09)
draw.map(maps, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75) 

```


