---
title: "overlay_sig_loci_Erucic_Oleic"
output: html_document
---

Purpose of this script is to 
1) visualize how fatty acid biosynthesis and metabolic related genes overlay with the significant SNPs that I identified from QTL and GWAS analysis. 
2) visualize how previous identified loci overlay with the significant SNPs and loci that I identified from GWAS and QTL analysis. 

### import lib
```{r}
library(tidyverse)
library(IRanges)
library(GenomicRanges)
library(GenomicFeatures)
library("rtracklayer")
# source("https://bioconductor.org/biocLite.R")
# biocLite("ggbio")
library(ggbio) 
```

### import GWAS and QTL analysis CIM result 
```{r}
setwd("~/Desktop/Brassica_project/KIAT_RNA_seq/overlay/")

# QTL result 
load("cim.qtl.43traits.Rdata")
load("cim.perm.43traits.Rdata")
cim.qtl.sig.Erucic <- cim.qtl$Erucic_acid[cim.qtl$Erucic_acid$lod > cim.perm$Erucic_acid,]
cim.qtl.sig.Oleic <- cim.qtl$Oleic_acid[cim.qtl$Oleic_acid$lod > cim.perm$Oleic_acid,]

# GWAS result 
Erucic_acid <- read.csv("GAPIT..Erucic_acid.GWAS.Results.csv", header = T)

Oleic_acid <- read.csv("GAPIT..Oleic_acid.GWAS.Results.csv", header = T) 

Erucic_Oleic <- rbind(Erucic_acid, Oleic_acid)

SNP.Erucic.Oleic <- data.frame(CHROM = gsub("(S)([[:print:]]+)(_)([[:digit:]]+)", "\\2", Erucic_Oleic[Erucic_Oleic$P.value < 0.00001,]$SNP), 
                         POS = gsub("(S)([[:print:]]+)(_)([[:digit:]]+)", "\\4", Erucic_Oleic[Erucic_Oleic$P.value < 0.00001,]$SNP)) 

SNP.Erucic.Oleic$CHROM <- paste("chr", SNP.Erucic.Oleic$CHROM, sep = "")
SNP.Erucic.Oleic$POS <- as.numeric(as.character(SNP.Erucic.Oleic$POS))

SNP.Erucic.Oleic.unique <- unique(SNP.Erucic.Oleic)
dim(SNP.Erucic.Oleic.unique) # 235 2 
```

### combine and display Erucic and Oleic acid GWAS and QTL analysis result 
```{r}
SNP.Erucic.Oleic 
```

### demo of simple example on how to use ggbio to draw plot of genomic data 
```{r}
data(hg19IdeogramCyto, package = "biovizBase")
head(hg19IdeogramCyto)
getOption("biovizBase")$cytobandColor
autoplot(hg19IdeogramCyto, layout = "karyogram", cytoband = TRUE)

library(GenomicRanges)
hg19 <- keepSeqlevels(hg19IdeogramCyto, paste0("chr", c(1:22, "X", "Y")))
head(hg19)
autoplot(hg19, layout = "karyogram", cytoband = TRUE) 
gffRangeData <- import.gff("~/Desktop/Brassica_project/IGV_file/Brassica_napus.annotation_v5.gff3")
gffRangeData
gff<-as(gffRangeData, "GRanges")
test <- head(gff, 50000)
autoplot(test, layout = "karyogram", aes(color = strand))  
``` 

### draw similar plot for the overlay with sinificant loci and fatty acid genes 
```{r}
# import Grange data of B.napus 
gff.mRNA <- read.table("~/Desktop/Brassica_project/IGV_file/gff.mRNA")
colnames(gff.mRNA) <- c("CHROM", "start", "end", "name") 

# import fatty acid genes info (based on GO and IPR info)
lipid_fatty_acid_genes <- read.table("lipid_fatty_acid_genes", header = F, stringsAsFactors = FALSE)
lipid_fatty_acid_genes %>% dim()  # 7346 
colnames(lipid_fatty_acid_genes) <- "name"

lipid_fatty_acid_genes.2 <- 
gff.mRNA %>% 
  inner_join(lipid_fatty_acid_genes, by = "name") %>% 
  mutate(name = "lipid_fatty_acid_genes")

lipid_fatty_acid_genes.2 %>% dim() # 7346    4 

SNP.Erucic.Oleic.2 <-  
SNP.Erucic.Oleic.unique %>% 
  mutate(start = POS, end = POS, name = "SNP") %>% 
  dplyr::select(-POS)

lipid_fatty_acid_genes.Erucic_Oleic <- 
rbind(lipid_fatty_acid_genes.2, SNP.Erucic.Oleic.2)

gff.lipid_fatty_acid_genes.Erucic_Oleic <- GRanges(seqnames = Rle(lipid_fatty_acid_genes.Erucic_Oleic$CHROM),ranges = IRanges(start = lipid_fatty_acid_genes.Erucic_Oleic$start, end = lipid_fatty_acid_genes.Erucic_Oleic$end), names = lipid_fatty_acid_genes.Erucic_Oleic$name)
 
gff.lipid_fatty_acid_genes
autoplot(gff.lipid_fatty_acid_genes.Erucic_Oleic, layout = "karyogram", aes(color = names))
# this plot shows that if just looking at all lipid and fatty acid genes, there are so many that not meaningful to check the result 

###### now just focus on fatty acid biosynthesis pathways genes (GO:0006633)
# fatty_acid_gene_index <- read.delim("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/fatty_acid_biosynthesis_genes_ID.txt", header = F)
# head(fatty_acid_gene_index)
# nrow(fatty_acid_gene_index) # 248 
# rownames(fatty_acid_gene_index) <- fatty_acid_gene_index$V1
# 
# # extract only fatty acid genes from dataset 
# fatty.acid.genes <- (merge(parent.read.count.one.small, fatty_acid_gene_index, by="row.names"))[,1:28]
# head(fatty.acid.genes)
# rownames(fatty.acid.genes) <- fatty.acid.genes$Row.names
# fatty.acid.genes <- fatty.acid.genes[,-1]
# dim(fatty.acid.genes) # 207 27
# fatty.acid.genes$total <- rowSums(fatty.acid.genes)
# dim(fatty.acid.genes)
```

# sessioninfo
```{r}
sessionInfo() 
# R version 3.2.4 (2016-03-10)
# Platform: x86_64-apple-darwin13.4.0 (64-bit)
# Running under: OS X 10.12.6 (unknown)
# 
# locale:
# [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
# 
# attached base packages:
# [1] stats     graphics  grDevices utils     datasets  methods   base     
# 
# loaded via a namespace (and not attached):
#  [1] Rcpp_0.12.13     cellranger_1.1.0 plyr_1.8.4       bindr_0.1        forcats_0.2.0   
#  [6] tools_3.2.4      jsonlite_1.5     lubridate_1.7.1  tibble_1.3.4     nlme_3.1-131    
# [11] gtable_0.2.0     lattice_0.20-35  pkgconfig_2.0.1  rlang_0.1.4      psych_1.7.8     
# [16] rstudioapi_0.7   parallel_3.2.4   haven_1.1.0      bindrcpp_0.2     knitr_1.17      
# [21] xml2_1.1.1       dplyr_0.7.4      stringr_1.2.0    httr_1.3.1       hms_0.3         
# [26] grid_3.2.4       glue_1.2.0       R6_2.2.2         readxl_1.0.0     foreign_0.8-69  
# [31] tidyr_0.7.2      ggplot2_2.2.1    reshape2_1.4.2   purrr_0.2.4      readr_1.1.1     
# [36] modelr_0.1.1     magrittr_1.5     scales_0.5.0     rvest_0.3.2      assertthat_0.2.0
# [41] mnormt_1.5-5     tidyverse_1.2.0  colorspace_1.3-2 stringi_1.1.5    lazyeval_0.2.1  
# [46] munsell_0.4.3    broom_0.4.2      crayon_1.3.4  
```

