---
title: "eQTL_batch_corrected"
output: html_document
---

Purpose of this script is to find loci controlling expression of different genes. We will firstly focus on fatty acid related genes. this version does not do scale and center, also batch effect were corrected. 

### import lib
```{r}
library(tidyverse) 
library(qtl)
library(snowfall)
library(ggrepel) 
library(Biostrings)
source("~/KIAT/function_BnRNAseq.R")
```

### import data 
```{r}
load("~/F2/output/network_analysis/vstMat.f2.batch.corrected.Rdata")
dim(vstMat.f2.batch.corrected) # 56180   166 
vstMat.f2.batch.corrected[1:10, 1:10]
write.csv(vstMat.f2.batch.corrected, file = "~/F2/output/network_analysis/vstMat.f2.batch.corrected.csv")
### modify this expression file to the right format 

cross.F2 <- read.cross("csvsr", genfile =  "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised.csv", 
                         phefile = "~/F2/output/network_analysis/vstMat.f2.batch.corrected_revised.csv",
                         genotypes = c("AA", "AB", "BB"))   # although the sample IDs are not matched in the original phe and gen file, I still get the right result. 

cross.F2 <- sim.geno(cross.F2,step=1,n.draws=32) # imputation?  
cross.F2 <- calc.genoprob(cross.F2,step=1) 

scanone_eQTL.F2 <- scanone(cross.F2, pheno.col = 2:ncol(cross.F2$pheno), 
	         method = "imp", use = "all.obs") 

### permutation...  
set.seed(12345)   
permtest.F2 <- scanone(cross.F2, pheno.col = 2, method = "imp", n.perm = 10000)  

alphas <- seq(0.01, 0.10, by = 0.01) 
lod.thrs <- summary(permtest.F2, alphas) 
lod.thrs   

save(cross.F2, scanone_eQTL.F2, permtest.F2, lod.thrs, file = "~/F2/output/eQTL/scanone-eqtl_F2.RData") 
```

### determine cis- and trans- eQTL 
```{r}


```

