---
title: "eQTL_no_scale_center"
output: html_document
---

The reason I want to do this analysis is because I found many many eQTL, which I am not sure whether because I used scale and centered expression data from genes. So this data are for genes with just voom transformed data 

# load data 
### load libs
```{r}
library(tidyverse)
library(qtl)
library(snowfall)
library(Biostrings)
source("~/KIAT/function_BnRNAseq.R")
```

### format data for QTL analysis (use v1 map)
```{r}
load("~/F2/output/vstMat.f2.Rdata")
F2_geno_data_2_Ae_Ol_new <- read.table("~/F2/data/F2_geno_data_2_Ae_Ol_new.txt")
dim(F2_geno_data_2_Ae_Ol_new) # 3443 166 
dim(vstMat.f2) # 56180   166

colnames(F2_geno_data_2_Ae_Ol_new) <- gsub("X", "ID_", colnames(F2_geno_data_2_Ae_Ol_new))
F2_geno_data_2_Ae_Ol_new.t <- as.data.frame(t(F2_geno_data_2_Ae_Ol_new))
F2_geno_data_2_Ae_Ol_new.t$ID <- rownames(F2_geno_data_2_Ae_Ol_new.t)

colnames(vstMat.f2) <- gsub("Sample_F2", "ID", colnames(vstMat.f2))
vstMat.f2.t <- as.data.frame(t(vstMat.f2))
vstMat.f2.t$ID <- rownames(vstMat.f2.t)

F2_geno_data_2_Ae_Ol_new.2 <- 
F2_geno_data_2_Ae_Ol_new.t %>% 
  left_join(vstMat.f2.t) %>% 
  dplyr::select(-ID) 

dim(F2_geno_data_2_Ae_Ol_new.2) # 166 59623 
F2_geno_data_2_Ae_Ol_new.2 <- t(F2_geno_data_2_Ae_Ol_new.2)
write.table(F2_geno_data_2_Ae_Ol_new.2, file = "~/F2/data/QTL_analysis/F2_geno_data_2_Ae_Ol_new.2_all_expressed_genes_no_scale_center.txt") 
```

```{r} 
# change file format in linux 
# cat F2_geno_data_2_Ae_Ol_new.2_all_expressed_genes.txt | sed 's/"//g' | awk '{first = $1; $1 = ""; print $0}' > tmp 
# tail -59623 tmp > tmp.1 

write.table(rownames(F2_geno_data_2_Ae_Ol_new.2), file = "~/F2/data/QTL_analysis/marker_info_all_expressed_genes.txt" )
# cat marker_info_all_expressed_genes.txt | awk '{print "*"$2}' | tail -59623 | sed 's/"//g' > marker_info_reform.txt
# paste marker_info_reform.txt tmp.1  | tr "\t" "\ " > F2_geno_for_one_map.txt 
# cat header_one_map_input F2_geno_for_one_map.txt >  F2_geno_for_one_map_final_all_expressed_gene.txt 
# change header info: maker number to the right marker number & phenotype data number 
```

### load data 
```{r}
LG.f2 <- read.cross("mm", file = "~/F2/data/QTL_analysis/F2_geno_for_one_map_final_all_expressed_gene_no_scale_center.txt", mapfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.map") ### takes one day to read in this data... 

save(LG.f2, file = "~/F2/output/QTL_analysis/LG.f2.all_expressed_genes_no_scale_center.Rdata") 
LG.f2$pheno %>% dim() 
```

### re-estimate map
```{r} 
load("~/F2/output/QTL_analysis/LG.f2.all_expressed_genes_no_scale_center.Rdata")
LG.f2
summaryMap(LG.f2)

LG.f2.before.crossover <- LG.f2

for (chr in names(LG.f2$geno)) { # for each chromosome in cross genotype data
  my.chr <- get(chr,LG.f2$geno) # return the genotype data, including data & map
  print(paste(chr,"NA before",sum(is.na(my.chr$data)))) 
  if(ncol(my.chr$data) > 3) { 
    my.chr$data[,2:(ncol(my.chr$data)-1)] <- sapply(2:(ncol(my.chr$data)-1),function(i) {
      apply(my.chr$data[,(i-1):(i+1)],1,function(gt) {
        if (any(is.na(gt))) return(gt[2]) #technically should be looking at the next genotyped marker.
        if ( (length(unique(gt)) == 2) & (gt[1] == gt[3])) return(NA)
        if ( length(unique(gt))  == 3) return(NA)
        return(gt[2])
      })
    })
  }
  LG.f2$geno <- within(LG.f2$geno,assign(chr,my.chr))
  print(paste(chr,"NA after",sum(is.na(get(chr,LG.f2$geno)$data))))
}  

map.new <- est.map(LG.f2,verbose=T,error.prob=.01) 
save(map.new, file = "~/F2/output/QTL_analysis/map.new.fatty_acid_lipid_no_scale_center.Rdata") 

LG.f2 <- replace.map(LG.f2, map.new)
LG.f2.after.crossover <- LG.f2

plot.map(LG.f2.after.crossover, alternate.chrid = T) # the old genetic map
plot.map(LG.f2.before.crossover,LG.f2.after.crossover, alternate.chrid = T) # genetic map comparison 

# fix LG10 problem using ripple 
set.seed(16)
LG.f2.after.crossover <- orderMarkers(LG.f2.after.crossover, chr = c(10), 
	                        window = 5, use.ripple = TRUE, maxit = 4000, 
	                        error.prob = 0.0001, verbose = T) 

plotMap(LG.f2.after.crossover, chr = '10')
summaryMap(LG.f2.after.crossover) # 2884
save(LG.f2.after.crossover, file = "~/F2/output/QTL_analysis/LG.f2.after.crossover_all_expressed_genes_no_scale_center.Rdata") 
```     

### eQTL analysis on cabernet 
1) running eQTL takes a long time, so I split the 50,000 genes into 10 subset, and run scanone on subset in parallel. 
https://github.com/leejimmy93/KIAT_cabernet/tree/master/F2/eQTL/scanone_all_expressed_genes_1.R ... https://github.com/leejimmy93/KIAT_cabernet/tree/master/F2/eQTL/scanone_all_expressed_genes_10.R 
https://github.com/leejimmy93/KIAT_cabernet/blob/master/F2/scanone_all_expressed_genes_1_10.slurm 

2) For permutation, since it takes too long time to run permutation on all genes, so I sampled 100 genes and ran 1000 times permutation on these genes so that median can be taken as the threshold for eQTL analysis 
https://github.com/leejimmy93/KIAT_cabernet/blob/master/F2/eQTL/scanone_all_expressed_genes_perm.R 

3) transfer to whitney for analysis 

### eQTL result analysis 
```{r}
# firstly check the two FAD genes 
# load data (scanone result)
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/data/eQTL_analysis/no_scalen_center/"
files <- list.files(path=path)
files 
setwd(path)
eQTL_result <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 
save(eQTL_result, file = "~/F2/data/eQTL_analysis/no_scalen_center/eQTL_result.Rdata")
# for the result, 1st 10 are eQTL analysis result, 11 is the permuation result for 100 sampled genes.
```

### determine cis- and trans- eQTL 
```{r}
load("~/F2/data/eQTL_analysis/no_scalen_center/eQTL_result.Rdata")

# get threshold 
lod.thres # [1] 4.159691 

###### a bunch of test 
plot(eQTL_result[[3]]$BnaA08g12780D,bandcol="gray90", main="BnaA08g12780D") # looks good 
plot(eQTL_result[[9]]$BnaA08g11140D,bandcol="gray90", main="BnaA08g11140D") # looks good
plot(eQTL_result[[9]]$BnaA08g11130D,bandcol="gray90", main="BnaA08g11130D") # as expected 
abline(h=lod.thres,lty=2) 
plot(eQTL_result[[7]]$BnaC03g65980D,bandcol="gray90", main="BnaC03g65980D") # as expected
abline(h=lod.thres,lty=2) 

bayesint(eQTL_result[[3]]$BnaA08g12780D, chr = 8, expandtomarkers = T) 
bayesint(scanone.imp$Erucic_acid, chr = "A08", expandtomarkers = T)

bayesint(eQTL_result[[9]]$BnaA08g11140D, chr = 8, expandtomarkers = T) 
bayesint(scanone.imp$Erucic_acid, chr = "A08", expandtomarkers = T)

# below is the list of fatty acid genes in the interval which are differentially expressed between Da-Ae and Da-Ol-1 
"BnaA08g11810D" %in% cis_trans_result_final$gene_ID # F
"BnaA08g12800D" %in% cis_trans_result_final$gene_ID
"BnaA08g13200D" %in% cis_trans_result_final$gene_ID
"BnaA08g14550D" %in% cis_trans_result_final$gene_ID
"BnaA08g12780D" %in% cis_trans_result_final$gene_ID
"BnaA08g11130D" %in% cis_trans_result_final$gene_ID
"BnaC03g67820D" %in% cis_trans_result_final$gene_ID 

####### abunch of test before analysis 

# get all eQTL based on this threshold 
system.time(
tmp <- 
lapply(names(eQTL_result[1:10]), function(subset) {
  print(subset)
  lapply(names(eQTL_result[[subset]]), function(gene) {
    tmp1 <- eQTL_result[[subset]][[gene]][eQTL_result[[subset]][[gene]]$lod > lod.thres,]
    if (nrow(tmp1) > 0){
    tmp1$gene_ID <- gene
    tmp1$snp_ID <- rownames(tmp1)
    tmp1
  }
})
}) 
) # 20 sec  

# test <- do.call("rbind", tmp) 
# test2 <- do.call("rbind", test)

####### re do # both the above two line code the code below work, but why the above works
test <- list()
for (subset in seq_along(tmp)){
  test[[subset]] <- do.call("rbind", tmp[[subset]])
}

test[[10]] %>% dim()

test2 <- do.call("rbind", test)
dim(test2) 
############# 

# extract only SNP info 
cis_trans_result <- test2[grepl("^chr", test2$snp_ID),]
dim(cis_trans_result) # 728717      5 
cis_trans_result %>% head()

unique(cis_trans_result$snp_ID) %>% length # 2013 
unique(cis_trans_result$gene_ID) %>% length() # 21823 

# get genome range
library(IRanges)
library(GenomicRanges)
library(GenomicFeatures)
library("rtracklayer")

### get gff file with gene chrom & pos info, gff3 file must be sorted 
gff.mRNA <- read.table("~/Reference/B.napus/gff.mRNA")
dim(gff.mRNA) # 101040      4 
head(gff.mRNA) 
colnames(gff.mRNA) <- c("CHROM", "start", "end", "gene_ID") 
gff.mRNA %>% head()

# merge two files
cis_trans_result_2 <- 
cis_trans_result %>% 
  left_join(gff.mRNA, by = "gene_ID") %>% 
  mutate(gene_chrom = CHROM, gene_start = start, gene_end = end) %>% 
  dplyr::select(-(CHROM:end)) %>% 
  separate(snp_ID, into = c("snp_chrom", "snp_pos"), sep = "_", remove = F) %>% 
  mutate(snp_pos = as.numeric(snp_pos)) %>% 
  mutate(LG = chr, genetic_pos = as.numeric(pos)) %>% 
  dplyr::select(-(chr:pos))

# get snp & chrom ID & pos, as well as gene  
cis_trans_result_2 %>% head()

# determine cis- and trans- eQTL only on chromosome 
# e-traits (genes) which are not on the same chrom as SNPs are trans- 
cis_trans_result_2$cis_trans <- ifelse(cis_trans_result_2$gene_chrom == cis_trans_result_2$snp_chrom, "cis_candidate", "trans")
sum(cis_trans_result_2$cis_trans == "cis_candidate") # 492708 
sum(cis_trans_result_2$cis_trans == "trans") # 236009 
dim(cis_trans_result_2) # 728717     11 

length(unique(cis_trans_result_2[cis_trans_result_2$cis_trans == "cis_candidate",]$gene_ID)) # 11118 
length(unique(cis_trans_result_2[cis_trans_result_2$cis_trans == "trans",]$gene_ID)) # 13295   
length(unique(cis_trans_result_2$gene_ID)) # 21823 

# use bayesint to find interval estimate of each gene --> for cis-eQTL 
# for each cis-candidate gene --> if there are markers go above the threshold --> check to see if there are markers on the same chromsome as the gene --> if yes, do bayesint on the QTL result of that gene --> and extract the marker result rownames --> use the min as the start of the sig loci, and max as the end of the sig loci --> 
cis_candidate <- unique(cis_trans_result_2[cis_trans_result_2$cis_trans == "cis_candidate",]$gene_ID)
length(cis_candidate) # 11118 

cis_candidate %>% head()
sum(cis_candidate %in% names(eQTL_result[[1]]), 
    cis_candidate %in% names(eQTL_result[[2]]),
    cis_candidate %in% names(eQTL_result[[3]]),
    cis_candidate %in% names(eQTL_result[[4]]),
    cis_candidate %in% names(eQTL_result[[5]]),
    cis_candidate %in% names(eQTL_result[[6]]),
    cis_candidate %in% names(eQTL_result[[7]]),
    cis_candidate %in% names(eQTL_result[[8]]),
    cis_candidate %in% names(eQTL_result[[9]]),
    cis_candidate %in% names(eQTL_result[[10]])) # 11118, correct 

system.time(
tmp.2 <- 
lapply(names(eQTL_result[1:10]), function(subset) {
  print(subset)
  lapply(names(eQTL_result[[subset]])[names(eQTL_result[[subset]]) %in% cis_candidate], function(gene) {
    sub = gsub("(Bna)(A|C)(01|02|03|04|05|06|07|08|09|10)([[:print:]]+)", "\\2", gene) 
    chrom = gsub("(Bna)(A|C)(01|02|03|04|05|06|07|08|09|10)([[:print:]]+)", "\\3", gene)
    chrom <- ifelse(sub == "A", as.numeric(as.character(chrom)), 10 + as.numeric(as.character(chrom)))
    tmp1 <- bayesint(eQTL_result[[subset]][[gene]], chr = chrom, expandtomarkers = T)
    tmp1 <- tmp1[grepl("^chr", rownames(tmp1)),] 
    tmp1$gene_ID <- gene
    tmp1$pos_2 <- gsub("([[:print:]]+)(_)([[:print:]])", "\\3", rownames(tmp1))
    tmp1$loci_start <- min(as.numeric(tmp1$pos_2))
    tmp1$loci_end <- max(as.numeric(tmp1$pos_2)) 
    tmp1[1,c("gene_ID", "loci_start", "loci_end")] # the last item is the one that was returned
    })
}
)
) # 20 sec  

# combine result 
test3 <- list()
for (subset in seq_along(tmp.2)){
  test3[[subset]] <- do.call("rbind", tmp.2[[subset]]) 
}

test4 <- do.call("rbind", test3)
dim(test4) # 11118 3 
head(test4)

sum(test4$loci_start < test4$loci_end) # 11087
sum(test4$loci_start == test4$loci_end) # 31

cis_candidate_2 <- cis_trans_result_2[cis_trans_result_2$cis_trans == "cis_candidate",]
dim(cis_candidate_2) # 492708     11 
cis_candidate_2 %>% head()

cis_candidate_3 <- 
cis_candidate_2 %>% 
  left_join(test4, by = "gene_ID")

cis_candidate_3 %>% head()
dim(cis_candidate_3) # 492708     13 

sum(cis_candidate_3$gene_start < cis_candidate_3$gene_end) # 492708
sum(cis_candidate_3$loci_start < cis_candidate_3$loci_end) # 490133
sum(cis_candidate_3$loci_start == cis_candidate_3$loci_end) # 2575 

# use loci, gene start/end, and marker genetic position to find cis-eQTL 
cis_candidate_4 <- 
cis_candidate_3 %>% 
  filter((gene_start < loci_start & gene_end > loci_start & snp_pos > loci_start & snp_pos < loci_end) |
         (gene_start > loci_start & gene_end < loci_end & snp_pos > loci_start & snp_pos < loci_end) |
         (gene_start < loci_end & gene_end > loci_end & snp_pos > loci_start & snp_pos < loci_end)) 


dim(cis_candidate_4)  # 75225     13 
"BnaA08g12780D" %in% cis_candidate_4$gene_ID # good
"BnaA08g11130D" %in% cis_candidate_4$gene_ID # as expected 
"BnaA08g11140D" %in% cis_candidate_4$gene_ID # good 

unique(cis_candidate_4$gene_ID) %>% length() # 7514 cis-genes

# get the final set of cis- and trans- eQTL, 
cis_candidate_4$gene_ID 


cis_trans_result_2$cis_trans <- ifelse(cis_trans_result_2$cis_trans == "cis_candidate", "cis", "trans") 
cis_trans_result_final <- cis_trans_result_2 
View(cis_candidate_4)
save(cis_trans_result_final, file = "~/F2/output/eQTL/cis_trans_result_final.Rdata")
``` 
