---
title: "eQTL_batch_corrected"
output: html_document
---

Purpose of this script is to find loci controlling expression of different genes. We will firstly focus on fatty acid related genes. this version does not do scale and center, also batch effect were corrected. 

### import lib
```{r} 
library(tidyverse) 
library(qtl)
library(snowfall) 
library(ggrepel) 
library(Biostrings)
source("~/KIAT/function_BnRNAseq.R")
```

### import data 
```{r}
load("~/F2/output/network_analysis/vstMat.f2.batch.corrected.Rdata")
dim(vstMat.f2.batch.corrected) # 56180   166 
vstMat.f2.batch.corrected[1:10, 1:10]
write.csv(vstMat.f2.batch.corrected, file = "~/F2/output/network_analysis/vstMat.f2.batch.corrected.csv")
### modify this expression file to the right format 

cross.F2 <- read.cross("csvsr", genfile =  "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised.csv", 
                         phefile = "~/F2/output/network_analysis/vstMat.f2.batch.corrected_revised.csv",
                         genotypes = c("AA", "AB", "BB"))   # although the sample IDs are not matched in the original phe and gen file, I still get the right result. 

cross.F2 <- sim.geno(cross.F2,step=1,n.draws=32) # imputation?  
cross.F2 <- calc.genoprob(cross.F2,step=1) 

scanone_eQTL.F2 <- scanone(cross.F2, pheno.col = 2:ncol(cross.F2$pheno), 
	         method = "imp", use = "all.obs") 

### permutation...  
set.seed(12345)   
permtest.F2 <- scanone(cross.F2, pheno.col = 2, method = "imp", n.perm = 10000)  

alphas <- seq(0.01, 0.10, by = 0.01) 
lod.thrs <- summary(permtest.F2, alphas) 
lod.thrs   

save(cross.F2, scanone_eQTL.F2, permtest.F2, lod.thrs, file = "~/F2/output/eQTL/scanone-eqtl_F2.RData") 
```

### determine cis- and trans- eQTL 
```{r}
load("~/F2/output/eQTL/scanone-eqtl_F2.RData")
scanone_eQTL.F2 %>% dim() # 4887 56182 

# get threshold 
threshold.95 <- lod.thrs[5,]

# get all eQTL based on this threshold  
eQTL_sign <- 
sapply(colnames(scanone_eQTL.F2), function(gene) {
  sum(scanone_eQTL.F2[,gene] > threshold.95) > 0 
}) 

sum(eQTL_sign) # 20,172 genes with eQTL 
scanone_eQTL.F2 <- scanone_eQTL.F2[,eQTL_sign]  
dim(scanone_eQTL.F2) # 4887 20174 

# get bayesint result for every gene 
scanone.gather <-  
scanone_eQTL.F2 %>% 
  gather(key = trait, value = LOD, -chr, -pos) 

sig.chrs <- scanone.gather %>% filter(LOD > threshold.95) %>%
  group_by(trait,chr) %>% 
  dplyr::summarise(count = n()) # this is to get the significant chr ID for each trait 

sig.chrs 
sig.chrs %>% dim() # 98580564        4

bayesint.list <- apply(sig.chrs,1,function(hit) { # for every row("trait, chr, count") in eigengene module 
    result <- bayesint(scanone_eQTL.F2[c("chr","pos",hit["trait"])],  
                     chr=hit["chr"], 
                     lodcolumn = 1, 
                     expandtomarkers = TRUE 
  )
  colnames(result)[3] <- "LOD" 
  result
}) 

names(bayesint.list) <- sig.chrs$trait

bayesint.list <- lapply(bayesint.list,function(x) 
                          x %>% 
                          as.data.frame() %>%
                          rownames_to_column(var="markername")  %>% # make rownames to column and use "markername" as the colname for the new colomn  
                          mutate(chr=as.character(chr))
)

bayesint.result <- as.tibble(bind_rows(bayesint.list,.id="trait")) %>% # combine list into tibble 
  dplyr::select(trait,chr,pos,markername,LOD) %>%  
  separate(markername,into=c("chr1","Mbp"),sep="_", convert=TRUE) %>% 
  group_by(trait,chr) %>% 
  dplyr::summarize(start=min(Mbp, na.rm = T),end=max(Mbp, na.rm = T),min_eQTL_LOD=min(LOD),max_eQTL_LOD=max(LOD)) %>% 
  #for the high QTL peaks the interval width is 0.  That is overly precise and need to widen those.
  mutate(start=ifelse(start==end, start-20000,start), end=ifelse(start==end,end+20000,end)) 
# the original code (below) doesn't work correctly for some reason 
# mutate(start=ifelse(start==end,max(0,start-20000),start), end=ifelse(start==end,end+20000,end))
  
bayesint.result %>% dim() # 22381     6 

# get genome range 
library(IRanges)
library(GenomicRanges)
library(GenomicFeatures)
library("rtracklayer")

### get gff file with gene chrom & pos info, gff3 file must be sorted 
gff.mRNA <- read.table("~/Reference/B.napus/gff.mRNA")
dim(gff.mRNA) # 101040      4 
head(gff.mRNA) 
colnames(gff.mRNA) <- c("CHROM", "start", "end", "gene_ID") 
gff.mRNA %>% head()  

# look for cis-eQTL 
bayesint.result.2 <- 
bayesint.result %>% 
  mutate(gene_ID = trait, eQTL_chr = chr, eQTL_start = start, eQTL_end = end) %>% 
  dplyr::select(trait, gene_ID, eQTL_chr, eQTL_start, eQTL_end, min_eQTL_LOD, max_eQTL_LOD) %>% 
  left_join(gff.mRNA, by = "gene_ID") 

bayesint.result.2$eQTL_chr <- paste("chr", bayesint.result.2$eQTL_chr, sep = "")
bayesint.result.2 %>% head()

cis_eQTL <- 
bayesint.result.2 %>% 
  filter(eQTL_chr == CHROM) %>% 
  filter((start < eQTL_start & end > eQTL_start) | # also need SNP pos... 
         (start >= eQTL_start & end <= eQTL_end) |
         (start < eQTL_end & end > eQTL_end)) 

dim(cis_eQTL) # 8512   10 

trans_eQTL <- 
bayesint.result.2 %>% 
  anti_join(cis_eQTL) 

save(cis_eQTL, trans_eQTL, file = "~/F2/output/eQTL/cis_trans_result_new.Rdata")
```

### find overlaps between eQTL & trait QTL  
```{r}
# trait QTL using scanone result 
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/eQTL/traitQTL.annotated.Rdata")
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/eQTL/cis_trans_result_new.Rdata")

cis_eQTL %>% dim() # 8512   10 
trans_eQTL %>% dim() # 13869    10 

cis_eQTL.qtl.combined <- inner_join(cis_eQTL,traitQTL.annotated,by=c("gene_ID"="name")) 
cis_eQTL.qtl.combined %>% head()    
cis_eQTL.qtl.combined %>% dim() # 2484   17  

colnames(cis_eQTL.qtl.combined) 

trans_eQTL.qtl.combined <- inner_join(trans_eQTL,traitQTL.annotated,by=c("gene_ID"="name")) 
trans_eQTL.qtl.combined %>% head()    
trans_eQTL.qtl.combined %>% dim() # 2165    17  
  
napus_GO <- read.table("~/Desktop/Brassica_project/reference/Brassica_napus_GO", stringsAsFactors = F)
colnames(napus_GO)[1] <- "gene_ID"

napus_GO_combined <- 
napus_GO %>% 
  group_by(gene_ID) %>% 
  summarise(paste(V2, collapse = ";"))  

colnames(napus_GO_combined)[2] <- "B.napus_GO_term"
colnames(cis_eQTL.qtl.combined)

cis_eQTL.qtl.combined.final <- 
cis_eQTL.qtl.combined %>% 
  left_join(napus_GO_combined) %>% 
  mutate(eQTL_start = start.x, eQTL_end = end.x, QTL_start = start.y, QTL_end = end.y) %>% 
  dplyr::select(-c(start.x, end.x, start.y, end.y)) 
  
trans_eQTL.qtl.combined.final <- 
trans_eQTL.qtl.combined %>% 
  left_join(napus_GO_combined) %>% 
  mutate(eQTL_start = start.x, eQTL_end = end.x, QTL_start = start.y, QTL_end = end.y) %>% 
  dplyr::select(-c(start.x, end.x, start.y, end.y)) 

trans_eQTL.qtl.combined.final$B.napus_GO_term 
write.csv(cis_eQTL.qtl.combined.final, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/eQTL/cis_eQTL.qtl.combined.final.csv")

write.csv(trans_eQTL.qtl.combined.final, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/eQTL/trans_eQTL.qtl.combined.final.csv") 

### make some plot for several important genes 
# the two FAE 

# the two 
```

### make a plot of cis- and trans- eQTL, x-axis being eQTL physical position, y-axis being transcript start position 
```{r}

```

### tran-eQTL hotspot 
```{r}

```



















