---
title: "eQTL"
output: html_document
---
Purpose of this script is to find loci controlling expression of different genes. We will firstly focus on fatty acid related genes.  

### import lib
```{r}
library(tidyverse) 
```

### import data, get genes with detectable expression  
```{r}
read_count_F2 <- read.table("~/F2/data/est.counts.F2.tsv.gz", header = T, row.names = 1)
dim(read_count_F2) # 101040    166

# remove lowly expressed genes 
read.count.small <- read_count_F2[rowSums(read_count_F2 > 10) >= 166*0.25,]
dim(read.count.small) # 59934   131  
colnames(read.count.small)
save(read.count.small, file="~/F2/output/read.count.small.Rdata") 
```

### voom transformation 
```{r}
# source("https://bioconductor.org/biocLite.R") 
# biocLite("DESeq2") 
library("DESeq2")

# define sample 
read.count.sample <- data.frame(group=factor(colnames(read_count_F2)))

dds.f2 <- DESeqDataSetFromMatrix(countData = round(read.count.small), colData = read.count.sample, design = ~ group) 
vsd.f2 <- varianceStabilizingTransformation(dds.f2)
vstMat.f2 <- assay(vsd.f2)
colnames(vstMat.f2) <- colnames(read.count.small)
save(vstMat.f2, file = "~/F2/output/vstMat.f2.Rdata")
```

### scale and center
```{r} 
dim(vstMat.f2) # 56180   166  
vstMat.f2.centered.scaled <- scale(vstMat.f2, center = TRUE, scale = TRUE) # center to zero
save(vstMat.f2.centered.scaled, file = "~/F2/output/vstMat.f2.centered.scaled.Rdata")
```

### get fatty acid & lipid genes 
```{r}
load("~/F2/output/vstMat.f2.centered.scaled.Rdata") 
vstMat.f2.centered.scaled[1:10, 1:10] 

# get lipid and fatty acid related genes using GO and IPR term 
# cat Brassica_napus_IPR.withdescription.gz | grep "fatty\|lipid" > fatty_acid_related/lipid_fatty_acid_IPR

# cat Brassica_napus_GO | awk '{print $2}' > GO_list 

# cat GO_list | sort | uniq > GO_list_unique 

# extract GO annotation from http://yeastmine.yeastgenome.org/yeastmine/bag.do save as GO_annotation.csv 

# cat GO_annotation.csv | grep "fatty\|lipid" | awk '{print $1}' | sort | uniq | sed 's/"//g' > fatty_acid_lipd_GO 

# cat lipid_fatty_acid_* | awk '{print $1}' | sort | uniq > lipid_fatty_acid_genes
lipid_fatty_acid_genes <- read.table("~/Reference/B.napus/fatty_acid_related/lipid_fatty_acid_genes", header = F, stringsAsFactors = FALSE)
dim(lipid_fatty_acid_genes) # 7346 
```

### load libs
```{r}
library(tidyverse)
library(qtl)
library(snowfall)
library(ggrepel) 
library(Biostrings)
source("~/KIAT/function_BnRNAseq.R")
```

### format data for QTL analysis (use v1 map)
```{r}
# geno data 
F2_geno_data_2_Ae_Ol_new <- read.table("~/F2/data/F2_geno_data_2_Ae_Ol_new.txt")
dim(F2_geno_data_2_Ae_Ol_new) # 3443 166 
F2_geno_data_2_Ae_Ol_new[1:10, 1:10]

# pheno (gene counts)
load("~/F2/output/vstMat.f2.centered.scaled.Rdata")
vstMat.f2.centered.scaled <- as.data.frame(vstMat.f2.centered.scaled)
vstMat.f2.centered.scaled[1:10, 1:10]

# get only fatty acid & lipid related genes 
vstMat.f2.centered.scaled$geneID <- rownames(vstMat.f2.centered.scaled)
vstMat.f2.centered.scaled$geneID
lipid_fatty_acid_genes <- read.table("~/Reference/B.napus/fatty_acid_related/lipid_fatty_acid_genes", header = F, stringsAsFactors = FALSE)
lipid_fatty_acid_genes$geneID <- lipid_fatty_acid_genes$V1 

lipid_fatty_acid_genes.expression <- 
  vstMat.f2.centered.scaled %>%
    semi_join(lipid_fatty_acid_genes) 

rownames(lipid_fatty_acid_genes.expression) <- lipid_fatty_acid_genes.expression$geneID
lipid_fatty_acid_genes.expression <- 
  lipid_fatty_acid_genes.expression %>% 
    dplyr::select(-geneID)

colnames(lipid_fatty_acid_genes.expression) <- gsub("Sample_F2", "ID", colnames(lipid_fatty_acid_genes.expression))
dim(lipid_fatty_acid_genes.expression) # 4943 166 

# combine geno & pheno for QTL analysis 
colnames(F2_geno_data_2_Ae_Ol_new) <- gsub("X", "ID_", colnames(F2_geno_data_2_Ae_Ol_new))

F2_geno_data_2_Ae_Ol_new.t <- as.data.frame(t(F2_geno_data_2_Ae_Ol_new))
F2_geno_data_2_Ae_Ol_new.t$ID <- rownames(F2_geno_data_2_Ae_Ol_new.t)
lipid_fatty_acid_genes.expression.t <- as.data.frame(t(lipid_fatty_acid_genes.expression)) 
lipid_fatty_acid_genes.expression.t$ID <- rownames(lipid_fatty_acid_genes.expression.t)

F2_geno_data_2_Ae_Ol_new.2 <- 
F2_geno_data_2_Ae_Ol_new.t %>% 
  left_join(lipid_fatty_acid_genes.expression.t) %>% 
  dplyr::select(-ID) 

dim(F2_geno_data_2_Ae_Ol_new.2) # 166 8386 
F2_geno_data_2_Ae_Ol_new.2[1:10, 3440:3450] 
F2_geno_data_2_Ae_Ol_new.2 <- t(F2_geno_data_2_Ae_Ol_new.2)
write.table(F2_geno_data_2_Ae_Ol_new.2, file = "~/F2/data/QTL_analysis/F2_geno_data_2_Ae_Ol_new.2.txt")

# change file format in linux 
# cat F2_geno_data_2_Ae_Ol_new.2.txt | sed 's/"//g' | awk '{first = $1; $1 = ""; print $0}' > tmp
# tail -8386 tmp > tmp.1

write.table(rownames(F2_geno_data_2_Ae_Ol_new.2), file = "~/F2/data/QTL_analysis/marker_info.txt" )
# cat marker_info.txt | awk '{print "*"$2}' | tail -8386 | sed 's/"//g' > marker_info_reform.txt
# paste marker_info_reform.txt tmp.1  | tr "\t" "\ " > F2_geno_for_one_map.txt 
# paste marker_info_reform.txt tmp.1 | awk '{print $1 " " $2}' > F2_geno_for_one_map.txt
# cat header_one_map_input F2_geno_for_one_map.txt > F2_geno_for_one_map_final.txt
# change header info: maker number to the right marker number & phenotype data number 
```

### load data (gen file & phe file)
```{r}
LG.f2 <- read.cross("mm", file = "~/F2/data/QTL_analysis/F2_geno_for_one_map_final.txt", mapfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.map") 
save(LG.f2, file = "~/F2/output/QTL_analysis/LG.f2.fatty_acid_lipid.Rdata")
```

### re-estimate map
```{r} 
LG.f2

LG.f2.before.crossover <- LG.f2

for (chr in names(LG.f2$geno)) { # for each chromosome in cross genotype data
  my.chr <- get(chr,LG.f2$geno) # return the genotype data, including data & map
  print(paste(chr,"NA before",sum(is.na(my.chr$data)))) 
  if(ncol(my.chr$data) > 3) { 
    my.chr$data[,2:(ncol(my.chr$data)-1)] <- sapply(2:(ncol(my.chr$data)-1),function(i) {
      apply(my.chr$data[,(i-1):(i+1)],1,function(gt) {
        if (any(is.na(gt))) return(gt[2]) #technically should be looking at the next genotyped marker.
        if ( (length(unique(gt)) == 2) & (gt[1] == gt[3])) return(NA)
        if ( length(unique(gt))  == 3) return(NA)
        return(gt[2])
      })
    })
  }
  LG.f2$geno <- within(LG.f2$geno,assign(chr,my.chr))
  print(paste(chr,"NA after",sum(is.na(get(chr,LG.f2$geno)$data))))
}  

map.new <- est.map(LG.f2,verbose=T,error.prob=.01) 
save(map.new, file = "~/F2/output/QTL_analysis/map.new.fatty_acid_lipid.Rdata")

LG.f2 <- replace.map(LG.f2, map.new)
LG.f2.after.crossover <- LG.f2

plot.map(LG.f2.after.crossover, alternate.chrid = T) # the old genetic map
plot.map(LG.f2.before.crossover,LG.f2.after.crossover, alternate.chrid = T) # genetic map comparison 
```    

### deal with LG10 
```{r}
load("~/F2/output/QTL_analysis/LG.f2.after.crossover_fatty_acid_lipid.Rdata")
summaryMap(LG.f2.after.crossover) # 2853. 2

set.seed(16)
LG.f2.after.crossover <- orderMarkers(LG.f2.after.crossover, chr = c(10), 
	                        window = 5, use.ripple = TRUE, maxit = 4000, 
	                        error.prob = 0.0001, verbose = T) 

plotMap(LG.f2.after.crossover, chr = '10')
summaryMap(LG.f2.after.crossover)
save(LG.f2.after.crossover, file = "~/F2/output/QTL_analysis/LG.f2.after.crossover_fatty_acid_lipid.Rdata")
```

### scanone eQTL (on the two FAD genes only)
```{r}
load("~/F2/output/QTL_analysis/LG.f2.after.crossover_fatty_acid_lipid.Rdata")
summaryMap(LG.f2.after.crossover)

LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation? 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) 
LG.f2.after.crossover$pheno <- LG.f2.after.crossover$pheno[,c("BnaA08g12780D", "BnaC03g67820D")] 

system.time(
scanone.perm.imp <- 
  sfLapply(seq_along(LG.f2.after.crossover$pheno), function(trait){
    print(trait) # print doesn't work in here 
    tmp <-scanone(LG.f2.after.crossover,pheno.col = trait, method="imp",n.perm=1000, n.cluster = 12, verbose = T)
    summary(tmp)[1] # #keep the 95th percentile for future use.This corresponds to p <0.05
  }) # takes 40 mins to finish 
) 

  #  user  system elapsed 
  # 1.020   1.200 394.109 

sfStop() 

names(scanone.perm.imp) <- colnames(LG.f2.after.crossover$pheno) 

system.time(
scanone.imp <- 
lapply(seq_along(LG.f2.after.crossover$pheno), function(trait) {
  print(trait)
  scanone(LG.f2.after.crossover,pheno.col=trait,method="imp", chr = 01)
}) 
) 
  #  user  system elapsed 
  # 6.124   3.420   8.849 
names(scanone.imp) <- colnames(LG.f2.after.crossover$pheno) 
# plot 
save(scanone.perm.imp, scanone.imp, file = "~/F2/output/QTL_analysis/eQTL_twogenes.Rdata")

png("~/F2/output/QTL_analysis/figure/scanone.imp.BnaA08g12780D.png", width = 8, height = 8, units = "in", res = 300)
scanone.imp.BnaA08g12780D <- scanone(LG.f2.after.crossover,pheno.col="BnaA08g12780D",method="imp") # 
plot(scanone.imp.BnaA08g12780D,bandcol="gray90", main="BnaA08g12780D")
abline(h=scanone.perm.imp[["BnaA08g12780D"]],lty=2) #add permuation threshold
dev.off()

png("~/F2/output/QTL_analysis/figure/scanone.imp.BnaC03g67820D.png", width = 8, height = 8, units = "in", res = 300)
scanone.imp.BnaC03g67820D <- scanone(LG.f2.after.crossover,pheno.col="BnaC03g67820D",method="imp") # 
plot(scanone.imp.BnaC03g67820D,bandcol="gray90", main="BnaC03g67820D", ylim=c(0, 30))
abline(h=scanone.perm.imp[["BnaC03g67820D"]],lty=2) #add permuation threshold
dev.off() 
```

### eQTL on all fatty acid and lipid genes 
```{r}
# https://github.com/leejimmy93/KIAT/blob/master/F2/scanone_fatty_acid_lipid.R
load("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/QTL_analysis/scanone.imp_fatty_acid_lipid.Rdata")
scanone.imp_fatty_acid_lipid <- scanone.imp
scanone.imp_fatty_acid_lipid

# permutation... just on some genes? 
```

### on all genes
### combine geno & expression data 
```{r}
load("~/F2/output/vstMat.f2.centered.scaled.Rdata")
F2_geno_data_2_Ae_Ol_new <- read.table("~/F2/data/F2_geno_data_2_Ae_Ol_new.txt")
dim(F2_geno_data_2_Ae_Ol_new) # 3443 166 
dim(vstMat.f2.centered.scaled) # 56180   166

colnames(F2_geno_data_2_Ae_Ol_new) <- gsub("X", "ID_", colnames(F2_geno_data_2_Ae_Ol_new))
F2_geno_data_2_Ae_Ol_new.t <- as.data.frame(t(F2_geno_data_2_Ae_Ol_new))
F2_geno_data_2_Ae_Ol_new.t$ID <- rownames(F2_geno_data_2_Ae_Ol_new.t)

colnames(vstMat.f2.centered.scaled) <- gsub("Sample_F2", "ID", colnames(vstMat.f2.centered.scaled))
vstMat.f2.centered.scaled.t <- as.data.frame(t(vstMat.f2.centered.scaled))
vstMat.f2.centered.scaled.t$ID <- rownames(vstMat.f2.centered.scaled.t)

F2_geno_data_2_Ae_Ol_new.2 <- 
F2_geno_data_2_Ae_Ol_new.t %>% 
  left_join(vstMat.f2.centered.scaled.t) %>% 
  dplyr::select(-ID) 

dim(F2_geno_data_2_Ae_Ol_new.2) # 166 59623 
F2_geno_data_2_Ae_Ol_new.2 <- t(F2_geno_data_2_Ae_Ol_new.2)
write.table(F2_geno_data_2_Ae_Ol_new.2, file = "~/F2/data/QTL_analysis/F2_geno_data_2_Ae_Ol_new.2_all_expressed_genes.txt") 
```

```{r} 
# change file format in linux 
# cat F2_geno_data_2_Ae_Ol_new.2_all_expressed_genes.txt | sed 's/"//g' | awk '{first = $1; $1 = ""; print $0}' > tmp 
# tail -59623 tmp > tmp.1

write.table(rownames(F2_geno_data_2_Ae_Ol_new.2), file = "~/F2/data/QTL_analysis/marker_info_all_expressed_genes.txt" )
# cat marker_info_all_expressed_genes.txt | awk '{print "*"$2}' | tail -59623 | sed 's/"//g' > marker_info_reform.txt
# paste marker_info_reform.txt tmp.1  | tr "\t" "\ " > F2_geno_for_one_map.txt 
# cat header_one_map_input F2_geno_for_one_map.txt >  F2_geno_for_one_map_final_all_expressed_gene.txt 
# change header info: maker number to the right marker number & phenotype data number 
```

### load data 
```{r}
LG.f2 <- read.cross("mm", file = "~/F2/data/QTL_analysis/F2_geno_for_one_map_final_all_expressed_gene.txt", mapfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.map") ### takes one day to read in this data... 
save(LG.f2, file = "~/F2/output/QTL_analysis/LG.f2.all_expressed_genes.Rdata")
important_eQTL <- LG.f2 
important_eQTL
LG.f2
load("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/QTL_analysis/LG.f2.fatty_acid_lipid.Rdata")
LG.f2$pheno %>% dim()
```

### re-estimate map
```{r} 
LG.f2

LG.f2.before.crossover <- LG.f2

for (chr in names(LG.f2$geno)) { # for each chromosome in cross genotype data
  my.chr <- get(chr,LG.f2$geno) # return the genotype data, including data & map
  print(paste(chr,"NA before",sum(is.na(my.chr$data)))) 
  if(ncol(my.chr$data) > 3) { 
    my.chr$data[,2:(ncol(my.chr$data)-1)] <- sapply(2:(ncol(my.chr$data)-1),function(i) {
      apply(my.chr$data[,(i-1):(i+1)],1,function(gt) {
        if (any(is.na(gt))) return(gt[2]) #technically should be looking at the next genotyped marker.
        if ( (length(unique(gt)) == 2) & (gt[1] == gt[3])) return(NA)
        if ( length(unique(gt))  == 3) return(NA)
        return(gt[2])
      })
    })
  }
  LG.f2$geno <- within(LG.f2$geno,assign(chr,my.chr))
  print(paste(chr,"NA after",sum(is.na(get(chr,LG.f2$geno)$data))))
}  

map.new <- est.map(LG.f2,verbose=T,error.prob=.01) 

LG.f2 <- replace.map(LG.f2, map.new)
LG.f2.after.crossover <- LG.f2

plot.map(LG.f2.after.crossover, alternate.chrid = T) # the old genetic map
summaryMap(LG.f2.after.crossover) # 2853 

# fix LG10 problem using ripple 
set.seed(16)
LG.f2.after.crossover <- orderMarkers(LG.f2.after.crossover, chr = c(10), 
	                        error.prob = 0.0001, verbose = T) 

plotMap(LG.f2.after.crossover, chr = '10')
summaryMap(LG.f2.after.crossover) # 2884
save(LG.f2.after.crossover, file = "~/F2/output/QTL_analysis/LG.f2.after.crossover_all_expressed_genes.Rdata") 
```     

### eQTL analysis on cabernet 
1) running eQTL takes a long time, so I split the 50,000 genes into 10 subset, and run scanone on subset in parallel. 
https://github.com/leejimmy93/KIAT_cabernet/blob/master/F2/scanone_all_expressed_genes_1.R ... https://github.com/leejimmy93/KIAT_cabernet/blob/master/F2/scanone_all_expressed_genes_10.R
https://github.com/leejimmy93/KIAT_cabernet/blob/master/F2/scanone_all_expressed_genes_1_10.slurm 

2) For permutation, since it takes too long time to run permutation on all genes, so I sampled 100 genes and ran 1000 times permutation on these genes so that median can be taken as the threshold for eQTL analysis 
https://github.com/leejimmy93/KIAT_cabernet/blob/master/F2/scanone_all_expressed_genes_perm.R 

3) transfer to whitney for analysis 

### eQTL result analysis 
```{r}
# firstly check the two FAD genes 
# load data (scanone result)
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/data/eQTL_analysis/"
files <- list.files(path=path)
setwd(path)
eQTL_result <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 
save(eQTL_result, file = "~/F2/data/eQTL_analysis/eQTL_result.Rdata")
# for the result, 1st 10 are eQTL analysis result, 11 is the permuation result for 100 sampled genes.
```

### determine cis- and trans- eQTL 
```{r}
load("~/F2/data/eQTL_analysis/eQTL_result.Rdata")

# get threshold 
lod.thres <- median(unlist(eQTL_result[[11]])) # 95% percentile as the threshold  
lod.thres # [1] 4.159691 

###### a bunch of test 
plot(eQTL_result[[3]]$BnaA08g12780D,bandcol="gray90", main="BnaA08g12780D") # looks good 
plot(eQTL_result[[9]]$BnaA08g11140D,bandcol="gray90", main="BnaA08g11140D") # looks good
plot(eQTL_result[[9]]$BnaA08g11130D,bandcol="gray90", main="BnaA08g11130D") # as expected 
abline(h=lod.thres,lty=2) 
plot(eQTL_result[[7]]$BnaC03g65980D,bandcol="gray90", main="BnaC03g65980D") # as expected
abline(h=lod.thres,lty=2) 

bayesint(eQTL_result[[3]]$BnaA08g12780D, chr = 8, expandtomarkers = T) 
bayesint(scanone.imp$Erucic_acid, chr = "A08", expandtomarkers = T)

bayesint(eQTL_result[[9]]$BnaA08g11140D, chr = 8, expandtomarkers = T) 
bayesint(scanone.imp$Erucic_acid, chr = "A08", expandtomarkers = T)

# below is the list of fatty acid genes in the interval which are differentially expressed between Da-Ae and Da-Ol-1 
"BnaA08g11810D" %in% cis_trans_result_final$gene_ID # F
"BnaA08g12800D" %in% cis_trans_result_final$gene_ID
"BnaA08g13200D" %in% cis_trans_result_final$gene_ID
"BnaA08g14550D" %in% cis_trans_result_final$gene_ID
"BnaA08g12780D" %in% cis_trans_result_final$gene_ID
"BnaA08g11130D" %in% cis_trans_result_final$gene_ID
"BnaC03g67820D" %in% cis_trans_result_final$gene_ID 

####### abunch of test before analysis 

# get all eQTL based on this threshold 
system.time(
tmp <- 
lapply(names(eQTL_result[1:10]), function(subset) {
  print(subset)
  lapply(names(eQTL_result[[subset]]), function(gene) {
    tmp1 <- eQTL_result[[subset]][[gene]][eQTL_result[[subset]][[gene]]$lod > lod.thres,]
    if (nrow(tmp1) > 0){
    tmp1$gene_ID <- gene
    tmp1$snp_ID <- rownames(tmp1)
    tmp1
  }
})
}) 
) # 20 sec  

# test <- do.call("rbind", tmp) 
# test2 <- do.call("rbind", test)

####### re do # both the above two line code the code below work, but why the above works
test <- list()
for (subset in seq_along(tmp)){
  test[[subset]] <- do.call("rbind", tmp[[subset]])
}

test2 <- do.call("rbind", test)
############# 

# extract only SNP info 
cis_trans_result <- test2[grepl("^chr", test2$snp_ID),]
dim(cis_trans_result) # 728717      5 
cis_trans_result %>% head()

unique(cis_trans_result$snp_ID) %>% length # 2013 
unique(cis_trans_result$gene_ID) %>% length() # 21823 

# get genome range
library(IRanges)
library(GenomicRanges)
library(GenomicFeatures)
library("rtracklayer")

### get gff file with gene chrom & pos info, gff3 file must be sorted 
gff.mRNA <- read.table("~/Reference/B.napus/gff.mRNA")
dim(gff.mRNA) # 101040      4 
head(gff.mRNA) 
colnames(gff.mRNA) <- c("CHROM", "start", "end", "gene_ID") 
gff.mRNA %>% head()

# merge two files
cis_trans_result_2 <- 
cis_trans_result %>% 
  left_join(gff.mRNA, by = "gene_ID") %>% 
  mutate(gene_chrom = CHROM, gene_start = start, gene_end = end) %>% 
  dplyr::select(-(CHROM:end)) %>% 
  separate(snp_ID, into = c("snp_chrom", "snp_pos"), sep = "_", remove = F) %>% 
  mutate(snp_pos = as.numeric(snp_pos)) %>% 
  mutate(LG = chr, genetic_pos = as.numeric(pos)) %>% 
  dplyr::select(-(chr:pos))

# get snp & chrom ID & pos, as well as gene  
cis_trans_result_2 %>% head()

# determine cis- and trans- eQTL only on chromosome 
# e-traits (genes) which are not on the same chrom as SNPs are trans- 
cis_trans_result_2$cis_trans <- ifelse(cis_trans_result_2$gene_chrom == cis_trans_result_2$snp_chrom, "cis_candidate", "trans")
sum(cis_trans_result_2$cis_trans == "cis_candidate") # 492708 
sum(cis_trans_result_2$cis_trans == "trans") # 236009 
dim(cis_trans_result_2) # 728717     11 

length(unique(cis_trans_result_2[cis_trans_result_2$cis_trans == "cis_candidate",]$gene_ID)) # 11118 
length(unique(cis_trans_result_2[cis_trans_result_2$cis_trans == "trans",]$gene_ID)) # 13295   
length(unique(cis_trans_result_2$gene_ID)) # 21823 

# use bayesint to find interval estimate of each gene --> for cis-eQTL 
# for each cis-candidate gene --> if there are markers go above the threshold --> check to see if there are markers on the same chromsome as the gene --> if yes, do bayesint on the QTL result of that gene --> and extract the marker result rownames --> use the min as the start of the sig loci, and max as the end of the sig loci --> 
cis_candidate <- unique(cis_trans_result_2[cis_trans_result_2$cis_trans == "cis_candidate",]$gene_ID)
length(cis_candidate) # 11118 

cis_candidate %>% head()
sum(cis_candidate %in% names(eQTL_result[[1]]), 
    cis_candidate %in% names(eQTL_result[[2]]),
    cis_candidate %in% names(eQTL_result[[3]]),
    cis_candidate %in% names(eQTL_result[[4]]),
    cis_candidate %in% names(eQTL_result[[5]]),
    cis_candidate %in% names(eQTL_result[[6]]),
    cis_candidate %in% names(eQTL_result[[7]]),
    cis_candidate %in% names(eQTL_result[[8]]),
    cis_candidate %in% names(eQTL_result[[9]]),
    cis_candidate %in% names(eQTL_result[[10]])) # 11118, correct 

system.time(
tmp.2 <- 
lapply(names(eQTL_result[1:10]), function(subset) {
  print(subset)
  lapply(names(eQTL_result[[subset]])[names(eQTL_result[[subset]]) %in% cis_candidate], function(gene) {
    sub = gsub("(Bna)(A|C)(01|02|03|04|05|06|07|08|09|10)([[:print:]]+)", "\\2", gene) 
    chrom = gsub("(Bna)(A|C)(01|02|03|04|05|06|07|08|09|10)([[:print:]]+)", "\\3", gene)
    chrom <- ifelse(sub == "A", as.numeric(as.character(chrom)), 10 + as.numeric(as.character(chrom)))
    tmp1 <- bayesint(eQTL_result[[subset]][[gene]], chr = chrom, expandtomarkers = T)
    tmp1 <- tmp1[grepl("^chr", rownames(tmp1)),] 
    tmp1$gene_ID <- gene
    tmp1$pos_2 <- gsub("([[:print:]]+)(_)([[:print:]])", "\\3", rownames(tmp1))
    tmp1$loci_start <- min(as.numeric(tmp1$pos_2))
    tmp1$loci_end <- max(as.numeric(tmp1$pos_2)) 
    tmp1$loci_start_genetic <- min(as.numeric(tmp1$pos))
    tmp1$loci_end_genetic <- max(as.numeric(tmp1$pos))
    tmp1[1,c("gene_ID", "loci_start", "loci_end", "loci_start_genetic", "loci_end_genetic")] # the last item is the one that was returned
    })
}
)
) # 20 sec  

# combine result 
test3 <- list()
for (subset in seq_along(tmp.2)){
  test3[[subset]] <- do.call("rbind", tmp.2[[subset]]) 
}

test4 <- do.call("rbind", test3)
dim(test4) # 11118 3 
head(test4)

sum(test4$loci_start < test4$loci_end) # 11087
sum(test4$loci_start == test4$loci_end) # 31

cis_candidate_2 <- cis_trans_result_2[cis_trans_result_2$cis_trans == "cis_candidate",]
dim(cis_candidate_2) # 492708     11 
cis_candidate_2 %>% head()

cis_candidate_3 <- 
cis_candidate_2 %>% 
  left_join(test4, by = "gene_ID")

cis_candidate_3 %>% head()
dim(cis_candidate_3) # 492708     13 

sum(cis_candidate_3$gene_start < cis_candidate_3$gene_end) # 492708
sum(cis_candidate_3$loci_start < cis_candidate_3$loci_end) # 490133
sum(cis_candidate_3$loci_start == cis_candidate_3$loci_end) # 2575 


# use loci, gene start/end, and marker genetic position to find cis-eQTL 
cis_candidate_4 <- 
cis_candidate_3 %>% 
  filter((gene_start < loci_start & gene_end > loci_start & snp_pos >= loci_start & snp_pos <= loci_end) |
         (gene_start > loci_start & gene_end < loci_end & snp_pos >= loci_start & snp_pos <= loci_end) |
         (gene_start < loci_end & gene_end > loci_end & snp_pos >= loci_start & snp_pos <= loci_end)) 

gff.cis_eQTL_candidate_gene <- GRanges(seqnames = Rle(cis_candidate_3$gene_chrom),ranges = IRanges(start = cis_candidate_3$gene_start, end = cis_candidate_3$gene_end), names = cis_candidate_3$gene_ID)
gff.cis_eQTL_candidate_loci <- GRanges(seqnames = Rle(cis_candidate_3$snp_chrom),ranges = IRanges(start = cis_candidate_3$loci_start, end = cis_candidate_3$loci_end))



dim(cis_candidate_4)  # 88407    13 
"BnaA08g12780D" %in% cis_candidate_4$gene_ID # good
"BnaA08g11130D" %in% cis_candidate_4$gene_ID # as expected 
"BnaA08g11140D" %in% cis_candidate_4$gene_ID # good 

unique(cis_candidate_4$gene_ID) %>% length() # 8110 cis-genes

# get the final set of cis- and trans- eQTL, 
cis_candidate_4 <- 
cis_candidate_4 %>% 
  mutate(cis_trans_2 = "cis") 

colnames(cis_candidate_4) 
colnames(cis_trans_result_2) 

cis_trans_result_3 <- 
cis_candidate_4 %>% 
  full_join(cis_trans_result_2) %>% 
  dplyr::select(-(cis_trans))

dim(cis_trans_result_3) # 728717     15 

cis_trans_result_3[is.na(cis_trans_result_3$cis_trans_2),][,"cis_trans_2"] <- rep("trans", sum(is.na(cis_trans_result_3$cis_trans_2)))
sum(cis_trans_result_3$cis_trans_2 == "cis") # 88407 
sum(cis_trans_result_3$cis_trans_2 == "trans") # 640310

length(unique(cis_trans_result_3[cis_trans_result_3$cis_trans_2 == "cis",]$gene_ID)) # 8110
length(unique(cis_trans_result_3[cis_trans_result_3$cis_trans_2 == "trans",]$gene_ID)) # 20702 

# get loci_start and loci_end for genes with trans-eQTL 
trans_candidate <- cis_trans_result_3[cis_trans_result_3$cis_trans_2 == "trans",] %>% dplyr::select(gene_ID, LG) 
trans_candidate <- unique(trans_candidate) # 23814 2

trans_candidate %>% head()
sum(trans_candidate$gene_ID %in% names(eQTL_result[[1]]), 
    trans_candidate$gene_ID %in% names(eQTL_result[[2]]),
    trans_candidate$gene_ID %in% names(eQTL_result[[3]]),
    trans_candidate$gene_ID %in% names(eQTL_result[[4]]),
    trans_candidate$gene_ID %in% names(eQTL_result[[5]]),
    trans_candidate$gene_ID %in% names(eQTL_result[[6]]),
    trans_candidate$gene_ID %in% names(eQTL_result[[7]]),
    trans_candidate$gene_ID %in% names(eQTL_result[[8]]),
    trans_candidate$gene_ID %in% names(eQTL_result[[9]]),
    trans_candidate$gene_ID %in% names(eQTL_result[[10]])) # 23814, correct 

trans_candidate_list <- list(trans_candidate[trans_candidate$gene_ID %in% names(eQTL_result[[1]]),],
                             trans_candidate[trans_candidate$gene_ID %in% names(eQTL_result[[2]]),],
                             trans_candidate[trans_candidate$gene_ID %in% names(eQTL_result[[3]]),],
                             trans_candidate[trans_candidate$gene_ID %in% names(eQTL_result[[4]]),],
                             trans_candidate[trans_candidate$gene_ID %in% names(eQTL_result[[5]]),],
                             trans_candidate[trans_candidate$gene_ID %in% names(eQTL_result[[6]]),],
                             trans_candidate[trans_candidate$gene_ID %in% names(eQTL_result[[7]]),],
                             trans_candidate[trans_candidate$gene_ID %in% names(eQTL_result[[8]]),],
                             trans_candidate[trans_candidate$gene_ID %in% names(eQTL_result[[9]]),],
                             trans_candidate[trans_candidate$gene_ID %in% names(eQTL_result[[10]]),])

system.time(  
tmp.2 <- 
lapply(seq_along(trans_candidate_list), function(subset) { 
  print(subset)
  lapply(seq_along(trans_candidate_list[[subset]]$gene_ID), function(index) {
    gene <- trans_candidate_list[[subset]]$gene_ID[index] 
    chrom <- as.numeric(as.character(trans_candidate_list[[subset]]$LG[index])) 
    tmp1 <- bayesint(eQTL_result[[subset]][[gene]], chr = chrom, expandtomarkers = T)
    tmp1 <- tmp1[grepl("^chr", rownames(tmp1)),] 
    tmp1$gene_ID <- gene
    tmp1$LG <- chrom 
    tmp1$pos_2 <- gsub("([[:print:]]+)(_)([[:print:]])", "\\3", rownames(tmp1))
    tmp1$loci_start_2 <- min(as.numeric(tmp1$pos_2))
    tmp1$loci_end_2 <- max(as.numeric(tmp1$pos_2)) 
    tmp1$loci_start_genetic_2 <- min(as.numeric(tmp1$pos))
    tmp1$loci_end_genetic_2 <- max(as.numeric(tmp1$pos))
    tmp1[1,c("gene_ID", "LG", "loci_start_2", "loci_end_2", "loci_start_genetic_2", "loci_end_genetic_2")] # the last item is the one that was returned
    })
}
)
) # 52 sec   

test3 <- list()
for (subset in seq_along(tmp.2)){
  test3[[subset]] <- do.call("rbind", tmp.2[[subset]]) 
}

test4 <- do.call("rbind", test3)
dim(test4) # 23814 6
head(test4)

sum(test4$loci_start_2 < test4$loci_end_2) # 23770 
sum(test4$loci_start_2 == test4$loci_end_2) # 44

trans_candidate_2 <- cis_trans_result_3[cis_trans_result_3$cis_trans_2 == "trans",]
dim(trans_candidate_2) # 640310     13
trans_candidate_2 %>% head()

trans_candidate_2$LG <- as.numeric(as.character(trans_candidate_2$LG))

trans_candidate_3 <- 
trans_candidate_2 %>% 
  left_join(test4) %>%
  mutate(loci_start = loci_start_2, loci_end = loci_end_2, loci_start_genetic = loci_start_genetic_2, loci_end_genetic = loci_end_genetic_2) %>% 
  dplyr::select(-(loci_start_2:loci_end_genetic_2))

trans_candidate_3 %>% head()
dim(trans_candidate_3) # 640310     15

sum(trans_candidate_3$gene_start < trans_candidate_3$gene_end) # 640310
sum(trans_candidate_3$loci_start < trans_candidate_3$loci_end) # 637051
sum(trans_candidate_3$loci_start == trans_candidate_3$loci_end) # 3259  

trans_candidate_3 %>% dim() 
cis_candidate_4 %>% dim() 

cis_trans_result_final <- 
cis_candidate_4 %>% 
  dplyr::select(-cis_trans) %>% 
  rbind(trans_candidate_3)  

####### also delete snps outside these ranges (in order to be consistent with cis-eQTL analysis), or this is unnecessary for now.. leave for later if needed... 
trans_candidate_3 %>% 
  filter(snp_chrom == "chrA09" & LG == 18) # need to delete these genes maybe, this is caused by the anchor of A09 marker to LG18 during genetic map construction   


# save result for next step 
save(cis_trans_result_final, file = "~/F2/output/eQTL/cis_trans_result_final.Rdata") 
``` 

### make a plot of cis- and trans- eQTL, x-axis being eQTL physical position, y-axis being transcript start position 
```{r}
load("~/F2/output/eQTL/cis_trans_result_final.Rdata")

dim(cis_trans_result_final) # 728717     15 
cis_trans_result_final %>% head() # give up for now... 

cis_trans_result_final_2 <- 
  cis_trans_result_final %>% 
  dplyr::select(gene_ID, snp_chrom, gene_chrom, gene_start, LG, loci_start_genetic, loci_end_genetic, cis_trans_2) %>% 
  unique() 

dim(cis_trans_result_final_2) # 33441     8 
cis_trans_result_final_2[cis_trans_result_final_2$cis_trans_2 == "cis",] %>% dim() # 8110
cis_trans_result_final_2[cis_trans_result_final_2$cis_trans_2 == "trans",]$gene_ID %>% unique() %>% length() # 20702

cis_eQTL_final <- cis_trans_result_final_2[cis_trans_result_final_2$cis_trans_2 == "cis",] 
save(cis_eQTL_final, file = "~/F2/output/eQTL/cis_eQTL_final.Rdata")
cis_eQTL_final %>% head()

cis_trans_result_final_2[cis_trans_result_final_2$cis_trans_2 == "trans",][,c("loci_start_genetic", "LG", "gene_ID")] %>% dim() # 25331     3  
cis_trans_result_final_2[cis_trans_result_final_2$cis_trans_2 == "trans",][,c("loci_start_genetic", "LG", "gene_ID")] %>% unique() %>% dim() # 23814     3  
cis_trans_result_final_2[cis_trans_result_final_2$cis_trans_2 == "cis",][,c("loci_start_genetic", "LG", "gene_ID")] %>% dim() # 8110    3 

View(cis_trans_result_final_2)

cis_trans_result_final_2 %>% 
  filter(snp_chrom == "chrA06" & gene_chrom == "chrA06") %>% 
  ggplot() + 
  geom_point(mapping = aes(x=loci_start_genetic, y=gene_start)) 
# problem with this is that some LG I have reversed direction. 

########## trans-eQTL hotspot 
cis_trans_result_final_3 <- 
  cis_trans_result_final %>% 
  dplyr::select(gene_ID, snp_chrom, gene_chrom, gene_start, LG, loci_start, loci_end, loci_start_genetic, loci_end_genetic, cis_trans_2) %>% 
  unique()  

cis_trans_result_final_3[cis_trans_result_final_3$cis_trans_2 == "trans",][,c("loci_start", "LG", "gene_ID")] %>% dim() # 25331     3  

trans_candidate_4 <- 
cis_trans_result_final_3[cis_trans_result_final_3$cis_trans_2 == "trans",][,c("loci_start", "LG", "gene_ID", "snp_chrom", "loci_start_genetic", "loci_end_genetic")] %>% unique()  # 23814     3 

trans_candidate_4$LG != trans_candidate_4$snp_chrom
trans_candidate_4 %>% head() 

trans_candidate_4$sub <- gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\2", trans_candidate_4$snp_chrom) 
trans_candidate_4$chrom <- gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\3", trans_candidate_4$snp_chrom)
trans_candidate_4$chrom <- ifelse(trans_candidate_4$sub == "A", as.numeric(as.character(trans_candidate_4$chrom)), 10 + as.numeric(as.character(trans_candidate_4$chrom)))

sum(as.numeric(as.character(trans_candidate_4$LG)) != trans_candidate_4$chrom) # problem... ignore for now, come back later? no, delete these now... 

trans_candidate_5 <- 
trans_candidate_4[!(as.numeric(as.character(trans_candidate_4$LG)) != trans_candidate_4$chrom),] 
trans_candidate_5 %>% head()
trans_candidate_5 %>% dim() # 23771 8  

trans_candidate_5$chrom2 <- gsub("(chr)(A|C)(01|02|03|04|05|06|07|08|09|10)", "\\3", trans_candidate_5$snp_chrom)
save(trans_candidate_5, file = "~/F2/output/eQTL/trans_candidate_5.Rdata") # still need to be refined... 

p.transeQTL <- 
  trans_candidate_5 %>% 
  ggplot() + 
  geom_histogram(aes(x = loci_start, fill = sub), binwidth = 1000000) + 
  facet_grid(chrom2 ~ sub) +
  labs(list(title = "", x = "trans-eQTL loci", y = "number of trans-eQTL per Mb")) +
  theme(legend.position = "none")

ggsave("~/F2/output/eQTL/figure/p.transeQTL.png", width = 9, height = 6)
```

### cis-eQTL overlap with trait-QTL --> cis-candidate genes 
```{r}
load("~/F2/output/QTL_analysis/cim.qtl.43traits.Rdata")
load("~/F2/output/QTL_analysis/cim.perm.43traits.Rdata")

# 1) get all the SNPs with lod score greater than the threshold 
system.time(
tmp <- 
lapply(names(cim.qtl), function(trait) {
  print(trait)
    tmp1 <- cim.qtl[[trait]][cim.qtl[[trait]]$lod > cim.perm[[trait]],]
    if (nrow(tmp1) > 0){
    tmp1$trait <- trait
    tmp1$snp_ID <- rownames(tmp1)
    tmp1
  }
})
)

test <- do.call("rbind", tmp)
test 
trait_qtl_result <- test

unique(trait_qtl_result$snp_ID) %>% length # 128 
unique(trait_qtl_result$trait) %>% length() # 17   

# 2) get loci interval info for all traits 
system.time(
tmp.2 <- 
  lapply(seq_along(trait_qtl_result$trait), function(index) {
    print(index)
    tmp1 <- bayesint(cim.qtl[[trait_qtl_result$trait[index]]], chr = trait_qtl_result$chr[index], expandtomarkers = T)
    # tmp1 <- tmp1[grepl("^chr", rownames(tmp1)),] 
    tmp1$trait <- trait_qtl_result$trait[index]
    tmp1$chr <- trait_qtl_result$chr[index] 
    # tmp1$pos_2 <- gsub("([[:print:]]+)(_)([[:print:]])", "\\3", rownames(tmp1))
    # tmp1$loci_start <- min(as.numeric(tmp1$pos_2))
    # tmp1$loci_end <- max(as.numeric(tmp1$pos_2)) 
    tmp1$loci_start_genetic <- min(as.numeric(tmp1$pos))
    tmp1$loci_end_genetic <- max(as.numeric(tmp1$pos))
    tmp1[1,c("trait", "chr", "loci_start_genetic", "loci_end_genetic")] # the last item is the one that was returned
    })
)

tmp.2[[14]] 

# 3) finalize  
test <- as.tibble(do.call("rbind", tmp.2))
trait_qtl_result.final <- unique(test) 
trait_qtl_result.final
save(trait_qtl_result.final, file = "~/F2/output/eQTL/trait_qtl_result.final.Rdata")

# 4) look for overlap between cis-eQTL and trait-eQTL 
load("~/F2/output/eQTL/cis_eQTL_final.Rdata")
cis_eQTL_final %>% head()

trait_qtl_result.final[1,]
gff.trait_qtl_result.final <- GRanges(seqnames = Rle(paste("chr", trait_qtl_result.final$chr, sep = "")),ranges = IRanges(start = trait_qtl_result.final$loci_start_genetic, end = trait_qtl_result.final$loci_end_genetic), trait = trait_qtl_result.final$trait)   

gff.cis_eQTL_final <- GRanges(seqnames = Rle(cis_eQTL_final$snp_chrom),ranges = IRanges(start = cis_eQTL_final$loci_start_genetic, end = cis_eQTL_final$loci_end_genetic), names = cis_eQTL_final$gene_ID) 

gff.cis_eQTL_final

cis_eQTL_trait <- mergeByOverlaps(gff.trait_qtl_result.final, gff.cis_eQTL_final) 
cis_eQTL_trait <- cis_eQTL_trait %>% as.data.frame()  

napus_vs_ara.non_reciprocal <- read.table("~/Reference/B.napus/napus_vs_ara.non_reciprocal.table")
head(napus_vs_ara.non_reciprocal)
dim(napus_vs_ara.non_reciprocal) # 64949 

colnames(napus_vs_ara.non_reciprocal)[1] <- "names"

cis_eQTL_trait.2 <- merge(cis_eQTL_trait, napus_vs_ara.non_reciprocal, by = "names", all.x=T)
cis_eQTL_trait.2 %>% colnames()

cis_eQTL_trait.final <- 
  cis_eQTL_trait.2 %>% 
  mutate(gene_ID = names, trait_qtl_chr = gff.trait_qtl_result.final.seqnames, trait_loci_start = gff.trait_qtl_result.final.start, trait_loci_end = gff.trait_qtl_result.final.end, cis_eQTL_chr = gff.cis_eQTL_final.seqnames, cis_eQTL_start = gff.cis_eQTL_final.start, cis_eQTL_end = gff.cis_eQTL_final.end) %>% 
  dplyr::select(trait, gene_ID, trait_qtl_chr, trait_loci_start, trait_loci_end, cis_eQTL_chr, cis_eQTL_start, cis_eQTL_end, ara_ID, description)

dim(cis_eQTL_trait.final) # 2369 10 
# View(cis_eQTL_trait.final)

View(cis_eQTL_trait.final[cis_eQTL_trait.final$trait == "Erucic_acid",])

# add GO term
B.napus.GO <- read.delim(file="~/Reference/B.napus/Brassica_napus_GO", header=F, sep=" ")
dim(B.napus.GO) # 150320      2 # GO annotation
colnames(B.napus.GO) <- c("gene_ID", "GO") # better to have GO description associated with ... 

tmp <- cis_eQTL_trait.final %>% 
  left_join(B.napus.GO, "gene_ID")

dim(tmp) # 5317 
View(tmp[tmp$trait == "Erucic_acid",])

GO_annotation <- read_delim("~/Reference/B.napus/GO_annotation.csv", delim = "\"")

GO_annotation <- 
GO_annotation %>% 
  dplyr::select(input, name)

dim(GO_annotation) # 1659 2 

colnames(GO_annotation) <- c("GO", "GO_des")

# add GO annotation 
tmp.2 <- tmp %>% 
  left_join(GO_annotation, "GO")

tmp.2 %>% dim() # 5317 12 
View(tmp.2[tmp.2$trait == "Erucic_acid",]) 

cis_eQTL_trait.final.final <- tmp.2 
save(cis_eQTL_trait.final.final, file = "~/F2/output/eQTL/cis_eQTL_trait.final.final.Rdata")
View(cis_eQTL_trait.final.final[grep("lipid|fatty", cis_eQTL_trait.final.final$GO_des),]) 
```

### plot out cis-eQTL result 
```{r}
load("~/F2/output/QTL_analysis/cim.qtl.43traits.Rdata")
# BnaA08g11060D & BnaA08g11140D two cis-eQTL gene for fatty acid 

"BnaA08g11060D" %in% names(eQTL_result[[9]]) 
plot(eQTL_result[[9]]$BnaA08g11140D,bandcol="gray90", main="BnaA08g11140D") # looks good
 
cim.qtl[["BnaA08g11140D"]] <- eQTL_result[[9]]$BnaA08g11140D
cim.qtl[["BnaA08g11060D"]] <- eQTL_result[[9]]$BnaA08g11060D

cim.qtl[["BnaA08g11140D"]]$chr <- ifelse(cim.qtl[["BnaA08g11140D"]]$chr == "8", "A08", cim.qtl[["BnaA08g11140D"]]$chr)
cim.qtl[["BnaA08g11140D"]]$chr <- ifelse(cim.qtl[["BnaA08g11140D"]]$chr == "13", "C03", cim.qtl[["BnaA08g11140D"]]$chr)

cim.qtl[["BnaA08g11060D"]]$chr <- ifelse(cim.qtl[["BnaA08g11060D"]]$chr == "8", "A08", cim.qtl[["BnaA08g11060D"]]$chr)
cim.qtl[["BnaA08g11060D"]]$chr <- ifelse(cim.qtl[["BnaA08g11060D"]]$chr == "13", "C03", cim.qtl[["BnaA08g11060D"]]$chr)

# oil related  
oil_cim_eQTL_1 <- 
qtl_plot(input = rbind(data.frame(cim.qtl[["Palmitic_acid"]], method = "Palmitic_acid"), 
                       data.frame(cim.qtl[["Stearic_acid"]], method = "Stearic_acid"), 
                       data.frame(cim.qtl[["Oleic_acid"]], method = "Oleic_acid"),
                       data.frame(cim.qtl[["vaccenic_acid"]], method = "vaccenic_acid"), 
                       data.frame(cim.qtl[["Linoleic_acid"]], method = "Linoleic_acid"),
                       data.frame(cim.qtl[["Arachidic_acid"]], method = "Arachidic_acid"),
                       data.frame(cim.qtl[["Linolenic_acid"]], method = "Linolenic_acid"),
                       data.frame(cim.qtl[["Erucic_acid"]], method = "Erucic_acid"),
                       data.frame(cim.qtl[["BnaA08g11140D"]], method = "BnaA08g11140D")
                       ),
         chrs = c("A08", "C03"), 
         # lod = cim.perm[[1]], 
         title = "BnaA08g11140D cis-eQTL",
         rug = TRUE)  

oil_cim_eQTL_2 <- 
qtl_plot(input = rbind(data.frame(cim.qtl[["Palmitic_acid"]], method = "Palmitic_acid"), 
                       data.frame(cim.qtl[["Stearic_acid"]], method = "Stearic_acid"), 
                       data.frame(cim.qtl[["Oleic_acid"]], method = "Oleic_acid"),
                       data.frame(cim.qtl[["vaccenic_acid"]], method = "vaccenic_acid"), 
                       data.frame(cim.qtl[["Linoleic_acid"]], method = "Linoleic_acid"),
                       data.frame(cim.qtl[["Arachidic_acid"]], method = "Arachidic_acid"),
                       data.frame(cim.qtl[["Linolenic_acid"]], method = "Linolenic_acid"),
                       data.frame(cim.qtl[["Erucic_acid"]], method = "Erucic_acid"),
                       data.frame(cim.qtl[["BnaA08g11060D"]], method = "BnaA08g11060D")
                       ),
         chrs = c("A08", "C03"), 
         # lod = cim.perm[[1]], 
         title = "BnaA08g11060D cis-eQTL",
         rug = TRUE)  

ggsave(oil_cim_eQTL_1$plot, filename = "~/F2/output/eQTL/figure/oil_cim_eQTL_1.png", width = 8, height = 5) 
ggsave(oil_cim_eQTL_2$plot, filename = "~/F2/output/eQTL/figure/oil_cim_eQTL_2.png", width = 8, height = 5)  
```

### trans-eQTL overlap with trait-QTL --> trans-candidate genes  
```{r}
load("~/F2/output/eQTL/trans_candidate_5.Rdata")
load("~/F2/output/eQTL/trait_qtl_result.final.Rdata")

# look for overlap between cis-eQTL and trait-eQTL 
trans_candidate_5 %>% head()

trait_qtl_result.final[1,]
gff.trait_qtl_result.final <- GRanges(seqnames = Rle(paste("chr", trait_qtl_result.final$chr, sep = "")),ranges = IRanges(start = trait_qtl_result.final$loci_start_genetic, end = trait_qtl_result.final$loci_end_genetic), trait = trait_qtl_result.final$trait)   

gff.trans_candidate_5 <- GRanges(seqnames = Rle(trans_candidate_5$snp_chrom),ranges = IRanges(start = trans_candidate_5$loci_start_genetic, end = trans_candidate_5$loci_end_genetic), names = trans_candidate_5$gene_ID) 

gff.trans_candidate_5

trans_eQTL_trait <- mergeByOverlaps(gff.trait_qtl_result.final, gff.trans_candidate_5) 
trans_eQTL_trait <- trans_eQTL_trait %>% as.data.frame() 

napus_vs_ara.non_reciprocal <- read.table("~/Reference/B.napus/napus_vs_ara.non_reciprocal.table")
head(napus_vs_ara.non_reciprocal)
dim(napus_vs_ara.non_reciprocal) # 64949 

colnames(napus_vs_ara.non_reciprocal)[1] <- "names"

trans_eQTL_trait.2 <- merge(trans_eQTL_trait, napus_vs_ara.non_reciprocal, by = "names", all.x=T)
trans_eQTL_trait.2 %>% colnames()

trans_eQTL_trait.final <- 
  trans_eQTL_trait.2 %>% 
  mutate(gene_ID = names, trait_qtl_chr = gff.trait_qtl_result.final.seqnames, trait_loci_start = gff.trait_qtl_result.final.start, trait_loci_end = gff.trait_qtl_result.final.end, trans_eQTL_chr = gff.trans_candidate_5.seqnames, trans_eQTL_start = gff.trans_candidate_5.start, trans_eQTL_end = gff.trans_candidate_5.end) %>% 
  dplyr::select(trait, gene_ID, trait_qtl_chr, trait_loci_start, trait_loci_end, trans_eQTL_chr, trans_eQTL_start, trans_eQTL_end, ara_ID, description)

dim(trans_eQTL_trait.final) # 7327 10 

View(trans_eQTL_trait.final[trans_eQTL_trait.final$trait == "Erucic_acid",])

# add GO term
B.napus.GO <- read.delim(file="~/Reference/B.napus/Brassica_napus_GO", header=F, sep=" ")
dim(B.napus.GO) # 150320      2 # GO annotation
colnames(B.napus.GO) <- c("gene_ID", "GO") # better to have GO description associated with ... 

tmp <- trans_eQTL_trait.final %>% 
  left_join(B.napus.GO, "gene_ID")

dim(tmp) # 16822 

GO_annotation <- read_delim("~/Reference/B.napus/GO_annotation.csv", delim = "\"")

GO_annotation <- 
GO_annotation %>% 
  dplyr::select(input, name)

dim(GO_annotation) # 1659 2 

colnames(GO_annotation) <- c("GO", "GO_des")

# add GO annotation 
tmp.2 <- tmp %>% 
  left_join(GO_annotation, "GO")

tmp.2 %>% dim() # 16822 12 

###### due to some problem in cis- and trans- eQTL analysis, some loci are defined as both cis- and trans- eQTL for a gene, those loci with trans- signature should be deleted... go back to the original anaysis afterwards 

tmp.2 %>% head()
load("~/F2/output/eQTL/cis_eQTL_trait.final.final.Rdata")

colnames(cis_eQTL_trait.final.final)[c(6, 7, 8)] <- c("trans_eQTL_chr", "trans_eQTL_start", "trans_eQTL_end") 

trans_eQTL_trait.final.final <- 
tmp.2 %>% 
  anti_join(cis_eQTL_trait.final.final)  

save(trans_eQTL_trait.final.final, file = "~/F2/output/eQTL/trans_eQTL_trait.final.final.Rdata")
View(trans_eQTL_trait.final.final[grep("lipid|fatty", trans_eQTL_trait.final.final$GO_des),])   
```

### Go enrichment of trans-eQTL for the different acid traits 
```{r}
source("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/analysis/function_BnRNAseq.R")

load("~/Desktop/2017_Nov_KIAT_report/trans_eQTL_trait.final.final.Rdata")
trans_eQTL_trait.final.final %>% head()
trans_eQTL_trait.final.final.acid <- trans_eQTL_trait.final.final[grep("acid", trans_eQTL_trait.final.final$trait),]
dim(trans_eQTL_trait.final.final.acid) # 5164 12 

trans_eQTL_trait.final.final.acid <- 
trans_eQTL_trait.final.final.acid %>% 
  filter(trait != "Behenic_acid") %>% 
  filter(trait != "cis_11_Eicosenoic_acid") %>% 
  filter(trait != "Myristic_acid")

dim(trans_eQTL_trait.final.final.acid) # 4408 12 
write.table(trans_eQTL_trait.final.final.acid$GO_des, file = "~/Desktop/2017_Nov_KIAT_report/test.txt") 

GOseq.Bn.ORA(unique(trans_eQTL_trait.final.final.acid$gene_ID)) # no enriched go terms 
unique(trans_eQTL_trait.final.final.acid$gene_ID) %>% length() # 384 

colnames(trans_eQTL_trait.final.final.acid) 
trans_eQTL_trait.final.final.acid %>% 
  dplyr::select(gene_ID, ara_ID, description) %>% 
  unique() %>% 
  dim()  

# fatty acid related genes 
trans_eQTL_trait.final.final.acid[grep("lipid|fatty", trans_eQTL_trait.final.final.acid$GO_des),] %>% 
  dplyr::select(gene_ID, description, GO_des) %>% 
  unique() %>% 
  View()

test <- 
trans_eQTL_trait.final.final.acid[grep("lipid|fatty", trans_eQTL_trait.final.final.acid$GO_des),] %>% 
  dplyr::select(gene_ID, description, GO_des) %>% 
  unique() 

test[-grep("phospholipid", test$GO_des),] %>% View

  
# transcription factor genes 
TF.trans_eQTL.acid <- 
trans_eQTL_trait.final.final.acid[grep("binding", trans_eQTL_trait.final.final.acid$GO_des),] %>% 
  dplyr::select(gene_ID, description, GO_des) %>% 
  unique() 

TF.trans_eQTL.acid$gene_ID %>% unique() %>% length() 

test <- 
trans_eQTL_trait.final.final.acid[grep("DNA|nucleotide|transcription", trans_eQTL_trait.final.final.acid$GO_des),] %>% 
  dplyr::select(gene_ID, description, GO_des) %>% 
  unique() 

test[grep("binding", test$GO_des),] %>% 
  View() 
	
# DNA binding/nucleotide binding/transcription factor activity/ 

save(TF.trans_eQTL.acid, file = "~/Desktop/2017_Nov_KIAT_report/TF.trans_eQTL.acid.Rdata")
```












