---
title: "eQTL_batch_corrected"
output: html_document
---

Purpose of this script is to find loci controlling expression of different genes. We will firstly focus on fatty acid related genes. this version does not do scale and center, also batch effect were corrected. 

### import lib
```{r}
library(tidyverse) 
library(qtl)
library(snowfall)
library(ggrepel) 
library(Biostrings)
source("~/KIAT/function_BnRNAseq.R")
```

### import data 
```{r}
load("~/F2/output/network_analysis/vstMat.f2.batch.corrected.Rdata")
dim(vstMat.f2.batch.corrected) # 56180   166 
vstMat.f2.batch.corrected[1:10, 1:10]
write.csv(vstMat.f2.batch.corrected, file = "~/F2/output/network_analysis/vstMat.f2.batch.corrected.csv")
### modify this expression file to the right format 

cross.F2 <- read.cross("csvsr", genfile =  "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised.csv", 
                         phefile = "~/F2/output/network_analysis/vstMat.f2.batch.corrected_revised.csv",
                         genotypes = c("AA", "AB", "BB"))   


cross.F2 <- sim.geno(cross.F2,step=1,n.draws=32) # imputation?  
cross.F2 <- calc.genoprob(cross.F2,step=1)

scanone_eQTL.F2 <- scanone(cross.F2, pheno.col = 2:ncol(cross.F2$pheno), 
	         method = "imp", use = "all.obs")

### permutation... 
set.seed(12345)  
system.time(
permtest.F2 <- scanone(cross.F2, method = "imp", 
                       pheno.col = 2:ncol(cross.F2$pheno),
                       n.perm = 1000,
                       n.cluster=8)) 

alphas <- seq(0.01, 0.10, by = 0.01)
lod.thrs <- summary(permtest.F2, alphas)
lod.thrs  

save(cross.F2, scanone_eigen.F2, permtest.F2, lod.thrs, file = "~/F2/output/network_analysis/scanone-eigengene-qtl_F2.RData") # this is the QTL result for all 95 modules, no matter whether significant correlation were identified with trait or not.  
```

