---
title: "F2_map_construction"
author: "Ruijuan Li"
date: "4/18/2017"
output: html_document 
---

# Goal
The goal of this analysis is to build genetic map using F2 genotyping information, F2 genotyping was done based on the SNPs we identified from its parents 

# install & load package 
```{r}
# install.packages("onemap")
# install.packages("tkrplot", type="source") 
library(tcltk)
library(tkrplot)
library(onemap)  
library(ggplot2)
library(reshape) 
library(tidyverse) 
library("qtl") 
```

# import & format data (MAC version)
```{r} 
F2_geno_data <- read.table("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/F2_Final_SNP_Calls", header = T)
head(F2_geno_data) 
dim(F2_geno_data) # 18226   172 
class(F2_geno_data)

rownames(F2_geno_data) <- paste(F2_geno_data$CHROM, F2_geno_data$POS, sep = "_")
F2_geno_data <- F2_geno_data[, -c(1:6)]
dim(F2_geno_data) # 18226   166 
colnames(F2_geno_data)
colnames(F2_geno_data) <- gsub("([[:print:]]+)(\\.)([[:digit:]])", "\\3", colnames(F2_geno_data))
colnames(F2_geno_data)  
```

# check missing rate & remove data with high missing rate above 0.85 
```{r}
F2_geno_data_with_missing_rate <- mutate(F2_geno_data, 
       missing_rate = round(apply(F2_geno_data, 1, function(x) sum(is.na(x)))/166, 2)) 

rownames(F2_geno_data_with_missing_rate) <- rownames(F2_geno_data) 
hist(F2_geno_data_with_missing_rate$missing_rate)
max(F2_geno_data_with_missing_rate$missing_rate) # 0.47, the highest missing rate is 0.47 

# remove markers with missing rate greater than 0.85 
F2_geno_data_with_missing_rate_filtered <-  F2_geno_data_with_missing_rate[F2_geno_data_with_missing_rate$missing_rate < 0.10,]
rownames(F2_geno_data_with_missing_rate_filtered)

dim(F2_geno_data_with_missing_rate_filtered) # 3608 167
hist(F2_geno_data_with_missing_rate_filtered$missing_rate)

F2_geno_data_with_missing_rate_filtered.final <- subset(F2_geno_data_with_missing_rate_filtered, select=-missing_rate)
rownames(F2_geno_data_with_missing_rate_filtered.final)

dim(F2_geno_data_with_missing_rate_filtered.final) # 3608 166 
rownames(F2_geno_data_with_missing_rate_filtered.final) 
```

# calculate pairwise corrlelation and remove SNPs with corrlelation greater or equal to 0.9
some SNPs should have exactly the same genotype across all individuals, they are redundant in terms of genetic map construction, so they should be removed. find those SNPs by doing correlation test. 
```{r}
F2_geno_data_t <- as.data.frame(t(F2_geno_data_with_missing_rate_filtered.final))
F2_geno_data_t.numeric <- data.matrix(F2_geno_data_t)

# delete markers with all same genotypes across individuals 
test <- apply(F2_geno_data_t.numeric, 2, function(x) length(unique(x[!is.na(x)])))
filter.polymorphsm <- test != 1 

F2_geno_data_t.numeric.2 <- F2_geno_data_t.numeric[,filter.polymorphsm]
dim(F2_geno_data_t.numeric.2) # 166 3606  

# also output save markers with same genotype across individuals
polymorphism <- test == 1
SNP.not.poly <- colnames(F2_geno_data_t.numeric)[polymorphism]
length(SNP.not.poly) # two markers are not polymorphism across individuals 
save(SNP.not.poly, file="SNP.not.poly.Rdata")

# correlation test
options(warn=-1) # suppress warning message
F2_SNP_correlation <- cor(F2_geno_data_t.numeric.2, use = "pairwise.complete.obs")
save(F2_SNP_correlation, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_SNP_correlation.Rdata")
options(warn=0) # unsuppress warning message
dim(F2_SNP_correlation) # 3606 3606 

# calcualte number of SNPs with correlation of 1
nrow(which(F2_SNP_correlation == 1 & lower.tri(F2_SNP_correlation), arr.ind = T, useNames = T)) # 172 pairs of SNPs have identical genotype data 

# find SNPs with correlation of 1 and remove them 
dup.cordinate <- which(F2_SNP_correlation == 1 & lower.tri(F2_SNP_correlation), arr.ind = T, useNames = F)
length(dup.cordinate) # 344
dup.cordinate.df <- as.data.frame(dup.cordinate)
sample.ID <- colnames(F2_SNP_correlation)

# extract duplicate pair information based on their coordicate
dup.pair <- data.frame(matrix(nrow = nrow(dup.cordinate), ncol = 2))
for (i in 1:nrow(dup.cordinate)){
 dup.pair[i,1] <- sample.ID[dup.cordinate[i,1]]
 dup.pair[i,2] <- sample.ID[dup.cordinate[i,2]]
}

length(dup.pair[,2]) # 171 
rem <- unique(dup.pair[,2])
length(rem) # 163 markers are removed due to pairwise correlation of 1 

# save data
save(dup.pair, file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/dup.pair.Rdata")
save(rem, file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/rem.Rdata") 

### 
F2_geno_data_1 <- F2_geno_data_with_missing_rate_filtered.final[!(rownames(F2_geno_data_with_missing_rate_filtered.final) %in% rem),]  

# remove non polymorphism SNPs 
F2_geno_data_2 <- F2_geno_data_1[!(rownames(F2_geno_data_1) %in% SNP.not.poly),]

dim(F2_geno_data_2) # 3443  166, end up with 2932 SNPs for map construction 

save(F2_geno_data_2, file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_data_2.Rdata") # this is non-redundant SNPs with missing data less than 0.1
``` 

# format data for onemap  
```{r}
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_data_2.Rdata")
dim(F2_geno_data_2) # 3443 166 
head(F2_geno_data_2) 

########### below should be corrected
# read parent data 
Ae_Ol <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/vcf.Ae.Ol.intersect.final.csv", stringsAsFactors = F)

# left join to filter Ae_Ol SNP based on F2 genotypes
Ae_Ol$index <- paste(Ae_Ol$CHROM, Ae_Ol$POS, sep = "_")
F2_geno_data_2$index <- rownames(F2_geno_data_2)
F2_geno_data_2_Ae_Ol <- 
left_join(F2_geno_data_2, Ae_Ol, by="index") %>% 
  select(-(X:ALT)) 

F2_geno_data_2_Ae_Ol <- as.matrix(F2_geno_data_2_Ae_Ol) 

# reassign genotype according to parents genotypes
F2_geno_data_2_Ae_Ol_new <- data.frame()

for (i in colnames(F2_geno_data_2_Ae_Ol)[1:166]) {
  print(i)
  for (j in 1:nrow(F2_geno_data_2_Ae_Ol)){
    if (is.na(F2_geno_data_2_Ae_Ol[j,i])){
    F2_geno_data_2_Ae_Ol_new[j,i] = "-"
    } else if (F2_geno_data_2_Ae_Ol[j,i] == "0/1"){
      F2_geno_data_2_Ae_Ol_new[j,i] = "H"
    } else if (F2_geno_data_2_Ae_Ol[j,i] == F2_geno_data_2_Ae_Ol[j,"Ae.gt"]){
      F2_geno_data_2_Ae_Ol_new[j,i] = "A"
    } else {
      F2_geno_data_2_Ae_Ol_new[j,i] = "B"
    }
  }
}

rownames(F2_geno_data_2_Ae_Ol_new) <- F2_geno_data_2_Ae_Ol[,"index"]
F2_geno_data_2_Ae_Ol_new[1:10, 1:10]
############ start working from here ############### 
dim(F2_geno_data_2_Ae_Ol_new) # 3443 166  
write.table(F2_geno_data_2_Ae_Ol_new, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_data_2_Ae_Ol_new.txt")
# change file format in linux 
# cat F2_geno_data_2_Ae_Ol_new.txt | sed 's/"//g' | awk '{first = $1; $1 = ""; print $0}' | sed 's/ //g' > tmp
# tail -3443 tmp > tmp.1 

write.table(rownames(F2_geno_data_2_Ae_Ol_new), file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/marker_info.txt" )
# cat marker_info.txt | awk '{print "*"$2}' | tail -5556 | sed 's/"//g' > marker_info_reform.txt

# paste marker_info_reform.txt tmp.1 | awk '{print $1 " " $2}' > F2_geno_for_one_map.txt 
# cat header_one_map_input F2_geno_for_one_map.txt > F2_geno_for_one_map_final.txt 
# change header info: maker number to the right marker number 
```

# calculate two point rf & assign to linkage groups (on whitney) 
```{r}
# F2_map_construction_LOD3_rf0.5_missing_0.10.R 
```

### assign markers to different linkage groups based on their chromsome assignment 
use this method to test if I get correct result, test it on one chromosome with the least number of markers 
```{r}
# get the number of markers for each chromosome 
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_data_2.Rdata")
dim(F2_geno_data_2) # 3443 166

F2_geno_data_2$Chr <- gsub("([[:print:]]+)(_)([[:digit:]]+)", "\\1", rownames(F2_geno_data_2)) 
F2_geno_data_2$Pos <- gsub("([[:print:]]+)(_)([[:digit:]]+)", "\\3", rownames(F2_geno_data_2)) 
F2_geno_data_2$index <- seq(1: nrow(F2_geno_data_2))

head(F2_geno_data_2[F2_geno_data_2$Chr=="chrA04",],1)
tail(F2_geno_data_2[F2_geno_data_2$Chr=="chrC09",],1)

sum(F2_geno_data_2$Chr=="chrA01") # 221 1-221  
sum(F2_geno_data_2$Chr=="chrA02") # 92 222-313
sum(F2_geno_data_2$Chr=="chrA03") # 303 314-616
sum(F2_geno_data_2$Chr=="chrA04") # 165 617-781
sum(F2_geno_data_2$Chr=="chrA05") # 156 782-937
sum(F2_geno_data_2$Chr=="chrA06") # 439 938-1376
sum(F2_geno_data_2$Chr=="chrA07") # 290 1377-1666
sum(F2_geno_data_2$Chr=="chrA08") # 178 1667-1844
sum(F2_geno_data_2$Chr=="chrA09") # 352 1845-2196
sum(F2_geno_data_2$Chr=="chrA10") # 295 2197-2491
sum(F2_geno_data_2$Chr=="chrC01") # 135 2492-2626
sum(F2_geno_data_2$Chr=="chrC02") # 67 2627-2693
sum(F2_geno_data_2$Chr=="chrC03") # 188 2694-2881
sum(F2_geno_data_2$Chr=="chrC04") # 161 2882-3042
sum(F2_geno_data_2$Chr=="chrC05") # 51 3043-3093
sum(F2_geno_data_2$Chr=="chrC06") # 105 3094-3198
sum(F2_geno_data_2$Chr=="chrC07") # 124 3199-3322
sum(F2_geno_data_2$Chr=="chrC08") # 73 3323-3395
sum(F2_geno_data_2$Chr=="chrC09") # 48 3396-3443  

i = 1
index = vector()
for (ch in unique(F2_geno_data_2$Chr)){
  while (F2_geno_data_2$Chr[i]==ch){
      i = i + 1 
  }
  print (i-1)
  index[ch] = i-1
}

index 
# test on C09 
# https://github.com/leejimmy93/KIAT/map_A01.R .... mapC09.R 
```

### use madmapper to assign markers to clusters/LGs 
[http://cgpdb.ucdavis.edu/XLinkage/MadMapper/Genetic_Map_MadMapper_Arabidopsis.html]

### running madmapper 
### 1) format marker data to madmapper accepted format (on MAC)
```{r}
# cat F2_geno_data_2_Ae_Ol_new.txt | sed 's/"//g' | tr " " "\t" > F2_geno_data_m.txt 
# vim edit add ";" to the 1st line 
# seq 166 | tr "\n" "\t" > number_line_header 
# vim edit add ";" to number_line_header
# cat number_line_header F2_geno_data_m.txt > mv F2_geno_data_m.txt 
# mv F2_geno_data_m.txt F2_geno_data_m.loc 
# wc -l F2_geno_data_m.loc 3444 
# cat F2_geno_data_m.loc | awk '{print NF}' | sort | uniq 167 
# scp to whitney 
```

* madmapper (on whitney)
http://cgpdb.ucdavis.edu/XLinkage/MadMapper/#Part_2
```{r} 
# https://github.com/leejimmy93/KIAT/blob/master/madmapper.sh
# the output file we should check is tree.clust file 
```

check madmapper output (export *.tree.clust file to Excel, and decide which SNP belong to which LG based on tree value) /Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/F2_geno_data_m_loc_mmout.x_tree_clust.xlsx 

* get linkage group from madmapper generated tree.clust file 
```{r}
madmapper.result <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/F2_geno_data_m_loc_mmout.x_tree_clust.csv", header = F)

LG.A01 <- madmapper.result$V26[1:212] # 212 
LG.C01 <- madmapper.result$V26[225:355] # 131
LG.A03 <- madmapper.result$V26[356:621] # 266
LG.C03 <- madmapper.result$V26[659:844] # 186
LG.A09 <- madmapper.result$V26[845:1186] # 342
LG.C08 <- madmapper.result$V26[1187:1248] # 62
LG.A06 <- madmapper.result$V26[1261:1699] # 439
LG.A05 <- madmapper.result$V26[1717:1871] # 155
LG.C05 <- madmapper.result$V26[1872:1909] # 38
LG.A02 <- madmapper.result$V26[1924:2015] # 92
LG.A04 <- madmapper.result$V26[2016:2178] # 163
LG.C04 <- madmapper.result$V26[2179:2335] # 157
LG.A07 <- madmapper.result$V26[2336:2626] # 291
LG.A10 <- madmapper.result$V26[2627:2921] # 295
LG.C06 <- madmapper.result$V26[2922:3026] # 105
LG.C09 <- madmapper.result$V26[3027:3062] # 36
LG.A08 <- madmapper.result$V26[3075:3251] # 177
LG.C02 <- madmapper.result$V26[3252:3319] # 68
LG.C07 <- madmapper.result$V26[3320:3442] # 123

######### unincorported LGs
LG.A01.2 <- madmapper.result$V26[213:224] # 12
LG.A03.2 <- madmapper.result$V26[622:658] # 37
LG.A09.2 <- madmapper.result$V26[1251:1259] # 9
LG.C04.C05 <- madmapper.result$V26[1703:1716] # 14
LG.C05.2 <- madmapper.result$V26[1910:1913] # 4
LG.C08.2 <- madmapper.result$V26[1914:1923] # 10 
LG.C09.2 <- madmapper.result$V26[3063:3073] # 11
#########

LG.A <- list(LG.A01, LG.A01.2, LG.A02, LG.A03, LG.A03.2, LG.A04, LG.A05, LG.A06, LG.A07, LG.A08, LG.A09, LG.A09.2, LG.A10)

LG.C <- list(LG.C01, LG.C02, LG.C03, LG.C04, LG.C04.C05, LG.C05, LG.C05.2, LG.C06, LG.C07, LG.C08, LG.C08.2, LG.C09, LG.C09.2)
length(LG.A) # 13 linkage groups 
length(LG.C) # 13 linkage groups 

LG.madmapper <- list(LG.A01, LG.A02, LG.A03, LG.A04, LG.A05, LG.A06, LG.A07, LG.A08, LG.A09, LG.A10, LG.C01, LG.C02, LG.C03, LG.C04, LG.C05, LG.C06, LG.C07, LG.C08, LG.C09) 
length(LG.madmapper) # 19 
```  

# get index for the above markers in different linkage group 
```{r}
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/twopts.LOD3_rf0.5.Rdata")
twopts.f2.LOD3_rf0.5$marnames

group.A <- list()

for (i in 1:length(LG.A)) {
  group.A[[i]] <-  which(twopts.f2.LOD3_rf0.5$marnames %in% LG.A[[i]])
}
group.A

group.C <- list()

for (i in 1:length(LG.C)) {
  group.C[[i]] <-  which(twopts.f2.LOD3_rf0.5$marnames %in% LG.C[[i]])
}

group.C
save(group.A, group.C, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/group.Rdata")

group.AC <- list()

for (i in 1:length(LG.madmapper)) {
  group.AC[[i]] <-  which(twopts.f2.LOD3_rf0.5$marnames %in% LG.madmapper[[i]])
}
group.AC 
# save(group.AC, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/group.AC.Rdata")  
```

#### order within each chromosome 
```{r}
# scp group.A & group.C to whitney 
# https://github.com/leejimmy93/KIAT/blob/master/F2/map_A.R
# https://github.com/leejimmy93/KIAT/blob/master/F2/map_C.R 
# https://github.com/leejimmy93/KIAT/blob/master/F2/map_AC.R 
``` 

#### analyze result 
C08 eg
```{r}
######## import data #######
# marker data 
F2.data <- read.mapmaker(file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_for_one_map_final.txt") 
F2.data 

# two-point rf data  
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/twopts.LOD3_rf0.5.Rdata")

# order data 
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/madmapper/LG.f2.ord.C8.Rdata")

# get order 
LG.f2.C8 <- make.seq(LG.f2.ord.C8[[10]], "force") 
LG.f2.C8 # 62 

rf.graph.table(LG.f2.C8)
par(mfrow=c(1,1))
draw.map(LG.f2.C8, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)  
write.map(LG.f2.C8, "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.C8.map") 

# use safe mode 
LG.f2.C8.safe <- make.seq(LG.f2.ord.C8[[10]], "safe") 
LG.f2.C8.safe # 40 
rf.graph.table(LG.f2.C8.safe)
par(mfrow=c(1,1))
draw.map(LG.f2.C8.safe, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)  
write.map(LG.f2.C8.safe, "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.C8.safe.map")

#### check alternative order 
ripple.seq(LG.f2.C8, ws = 5, LOD = 3)

```
* Is there need to use ripple.seq() to compare all possible orders, cannot output the result... 

### Now import C subgenome 
### analyze result C sub genome 
### import data 
```{r}
######## import data #######
# marker data 
F2.data <- read.mapmaker(file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_for_one_map_final.txt") 
F2.data 
# This is an object of class 'f2.onemap'
#     No. individuals:     166 
#     No. markers:         3443 
#     Percent genotyped:   95 
# 
#     Number of markers per type:
#        AA : AB : BB -->  3443
# 
# This data contains no phenotypic information

# two-point rf data  
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/twopts.LOD3_rf0.5.Rdata")

# order data 
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/madmapper/LG.f2.ord.C.Rdata")

# get order 
LG.f2.C <- list()

for (i in 1:length(LG.f2.ord.C)){
  LG.f2.C[[i]] <- make.seq(LG.f2.ord.C[[i]], "force") 
}

# rf.graph.table(LG.f2.C[[i]])
par(mfrow=c(1,1))
draw.map(LG.f2.C, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)  
write.map(LG.f2.C, "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.C.map") 

# use safe mode 
LG.f2.C.safe <- list()

for (i in 1:length(LG.f2.ord.C)){
  LG.f2.C.safe[[i]] <- make.seq(LG.f2.ord.C[[i]], "safe") 
} 

# rf.graph.table(LG.f2.C.safe[[13]])
par(mfrow=c(1,1))
draw.map(LG.f2.C.safe, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)  
write.map(LG.f2.C.safe, "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.C.safe.map")
```

### Now import A subgenome 
### analyze result C sub genome 
### import data 
```{r}
######## import data #######
# marker data 
F2.data <- read.mapmaker(file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_for_one_map_final.txt") 
F2.data

# two-point rf data  
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/twopts.LOD3_rf0.5.Rdata")

# order data 
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/madmapper/LG.f2.ord.A.Rdata")

# get order 
LG.f2.A <- list()

for (i in 1:length(LG.f2.ord.A)){
  LG.f2.A[[i]] <- make.seq(LG.f2.ord.A[[i]], "force") 
}

rf.graph.table(LG.f2.A[[7]])
par(mfrow=c(1,1))
draw.map(LG.f2.A, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)  
write.map(LG.f2.A, "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.A.map") 

# use safe mode 
LG.f2.A.safe <- list()

for (i in 1:length(LG.f2.ord.A)){
  LG.f2.A.safe[[i]] <- make.seq(LG.f2.ord.A[[i]], "safe") 
} 

# rf.graph.table(LG.f2.C.safe[[13]])
par(mfrow=c(1,1))
draw.map(LG.f2.A.safe, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)  
write.map(LG.f2.A.safe, "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.A.safe.map")
```

### combine A & C subgenome 
```{r}
F2.data <- read.mapmaker(file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_for_one_map_final.txt") 
F2.data

# two-point rf data  
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/twopts.LOD3_rf0.5.Rdata")

# order data 
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/madmapper/LG.f2.ord.A.Rdata")
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/madmapper/LG.f2.ord.C.Rdata")

# get order for A sub
LG.f2 <- list() 

for (i in c(1, 3, 4, 6, 7, 8, 9, 10, 11, 13)){
  LG.f2[[i]] <- make.seq(LG.f2.ord.A[[i]], "safe") 
}

for (i in c(1, 2, 3, 4, 6, 8, 9, 10, 12)){
  LG.f2[[13+i]] <- make.seq(LG.f2.ord.C[[i]], "safe") 
} 

# remove null LGs 
LG.f2 = LG.f2[-which(sapply(LG.f2, is.null))] 
draw.map(LG.f2, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)  
write.map(LG.f2, "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.madmapper.map") 
```

# R/qtl package to plot LOD & rf headmap to check marker order & assignment to LG correctness  
```{r}
LG.f2.C <- read.cross("mm", file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_for_one_map_final.txt", mapfile = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.C.map")

summary(LG.f2.C)
plotMap(LG.f2.C)
# plot.rf(LG.f2.C) 

totmar(LG.f2.C) # 945

LG.f2.C.safe <- read.cross("mm", file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_for_one_map_final.txt", mapfile = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.C.safe.map")

summary(LG.f2.C.safe)
plotMap(LG.f2.C.safe)
# plot.rf(LG.f2.C.safe) 

totmar(LG.f2.C.safe) # 631 

# geno.image(LG.f2.C, alternate.chrid = T) # grid with color pixels for different genotypes
# geno.image(LG.f2.A, alternate.chrid = T) # grid with color pixels for different genotypes

# save(newmap.drop.marker, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/newmap.drop.marker.Rdata")

####### A sub
LG.f2.A <- read.cross("mm", file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_for_one_map_final.txt", mapfile = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.A.map")

summary(LG.f2.A)
plotMap(LG.f2.A)
# plot.rf(LG.f2.A)

totmar(LG.f2.C) # 945

LG.f2.A.safe <- read.cross("mm", file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_for_one_map_final.txt", mapfile = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.A.safe.map")

summary(LG.f2.A.safe) 
plotMap(LG.f2.A.safe) 
totmar(LG.f2.C.safe) # 
# plot.rf(LG.f2.A.safe) 

## all markers on A & C 
LG.f2.madmapper <- read.cross("mm", file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_for_one_map_final.txt", mapfile = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.madmapper.map")

summary(LG.f2.madmapper) # 2013 
plotMap(LG.f2.madmapper) 
totmar(LG.f2.madmapper)  
# plot.rf(LG.f2.madmapper) 

# write.csv(pull.geno(LG.f2.madmapper),file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.madmapper.before.doublecrossover.csv") 
```

### refine map 
```{r} 
# remove double crossover 
LG.f2.madmapper <- read.cross("mm", file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_for_one_map_final.txt", mapfile = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.madmapper.map")

LG.f2.madmapper.before.crossover <- LG.f2.madmapper
save(LG.f2.madmapper.before.crossover, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.madmapper.before.crossover.Rdata")

for (chr in names(LG.f2.madmapper$geno)) { # for each chromosome in cross genotype data
  my.chr <- get(chr,LG.f2.madmapper$geno) # return the genotype data, including data & map
  print(paste(chr,"NA before",sum(is.na(my.chr$data)))) 
  if(ncol(my.chr$data) > 3) { 
    my.chr$data[,2:(ncol(my.chr$data)-1)] <- sapply(2:(ncol(my.chr$data)-1),function(i) {
      apply(my.chr$data[,(i-1):(i+1)],1,function(gt) {
        if (any(is.na(gt))) return(gt[2]) #technically should be looking at the next genotyped marker.
        if ( (length(unique(gt)) == 2) & (gt[1] == gt[3])) return(NA)
        if ( length(unique(gt))  == 3) return(NA)
        return(gt[2])
      })
    })
  }
  LG.f2.madmapper$geno <- within(LG.f2.madmapper$geno,assign(chr,my.chr))
  print(paste(chr,"NA after",sum(is.na(get(chr,LG.f2.madmapper$geno)$data))))
}  

# [1] "1 NA before 969"
# [1] "1 NA after 1270"
# [1] "2 NA before 425"
# [1] "2 NA after 617"
# [1] "3 NA before 1095"
# [1] "3 NA after 1452"
# [1] "4 NA before 841"
# [1] "4 NA after 1092"
# [1] "5 NA before 823"
# [1] "5 NA after 1109"
# [1] "6 NA before 1849"
# [1] "6 NA after 2487"
# [1] "7 NA before 1210"
# [1] "7 NA after 1600"
# [1] "8 NA before 827"
# [1] "8 NA after 1095"
# [1] "9 NA before 1527"
# [1] "9 NA after 2046"
# [1] "10 NA before 1283"
# [1] "10 NA after 1686"
# [1] "11 NA before 724"
# [1] "11 NA after 957"
# [1] "12 NA before 325"
# [1] "12 NA after 452"
# [1] "13 NA before 952"
# [1] "13 NA after 1309"
# [1] "14 NA before 785"
# [1] "14 NA after 1160"
# [1] "15 NA before 232"
# [1] "15 NA after 311"
# [1] "16 NA before 463"
# [1] "16 NA after 679"
# [1] "17 NA before 740"
# [1] "17 NA after 1033"
# [1] "18 NA before 309"
# [1] "18 NA after 419"
# [1] "19 NA before 140"
# [1] "19 NA after 235" 

write.csv(pull.geno(LG.f2.madmapper),file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.madmapper.rm.doublecrossover.csv") 
```

#### work on maps with double crossover removed 
```{r}
f2.map.new <- est.map(LG.f2.madmapper,verbose=T,error.prob=.01) # why going through 
LG.f2.madmapper <- replace.map(LG.f2.madmapper, f2.map.new)
LG.f2.madmapper.after.crossover <- LG.f2.madmapper
plot.map(LG.f2.madmapper.after.crossover, alternate.chrid = T) # the old genetic map
summaryMap(LG.f2.madmapper.after.crossover)
summaryMap(LG.f2.madmapper.before.crossover)

plot.map(LG.f2.madmapper.before.crossover,LG.f2.madmapper.after.crossover, alternate.chrid = T) # genetic map comparison

# plot.rf(LG.f2.madmapper, col.scheme = "redblue", alternate.chrid = T) 
# par(mfrow=c(1,1)) # reset to 1:1
save(LG.f2.madmapper.after.crossover, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.madmapper.after.crossover.Rdata")
```

### check rf graph for map 
```{r}
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.madmapper.before.crossover.Rdata")
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.madmapper.after.crossover.Rdata")

png("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/figure /rf_chr10.png", width=8, height=8, units="in", res=300)
plot.rf(LG.f2.madmapper.before.crossover, chr = '10', main = "RF and LOD scores for chrom A10")
dev.off()

png("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/figure /rf_madmapper.png", width=8, height=8, units="in", res=300)
plot.rf(LG.f2.madmapper.before.crossover, main = "RF and LOD scores")
dev.off() 

png("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/figure /rf_chr10.png", width=8, height=8, units="in", res=300)
plot.rf(LG.f2.madmapper.after.crossover, chr = '10', main = "RF and LOD scores for chrom A10")
dev.off() 

png("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/figure /map_before_crossover.png", width=8, height=8, units="in", res=300)
plot.map(LG.f2.madmapper.before.crossover, alternate.chrid = T) # the old genetic map
dev.off()   

png("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/figure /map_after_crossover.png", width=8, height=8, units="in", res=300)
plot.map(LG.f2.madmapper.after.crossover, alternate.chrid = T) # genetic map comparison
dev.off()

png("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/figure /map_comparison_1.png", width=8, height=8, units="in", res=300)
plot.map(LG.f2.madmapper.before.crossover,LG.f2.madmapper.after.crossover, alternate.chrid = T) # genetic map comparison
dev.off() 
```

### continue refine map, ripple 
```{r}
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.madmapper.after.crossover.Rdata")
summaryMap(LG.f2.madmapper.after.crossover)
plotMap(LG.f2.madmapper.after.crossover)

totmar(LG.f2.madmapper.after.crossover) # 2013 

# work on A10 
plot.rf(LG.f2.madmapper.after.crossover, chr = "10") # this 
plotMap(LG.f2.madmapper.after.crossover, chr = '10') 

set.seed(16)
LG.f2.madmapper.after.crossover <- orderMarkers(LG.f2.madmapper.after.crossover, chr = c(10), 
	                        window = 5, use.ripple = TRUE, maxit = 4000, 
	                        error.prob = 0.0001, verbose = T)

plotMap(LG.f2.madmapper.after.crossover, chr = '10') 
plot.rf(LG.f2.madmapper.after.crossover, chr = '10') # plot appears better but length increased... doesn't seem to be a good way. 
```

# follow Julin's suggestion, to remove markers with high double crossover (genotyping error)
```{r}
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.madmapper.after.crossover.Rdata")
summaryMap(LG.f2.madmapper.after.crossover)
plotMap(LG.f2.madmapper.after.crossover)

totmar(LG.f2.madmapper.after.crossover) # 2013 

geno_new <- pull.geno(LG.f2.madmapper.after.crossover)
geno <- pull.geno(LG.f2.madmapper.before.crossover)
dim(geno); dim(geno_new)
geno[1:10, 1:10]; geno_new[1:10, 1:10]

double.crossover.count <- sapply(colnames(geno), function(marker){
  sum(is.na(geno_new[,marker]))-sum(is.na(geno[,marker]))
} 
)

double.crossover.count <- as.data.frame(double.crossover.count)
double.crossover.count
hist(double.crossover.count)
marker.drop <- rownames(subset(double.crossover.count, double.crossover.count>10))
marker.drop

# drop marker with double crossover more than 10 times 
LG.f2.madmapper.after.crossover.drop <- drop.markers(LG.f2.madmapper.after.crossover, marker.drop)
totmar(LG.f2.madmapper.after.crossover.drop) # 1965 

LG.f2.madmapper.after.crossover.drop <- 
                          orderMarkers(LG.f2.madmapper.after.crossover.drop, 
	                        window = 4, use.ripple = TRUE, maxit = 4000, 
	                        error.prob = 0.0001)

map.new.new <- est.map(LG.f2.madmapper.after.crossover.drop)
plot.map(map.new.new, alternate.chrid = TRUE)

# even longer... bad.... 
```

# continue refine later... 
```{r}
# identify markers with possible segregation distortion
gt <- geno.table(LG.f2.madmapper.after.crossover)
seg_dist <- gt[gt$P.value < 0.05/totmar(LG.f2.madmapper.after.crossover),]
seg_dist
dim(seg_dist) # 67 8 

#                 chr missing AA  AB BB not.BB not.AA      P.value
# chrA01_684630     1      13  9 106 38      0      0 4.702676e-08
# chrA01_4318350    1      25  7  99 35      0      0 3.815686e-08
# chrA01_4081437    1      12 12 114 28      0      0 3.602786e-09
# chrA02_6523578    2       8 13  98 47      0      0 6.885616e-06
# chrA02_6264806    2      20  6  93 47      0      0 4.168391e-08 .... this many might be caused by missing data replace SNPs with double cross over 

# now remove markers that obviously do not belong
# make a copy so I can go back and not have to reestimate the map
LG.f2.madmapper.2 <- LG.f2.madmapper.after.crossover
rffull <- pull.rf(LG.f2.madmapper.2) 
```


######### above is for madmapper 

########### what if I use mapping order to order markers ################
get marker index within each chromosome 
```{r}
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_data_2.Rdata")
dim(F2_geno_data_2) # 3443 166

F2_geno_data_2$Chr <- gsub("([[:print:]]+)(_)([[:digit:]]+)", "\\1", rownames(F2_geno_data_2)) 
F2_geno_data_2$Pos <- gsub("([[:print:]]+)(_)([[:digit:]]+)", "\\3", rownames(F2_geno_data_2)) 
F2_geno_data_2$index <- seq(1: nrow(F2_geno_data_2))

# figure out the index positon within each chromosome for all chromosomes 
F2_geno_data_info <- F2_geno_data_2[,c("Chr", "Pos", "index")]
dim(F2_geno_data_info) # 3443 3 
F2_geno_data_info[1:10,]

out <- split( F2_geno_data_info, f = F2_geno_data_info$Chr)
length(out) # 19 
out

order <- list()

for (i in 1:length(out)){
  out[[i]]$Pos <- as.numeric(out[[i]]$Pos)
  order[[i]] <- out[[i]][order(out[[i]]$Pos),]$index 
}

order 
save(order, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/order.Rdata")
```

use mapping order to order markers  
```{r}
# load data 
F2.data <- read.mapmaker(file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_for_one_map_final.txt") 
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/twopts.LOD3_rf0.5.Rdata") # LOD=3, max rf=0.5 
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/missing_rate_0.10/order.Rdata")

# must unload tydiverse, otherwise error, still need to fix these two lines 
unloadNamespace("tidyverse")
detach("package:purrr", unload=TRUE) 

# mapping order 
LG.f2.arbi <- list()

for (i in 1:length(order)){
  tmp <- make.seq(twopts.f2.LOD3_rf0.5, order[[i]])
  LG.f2.arbi[[i]] <- map(tmp)
} 

rf.graph.table(LG.f2.arbi[[19]])
par(mfrow=c(1,1)) 
draw.map(LG.f2.arbi, names= F, grid=F, cex.mrk=0.5, cex.grp=0.75, horizontal=T)  
# write.map(LG.f2.arbi, "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.arbi.map") 
```

remove markers with double crossover 
```{r}
########### use R/qtl ########################  
# import data 
map.F2 <- read.cross("mm", file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_for_one_map_final.txt", mapfile = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.arbi.map")

# get rf figure & map figure for mapping order based map 
summary(map.F2)
plotMap(map.F2)
# plot.rf(map.F2) 

LG.f2.mappingpos.before.crossover <- map.F2

LG <- map.F2

  for (chr in names(LG$geno)) { # for each chromosome in cross genotype data
    my.chr <- get(chr,LG$geno) # return the genotype data, including data & map
    print(paste(chr,"NA before",sum(is.na(my.chr$data)))) 
    if(ncol(my.chr$data) > 3) { 
      my.chr$data[,2:(ncol(my.chr$data)-1)] <- sapply(2:(ncol(my.chr$data)-1),function(i) {
        apply(my.chr$data[,(i-1):(i+1)],1,function(gt) {
          if (any(is.na(gt))) return(gt[2]) #technically should be looking at the next genotyped marker.
          if ( (length(unique(gt)) == 2) & (gt[1] == gt[3])) return(NA)
          if ( length(unique(gt))  == 3) return(NA)
          return(gt[2])
        })
      })
    }
    LG$geno <- within(LG$geno,assign(chr,my.chr))
    print(paste(chr,"NA after",sum(is.na(get(chr,LG$geno)$data))))
  } 

# [1] "1 NA before 1950"
# [1] "1 NA after 2744"
# [1] "2 NA before 842"
# [1] "2 NA after 1205"
# [1] "3 NA before 2448"
# [1] "3 NA after 3412"
# [1] "4 NA before 1344"
# [1] "4 NA after 1932"
# [1] "5 NA before 1371"
# [1] "5 NA after 1958"
# [1] "6 NA before 3573"
# [1] "6 NA after 4857"
# [1] "7 NA before 2511"
# [1] "7 NA after 3337"
# [1] "8 NA before 1460"
# [1] "8 NA after 2044"
# [1] "9 NA before 3015"
# [1] "9 NA after 4211"
# [1] "10 NA before 2510"
# [1] "10 NA after 3401"
# [1] "11 NA before 1188"
# [1] "11 NA after 1805"
# [1] "12 NA before 495"
# [1] "12 NA after 724"
# [1] "13 NA before 1565"
# [1] "13 NA after 2282"
# [1] "14 NA before 1301"
# [1] "14 NA after 1933"
# [1] "15 NA before 439"
# [1] "15 NA after 694"
# [1] "16 NA before 924"
# [1] "16 NA after 1343"
# [1] "17 NA before 1069"
# [1] "17 NA after 1642"
# [1] "18 NA before 660"
# [1] "18 NA after 965"
# [1] "19 NA before 340"
# [1] "19 NA after 554"

map.F2 <- LG

f2.map.new <- est.map(map.F2,verbose=T,error.prob=.01) 
save(f2.map.new, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/f2.map.new.mapping.Rdata")
map.F2 <- replace.map(map.F2, f2.map.new)
LG.f2.mappingpos.after.crossover <- map.F2

plot.map(LG.f2.mappingpos.after.crossover, alternate.chrid = T) # the old genetic map
plot.map(LG.f2.mappingpos.before.crossover, LG.f2.mappingpos.after.crossover, alternate.chrid = T) # genetic map comparison 
```

# a bunch of checks in R/qtl 
```{r}
summary(map.F2)
plotMap(map.F2)
# plot.rf(map.F2)

totmar(map.F2)

# 1) segregation distortion 
test <- geno.table(map.F2)
seg_dist <- test[test$P.value < 0.01,]
seg_dist
dim(seg_dist) # 340 markers 

##### drop markers with segregation distortion 
# map.F2.after.drop <- drop.marker(map.F2, rownames(seg_dist)) # doesn't work... 
map.F2

# rffull.Asub <- pull.rf(map.F2)

# work with each chromosome seperately 
# plot.rf(map.F2, chr = "chrA02")
rfA02 <- pull.rf(map.F2, chr = 'chrA02')
chrA02 <- markernames(map.F2, chr = 'chrA02')
chrA02

# 2) compare individual's genotype similarity. Assumption: no two individuals with very similary genotypes
cg <- comparegeno(map.F2) 
dim(cg) 
hist(cg, breaks = 200,
     xlab="Proportion of identical genotypes")
rug(cg)
which(cg >0.9, arr.ind = TRUE) # no individuals with high genotype similarities

# No apparent problem, make a plot to see the data
plot.rf(map.F2, col.scheme = "redblue", alternate.chrid = T)

# looks like they are on the right chromosomes but wrong order within each chromosome

# estimate the genetic map
nm <- est.map(map.F2)
plot(nm, alternate.chrid = T) 
pull.map(cross_m, chr = c("A03","A05")) # longer length was plot than what was indicated in the data? 
pull.map(cross_m) 
``` 

### re-build the map 
While I used safe mode to secure all the markers to different LGs, over 1000 markers were lost during the process, because they could not satisfy the threshold (max rf = 0.5 and LOD score of 3). Julin suggested that we might want to increase the threshold to map markers, like max rf = 0.55, because there can be some makers by chance have a max rf greater than 0.5 but still can be mapped to different LGs. 
```{r}
# re calculate recombination fraction using 
# https://github.com/leejimmy93/KIAT/blob/master/F2/F2_map_construction_LOD3_rf0.55_missing_0.10.R  
``` 

### appendix 
### check double cross over 
```{r}
rm.double.crossover <- function(LG){
for (chr in names(LG$geno)) { # for each chromosome in cross genotype data
  my.chr <- get(chr,LG$geno) # return the genotype data, including data & map
  print(paste(chr,"NA before",sum(is.na(my.chr$data)))) 
  if(ncol(my.chr$data) > 3) { 
    my.chr$data[,2:(ncol(my.chr$data)-1)] <- sapply(2:(ncol(my.chr$data)-1),function(i) {
      apply(my.chr$data[,(i-1):(i+1)],1,function(gt) {
        if (any(is.na(gt))) return(gt[2]) #technically should be looking at the next genotyped marker.
        if ( (length(unique(gt)) == 2) & (gt[1] == gt[3])) return(NA)
        if ( length(unique(gt))  == 3) return(NA)
        return(gt[2])
      })
    })
  }
  LG$geno <- within(LG$geno,assign(chr,my.chr))
  print(paste(chr,"NA after",sum(is.na(get(chr,LG$geno)$data))))
} 
}

# write.csv(pull.geno(cross.drop.marker),file="~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/cross.drop.marker.csv")

# newmap.drop.marker <- est.map(cross.drop.marker,verbose=T,error.prob=.01) # why going through 
# # the last section of code doesn't assign new map 
# plot.map(map.F2, alternate.chrid = T) # the old genetic map
# plot.map(map.F2,newmap.drop.marker, alternate.chrid = T) # genetic map comparison
# 
# geno.image(cross.drop.marker, alternate.chrid = T) # grid with color pixels for different genotypes
# 
# save(newmap.drop.marker, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/newmap.drop.marker.Rdata")
``` 

