---
title: "F2_map_construction_unused"
author: "Ruijuan Li"
date: "9/21/2017"
output: html_document
---

# MDS 
The idea is that because Alex's pipeline seperate my marker into multiple LGs for each chromosome, although markers from the same chromosome do cluster together. By checking the LOD score & rf matrix, I found that markers from the same chromosome, which were assigned to different LGs are genetically linked by LOD score, but their rf is very low too. 

Julin hypothezied that this might be caused by the different rf between markers that were present in the chromosome through evolution and markers that were integrated into the chromosome in this newly synthetic polyploid B.napus. We will arrange a meeting with Luca Comai to discuss this possibility. 

At the same time, Julin asked me to do MDS using markers from the same chromosome but assigned to different LGs, to see whether they form two clusters by LGs. 
```{r}
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_data_2.Rdata")
dim(F2_geno_data_2) # 2932 166

# use C subgenome LG15 & LG16 
LG.A_15_16 <- F2_geno_data_2[rownames(F2_geno_data_2) %in% LG.A[[5]] | rownames(F2_geno_data_2) %in% LG.A[[6]],] 

LG.A_15_16.trans <- transform_geno(temp2 = LG.A_15_16)

# MDS 
MDS.result <- MDS(genotype_file = LG.A_15_16.trans)
plot(MDS(genotype_file = LG.A_15_16.trans))

# plot 
MDS.result$LG <- ifelse(rownames(MDS.result) %in% LG.A[[5]], "5", "6")
ggplot(data = MDS.result, mapping = aes(x=V1, y = V2, color=LG)) + geom_point() + labs(title="A03")
ggsave("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/ChrA03.png", width = 7, height = 5)  

# use A subgenome LG15 & LG16 
LG.C_15_16 <- F2_geno_data_2[rownames(F2_geno_data_2) %in% LG.C[[15]] | rownames(F2_geno_data_2) %in% LG.C[[16]],] 

LG.C_15_16.trans <- transform_geno(temp2 = LG.C_15_16)

# MDS 
MDS.result <- MDS(genotype_file = LG.C_15_16.trans)
plot(MDS(genotype_file = LG.C_15_16.trans))

# plot 
MDS.result$LG <- ifelse(rownames(MDS.result) %in% LG.C[[15]], "15", "16")
ggplot(data = MDS.result, mapping = aes(x=V1, y = V2, color=LG)) + geom_point() + labs(title="C07")
ggsave("/Users/ruijuanli/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/ChrC07.png", width = 7, height = 5)

### to demonstrate this more, plot all the C subgenome markers in MDS 1stly 
length(unlist(LG.C)) 

LG.C_all <- F2_geno_data_2[rownames(F2_geno_data_2) %in% unlist(LG.C),] 

### Do LG15 & LG16 really belong to the same chromosome? yes 
MDS.F2.map(F2_geno_data = F2_geno_data_2, LG = LG.A, groupA = 5, groupB = 6)
```

# appendix 
```{r} 
transform_geno <- function(temp2){
  rnames <- rownames(temp2)
  temp2 <- data.frame(sapply(temp2, function(x) sub("0/0","0",x)))
  temp2 <- data.frame(sapply(temp2, function(x) sub("0/1","1",x)))
  temp2 <- data.frame(sapply(temp2, function(x) sub("1/1","2",x)))
  row.names(temp2) <- rnames 
  return(temp2)
} 

MDS <- function(genotype_file){
  genDist <- as.matrix(dist(genotype_file))
  #perform the multi-dimensional scaling
  geno.mds <- as.data.frame(cmdscale(genDist))
  plot(geno.mds)  

  return(geno.mds) 
} 

# MDS of SNPs from same chromosome but differnet LGs 
MDS.F2.map <- function(F2_geno_data, LG, groupA, groupB){
  LG.groupA_groupB <- F2_geno_data[rownames(F2_geno_data) %in% LG[[groupA]] | rownames(F2_geno_data) %in% LG.C[[groupB]],]

  LG.groupA_groupB.trans <- transform_geno(temp2 = LG.groupA_groupB)

  # MDS 
  MDS.result <- MDS(genotype_file = LG.groupA_groupB.trans)
  plot(MDS(genotype_file = LG.groupA_groupB.trans))

  # plot 
  MDS.result$LG <- ifelse(rownames(MDS.result) %in% LG.C[[groupA]], groupA, groupB)
  plot <- ggplot(data = MDS.result, mapping = aes(x=V1, y = V2, color=LG)) + geom_point()
  
  return(plot)
} 
``` 
