---
title: "linolenic_acid_genes_analysis"
author: "Ruijuan Li"
date: "2/22/2018"
output: html_document
---

Purpose of this script is to find genes (or maybe fatty acid related genes) whose allelic pattern match the linolenic acid content level (the interaction plot). 

based on the information from the interaction plot, I am looking for genes 
1) fall into the two major QTL interval on A08 and C03
2) have SNPs in them 
3) when both genes are homozygous identical as either parent, the linolenic acid level is very low; when both genes are heterzygous, the linolenic acid level is medium, when one gene is homozygous identical as one parent, whereas the other gene is homozygous identical as the other parent, the linolenic acid level is the highest. 

To do this, I will 
1) get John's F2 genotyping data where there are allelic information for many more SNPs
2) polarize them based on the parental genotype information 
3) only extract SNPs in the QTL interval 
4) get annotation 
5) check the pattern for several fatty acid genes firslty (plot the linolenic acid level based on the genotype for the two genes, maybe check the erucic and oleic acid etc. level as well if such gene is found for linolenic acid) 
6) more maybe afterwards if no genes found follow this pattern. 

### load data and polarize according to parental information 
```{r}
library(tidyverse) 

F2_geno_data <- read.table("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/F2_Final_SNP_Calls", header = T)
head(F2_geno_data) 
dim(F2_geno_data) # 18226   172 

rownames(F2_geno_data) <- paste(F2_geno_data$CHROM, F2_geno_data$POS, sep = "_")
F2_geno_data <- F2_geno_data[, -c(1:6)]
dim(F2_geno_data) # 18226   166 
colnames(F2_geno_data) <- gsub("([[:print:]]+)(\\.)([[:digit:]])", "\\3", colnames(F2_geno_data))

# read parent data 
Ae_Ol <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/vcf.Ae.Ol.intersect.final.csv", stringsAsFactors = F)
Ae_Ol %>% dim()

# left join to filter Ae_Ol SNP based on F2 genotypes
Ae_Ol$index <- paste(Ae_Ol$CHROM, Ae_Ol$POS, sep = "_")
F2_geno_data$index <- rownames(F2_geno_data)
F2_geno_data_2_Ae_Ol <- 
left_join(F2_geno_data, Ae_Ol, by="index") %>% 
  dplyr::select(-(X:ALT)) 

F2_geno_data_Ae_Ol <- as.matrix(F2_geno_data_Ae_Ol)  

# reassign genotype according to parents genotypes
F2_geno_data_2_Ae_Ol_new <- data.frame()

for (i in colnames(F2_geno_data_2_Ae_Ol)[1:166]) {
  print(i)
  for (j in 1:nrow(F2_geno_data_2_Ae_Ol)){
    if (is.na(F2_geno_data_2_Ae_Ol[j,i])){
    F2_geno_data_2_Ae_Ol_new[j,i] = "-"
    } else if (F2_geno_data_2_Ae_Ol[j,i] == "0/1"){
      F2_geno_data_2_Ae_Ol_new[j,i] = "H"
    } else if (F2_geno_data_2_Ae_Ol[j,i] == F2_geno_data_2_Ae_Ol[j,"Ae.gt"]){
      F2_geno_data_2_Ae_Ol_new[j,i] = "A"
    } else {
      F2_geno_data_2_Ae_Ol_new[j,i] = "B"
    }
  }
} 


```

