---
title: "matrix_eQTL"
author: "Ruijuan Li"
date: "11/2/2017"
output: html_document
---

There is this Matrix eQTL R package that seems useful for eQTL analysis, and I want myself to get familarity with this package, if possible, use it for my analysis. Refer to the webpage below for details... 
http://www.bios.unc.edu/research/genomic_software/Matrix_eQTL/runit.html 

### intalling and load lib
```{r}
# install.packages("MatrixEQTL")
library("MatrixEQTL")
```

### sample code and toy data set 

The sample code performs eQTL analysis of a toy data set consisting of three files: genotype, expression, and covariates. For every gene-SNP pair it runs linear regression analysis accounting for the set of covariates.

```{r}
# input data and output data setup 
base.dir = find.package("MatrixEQTL")
useModel = modelLINEAR; # modelANOVA or modelLINEAR or modelLINEAR_CROSS
SNP_file_name = paste(base.dir, "/data/SNP.txt", sep="")
expression_file_name = paste(base.dir, "/data/GE.txt", sep="")
SNP_file_name
expression_file_name

covariates_file_name = paste(base.dir, "/data/Covariates.txt", sep="")
output_file_name = tempfile()

# threshold 
pvOutputThreshold = 1e-2
errorCovariance = numeric()

# geno data 
snps = SlicedData$new()
snps$fileDelimiter = "\t"      # the TAB character
snps$fileOmitCharacters = "NA" # denote missing values;
snps$fileSkipRows = 1          # one row of column labels
snps$fileSkipColumns = 1       # one column of row labels
snps$fileSliceSize = 2000      # read file in slices of 2,000 rows
snps$LoadFile(SNP_file_name)

## Load gene expression data
gene = SlicedData$new()
gene$fileDelimiter = "\t"      # the TAB character
gene$fileOmitCharacters = "NA" # denote missing values;
gene$fileSkipRows = 1          # one row of column labels
gene$fileSkipColumns = 1       # one column of row labels
gene$fileSliceSize = 2000      # read file in slices of 2,000 rows
gene$LoadFile(expression_file_name)

## Load covariates 
cvrt = SlicedData$new()
cvrt$fileDelimiter = "\t"      # the TAB character
cvrt$fileOmitCharacters = "NA" # denote missing values;
cvrt$fileSkipRows = 1          # one row of column labels
cvrt$fileSkipColumns = 1       # one column of row labels
if(length(covariates_file_name)>0) {
cvrt$LoadFile(covariates_file_name)
}

## Run the analysis

me = Matrix_eQTL_engine(
snps = snps,
gene = gene,
cvrt = cvrt,
output_file_name = output_file_name,
pvOutputThreshold = pvOutputThreshold,
useModel = useModel, 
errorCovariance = errorCovariance, 
verbose = TRUE,
pvalue.hist = TRUE,
min.pv.by.genesnp = FALSE,
noFDRsaveMemory = FALSE)

unlink(output_file_name)

## Results:

cat('Analysis done in: ', me$time.in.sec, ' seconds', '\n')
cat('Detected eQTLs:', '\n')
show(me$all$eqtls)

## Plot the histogram of all p-values

plot(me) 

# Test local and distand gene-SNP pairs separately and plot Q-Q plots of local and distant p-values
source("Matrix_eQTL_R/Matrix_eQTL_engine.r")
library(MatrixEQTL)

base.dir = find.package('MatrixEQTL')
useModel = modelLINEAR
SNP_file_name = paste(base.dir, "/data/SNP.txt", sep="")
snps_location_file_name = paste(base.dir, "/data/snpsloc.txt", sep="")

expression_file_name = paste(base.dir, "/data/GE.txt", sep="")
gene_location_file_name = paste(base.dir, "/data/geneloc.txt", sep="")

covariates_file_name = paste(base.dir, "/data/Covariates.txt", sep="")

output_file_name_cis = tempfile()
output_file_name_tra = tempfile()

pvOutputThreshold_cis = 2e-2
pvOutputThreshold_tra = 1e-2

errorCovariance = numeric()

cisDist = 1e6

snps = SlicedData$new()
snps$fileDelimiter = "\t"      # the TAB character
snps$fileOmitCharacters = "NA" # denote missing values;
snps$fileSkipRows = 1          # one row of column labels
snps$fileSkipColumns = 1       # one column of row labels
snps$fileSliceSize = 2000      # read file in slices of 2,000 rows
snps$LoadFile(SNP_file_name)

gene = SlicedData$new()
gene$fileDelimiter = "\t"      # the TAB character
gene$fileOmitCharacters = "NA" # denote missing values;
gene$fileSkipRows = 1          # one row of column labels
gene$fileSkipColumns = 1       # one column of row labels
gene$fileSliceSize = 2000      # read file in slices of 2,000 rows
gene$LoadFile(expression_file_name)

cvrt = SlicedData$new()
cvrt$fileDelimiter = "\t"      # the TAB character
cvrt$fileOmitCharacters = "NA" # denote missing values;
cvrt$fileSkipRows = 1          # one row of column labels
cvrt$fileSkipColumns = 1       # one column of row labels
if(length(covariates_file_name)>0) {
cvrt$LoadFile(covariates_file_name)
}

## Run the analysis
snpspos = read.table(snps_location_file_name, header = TRUE, stringsAsFactors = FALSE);
genepos = read.table(gene_location_file_name, header = TRUE, stringsAsFactors = FALSE);

me = Matrix_eQTL_main(
snps = snps, 
gene = gene, 
cvrt = cvrt,
output_file_name     = output_file_name_tra,
pvOutputThreshold     = pvOutputThreshold_tra,
useModel = useModel, 
errorCovariance = errorCovariance, 
verbose = TRUE, 
output_file_name.cis = output_file_name_cis,
pvOutputThreshold.cis = pvOutputThreshold_cis,
snpspos = snpspos, 
genepos = genepos,
cisDist = cisDist,
pvalue.hist = "qqplot",
min.pv.by.genesnp = FALSE,
noFDRsaveMemory = FALSE)

unlink(output_file_name_tra)
unlink(output_file_name_cis)

## Results:

cat('Analysis done in: ', me$time.in.sec, ' seconds', '\n');
cat('Detected local eQTLs:', '\n')
show(me$cis$eqtls)
cat('Detected distant eQTLs:', '\n')
show(me$trans$eqtls)

## Plot the Q-Q plot of local and distant p-values

plot(me)
```

