---
title: "QTL_analysis"
author: "Ruijuan Li"
date: "9/26/2017"
output: 
  html_document: 
    keep_md: yes
---

Goal of this project: to map QTL using the genetic map that I constructed from F2 mapping population created from Da-Ae & Da-Ol-1   
```{r}
library(tidyverse)
library(qtl)
library(snowfall)
library(ggrepel)  
library(Biostrings)
source("~/KIAT/function_BnRNAseq.R")    
```

### reformat file with geno and pheno info
```{r}
# geno data 
F2_geno_data_2_Ae_Ol_new <- read.table("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_data_2_Ae_Ol_new.txt")
dim(F2_geno_data_2_Ae_Ol_new) # 3443 166 
F2_geno_data_2_Ae_Ol_new[1:10, 1:10] 

# pheno data 
pheno <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/F2_Phenotype_Bnapus_all.csv", header = T, row.names = 1)
dim(pheno) # 42 166 
pheno[1:10, 1:10]
rownames(pheno) <- gsub("\ ", "_", rownames(pheno))
rownames(pheno) <- gsub("-", "_", rownames(pheno))
rownames(pheno) <- gsub("\\.", "_", rownames(pheno)) 
save(pheno, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/pheno.Rdata")

# combine geno & pheno data for QTL analysis 
colnames(F2_geno_data_2_Ae_Ol_new) <- gsub("X", "ID_", colnames(F2_geno_data_2_Ae_Ol_new))
F2_geno_data_2_Ae_Ol_new[1:10, 1:10]

F2_geno_data_2_Ae_Ol_new.t <- as.data.frame(t(F2_geno_data_2_Ae_Ol_new))
F2_geno_data_2_Ae_Ol_new.t$ID <- rownames(F2_geno_data_2_Ae_Ol_new.t)
pheno.t <- as.data.frame(t(pheno)) 
pheno.t$ID <- rownames(pheno.t) 

F2_geno_data_2_Ae_Ol_new.2 <- 
F2_geno_data_2_Ae_Ol_new.t %>% 
  left_join(pheno.t) %>% 
  select(-ID)  

dim(F2_geno_data_2_Ae_Ol_new.2) # 166 3485 
F2_geno_data_2_Ae_Ol_new.2[1:10, 3440:3485] 
F2_geno_data_2_Ae_Ol_new.2 <- t(F2_geno_data_2_Ae_Ol_new.2)

write.table(F2_geno_data_2_Ae_Ol_new.2, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/QTL_analysis/F2_geno_data_2_Ae_Ol_new.2.txt")
# change file format in linux 
# cat F2_geno_data_2_Ae_Ol_new.2.txt | sed 's/"//g' | awk '{first = $1; $1 = ""; print $0}' > tmp
# tail -3485 tmp > tmp.1 

write.table(rownames(F2_geno_data_2_Ae_Ol_new.2), file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/QTL_analysis/marker_info.txt" )
# cat marker_info.txt | awk '{print "*"$2}' | tail -3485 | sed 's/"//g' > marker_info_reform.txt

# paste marker_info_reform.txt tmp.1  | tr "\t" "\ " > F2_geno_for_one_map.txt
# cat ../../header_one_map_input F2_geno_for_one_map.txt > F2_geno_for_one_map_final.txt 
# change header info: maker number to the right marker number 
```

### load data 
```{r} 
LG.f2 <- read.cross("mm", file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/QTL_analysis/F2_geno_for_one_map_final.txt", mapfile = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.madmapper.map")
```

### re-estimate map
```{r}
LG.f2 <- LG.f2.before.crossover
LG.f2.before.crossover <- LG.f2

for (chr in names(LG.f2$geno)) { # for each chromosome in cross genotype data
  my.chr <- get(chr,LG.f2$geno) # return the genotype data, including data & map
  print(paste(chr,"NA before",sum(is.na(my.chr$data)))) 
  if(ncol(my.chr$data) > 3) { 
    my.chr$data[,2:(ncol(my.chr$data)-1)] <- sapply(2:(ncol(my.chr$data)-1),function(i) {
      apply(my.chr$data[,(i-1):(i+1)],1,function(gt) {
        if (any(is.na(gt))) return(gt[2]) #technically should be looking at the next genotyped marker.
        if ( (length(unique(gt)) == 2) & (gt[1] == gt[3])) return(NA)
        if ( length(unique(gt))  == 3) return(NA)
        return(gt[2])
      })
    })
  }
  LG.f2$geno <- within(LG.f2$geno,assign(chr,my.chr))
  print(paste(chr,"NA after",sum(is.na(get(chr,LG.f2$geno)$data))))
}  

map.new <- est.map(LG.f2,verbose=T,error.prob=.01) 

LG.f2 <- replace.map(LG.f2, map.new)
LG.f2.after.crossover <- LG.f2

plot.map(LG.f2.after.crossover, alternate.chrid = T) # after removing double crossover 
save(LG.f2.after.crossover, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/QTL_analysis/LG.f2.after.crossover.Rdata")
```   

# deal with LG10 
```{r}
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/QTL_analysis/LG.f2.after.crossover.Rdata")
summaryMap(LG.f2.after.crossover) # 2853. 2

set.seed(16)
LG.f2.after.crossover <- orderMarkers(LG.f2.after.crossover, chr = c(10), 
	                        window = 5, use.ripple = TRUE, maxit = 4000, 
	                        error.prob = 0.0001, verbose = T) 

plotMap(LG.f2.after.crossover, chr = '10')
summaryMap(LG.f2.after.crossover) # 2884.2 cM 
save(LG.f2.after.crossover, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/QTL_analysis/LG.f2.after.crossover.final.Rdata") 
```

### follow Julin's CSHL code for QTL mapping 
```{r}
load("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/missing_rate_0.10/LG.f2.after.crossover.final.Rdata")
summary(LG.f2.after.crossover)
# take a look at the geno and pheno data as well as the rf plot (rf plot shows further refinement is required)
plot.map(LG.f2.after.crossover)
# plot.rf(LG.f2.after.crossover)
summaryMap(LG.f2.after.crossover) # 2884 

# look for possible genotyping errors
# LG.f2.after.crossover <- calc.errorlod(LG.f2.after.crossover, error.prob=0.01) # takes too long time, leave for later... 
# top.errorlod(LG.f2.after.crossover)

# get new phenotype, difference between flower time and bolting time, that is the time to have flower after bolting
LG.f2.after.crossover$pheno$bolting_to_flowering <- LG.f2.after.crossover$pheno$days_to_flower -  LG.f2.after.crossover$pheno$days_to_bolt

# run QTL analysis 
# rename the map to their real chromosome name 
names(LG.f2.after.crossover$geno)[1:10] <- paste("A", c("01", "02", "03", "04", "05", "06", "07", "08", "09", "10"), sep = "")
names(LG.f2.after.crossover$geno)[11:19] <- paste("C", c("01", "02", "03", "04", "05", "06", "07", "08", "09"), sep = "") 
# save this new data 
save(LG.f2.after.crossover, file = "~/F2/output/QTL_analysis/LG.f2.after.crossover.43traits.Rdata") 

LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation? 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate the probability of the true underlying genotypes given the observed multipoint marker data --> for each imputed data, give a probability? 

# map
# start from one dimensional scan on imputed genotypes 
# permute the data to set a significance threshold 

sfInit(parallel = TRUE, cpus = parallel::detectCores()) 
sfExport("LG.f2.after.crossover")
sfLibrary(qtl)

system.time(
scanone.perm.imp <- 
  sfLapply(seq_along(LG.f2.after.crossover$pheno), function(trait){
    print(trait) # print doesn't work in here 
    tmp <-scanone(LG.f2.after.crossover,pheno.col = trait, method="imp",n.perm=1000, n.cluster = 16)
    summary(tmp)[1] # #keep the 95th percentile for future use.This corresponds to p <0.05
  }) # takes 40 mins to finish 
)
sfStop() 

names(scanone.perm.imp) <- colnames(LG.f2.after.crossover$pheno)
save(scanone.perm.imp, file = "~/F2/output/QTL_analysis/scanone.perm.imp.43traits")

#note method "em" = interval mapping; "mr" = marker regression
#here we use method "imp" for multiple imputation.
  #advantage of "imp" is better for missing genotype data
  #disadvantage is that it is slower.  Not a big deal on current computers.
# Erucic acid 
scanone.imp.Erucic <- scanone(LG.f2.after.crossover,pheno.col=15,method="imp") # 
plot(scanone.imp.Erucic,bandcol="gray90", main="Erucic_acid")
abline(h=scanone.perm.imp[["Erucic_acid"]],lty=2) #add permuation threshold

# Oleic acid 
scanone.imp.Oleic <- scanone(LG.f2.after.crossover,pheno.col=8,method="imp") # 
plot(scanone.imp.Oleic,bandcol="gray90", main="Oleic_acid")
abline(h=scanone.perm.imp[["Oleic_acid"]],lty=2) #add permuation threshold
```

### run one dimeonsional mapping for all traits at once 
```{r}
load("~/F2/output/QTL_analysis/LG.f2.after.crossover.43traits.Rdata")
summaryMap(LG.f2.after.crossover)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation? 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate the probability of the true underlying genotypes given the observed multipoint marker data --> for each imputed data, give a probability?  

system.time(
scanone.imp <- 
lapply(seq_along(LG.f2.after.crossover$pheno), function(trait) {
  print(trait)
  scanone(LG.f2.after.crossover,pheno.col=trait,method="imp")
}) 
)
names(scanone.imp) <- colnames(LG.f2.after.crossover$pheno)

png("~/F2/output/QTL_analysis/figure/QTL_one_dim.png", width=25, height=15, units="in", res=300)
par(mfrow=c(6,7))

for (i in names(scanone.imp)){
  plot(scanone.imp[[i]],bandcol="gray90", main=i)
  abline(h=scanone.perm.imp[[i]],lty=2)
}

dev.off()  

save(scanone.imp, file = "~/F2/output/QTL_analysis/scanone.imp.43traits")
```

### cim mapping (composite interval mapping)
a more sophisticated model is to use the most significant markers as covariates during a 1D scan.  This is known as composite interval mapping (CIM) and is implmented using the function cim  
```{r}
cim.qtl <- 
lapply(seq_along(LG.f2.after.crossover$pheno), function(trait) {
  print(trait)
  cim(LG.f2.after.crossover, n.marcovar=5, pheno.col=trait,method="em")
}) 
# here we use the interval mapping method "em" as this is how cim was originaly implmented.the n.marcovar= argument defines the maximum number of marker covariates to use.
names(cim.qtl) <- colnames(LG.f2.after.crossover$pheno)

sfInit(parallel = TRUE, cpus = parallel::detectCores())  
sfExport("LG.f2.after.crossover")
sfLibrary(qtl)  

cim.perm <- 
  sfLapply(seq_along(LG.f2.after.crossover$pheno), function(trait){
    message(trait) # message doesn't work in here either 
    tmp <- cim(LG.f2.after.crossover,
               pheno.col = trait, 
               n.marcovar=5, 
               method="em",
               n.perm=1000)
    summary(tmp)[1] # #keep the 95th percentile for future use.This corresponds to p <0.05
  }) # takes almost 4 hours to finish 

sfStop()  

names(cim.perm) <- colnames(LG.f2.after.crossover$pheno)

# plot out result and save plot 
png("~/F2/output/QTL_analysis/figure/QTL_cim_1.png", width=25, height=15, units="in", res=300)
par(mfrow=c(6,7))

for (i in names(cim.qtl)){
  plot(cim.qtl[[i]],bandcol="gray90", main=i)
  abline(h=cim.perm[[i]],lty=2)
}

dev.off()  
# more QTL found for cim method 

save(cim.qtl, file = "~/F2/output/QTL_analysis/cim.qtl.43traits.Rdata")
save(cim.perm, file = "~/F2/output/QTL_analysis/cim.perm.43traits.Rdata")
```

### correct phenotype data 
```{r}
load("~/F2/output/QTL_analysis/LG.f2.after.crossover.43traits.Rdata")

# by checking the QTL result, cis_11_Eicosenoic_acid/Palmitoliec acid/Behenic acid don't seem to have the right result. 
hist(LG.f2.after.crossover$pheno$cis_11_Eicosenoic_acid) # greater than 5
hist(LG.f2.after.crossover$pheno$Palmitoliec_aicd) # greater than 1
hist(LG.f2.after.crossover$pheno$Behenic_acid) # greater than 5 

LG.f2.after.crossover$pheno$cis_11_Eicosenoic_acid <- ifelse(LG.f2.after.crossover$pheno$cis_11_Eicosenoic_acid > 5, NA, LG.f2.after.crossover$pheno$cis_11_Eicosenoic_acid)

LG.f2.after.crossover$pheno$Palmitoliec_aicd <- ifelse(LG.f2.after.crossover$pheno$Palmitoliec_aicd > 1, NA, LG.f2.after.crossover$pheno$Palmitoliec_aicd)

LG.f2.after.crossover$pheno$Behenic_acid <- ifelse(LG.f2.after.crossover$pheno$Behenic_acid > 5, NA, LG.f2.after.crossover$pheno$Behenic_acid) 

save(LG.f2.after.crossover, file = "LG.f2.after.crossover.43traits.corrected.Rdata")

# rerun scanone & cim on these three traits 
summaryMap(LG.f2.after.crossover)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate the probability of the true underlying genotypes given the observed multipoint marker data --> for each imputed data, give a probability 

# permutation 
sfInit(parallel = TRUE, cpus = parallel::detectCores()) 
sfExport("LG.f2.after.crossover")
sfLibrary(qtl)

system.time(
scanone.perm.imp.3.traits <- 
  sfLapply(c(5, 12, 14), function(trait){
    print(trait) # print doesn't work in here 
    tmp <-scanone(LG.f2.after.crossover,pheno.col = trait, method="imp",n.perm=1000, n.cluster = 16)
    summary(tmp)[1] # #keep the 95th percentile for future use.This corresponds to p <0.05
  }) 
)

  #  user  system elapsed 
  # 0.204   0.132 127.701

sfStop() 

names(scanone.perm.imp.3.traits) <- colnames(LG.f2.after.crossover$pheno[,c(5, 12, 14)])

system.time(
scanone.imp.3.traits <- 
lapply(c(5, 12, 14), function(trait) {
  print(trait)
  scanone(LG.f2.after.crossover,pheno.col=trait,method="imp")
}) 
) 
names(scanone.imp.3.traits) <- colnames(LG.f2.after.crossover$pheno[,c(5, 12, 14)])

plot(scanone.imp.3.traits[[1]],bandcol="gray90", main=i)
abline(h=scanone.perm.imp.3.traits[[1]],lty=2)

plot(scanone.imp.3.traits[[2]],bandcol="gray90", main=i)
abline(h=scanone.perm.imp.3.traits[[2]],lty=2)

plot(scanone.imp.3.traits[[3]],bandcol="gray90", main=i)
abline(h=scanone.perm.imp.3.traits[[3]],lty=2)

load("~/F2/output/QTL_analysis/scanone.perm.imp.43traits")
load("~/F2/output/QTL_analysis/scanone.imp.43traits")

scanone.perm.imp$Palmitoliec_aicd <- scanone.perm.imp.3.traits$Palmitoliec_aicd
scanone.perm.imp$cis_11_Eicosenoic_acid <- scanone.perm.imp.3.traits$cis_11_Eicosenoic_acid
scanone.perm.imp$Behenic_acid <- scanone.perm.imp.3.traits$Behenic_acid

scanone.imp$Palmitoliec_aicd <- scanone.imp$Palmitoliec_aicd
scanone.imp$cis_11_Eicosenoic_acid <- scanone.imp.3.traits$cis_11_Eicosenoic_acid
scanone.imp$Behenic_acid <- scanone.imp.3.traits$Behenic_acid

save(scanone.imp, file = "~/F2/output/QTL_analysis/scanone.imp.43traits.corrected")  
save(scanone.perm.imp, file = "~/F2/output/QTL_analysis/scanone.perm.imp.43traits.corrected")

# check result by plotting 
load("~/F2/output/QTL_analysis/scanone.perm.imp.43traits.corrected")
load("~/F2/output/QTL_analysis/scanone.imp.43traits.corrected")

plot(scanone.imp[["Palmitoliec_aicd"]],bandcol="gray90")
abline(h=scanone.perm.imp[["Palmitoliec_aicd"]],lty=2)

plot(scanone.imp[["cis_11_Eicosenoic_acid"]],bandcol="gray90")
abline(h=scanone.perm.imp[["cis_11_Eicosenoic_acid"]],lty=2)

plot(scanone.imp[["Behenic_acid"]],bandcol="gray90")
abline(h=scanone.perm.imp[["Behenic_acid"]],lty=2)  

# cim 
cim.qtl.3.traits <- 
lapply(c(5, 12, 14), function(trait) {
  print(trait)
  cim(LG.f2.after.crossover, n.marcovar=5, pheno.col=trait,method="em")
}) 
# here we use the interval mapping method "em" as this is how cim was originaly implmented.the n.marcovar= argument defines the maximum number of marker covariates to use.
names(cim.qtl.3.traits) <- colnames(LG.f2.after.crossover$pheno[c(5, 12, 14)])

sfInit(parallel = TRUE, cpus = parallel::detectCores()) 
sfExport("LG.f2.after.crossover")
sfLibrary(qtl)

system.time( 
cim.perm.3.traits <-
  sfLapply(c(5, 12, 14), function(trait){
    message(trait) # message doesn't work in here either
    tmp <- cim(LG.f2.after.crossover,
               pheno.col = trait,
               n.marcovar=5,
               method="em",
               n.perm=1000)
    summary(tmp)[1] # #keep the 95th percentile for future use.This corresponds to p <0.05
  }) # takes almost 40 mins to finish
)

sfStop()   
names(cim.perm.3.traits) <- colnames(LG.f2.after.crossover$pheno[c(5, 12, 14)])  

plot(cim.qtl.3.traits[[1]],bandcol="gray90")
abline(h=cim.perm.3.traits[[1]],lty=2)

plot(cim.qtl.3.traits[[2]],bandcol="gray90")
abline(h=cim.perm.3.traits[[2]],lty=2)

plot(cim.qtl.3.traits[[3]],bandcol="gray90", main=i)
abline(h=cim.perm.3.traits[[3]],lty=2)

load("~/F2/output/QTL_analysis/cim.perm.43traits.Rdata")
load("~/F2/output/QTL_analysis/cim.qtl.43traits.Rdata")

cim.perm$Palmitoliec_aicd <- cim.perm.3.traits$Palmitoliec_aicd
cim.perm$cis_11_Eicosenoic_acid <- cim.perm.3.traits$cis_11_Eicosenoic_acid
cim.perm$Behenic_acid <- cim.perm.3.traits$Behenic_acid

cim.qtl$Palmitoliec_aicd <- cim.qtl.3.traits$Palmitoliec_aicd
cim.qtl$cis_11_Eicosenoic_acid <- cim.qtl.3.traits$cis_11_Eicosenoic_acid
cim.qtl$Behenic_acid <- cim.qtl.3.traits$Behenic_acid

save(cim.qtl, file = "~/F2/output/QTL_analysis/cim.qtl.43traits.corrrected.Rdata")  
save(cim.perm, file = "~/F2/output/QTL_analysis/cim.perm.43traits.corrected.Rdata")

# check result by plotting 
plot(cim.qtl[["Palmitoliec_aicd"]],bandcol="gray90")
abline(h=cim.perm[["Palmitoliec_aicd"]],lty=2)

plot(cim.qtl[["cis_11_Eicosenoic_acid"]],bandcol="gray90")
abline(h=cim.perm[["cis_11_Eicosenoic_acid"]],lty=2)

plot(cim.qtl[["Behenic_acid"]],bandcol="gray90")
abline(h=cim.perm[["Behenic_acid"]],lty=2)  
```

### get loci estimated effect for all these traits  
```{r}
<<<<<<< HEAD
load("~/F2/output/QTL_analysis/cim.qtl.43traits.Rdata")
=======
# Palmitolec acid has differnet QTL, which might be caused by its incorrect phenotype data? 

load("~/F2/output/QTL_analysis/cim.qtl.43traits.corrrected.Rdata") 
>>>>>>> f58eccb8afa7610dc3b754f57f96453bc837b321
# Palmitic
bayesint(cim.qtl$Palmitic_acid, chr = "A08", expandtomarkers = T)
bayesint(cim.qtl$Palmitic_acid, chr = "C03", expandtomarkers = T)

load("~/F2/output/scantwo/LG.f2.after.crossover.final.rename.Rdata")
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Palmitic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

qtlm.Palmitic <- makeqtl(LG.f2.after.crossover,chr=c("A08", "C03"),pos=c(73, 36))
qtl.fit.Palmitic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Palmitic,formula = y ~ Q1 + Q2, get.ests=T)
summary(qtl.fit.Palmitic) # two positive effect 

# Arachidic  
bayesint(cim.qtl$Arachidic_acid, chr = "A08", expandtomarkers = T)

load("~/F2/output/scantwo/LG.f2.after.crossover.final.rename.Rdata")
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Arachidic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

qtlm.Arachidic <- makeqtl(LG.f2.after.crossover,chr="A08",pos=78)
qtl.fit.Arachidic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Arachidic,formula = y ~ Q1, get.ests=T)
summary(qtl.fit.Arachidic) # one positive effect 

# cis_11_Eicosenoic_acid (1 loci)
bayesint(cim.qtl$cis_11_Eicosenoic_acid, chr = "A08", expandtomarkers = T)

load("~/F2/output/scantwo/LG.f2.after.crossover.final.rename.Rdata")
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$cis_11_Eicosenoic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

qtlm.cis_11_Eicosenoic_acid <- makeqtl(LG.f2.after.crossover,chr="A08",pos=75)
qtl.fit.cis_11_Eicosenoic_acid <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.cis_11_Eicosenoic_acid,formula = y ~ Q1, get.ests=T)
summary(qtl.fit.cis_11_Eicosenoic_acid) # one negative effect 

# Behenic 
bayesint(cim.qtl$Behenic_acid, chr = "A08", expandtomarkers = T)
bayesint(cim.qtl$Behenic_acid, chr = "C03", expandtomarkers = T)

load("~/F2/output/scantwo/LG.f2.after.crossover.final.rename.Rdata")
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Behenic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

qtlm.Behenic <- makeqtl(LG.f2.after.crossover,chr=c("A08", "C03"),pos=c(73, 35))
qtl.fit.Behenic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Behenic,formula = y ~ Q1 + Q2, get.ests=T)
summary(qtl.fit.Behenic) # two negative effect 

# Linoleic 
bayesint(cim.qtl$Linoleic_acid, chr = "A08", expandtomarkers = T)
bayesint(cim.qtl$Linoleic_acid, chr = "C03", expandtomarkers = T)

load("~/F2/output/scantwo/LG.f2.after.crossover.final.rename.Rdata")
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Linoleic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

qtlm.Linoleic <- makeqtl(LG.f2.after.crossover,chr=c("A08", "C03"),pos=c(75, 39))
qtl.fit.Linoleic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Linoleic,formula = y ~ Q1 + Q2, get.ests=T)
summary(qtl.fit.Linoleic) # two positive effect 

# Linolenic 
bayesint(scanone.imp$Linolenic_acid, chr = "A08", expandtomarkers = T)
bayesint(scanone.imp$Linolenic_acid, chr = "C03", expandtomarkers = T)

load("~/F2/output/scantwo/LG.f2.after.crossover.final.rename.Rdata")
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Linolenic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

qtlm.Linolenic <- makeqtl(LG.f2.after.crossover,chr=c("A08", "C03"),pos=c(73, 36))
qtl.fit.Linolenic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Linolenic,formula = y ~ Q1 + Q2, get.ests=T)
summary(qtl.fit.Linolenic)  # strong interaction only plot can tell 

# vaccenic
bayesint(cim.qtl$vaccenic_acid, chr = "A08", expandtomarkers = T)
bayesint(cim.qtl$vaccenic_acid, chr = "C03", expandtomarkers = T)

load("~/F2/output/scantwo/LG.f2.after.crossover.final.rename.Rdata")
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$vaccenic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

qtlm.vaccenic <- makeqtl(LG.f2.after.crossover,chr=c("A08", "C03"),pos=c(72, 34))
qtl.fit.vaccenic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Linolenic,formula = y ~ Q1 + Q2, get.ests=T)
summary(qtl.fit.vaccenic) # two positive effect   

# summary: except for Behenic acid, Erucic acid, and Escosinic acid for which B allele has negative effect, and Linolenic acid for which there is strong interaction between the two loci, all the other fatty acid trait get positive effect from B allele. B allele is the allele from Da-Ol-1. 
```

### output result from scanone & cim, QTL result visulization
```{r}
load("~/F2/output/QTL_analysis/scanone.perm.imp.43traits.corrected")
load("~/F2/output/QTL_analysis/scanone.imp.43traits.corrected")
load("~/F2/output/QTL_analysis/cim.qtl.43traits.corrrected.Rdata")
load("~/F2/output/QTL_analysis/cim.perm.43traits.corrected.Rdata")
load("~/F2/output/QTL_analysis/LG.f2.after.crossover.43traits.Rdata")

plot(cim.qtl[[1]], chr="C09")  
colnames(LG.f2.after.crossover$pheno)

# flower related 
flower_cim <- 
qtl_plot(input = rbind(data.frame(cim.qtl[["days_to_flower"]], method = "days_to_flower"), 
                       data.frame(cim.qtl[["days_to_bolt"]], method = "days_to_bolt")),
         chrs = c("A10", "C06"), 
         # lod = c(cim.perm[["days_to_bolt"]]),
         rug = TRUE, 
         title = "flower bolting related")

flower_cim$plot
ggsave(flower_cim$plot, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/figure /flower_cim.png", width = 7, height = 4)
ggsave(flower_cim$plot, filename = "~/F2/output/QTL_analysis/figure/flower_cim.png", width = 7, height = 4) 

# oil related  
# library(RColorBrewer)
# library(colorRamps)
# 
# colourCount = 10 
# getPalette = colorRampPalette(brewer.pal(9, "Set1"))  # can be alternate way to plot the result for paper 

oil_cim <- 
qtl_plot(input = rbind(data.frame(cim.qtl[["Palmitic_acid"]], method = "Palmitic_acid"), 
                       data.frame(cim.qtl[["Stearic_acid"]], method = "Stearic_acid"), 
                       data.frame(cim.qtl[["Oleic_acid"]], method = "Oleic_acid"),
                       data.frame(cim.qtl[["vaccenic_acid"]], method = "vaccenic_acid"), 
                       data.frame(cim.qtl[["Linoleic_acid"]], method = "Linoleic_acid"),
                       data.frame(cim.qtl[["Arachidic_acid"]], method = "Arachidic_acid"),
                       data.frame(cim.qtl[["Linolenic_acid"]], method = "Linolenic_acid"),
                       data.frame(cim.qtl[["Erucic_acid"]], method = "Erucic_acid"),
                       data.frame(cim.qtl[["cis_11_Eicosenoic_acid"]], method = "cis_11_Eicosenoic_acid"),
                       data.frame(cim.qtl[["Behenic_acid"]], method = "Behenic_acid")),
         chrs = c("A08", "C03"), 
         # lod = cim.perm[[1]], 
         title = "Fatty acid traits",
         rug = TRUE) 

# ggsave(oil_cim$plot, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/figure /oil_cim.png", width = 8, height = 5) 
ggsave(oil_cim$plot, filename = "~/F2/output/QTL_analysis/figure/oil_cim.png", width = 8, height = 5)

# growth related  
growth_cim <- 
qtl_plot(input = rbind(data.frame(cim.qtl[["root_weight_2016_05_13"]], method = "root_weight_2016_05_13"), 
                       data.frame(cim.qtl[["plant_height_2016_05_13"]], method = "plant_height_2016_05_13"), 
                       data.frame(cim.qtl[["leaf_number_2016_03_21"]], method = "leaf_number_2016_03_21")),
         chrs = c("A10"), 
         # lod = cim.perm[[1]], 
         rug = TRUE,
         title = "growth_related")

ggsave(growth_cim$plot, filename = "~/F2/output/QTL_analysis/figure/growth_related_cim.png", width = 8, height = 5)

# ggsave(growth_cim$plot, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/figure /growth_cim.png", width = 7, height = 4) 

# compare between different models 
# flower 
flower_cim_scanone <- 
qtl_plot(input = rbind(data.frame(cim.qtl[["days_to_flower"]], method = "cim"), 
                       data.frame(scanone.imp[["days_to_flower"]], method = "scanone")),
         chrs = c("A10", "C06"), 
         lod = data.frame(group = c("cim_threshold", "scanone_threshold"),
                          lod = c(cim.perm[["days_to_flower"]], scanone.perm.imp[["days_to_flower"]])), 
         rug = TRUE)

ggsave(flower_cim_scanone$plot, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/figure /flower_cim_scanone.png", width = 7, height = 4) 

# bolt
bolt_cim_scanone <- 
qtl_plot(input = rbind(data.frame(cim.qtl[["days_to_bolt"]], method = "cim"), 
                       data.frame(scanone.imp[["days_to_bolt"]], method = "scanone")),
         chrs = c("A10", "C06"), 
         lod = data_frame(group = c("cim_threshold", "scanone_threshold"),
                          lod = c(cim.perm[["days_to_bolt"]],
                                  scanone.perm.imp[["days_to_bolt"]])), 
         rug = TRUE) 

ggsave(bolt_cim_scanone$plot, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/figure /bolt_cim_scanone.png", width = 7, height = 4)  

# Erucic acid 
Erucic_cim_scanone <- 
qtl_plot(input = rbind(data.frame(cim.qtl[["Erucic_acid"]], method = "cim"), 
                       data.frame(scanone.imp[["Erucic_acid"]], method = "scanone")),
         chrs = c("A08", "C03"), 
         lod = data_frame(group = c("cim_threshold", "scanone_threshold"),
                          lod = c(cim.perm[["Erucic_acid"]],
                                  scanone.perm.imp[["Erucic_acid"]])), 
         rug = TRUE) 

ggsave(Erucic_cim_scanone$plot, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/figure /Erucic_cim_scanone.png", width = 7, height = 4)    


qtl_plot(input = rbind(data.frame(cim.qtl[["vaccenic_acid"]], method = "cim"), 
                       data.frame(scanone.imp[["vaccenic_acid"]], method = "scanone")),
         chrs = c("A08", "C03"), 
         lod = data_frame(group = c("cim_threshold", "scanone_threshold"),
                          lod = c(cim.perm[["vaccenic_acid"]],
                                  scanone.perm.imp[["vaccenic_acid"]])), 
         rug = TRUE)  

colnames(LG.f2.after.crossover$pheno)

# make scanone & cim plot for each trait 
qtl_all <- 
  lapply(names(cim.qtl), function(trait) {
    qtl_plot(input = rbind(data.frame(cim.qtl[[trait]], method = "cim"), 
                       data.frame(scanone.imp[[trait]], method = "scanone")),
         lod = data_frame(group = c("cim_threshold", "scanone_threshold"),
                          lod = c(cim.perm[[trait]],
                                  scanone.perm.imp[[trait]])), 
         rug = TRUE, 
         title = trait
         )  
  })

plots <- lapply(1:length(qtl_all), function(trait) qtl_all[[trait]]$plot)

paths <- stringr::str_c(names(cim.qtl), ".pdf") 
paths
pwalk(list(paths, plots), ggsave, path = "~/F2/output/QTL_analysis/figure/", width = 15, height = 7) # save on whitney

pwalk(list(paths, plots), ggsave, path = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/figure /", width = 15, height = 7) # save on MAC 
``` 

### need a script to tell me which LG has significant QTL for different traits 
```{r}
cim.qtl.result <- 
lapply(names(cim.qtl), function(trait) {
  as.numeric(unique(cim.qtl[[trait]][(cim.qtl[[trait]]$lod > cim.perm[[trait]]),]$chr))
})

im.qtl.result <- 
lapply(names(scanone.imp), function(trait) {
  as.numeric(unique(scanone.imp[[trait]][(scanone.imp[[trait]]$lod > scanone.perm.imp[[trait]]),]$chr))
})

names(cim.qtl.result) <- names(cim.qtl)
names(im.qtl.result) <- names(scanone.imp) 
unlist(cim.qtl.result)
unlist(im.qtl.result) 
``` 

### find significant SNPs for certain trait and design primers 
```{r}
SNP_Erucic <-  
rbind(bayesint(cim.qtl[["Erucic_acid"]], chr=8, prob=0.95, expandtomarkers=TRUE),
      bayesint(cim.qtl[["Erucic_acid"]], chr=13, prob=0.95, expandtomarkers=TRUE)) 

SNP_flower <- 
rbind(bayesint(cim.qtl[["days_to_flower"]], chr=10, prob=0.95, expandtomarkers=TRUE), 
      bayesint(cim.qtl[["days_to_flower"]], chr=16, prob=0.95, expandtomarkers=TRUE))

SNP_height <- bayesint(cim.qtl[["plant_height_2016_05_13"]], chr=10, prob=0.95, expandtomarkers=TRUE) 

# get genome fasta sequence
napus <-readDNAStringSet("~/Desktop/Brassica_project/IGV_file/Brassica_napus_v4.1.chromosomes.fa") 

# function to reform SNP 
reform.SNP <- function(input){
  data.frame(CHROM = gsub("([[:print:]]+)(_)([[:print:]]+)", "\\1", rownames(input)), 
                         POS = gsub("([[:print:]]+)(_)([[:print:]]+)", "\\3", rownames(input))
)
}
  
SNP_qtl <- list(SNP_Erucic, SNP_flower, SNP_height) 
names(SNP_qtl) <- c("Erucic", "flower", "height")
  
SNP_qtl_seq <- 
lapply(names(SNP_qtl), function(trait) {
  reformed <- reform.SNP(SNP_qtl[[trait]])
  get.fasta(reformed, napus)
})

names(SNP_qtl_seq) <- c("Erucic", "flower", "height") 

writeXStringSet(SNP_qtl_seq[[1]], filepath = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/QTL_analysis/Erucic.fa") 

writeXStringSet(SNP_qtl_seq[[2]], filepath = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/QTL_analysis/flower.fa") 

writeXStringSet(SNP_qtl_seq[[3]], filepath = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/QTL_analysis/height.fa") 
```

### sanger sequencing PCR & CAPS PCR for the sequences above 
use NCBI primer_BLAST for sanger sequening primer design (refer to https://github.com/leejimmy93/KIAT/blob/master/parent/KIAT_SNP.Rmd)

use apeE for caps primer design. 
1) paste fasta seq to apeE window
2) highlight SNP position --> click on tools --> dCAPs caculator 
3) if there is primer, take it, otherwise, add mutation to the sequence to get primer
<<<<<<< HEAD
4) for the other pair of primer, extract from the sequence itself, make sure the final product is not too short, greater than 100bp?   

### extract markers which are in the QTL interval for Erucic and Oleic acid 
```{r}
load("~/F2/output/QTL_analysis/scanone.perm.imp.43traits")
load("~/F2/output/QTL_analysis/scanone.imp.43traits")
load("~/F2/output/QTL_analysis/cim.qtl.43traits.Rdata")
load("~/F2/output/QTL_analysis/cim.perm.43traits.Rdata")
load("~/F2/output/QTL_analysis/LG.f2.after.crossover.43traits.Rdata")

bayesint(scanone.imp$Erucic_acid, chr = "A08", expandtomarkers = T)
bayesint(scanone.imp$Erucic_acid, chr = "C03", expandtomarkers = T)

bayesint(scanone.imp$Oleic_acid, chr = "A08", expandtomarkers = T)
bayesint(scanone.imp$Oleic_acid, chr = "C03", expandtomarkers = T)

bayesint(cim.qtl$Erucic_acid, chr = "A08", expandtomarkers = T)
bayesint(cim.qtl$Erucic_acid, chr = "C03", expandtomarkers = T)

bayesint(cim.qtl$Oleic_acid, chr = "A08", expandtomarkers = T)
bayesint(cim.qtl$Oleic_acid, chr = "C03", expandtomarkers = T) 

# I decided to keep the largest boundary
# Erucic A08 72-75; C03: 29-55
# Oleic A08 70-77; C03: 29-43

test <- 
scanone.imp$Erucic_acid[(scanone.imp$Erucic_acid$chr == "A08" & scanone.imp$Erucic_acid$pos > 70 & scanone.imp$Erucic_acid$pos < 77) | (scanone.imp$Erucic_acid$chr == "C03" & scanone.imp$Erucic_acid$pos > 29 & scanone.imp$Erucic_acid$pos < 55),] 

test2 <- test[grepl("^chr", rownames(test)),]

test2 %>% dim() # 23 
SNP.Erucic.Oleic.unique.F2 <- data.frame(CHROM = gsub("(chrA08|chrC03)(_)([[:digit:]]+)", "\\1", rownames(test2)), POS = gsub("(chrA08|chrC03)(_)([[:digit:]])", "\\3", rownames(test2)))

SNP.Erucic.Oleic.unique.F2 
save(SNP.Erucic.Oleic.unique.F2, file = "~/F2/output/QTL_analysis/SNP.Erucic.Oleic.unique.F2.Rdata")
```

Looking at the prediction accuracy by using all markers and only markers at sig loci, the latter got higher accuracy... to explore this more, I want to see how days to flower and days to bolt data look like... 
```{r}
bayesint(scanone.imp$days_to_flower, chr = "A10", expandtomarkers = T)
bayesint(scanone.imp$days_to_flower, chr = "C06", expandtomarkers = T)

bayesint(scanone.imp$days_to_bolt, chr = "A10", expandtomarkers = T)
bayesint(scanone.imp$days_to_bolt, chr = "C06", expandtomarkers = T)

bayesint(cim.qtl$days_to_flower, chr = "A10", expandtomarkers = T)
bayesint(cim.qtl$days_to_flower, chr = "C06", expandtomarkers = T)

bayesint(cim.qtl$days_to_bolt, chr = "A10", expandtomarkers = T)
bayesint(cim.qtl$days_to_bolt, chr = "C06", expandtomarkers = T)

# I decided to keep the largest boundary
# days_to_flower A10 173-188; C06: 66-91
# Oleic A08 152-210; C03: 66-91

test <- 
scanone.imp$Erucic_acid[(scanone.imp$Erucic_acid$chr == "A10" & scanone.imp$Erucic_acid$pos > 152 & scanone.imp$Erucic_acid$pos < 210) | (scanone.imp$Erucic_acid$chr == "C06" & scanone.imp$Erucic_acid$pos > 66 & scanone.imp$Erucic_acid$pos < 91),] 

test2 <- test[grepl("^chr", rownames(test)),]

test2 %>% dim() # 56 
SNP.flower.bolt.unique.F2 <- data.frame(CHROM = gsub("(chrA10|chrC06)(_)([[:digit:]]+)", "\\1", rownames(test2)), POS = gsub("(chrA10|chrC06)(_)([[:digit:]])", "\\3", rownames(test2)))

SNP.flower.bolt.unique.F2 
save(SNP.flower.bolt.unique.F2, file = "~/F2/output/QTL_analysis/SNP.flower.bolt.unique.F2.Rdata")

# 4) for the other pair of primer, extract from the sequence itself, make sure the final product is not too short, greater than 100bp?  
```

### look at the phenotype data 
In order to understand the relationship between different fatty acid traits, Julin asked me to plot a pairs plot for the phenotype data of the various fatty acid traits. 

```{r}
load("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/pheno.Rdata")

pheno_fatty_acid <- 
  as.data.frame(t(pheno)) %>% 
  dplyr::select(c(Palmitic_acid, Stearic_acid, Oleic_acid, vaccenic_acid, Linoleic_acid, Arachidic_acid, Linolenic_acid, Erucic_acid))

png("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/QTL_analysis/figure/pheno_fatty_acid_pair.png", width = 8, height = 8, units = "in", res = 300)  
pairs(pheno_fatty_acid) 
dev.off() 

# after check the pairs plot, Julin said: "I am curious about the outliers in the linolenic vs erucic and linolenic vs oleic plots.  I'd like to to get those two plots with the points colored by genotype at the two loci. 

colnames(pheno_fatty_acid)
pheno_Erucic_Oleic <- pheno_fatty_acid[,c("Erucic_acid",  "Linolenic_acid", "Oleic_acid")]
pairs(pheno_Erucic_Oleic)

# QTL result 
load("~/Desktop/Brassica_project/KIAT_RNA_seq/overlay/cim.qtl.43traits.Rdata")
bayesint(cim.qtl$Erucic_acid, chr = "A08")
#                  chr      pos      lod
# chrA08_10120555  A08 72.33080 44.25032
# chrA08_10120555  A08 72.33080 44.25032
# chrA08_10043132  A08 73.25008 42.80078

bayesint(cim.qtl$Erucic_acid, chr = "C03")
#                 chr      pos      lod
# cC03.loc30      C03 30.00000 22.04239
# chrC03_53755234 C03 34.87978 23.41476
# chrC03_53748114 C03 35.05260 22.10259

bayesint(cim.qtl$Oleic_acid, chr = "A08")
#                 chr     pos      lod
# cA08.loc72      A08 72.0000 27.81306
# chrA08_10120555 A08 72.3308 28.64849
# cA08.loc76      A08 76.0000 27.53047

bayesint(cim.qtl$Oleic_acid, chr = "C03", expandtomarkers = T)
#                 chr      pos        lod
# chrC03_56259623 C03 29.17130  1.2902093
# cC03.loc37      C03 37.00000 17.4676054
# chrC03_50371054 C03 42.71985  0.4248082

bayesint(cim.qtl$Linolenic_acid, chr = "A08")
#                 chr      pos      lod
# chrA08_10120555 A08 72.33080 6.326089
# chrA08_10043132 A08 73.25008 7.183112
# cA08.loc78      A08 78.00000 6.423398

bayesint(cim.qtl$Linolenic_acid, chr = "C03")
#                 chr      pos      lod
# chrC03_53755234 C03 34.87978 4.649860
# chrC03_53442609 C03 36.06502 5.692567
# cC03.loc41      C03 41.00000 4.599865

# So I pick SNP chrA08_10120555 to represent the loci on chrA08,and SNP chrC03_53755234 to represent the loci on C03 

F2_geno_data_2_Ae_Ol_new <- read.table("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_data_2_Ae_Ol_new.txt")
F2_geno_data_2_Ae_Ol_new
colnames(F2_geno_data_2_Ae_Ol_new) <- gsub("X", "ID_", colnames(F2_geno_data_2_Ae_Ol_new))
colnames(F2_geno_data_2_Ae_Ol_new)

F2_geno_data_2_Ae_Ol_new[1:10, 1:10]
geno_A08 <- F2_geno_data_2_Ae_Ol_new[rownames(F2_geno_data_2_Ae_Ol_new) %in% "chrA08_10120555",]
geno_C03 <- F2_geno_data_2_Ae_Ol_new[rownames(F2_geno_data_2_Ae_Ol_new) %in% "chrC03_53755234",]

install.packages("GGally")
ggpairs(pheno_Erucic_Oleic)

```

<<<<<<< HEAD
### check the loci effect to the different fatty acid traits 
```{r}
test <- data.frame(x = c("AA", "AB", "BB"),
           y = c(10, 20, 30))
test

p1.test <- 
test %>% 
ggplot() + 
  geom_col(mapping = aes(x = x, y = y)) + 
  labs(list(x = "genotype at loci", y = "average phenotype", title = "no dominance effect"))
ggsave(p1.test, filename = "~/Desktop/2017_Nov_KIAT_report/p1.test.png", width = 3, height = 4)

test2 <- data.frame(x = c("AA", "AB", "BB"),
           y = c(10, 12, 30))

p1.test2 <- 
test2 %>% 
ggplot() + 
  geom_col(mapping = aes(x = x, y = y)) + 
  labs(list(x = "genotype at loci", y = "average phenotype", title = "with dominance effect"))
ggsave(p1.test2, filename = "~/Desktop/2017_Nov_KIAT_report/p1.test2.png", width = 3, height = 4) 
```

### check revised genetic map with ID added 
use the revised genetic map with individual's name added, check to see whether it is correct by running scanone 
```{r}
F2.map <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised.csv", 
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv", 
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later  

plot.map(F2.map)
summaryMap(F2.map) # 2884 

F2.map <- sim.geno(F2.map,step=1,n.draws=32) # imputation  
F2.map <- calc.genoprob(F2.map,step=1) # calculate the probability of the true underlying genotypes given the observed multipoint marker data --> for each imputed data, give a probability?  

F2.map$pheno %>% colnames()
# Erucic acid 
scanone.imp.Erucic <- scanone(F2.map,pheno.col=15,method="imp") # 
plot(scanone.imp.Erucic,bandcol="gray90", main="Erucic_acid")

# Oleic acid 
scanone.imp.Oleic <- scanone(F2.map,pheno.col=8,method="imp") # 
plot(scanone.imp.Oleic,bandcol="gray90", main="Oleic_acid") 
# OK, this looks correct! 
```

### get genes under QTL for different traits
```{r}
load("~/F2/output/QTL_analysis/scanone.imp.43traits.corrected")
load("~/F2/output/QTL_analysis/scanone.perm.imp.43traits.corrected")
# QTL mapping result with phenotype data corrected 

### I will use scanone result as the QTL result for different traits 
threshold.95 <- tibble(perm.threshold = bind_rows(scanone.perm.imp) %>% as.numeric(), 
                       trait = colnames(bind_rows(scanone.perm.imp)))
threshold.95

scanone.qtl.2 <-  
bind_cols(scanone.imp) %>% 
  dplyr::select(chr, pos, starts_with("lod"))
rownames(scanone.qtl.2) <- rownames(scanone.imp$Crude_oil_contents)
colnames(scanone.qtl.2)[3:45] <- names(scanone.imp)

scanone.gather <- scanone.qtl.2 %>%
  gather(key = trait, value = LOD, -chr, -pos) %>%
  left_join(threshold.95)

scanone.gather %>% head() 

# look for overlap, for each trait, find QTL border and look for genes under QTL peaks 
sig.chrs <- scanone.gather %>% dplyr::filter(LOD > perm.threshold) %>%
  group_by(trait,chr) %>%
  dplyr::summarise(count = n())
sig.chrs 

# now for each significant chromosome/trait combo run bayesint
bayesint.list <- apply(sig.chrs,1,function(hit) { # for every row("trait, chr, count") in eigengene module 
    result <- bayesint(scanone.qtl.2[c("chr","pos",hit["trait"])],  
                     chr=hit["chr"], 
                     lodcolumn = 1, 
                     expandtomarkers = TRUE 
  )
  colnames(result)[3] <- "LOD" 
  result
})  
 

names(bayesint.list) <- sig.chrs$trait

bayesint.list <- lapply(bayesint.list,function(x)  
                          x %>% 
                          as.data.frame() %>%
                          rownames_to_column(var="markername")  %>% # make rownames to column and use "markername" as the colname for the new colomn  
                          mutate(chr=as.character(chr))
) 

bayesint.list 
bayesint.list %>% length() # 28 

bayesint.result <- as.tibble(bind_rows(bayesint.list,.id="trait")) %>% # combine list into tibble 
  dplyr::select(trait,chr,pos,markername,LOD) %>% 
  separate(markername,into=c("chr1","Mbp"),sep="_", convert=TRUE) %>% 
  group_by(trait,chr) %>% 
  dplyr::summarize(start=min(Mbp, na.rm = T),end=max(Mbp, na.rm = T),min_eQTL_LOD=min(LOD),max_eQTL_LOD=max(LOD)) %>% 
  #for the high QTL peaks the interval width is 0.  That is overly precise and need to widen those.
  mutate(start=ifelse(start==end,max(0,start-20000),start), end=ifelse(start==end,end+20000,end))

bayesint.result %>% dim() # 28 6 

# annotate QTL  
load("~/Reference/B.napus/BnapusAnnotation.Rdata") 

traitQTL.annotated <- lapply(1:nrow(bayesint.result),function(row) { # for each trait/module 
  qtl <- bayesint.result[row,]  
  results <- subset(BnapusAnnotation, chrom==qtl$chr &
                    start >= qtl$start & # genes which fall into the QTL interval 
                    end <= qtl$end)
} 
)  
names(traitQTL.annotated) <- bayesint.result$trait 

traitQTL.annotated <- bind_rows(traitQTL.annotated,.id="trait") %>% # combine list into data.frame 
  mutate(chrom=as.character(chrom)) %>%
  left_join(bayesint.result,by=c("trait","chrom"="chr")) #get eQTL LOD

traitQTL.annotated <- 
traitQTL.annotated %>% 
  mutate(start = start.y, end = end.y) %>% 
  dplyr::select(-start.x, -end.x, -start.y, -end.y, -min_eQTL_LOD, -max_eQTL_LOD) 

save(traitQTL.annotated, file = "~/F2/output/QTL_analysis/traitQTL.annotated.Rdata") 
```

### check important genes
```{r}
load("~/F2/output/QTL_analysis/traitQTL.annotated.Rdata")
traitQTL.annotated %>% 
  filter(trait == "days_to_flower" | trait == "days_to_bolt") %>% 
  filter(AGI == "AT5G10140") # cool, FLC gene is on A10 

traitQTL.annotated %>% 
  # filter(trait == "days_to_flower" | trait == "days_to_bolt") %>% 
  filter(AGI == "AT5G10140") # cool, FLC gene is on A10  
```

```{r}
phefile = read.csv("~/F2/data/QTL_analysis/F2.pheno.csv", stringsAsFactors = F, header = F) 
phefile %>% dim()

phefile[phefile$V1 == "Behenic_acid",][,2:167] %>% as.numeric() %>% hist()
phefile[phefile$V1 == "cis_11_Eicosenoic_acid",][,2:167] %>% as.numeric() %>% hist()
phefile[phefile$V1 == "Palmitoliec_aicd",][,2:167] %>% as.numeric() %>% hist()  
```














