---
title: "QTL_analysis_flipped"
output: html_document
---

### purpose of this script is to finish QTL analysis with several LG flipped 

```{r}
### load lib
library(tidyverse)
library(qtl)
library(snowfall)
library(ggrepel)
library(Biostrings)
source("~/KIAT/function_BnRNAseq.R")

### load data and run scanone & cim
LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

plot.map(LG.f2.after.crossover)
summaryMap(LG.f2.after.crossover) # 2884

LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate the probability of the true underlying genotypes given the observed multipoint marker data --> for each imputed data, give a probability?

sfInit(parallel = TRUE, cpus = 4)
sfExport("LG.f2.after.crossover")
sfLibrary(qtl)

system.time(
scanone.perm.imp <-
  sfLapply(seq_along(LG.f2.after.crossover$pheno), function(trait){
    print(trait) # print doesn't work in here
    tmp <-scanone(LG.f2.after.crossover,pheno.col = trait, method="imp",n.perm=1000, n.cluster = 16)
    summary(tmp)[1] # #keep the 95th percentile for future use.This corresponds to p <0.05
  }) # takes 40 mins to finish
)
sfStop()

names(scanone.perm.imp) <- colnames(LG.f2.after.crossover$pheno)
save(scanone.perm.imp, file = "~/F2/output/QTL_analysis/scanone.perm.imp.43traits.flipped")

# Erucic acid 
scanone.imp.Erucic <- scanone(LG.f2.after.crossover,pheno.col=15,method="imp") #
plot(scanone.imp.Erucic,bandcol="gray90", main="Erucic_acid")
abline(h=scanone.perm.imp[["Erucic_acid"]],lty=2) #add permuation threshold

# Oleic acid
scanone.imp.Oleic <- scanone(LG.f2.after.crossover,pheno.col=8,method="imp") #
plot(scanone.imp.Oleic,bandcol="gray90", main="Oleic_acid")
abline(h=scanone.perm.imp[["Oleic_acid"]],lty=2) #add permuation threshold

system.time(
scanone.imp <-
lapply(seq_along(LG.f2.after.crossover$pheno[1:43]), function(trait) {
  print(trait)
  scanone(LG.f2.after.crossover,pheno.col=trait,method="imp")
})
)
names(scanone.imp) <- colnames(LG.f2.after.crossover$pheno)[1:43]

png("~/F2/output/QTL_analysis/figure/QTL_one_dim_flipped.png", width=25, height=15, units="in", res=300)
par(mfrow=c(6,7))

for (i in names(scanone.imp)){
  plot(scanone.imp[[i]],bandcol="gray90", main=i)
  abline(h=scanone.perm.imp[[i]],lty=2)
}

dev.off()

save(scanone.imp, file = "~/F2/output/QTL_analysis/scanone.imp.43traits.flipped")

cim.qtl <-
lapply(seq_along(LG.f2.after.crossover$pheno)[1:43], function(trait) {
  print(trait)
  cim(LG.f2.after.crossover, n.marcovar=5, pheno.col=trait,method="em")
})

cim.qtl[["vaccenic_acid"]] 

# here we use the interval mapping method "em" as this is how cim was originaly implmented.the n.marcovar= argument defines the maximum number of marker covariates to use.
names(cim.qtl) <- colnames(LG.f2.after.crossover$pheno)[1:43]

sfInit(parallel = TRUE, cpus = 4)
sfExport("LG.f2.after.crossover")
sfLibrary(qtl)

cim.perm <-
  sfLapply(seq_along(LG.f2.after.crossover$pheno)[1:43], function(trait){
    message(trait) # message doesn't work in here either
    tmp <- cim(LG.f2.after.crossover,
               pheno.col = trait,
               n.marcovar=5,
               method="em",
               n.perm=1000)
    summary(tmp)[1] # #keep the 95th percentile for future use.This corresponds to p <0.05
  }) # takes almost 4 hours to finish

sfStop() 

names(cim.perm) <- colnames(LG.f2.after.crossover$pheno)[1:43]

# plot out result and save plot
png("~/F2/output/QTL_analysis/figure/QTL_cim_1_flipped.png", width=25, height=15, units="in", res=300)
par(mfrow=c(6,7))

for (i in names(cim.qtl)){
  plot(cim.qtl[[i]],bandcol="gray90", main=i)
  abline(h=cim.perm[[i]],lty=2)
}

dev.off()
# more QTL found for cim method

save(cim.qtl, file = "~/F2/output/QTL_analysis/cim.qtl.43traits.flipped.Rdata")
save(cim.perm, file = "~/F2/output/QTL_analysis/cim.perm.43traits.flipped.Rdata") 
```

### need a script to tell me which LG has significant QTL for different traits 
```{r}
load("~/F2/output/QTL_analysis/cim.perm.43traits.flipped.Rdata")
load("~/F2/output/QTL_analysis/cim.qtl.43traits.flipped.Rdata")
load("~/F2/output/QTL_analysis/scanone.perm.imp.43traits.flipped")
load("~/F2/output/QTL_analysis/scanone.imp.43traits.flipped")

cim.qtl.result <- 
lapply(names(cim.qtl), function(trait) {
  as.numeric(unique(cim.qtl[[trait]][(cim.qtl[[trait]]$lod > cim.perm[[trait]]),]$chr))
})

im.qtl.result <- 
lapply(names(scanone.imp), function(trait) {
  as.numeric(unique(scanone.imp[[trait]][(scanone.imp[[trait]]$lod > scanone.perm.imp[[trait]]),]$chr))
})

names(cim.qtl.result) <- names(cim.qtl)
names(im.qtl.result) <- names(scanone.imp) 
unlist(cim.qtl.result)
unlist(im.qtl.result)  
``` 

### get genes under QTL for different traits
```{r}
load("~/F2/output/QTL_analysis/scanone.imp.43traits.flipped")
load("~/F2/output/QTL_analysis/scanone.perm.imp.43traits.flipped")
# QTL mapping result with phenotype data corrected 

### I will use scanone result as the QTL result for different traits 
threshold.95 <- tibble(perm.threshold = bind_rows(scanone.perm.imp) %>% as.numeric(), 
                       trait = colnames(bind_rows(scanone.perm.imp)))
threshold.95

scanone.qtl.2 <-  
bind_cols(scanone.imp) %>% 
  dplyr::select(chr, pos, starts_with("lod"))
rownames(scanone.qtl.2) <- rownames(scanone.imp$Crude_oil_contents)
colnames(scanone.qtl.2)[3:45] <- names(scanone.imp)

scanone.gather <- scanone.qtl.2 %>%
  gather(key = trait, value = LOD, -chr, -pos) %>%
  left_join(threshold.95)

scanone.gather %>% head() 

# look for overlap, for each trait, find QTL border and look for genes under QTL peaks 
sig.chrs <- scanone.gather %>% dplyr::filter(LOD > perm.threshold) %>%
  group_by(trait,chr) %>%
  dplyr::summarise(count = n())
sig.chrs 

# now for each significant chromosome/trait combo run bayesint
bayesint.list <- apply(sig.chrs,1,function(hit) { # for every row("trait, chr, count") in eigengene module 
    result <- bayesint(scanone.qtl.2[c("chr","pos",hit["trait"])],  
                     chr=hit["chr"], 
                     lodcolumn = 1, 
                     expandtomarkers = TRUE 
  )
  colnames(result)[3] <- "LOD" 
  result
})  
 

names(bayesint.list) <- sig.chrs$trait

bayesint.list <- lapply(bayesint.list,function(x)  
                          x %>% 
                          as.data.frame() %>%
                          rownames_to_column(var="markername")  %>% # make rownames to column and use "markername" as the colname for the new colomn  
                          mutate(chr=as.character(chr))
) 

bayesint.list 
bayesint.list %>% length() # 30 

bayesint.result <- as.tibble(bind_rows(bayesint.list,.id="trait")) %>% # combine list into tibble 
  dplyr::select(trait,chr,pos,markername,LOD) %>% 
  separate(markername,into=c("chr1","Mbp"),sep="_", convert=TRUE) %>% 
  group_by(trait,chr) %>% 
  dplyr::summarize(start=min(Mbp, na.rm = T),end=max(Mbp, na.rm = T),min_eQTL_LOD=min(LOD),max_eQTL_LOD=max(LOD)) %>% 
  #for the high QTL peaks the interval width is 0.  That is overly precise and need to widen those.
  mutate(start=ifelse(start==end,max(0,start-20000),start), end=ifelse(start==end,end+20000,end))

bayesint.result %>% dim() # 30 6 

# annotate QTL  
load("~/Reference/B.napus/BnapusAnnotation.Rdata") 

traitQTL.annotated <- lapply(1:nrow(bayesint.result),function(row) { # for each trait/module 
  qtl <- bayesint.result[row,]  
  results <- subset(BnapusAnnotation, chrom==qtl$chr &
                    start >= qtl$start & # genes which fall into the QTL interval 
                    end <= qtl$end)
} 
)  
names(traitQTL.annotated) <- bayesint.result$trait 

traitQTL.annotated <- bind_rows(traitQTL.annotated,.id="trait") %>% # combine list into data.frame 
  mutate(chrom=as.character(chrom)) %>%
  left_join(bayesint.result,by=c("trait","chrom"="chr")) #get eQTL LOD

traitQTL.annotated <- 
traitQTL.annotated %>% 
  mutate(start = start.y, end = end.y) %>% 
  dplyr::select(-start.x, -end.x, -start.y, -end.y, -min_eQTL_LOD, -max_eQTL_LOD) 

save(traitQTL.annotated, file = "~/F2/output/QTL_analysis/traitQTL.annotated.flipped.Rdata") 
```

### check important genes
```{r}
load("~/F2/output/QTL_analysis/traitQTL.annotated.flipped.Rdata")
traitQTL.annotated %>% 
  filter(trait == "days_to_flower" | trait == "days_to_bolt") %>% 
  filter(AGI == "AT5G10140") # cool, FLC gene is on A10 

traitQTL.annotated %>% 
  # filter(trait == "days_to_flower" | trait == "days_to_bolt") %>% 
  filter(AGI == "AT5G10140") # cool, FLC gene is on A10    
```  

### plot for paper 
```{r}
load("~/F2/output/QTL_analysis/cim.perm.43traits.flipped.Rdata")
load("~/F2/output/QTL_analysis/cim.qtl.43traits.flipped.Rdata")
load("~/F2/output/QTL_analysis/scanone.perm.imp.43traits.flipped")
load("~/F2/output/QTL_analysis/scanone.imp.43traits.flipped")

cim.perm.all <- cim.perm
cim.qtl.all <- cim.qtl
scanone.perm.imp.all <- scanone.perm.imp
scanone.imp.all <- scanone.imp

load("~/F2/output/growth_model/scanone_growth_model_trait_flipped.Rdata")
load("~/F2/output/growth_model/cim_growth_model_trait_flipped.Rdata") 

cim.perm.all <- c(cim.perm.all, cim.perm)[c(1:15, 17, 21:55)] 
cim.qtl.all <- c(cim.qtl.all, cim_growth_model_trait.F2)[c(1:15, 17, 21:55)]
scanone.perm.imp.all <- c(scanone.perm.imp.all[1:43], as.list(lod.thrs[5,]))[c(1:15, 17, 21:55)] 
scanone.imp.all <- c(scanone.imp.all, scanone_growth_model_trait.F2)[c(1:15, 17, 21:55)] 

png("~/F2/output/QTL_analysis/figure/QTL_one_dim_flipped.png", width=30, height=30, units="in", res=300)
par(mfrow=c(11,5))

for (i in names(scanone.imp.all)){
  plot(scanone.imp.all[[i]],bandcol="gray90", main=i)
  abline(h=scanone.perm.imp.all[[i]],lty=2)
}

dev.off()  

png("~/F2/output/QTL_analysis/figure/QTL_cim_flipped.png", width=30, height=30, units="in", res=300)
par(mfrow=c(11,5))

for (i in names(cim.qtl.all)){
  plot(cim.qtl.all[[i]],bandcol="gray90", main=i)
  abline(h=cim.perm.all[[i]],lty=2)
}

dev.off()  

cim.qtl.all[["vaccenic_acid"]] 
cim.perm.all[["vaccenic_acid"]] 

oil_ID <- c("Palmitic_acid", "Stearic_acid", "Oleic_acid", "vaccenic_acid", "Linoleic_acid", "Arachidic_acid", "Linolenic_acid", "Erucic_acid", "cis_11_Eicosenoic_acid", "Behenic_acid")

oil.plot <- list()
for (i in oil_ID){
  oil.plot[[i]] <- qtl_plot(input = rbind(data.frame(cim.qtl.all[[i]], method = "cim"), 
                       data.frame(scanone.imp.all[[i]], method = "scanone")),
         chrs = c("A08", "C03"), 
         lod = data_frame(group = c("cim_threshold", "scanone_threshold"),
                          lod = c(cim.perm.all[[i]],
                                  scanone.perm.imp.all[[i]])), 
         rug = TRUE, 
         labels = NA, 
         title = i) 
}

library(cowplot)
get_legend<-function(a.gplot){
  tmp <- ggplot_gtable(ggplot_build(a.gplot))
  leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box")
  legend <- tmp$grobs[[leg]]
  return(legend)}
legend <- get_legend(oil.plot[[1]]$plot)

oil.plot.2 <-plot_grid(
  oil.plot[[1]]$plot+theme(legend.position="none"),
  oil.plot[[2]]$plot+theme(legend.position="none"), 
  oil.plot[[3]]$plot+theme(legend.position="none"),
  oil.plot[[4]]$plot+theme(legend.position="none"),
  oil.plot[[5]]$plot+theme(legend.position="none"),
  oil.plot[[6]]$plot+theme(legend.position="none"),
  oil.plot[[7]]$plot+theme(legend.position="none"),
  oil.plot[[8]]$plot+theme(legend.position="none"),
  oil.plot[[9]]$plot+theme(legend.position="none"),
  oil.plot[[10]]$plot+theme(legend.position="none"),
  ncol=2, nrow = 5,labels=c("","","",""))  

oil.plot.3 <- plot_grid(oil.plot.2, legend, rel_heights = c(45, 1), ncol = 1, nrow = 2)
oil.plot.3 

ggsave(oil.plot.3, filename = "~/F2/output/QTL_analysis/figure/oil.plot.3.png", width = 12, height = 12)

growth_ID <- c("days_to_bolt", "days_to_flower", "leaf_number_2015_12_28", "leaf_number_2016_03_21", "lobe_number_2016_01_21", "root_weight_2016_05_13", "height_Hmax", "leaf_number_I", "lobe_number_I", "plant_height_2016_05_13") 
               
growth.plot <- list() 
for (i in growth_ID){
  growth.plot[[i]] <- qtl_plot(input = rbind(data.frame(cim.qtl.all[[i]], method = "cim"), 
                       data.frame(scanone.imp.all[[i]], method = "scanone")),
         chrs = c("A10", "C06"), 
         lod = data_frame(group = c("cim_threshold", "scanone_threshold"),
                          lod = c(cim.perm.all[[i]],
                                  scanone.perm.imp.all[[i]])), 
         rug = TRUE, 
         labels = NA, 
         title = i) 
}   

growth.plot.2 <-plot_grid( 
  growth.plot[[1]]$plot+theme(legend.position="none"),
  growth.plot[[2]]$plot+theme(legend.position="none"), 
  growth.plot[[3]]$plot+theme(legend.position="none"),
  growth.plot[[4]]$plot+theme(legend.position="none"),
  growth.plot[[5]]$plot+theme(legend.position="none"), 
  growth.plot[[6]]$plot+theme(legend.position="none"),
  growth.plot[[7]]$plot+theme(legend.position="none"),
  growth.plot[[8]]$plot+theme(legend.position="none"),
  growth.plot[[9]]$plot+theme(legend.position="none"),
  growth.plot[[10]]$plot+theme(legend.position="none"),
  # legend, 
  ncol=2, nrow = 5,labels=c("","","",""))  

growth.plot.3 <- plot_grid(growth.plot.2, legend, rel_heights = c(45, 1), ncol = 1, nrow = 2)
growth.plot.3 

ggsave(growth.plot.3, filename = "~/F2/output/QTL_analysis/figure/growth.plot.3.png", width = 12, height = 12)
``` 

### redo QTL analysis for several traits which don't follow normal distribution 
```{r}
### load data and run scanone & cim
LG.f2.after.crossover <- read.cross("csvsr", genfile = "~/F2/data/QTL_analysis/LG.f2.madmapper.final_gen_revised_flipped.csv",
                     phefile = "~/F2/data/QTL_analysis/F2.pheno.csv",
                     genotypes = c("AA", "AB", "BB")) # the only problem is that 44 phenotypes were read instead of 43, need to figure out why later

plot.map(LG.f2.after.crossover)
summaryMap(LG.f2.after.crossover) # 2884

LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate the probability of the true underlying genotypes given the observed multipoint marker data --> for each imputed data, give a probability?

LG.f2.after.crossover$pheno %>% colnames() # Caprylic_acid, Myristic_acid, Heptadecanoic_acid, should try "2 part" model in scanone(), Survival_rate use log transformed data to get normality 

scanone_Caprylic <- 
scanone(LG.f2.after.crossover, pheno.col = "Caprylic_acid", method = "em", model = "2part", upper = F)

##### run till here... 
perm_Caprylic


# permutation 
perm_Caprylic <- 
scanone(LG.f2.after.crossover, pheno.col = "Caprylic_acid", method = "em", model = "2part", upper = F, n.perm = 1000)

scanone_Myristic <-
scanone(LG.f2.after.crossover, pheno.col = "Myristic_acid", method = "em", model = "2part", upper = F)

LG.f2.after.crossover$pheno$Heptadecanoic_acid_new <- ifelse(LG.f2.after.crossover$pheno$Heptadecanoic_acid == 0, 0, 1) 

scanone_Heptadecanoic <-
scanone(LG.f2.after.crossover, pheno.col = "Heptadecanoic_acid_new", method = "em", model = "binary") # binary

test <-
scanone(LG.f2.after.crossover, pheno.col = "", method = "em", model = "2part") # transformation 

plot(scanone_Caprylic)
plot(scanone_Myristic)
plot(scanone_Heptadecanoic) # no significant loci 

ggplot(LG.f2.after.crossover$pheno, aes(x = Survival_rate)) + geom_histogram(bins=15) + scale_x_log10() 
min(LG.f2.after.crossover$pheno$Survival_rate, na.rm = T) 

system.time(
scanone.imp <-
lapply(seq_along(LG.f2.after.crossover$pheno[1:43]), function(trait) {
  print(trait)
  scanone(LG.f2.after.crossover,pheno.col=trait,method="imp")
})
)
names(scanone.imp) <- colnames(LG.f2.after.crossover$pheno)[1:43]


```





















