---
title: "QTL_analysis"
author: "Ruijuan Li"
date: "9/26/2017"
output: 
  html_document: 
    keep_md: yes
---

Goal of this project: to map QTL using the genetic map that I constructed from F2 mapping population created from Da-Ae & Da-Ol-1 

```{r}
library(tidyverse)
library(qtl)
library(snowfall)
library(ggrepel) 
library(Biostrings)
source("~/Desktop/Brassica_project/KIAT_RNA_seq/analysis/function_BnRNAseq.R")
```

### reformat file with geno and pheno info
```{r}
# geno data 
F2_geno_data_2_Ae_Ol_new <- read.table("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/F2_geno_data_2_Ae_Ol_new.txt")
dim(F2_geno_data_2_Ae_Ol_new) # 3443 166 
F2_geno_data_2_Ae_Ol_new[1:10, 1:10]

# pheno data 
pheno <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/F2_Phenotype_Bnapus_all.csv", header = T, row.names = 1)
dim(pheno) # 42 166 
pheno[1:10, 1:10]
rownames(pheno) <- gsub("\ ", "_", rownames(pheno))
rownames(pheno) <- gsub("-", "_", rownames(pheno))
rownames(pheno) <- gsub("\\.", "_", rownames(pheno)) 

# combine geno & pheno data for QTL analysis 
colnames(F2_geno_data_2_Ae_Ol_new) <- gsub("X", "ID_", colnames(F2_geno_data_2_Ae_Ol_new))
F2_geno_data_2_Ae_Ol_new[1:10, 1:10]

F2_geno_data_2_Ae_Ol_new.t <- as.data.frame(t(F2_geno_data_2_Ae_Ol_new))
F2_geno_data_2_Ae_Ol_new.t$ID <- rownames(F2_geno_data_2_Ae_Ol_new.t)
pheno.t <- as.data.frame(t(pheno)) 
pheno.t$ID <- rownames(pheno.t)

F2_geno_data_2_Ae_Ol_new.2 <- 
F2_geno_data_2_Ae_Ol_new.t %>% 
  left_join(pheno.t) %>% 
  select(-ID)  

dim(F2_geno_data_2_Ae_Ol_new.2) # 166 3485 
F2_geno_data_2_Ae_Ol_new.2[1:10, 3440:3485] 
F2_geno_data_2_Ae_Ol_new.2 <- t(F2_geno_data_2_Ae_Ol_new.2)

write.table(F2_geno_data_2_Ae_Ol_new.2, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/QTL_analysis/F2_geno_data_2_Ae_Ol_new.2.txt")
# change file format in linux 
# cat F2_geno_data_2_Ae_Ol_new.2.txt | sed 's/"//g' | awk '{first = $1; $1 = ""; print $0}' > tmp
# tail -3485 tmp > tmp.1 

write.table(rownames(F2_geno_data_2_Ae_Ol_new.2), file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/QTL_analysis/marker_info.txt" )
# cat marker_info.txt | awk '{print "*"$2}' | tail -3485 | sed 's/"//g' > marker_info_reform.txt

# paste marker_info_reform.txt tmp.1  | tr "\t" "\ " > F2_geno_for_one_map.txt
# cat ../../header_one_map_input F2_geno_for_one_map.txt > F2_geno_for_one_map_final.txt 
# change header info: maker number to the right marker number 
```

### load data 
```{r} 
LG.f2 <- read.cross("mm", file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/QTL_analysis/F2_geno_for_one_map_final.txt", mapfile = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/LG.f2.madmapper.map")
```

### re-estimate map
```{r}
LG.f2

LG.f2.before.crossover <- LG.f2

for (chr in names(LG.f2$geno)) { # for each chromosome in cross genotype data
  my.chr <- get(chr,LG.f2$geno) # return the genotype data, including data & map
  print(paste(chr,"NA before",sum(is.na(my.chr$data)))) 
  if(ncol(my.chr$data) > 3) { 
    my.chr$data[,2:(ncol(my.chr$data)-1)] <- sapply(2:(ncol(my.chr$data)-1),function(i) {
      apply(my.chr$data[,(i-1):(i+1)],1,function(gt) {
        if (any(is.na(gt))) return(gt[2]) #technically should be looking at the next genotyped marker.
        if ( (length(unique(gt)) == 2) & (gt[1] == gt[3])) return(NA)
        if ( length(unique(gt))  == 3) return(NA)
        return(gt[2])
      })
    })
  }
  LG.f2$geno <- within(LG.f2$geno,assign(chr,my.chr))
  print(paste(chr,"NA after",sum(is.na(get(chr,LG.f2$geno)$data))))
}  

map.new <- est.map(LG.f2,verbose=T,error.prob=.01) 
save(map.new, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/QTL_analysis/map.new.Rdata")

LG.f2 <- replace.map(LG.f2, map.new)
LG.f2.after.crossover <- LG.f2

plot.map(LG.f2.after.crossover, alternate.chrid = T) # the old genetic map
plot.map(LG.f2.before.crossover,LG.f2.after.crossover, alternate.chrid = T) # genetic map comparison 
save(LG.f2.after.crossover, file = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/data/QTL_analysis/LG.f2.after.crossover.Rdata")
```   

### follow Julin's CSHL code for QTL mapping 
```{r}
load("~/F2/data/QTL_analysis/LG.f2.after.crossover.Rdata")
summary(LG.f2.after.crossover)
# take a look at the geno and pheno data as well as the rf plot (rf plot shows further refinement is required)
plot(LG.f2.after.crossover)
plot.rf(LG.f2.after.crossover)

# look for possible genotyping errors
# LG.f2.after.crossover <- calc.errorlod(LG.f2.after.crossover, error.prob=0.01) # takes too long time, leave for later... 
# top.errorlod(LG.f2.after.crossover)

LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation? 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate the probability of the true underlying genotypes given the observed multipoint marker data --> for each imputed data, give a probability? 

# map
# start from one dimensional scan on imputed genotypes 
# permute the data to set a significance threshold 

sfInit(parallel = TRUE, cpus = parallel::detectCores()) 
sfExport("LG.f2.after.crossover")
sfLibrary(qtl)

scanone.perm.imp <- 
  sfLapply(seq_along(LG.f2.after.crossover$pheno), function(trait){
    print(trait) # print doesn't work in here 
    tmp <-scanone(LG.f2.after.crossover,pheno.col = trait, method="imp",n.perm=1000)
    summary(tmp)[1] # #keep the 95th percentile for future use.This corresponds to p <0.05
  }) # takes 40 mins to finish 

sfStop() 

names(scanone.perm.imp) <- colnames(LG.f2.after.crossover$pheno)

#note method "em" = interval mapping; "mr" = marker regression
#here we use method "imp" for multiple imputation.
  #advantage of "imp" is better for missing genotype data
  #disadvantage is that it is slower.  Not a big deal on current computers.
# Erucic acid 
scanone.imp.Erucic <- scanone(LG.f2.after.crossover,pheno.col=15,method="imp") # 
plot(scanone.imp.Erucic,bandcol="gray90", main="Erucic_acid")
abline(h=scanone.perm.imp[["Erucic_acid"]],lty=2) #add permuation threshold

# Oleic acid 
scanone.imp.Oleic <- scanone(LG.f2.after.crossover,pheno.col=8,method="imp") # 
plot(scanone.imp.Oleic,bandcol="gray90", main="Oleic_acid")
abline(h=scanone.perm.imp[["Oleic_acid"]],lty=2) #add permuation threshold
```

### run one dimeonsional mapping for all traits at once 
```{r}
scanone.imp <- 
lapply(seq_along(LG.f2.after.crossover$pheno), function(trait) {
  print(trait)
  scanone(LG.f2.after.crossover,pheno.col=trait,method="imp")
}) 

names(scanone.imp) <- colnames(LG.f2.after.crossover$pheno)

png("~/F2/output/QTL_analysis/figure/QTL_one_dim.png", width=25, height=15, units="in", res=300)
par(mfrow=c(6,7))

for (i in names(scanone.imp)){
  plot(scanone.imp[[i]],bandcol="gray90", main=i)
  abline(h=scanone.perm.imp[[i]],lty=2)
}

dev.off()  
```

### cim mapping (composite interval mapping)
a more sophisticated model is to use the most significant markers as covariates during a 1D scan.  This is known as composite interval mapping (CIM) and is implmented using the function cim 

```{r}
cim.qtl <- 
lapply(seq_along(LG.f2.after.crossover$pheno), function(trait) {
  print(trait)
  cim(LG.f2.after.crossover, n.marcovar=5, pheno.col=trait,method="em")
}) 
# here we use the interval mapping method "em" as this is how cim was originaly implmented.the n.marcovar= argument defines the maximum number of marker covariates to use.
names(cim.qtl) <- colnames(LG.f2.after.crossover$pheno)

cim.perm <- cim(LG.f2.after.crossover,n.marcovar=5,method="em",n.perm=1000) #takes almost an hour 
save(cim.perm, file = "~/F2/output/QTL_analysis/cim.perm.Rdata")
summary(cim.perm)
cim.perm95 <- summary(cim.perm)[1] 

sfInit(parallel = TRUE, cpus = parallel::detectCores()) 
sfExport("LG.f2.after.crossover")
sfLibrary(qtl)  

cim.perm <- 
  sfLapply(seq_along(LG.f2.after.crossover$pheno), function(trait){
    message(trait) # message doesn't work in here either 
    tmp <- cim(LG.f2.after.crossover,
               pheno.col = trait, 
               n.marcovar=5, 
               method="em",
               n.perm=1000)
    summary(tmp)[1] # #keep the 95th percentile for future use.This corresponds to p <0.05
  }) # takes almost 4 hours to finish 

sfStop() 

names(cim.perm) <- colnames(LG.f2.after.crossover$pheno)

# plot out result and save plot 
png("~/F2/output/QTL_analysis/figure/QTL_cim_1.png", width=25, height=15, units="in", res=300)
par(mfrow=c(6,7))

for (i in names(cim.qtl)){
  plot(cim.qtl[[i]],bandcol="gray90", main=i)
  abline(h=cim.perm[[i]],lty=2)
}

dev.off()  
# more QTL found for cim method 

save(cim.qtl, cim.perm, file = "~/F2/output/QTL_analysis/cim.Rdata")
save(scanone.imp, scanone.perm.imp, file = "~/F2/output/QTL_analysis/scanone.Rdata")
```

### output result from scanone & cim, QTL result visulization
```{r}
# devtools::install_github("kbroman/qtlcharts")
library(qtlcharts)

load("./F2/output/QTL_analysis/cim.Rdata")
load("./F2/output/QTL_analysis/scanone.Rdata")
load("./F2/output/QTL_analysis/LG.f2.after.crossover.Rdata")

plot(cim.qtl[[1]], chr=c(19))  
colnames(LG.f2.after.crossover$pheno)

# flower related 
flower_cim <- 
qtl_plot(input = rbind(data.frame(cim.qtl[["days_to_flower"]], method = "days_to_flower"), 
                       data.frame(cim.qtl[["days_to_bolt"]], method = "days_to_bolt")),
         chrs = c(10, 16), 
         # lod = c(cim.perm[["days_to_bolt"]]),
         rug = TRUE)

flower_cim$plot
ggsave(flower_cim$plot, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/figure /flower_cim.png", width = 7, height = 4)

# oil related  
oil_cim <- 
qtl_plot(input = rbind(data.frame(cim.qtl[["Palmitic_acid"]], method = "Palmitic_acid"), 
                       data.frame(cim.qtl[["Stearic_acid"]], method = "Stearic_acid"), 
                       data.frame(cim.qtl[["Oleic_acid"]], method = "Oleic_acid"),
                       data.frame(cim.qtl[["vaccenic_acid"]], method = "vaccenic_acid"), 
                       data.frame(cim.qtl[["Linoleic_acid"]], method = "Linoleic_acid"),
                       data.frame(cim.qtl[["Arachidic_acid"]], method = "Arachidic_acid"),
                       data.frame(cim.qtl[["Linolenic_acid"]], method = "Linolenic_acid"),
                       data.frame(cim.qtl[["Erucic_acid"]], method = "Erucic_acid")),
         chrs = c(8, 13), 
         # lod = cim.perm[[1]], 
         rug = TRUE)

ggsave(oil_cim$plot, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/figure /oil_cim.png", width = 8, height = 5)

# growth related  
growth_cim <- 
qtl_plot(input = rbind(data.frame(cim.qtl[["root_weight_2016_05_13"]], method = "root_weight_2016_05_13"), 
                       data.frame(cim.qtl[["plant_height_2016_05_13"]], method = "plant_height_2016_05_13"), 
                       data.frame(cim.qtl[["leaf_number_2016_03_21"]], method = "leaf_number_2016_03_21")),
         chrs = c(10), 
         # lod = cim.perm[[1]], 
         rug = TRUE)

ggsave(growth_cim$plot, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/figure /growth_cim.png", width = 7, height = 4) 

# compare between different models 
# flower 
flower_cim_scanone <- 
qtl_plot(input = rbind(data.frame(cim.qtl[["days_to_flower"]], method = "cim"), 
                       data.frame(scanone.imp[["days_to_flower"]], method = "scanone")),
         chrs = c(10, 16), 
         lod = data.frame(group = c("cim_threshold", "scanone_threshold"),
                          lod = c(cim.perm[["days_to_flower"]], scanone.perm.imp[["days_to_flower"]])), 
         rug = TRUE)

ggsave(flower_cim_scanone$plot, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/figure /flower_cim_scanone.png", width = 7, height = 4) 

# bolt
bolt_cim_scanone <- 
qtl_plot(input = rbind(data.frame(cim.qtl[["days_to_bolt"]], method = "cim"), 
                       data.frame(scanone.imp[["days_to_bolt"]], method = "scanone")),
         chrs = c(10, 16), 
         lod = data_frame(group = c("cim_threshold", "scanone_threshold"),
                          lod = c(cim.perm[["days_to_bolt"]],
                                  scanone.perm.imp[["days_to_bolt"]])), 
         rug = TRUE) 

ggsave(bolt_cim_scanone$plot, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/figure /bolt_cim_scanone.png", width = 7, height = 4)  

# Erucic acid 
Erucic_cim_scanone <- 
qtl_plot(input = rbind(data.frame(cim.qtl[["Erucic_acid"]], method = "cim"), 
                       data.frame(scanone.imp[["Erucic_acid"]], method = "scanone")),
         chrs = c(8, 13), 
         lod = data_frame(group = c("cim_threshold", "scanone_threshold"),
                          lod = c(cim.perm[["Erucic_acid"]],
                                  scanone.perm.imp[["Erucic_acid"]])), 
         rug = TRUE) 

ggsave(Erucic_cim_scanone$plot, filename = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/figure /Erucic_cim_scanone.png", width = 7, height = 4)    

qtl_plot(input = rbind(data.frame(cim.qtl[["vaccenic_acid"]], method = "cim"), 
                       data.frame(scanone.imp[["vaccenic_acid"]], method = "scanone")),
         chrs = c(8, 13), 
         lod = data_frame(group = c("cim_threshold", "scanone_threshold"),
                          lod = c(cim.perm[["vaccenic_acid"]],
                                  scanone.perm.imp[["vaccenic_acid"]])), 
         rug = TRUE)  

colnames(LG.f2.after.crossover$pheno)

# make scanone & cim plot for each trait 
qtl_plot_2(input = rbind(data.frame(cim.qtl[["vaccenic_acid"]], method = "cim"), 
                       data.frame(scanone.imp[["vaccenic_acid"]], method = "scanone")),
         lod = data_frame(group = c("cim_threshold", "scanone_threshold"),
                          lod = c(cim.perm[["vaccenic_acid"]],
                                  scanone.perm.imp[["vaccenic_acid"]])), 
         rug = TRUE, 
         title = "vaccenic_acid"
         )  



``` 

### 2D scan
The above methods only evaluate the evidence for a QTL one location a time.  It is possible to instead scan two locations at once.  This can be particularly advantageous in identifying QTL if there is epistasis.

```{r}

```

### find significant SNPs for certain trait and design primers 
```{r}
SNP_Erucic <-  
rbind(bayesint(cim.qtl[["Erucic_acid"]], chr=8, prob=0.95, expandtomarkers=TRUE),
      bayesint(cim.qtl[["Erucic_acid"]], chr=13, prob=0.95, expandtomarkers=TRUE))

SNP_flower <- 
rbind(bayesint(cim.qtl[["days_to_flower"]], chr=10, prob=0.95, expandtomarkers=TRUE), 
      bayesint(cim.qtl[["days_to_flower"]], chr=16, prob=0.95, expandtomarkers=TRUE))

SNP_height <- bayesint(cim.qtl[["plant_height_2016_05_13"]], chr=10, prob=0.95, expandtomarkers=TRUE) 

# get genome fasta sequence
napus <-readDNAStringSet("~/Desktop/Brassica_project/IGV_file/Brassica_napus_v4.1.chromosomes.fa") 

# function to reform SNP 
reform.SNP <- function(input){
  data.frame(CHROM = gsub("([[:print:]]+)(_)([[:print:]]+)", "\\1", rownames(input)), 
                         POS = gsub("([[:print:]]+)(_)([[:print:]]+)", "\\3", rownames(input))
)
}
  
SNP_qtl <- list(SNP_Erucic, SNP_flower, SNP_height) 
names(SNP_qtl) <- c("Erucic", "flower", "height")
  
SNP_qtl_seq <- 
lapply(names(SNP_qtl), function(trait) {
  reformed <- reform.SNP(SNP_qtl[[trait]])
  get.fasta(reformed, napus)
})

names(SNP_qtl_seq) <- c("Erucic", "flower", "height") 

writeXStringSet(SNP_qtl_seq[[1]], filepath = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/QTL_analysis/Erucic.fa") 

writeXStringSet(SNP_qtl_seq[[2]], filepath = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/QTL_analysis/flower.fa") 

writeXStringSet(SNP_qtl_seq[[3]], filepath = "~/Desktop/Brassica_project/KIAT_RNA_seq/F2/output/QTL_analysis/height.fa") 
```

### sanger sequencing PCR & CAPS PCR for the sequences above 
use NCBI primer_BLAST for sanger sequening primer design (refer to https://github.com/leejimmy93/KIAT/blob/master/parent/KIAT_SNP.Rmd)

use apeE for caps primer design. 
1) paste fasta seq to apeE window
2) highlight SNP position --> click on tools --> dCAPs caculator 
3) if there is primer, take it, otherwise, add mutation to the sequence to get primer
4) for the other pair of primer, extract from the sequence itself, make sure the final product is not too short, greater than 100bp?  

