---
title: "scantwo_43traits"
output: html_document
---

### 2D scan
scanone and cim only evaluate the evidence for a QTL one location a time.  It is possible to instead scan two locations at once. This can be particularly advantageous in identifying QTL if there is epistasis. 

### load data 
```{r}
# running scantwo analysis and permutation seprately
# 1) scantwo analysis on cabernet 
# https://github.com/leejimmy93/KIAT_cabernet/blob/master/F2/scantwo_v2.R 

# scantwo permutation 100 times for 43 traits, split traits file so that each file has 2 traits, in total 20 files, and each run is only 10 permutation, combine permutation result after finish running. 
# https://github.com/leejimmy93/KIAT_cabernet/blob/master/F2/scantwo_perm_1.R ... scantwo_perm_10.R, submit to slurm queue using https://github.com/leejimmy93/KIAT_cabernet/blob/master/F2/scantwo_perm_1_10.slurm 

# scp result to whitney
# QTL result 
load("~/F2/output/scantwo/scantwo.imp.v2.Rdata") 
scantwo.imp.42traits <- scantwo.imp 
load("~/F2/output/scantwo/scantwo.imp.v2.bolting_to_flowering.Rdata")

scantwo.imp.43traits <- c(scantwo.imp.42traits, list(scantwo.imp))
names(scantwo.imp.43traits)[[43]] <- "bolting_to_flowering"

length(scantwo.imp.43traits) # 43 

# also load scanone & CIM result 
load("~/F2/output/QTL_analysis/scanone.imp.43traits")
load("~/F2/output/QTL_analysis/cim.qtl.43traits.Rdata") 

load("~/F2/output/scantwo/LG.f2.after.crossover.final.Rdata")
names(LG.f2.after.crossover$geno)[1:10] <- paste("A", c("01", "02", "03", "04", "05", "06", "07", "08", "09", "10"), sep = "")
names(LG.f2.after.crossover$geno)[11:19] <- paste("C", c("01", "02", "03", "04", "05", "06", "07", "08", "09"), sep = "") 
save(LG.f2.after.crossover, file = "~/F2/output/scantwo/LG.f2.after.crossover.final.rename.Rdata")
```

### Erucic acid & Oleic acid 
```{r}
######## start from here tonight...
# Erucic acid 
# permutation result for Erucic acid and Oleic acid import 
# load data 
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/Erucic_Oleic"
files <- list.files(path=path)
setwd(path)
scantwo_perm_Erucic_Oleic <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

# permutation data check 
scantwo_perm_Erucic_Oleic.t <- transpose(scantwo_perm_Erucic_Oleic) 
scantwo_perm_Erucic <- do.call(c, scantwo_perm_Erucic_Oleic.t[[2]])
summary(scantwo_perm_Erucic)

png("~/F2/output/scantwo/figure/Erucic_scantwo_int_full.png", width=12, height=10, units="in", res=300)
plot(scantwo.imp.42traits[["Erucic_acid"]], main = "Erucic_acid_int_full")
dev.off()

png("~/F2/output/scantwo/figure/Erucic_scantwo_int_fv1.png", width=12, height=10, units="in", res=300)
plot(scantwo.imp.42traits[["Erucic_acid"]], main = "Erucic_acid_int_fv1", lower = "fv1") 
dev.off()

png("~/F2/output/scantwo/figure/Erucic_scantwo_add_fv1.png", width=12, height=10, units="in", res=300)
plot(scantwo.imp.42traits[["Erucic_acid"]], main = "Erucic_acid_add_fv1", upper = "add", lower = "fv1") 
dev.off()

summary(scantwo.imp.42traits[["Erucic_acid"]], thresholds = c(10.28, 8.08, 6.99, 7.30, 4.33)) # signal for just additive 

# review previous result 
### load scanone/cim/scantwo result
summary(scanone.imp[["Erucic_acid"]])
summary(cim.qtl[["Erucic_acid"]]) 
summary(scantwo.imp.42traits[["Erucic_acid"]],perm=scantwo_perm_Erucic,alpha=.05,pval=T)

# load LG 
#create an object to hold QTL of interest
load("~/F2/output/scantwo/LG.f2.after.crossover.final.rename.Rdata")
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Erucic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

qtlm.Erucic <- makeqtl(LG.f2.after.crossover,chr=c("A08", "C03"),pos=c(73, 30)) # Ruijuan: QTL with interactions
### position from additive model if no interaction identified 

qtlm.Erucic # n.gen is "number of possible genotypes for each chromosome", if F2, n.gen = 3 when doing sim.geno(). 

#make a model for QTL action.  Since there is no evidence of interaction, start with an additive model
qtl.fit.Erucic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Erucic,formula = y ~ Q1 + Q2, get.ests=T)
### Ruijuan: fv1 - av1 indicate interaction, very small here... ??? 

#examine our model
summary(qtl.fit.Erucic) # not supported...??? 

#we can now ask if all qtls are at their most likely positions
#this also provides a nice map
qtl.refine.Erucic <- refineqtl(LG.f2.after.crossover,
                        qtl=qtlm.Erucic,
                        formula= y ~ Q1 + Q2)
#you can see from the LOD scores that the new model is not much better than the old model.  Nevertheless we can keep this as our new model

#we can test for interactions specifically among these loci
addint(LG.f2.after.crossover,qtl=qtl.refine.Erucic)
qtl.add.Erucic <- addint(LG.f2.after.crossover,qtl=qtl.refine.Erucic) # evidence for interactions
summary(qtl.add.Erucic)

# scan the genome for additional QTL to add to this model 
qtl.add.Erucic <- addqtl(LG.f2.after.crossover,qtl=qtl.refine.Erucic)
summary(qtl.add.Erucic) # no additional QTL 

# so the final model 
qtl.fit2.Erucic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Erucic,formula = y ~ Q1 * Q2, get.ests=T)
summary(qtl.fit2.Erucic) 

# refine position
qtl.refine2.Erucic <- refineqtl(LG.f2.after.crossover,
                        qtl=qtlm.Erucic,
                        formula= y ~ Q1 * Q2)

# plot out 
png("~/F2/output/scantwo/figure/Erucic_scantwo_result.png", width=20, height=10, units="in", res=300)
par(mfrow=c(1,2)) 
#positions can be plotted on the map
plot(qtl.refine.Erucic)
#or showing lod scores
plotLodProfile(qtl.refine.Erucic) 
dev.off()
save(qtl.fit2.Erucic, file = "~/F2/output/scantwo/qtl.fit2.Erucic.Rdata")

# summary: two loci A08@72.9 and C03@30.0 with interaction 

################ Oleic acid ######################  
# permutation result 
scantwo_perm_Oleic <- do.call(c, scantwo_perm_Erucic_Oleic.t[[1]])
summary(scantwo_perm_Oleic)

# plot 
png("~/F2/output/scantwo/figure/Oleic_scantwo_int_full.png", width=12, height=10, units="in", res=300)
plot(scantwo.imp.42traits[["Oleic_acid"]], main = "Oleic_acid_int_full")
dev.off()

png("~/F2/output/scantwo/figure/Oleic_scantwo_int_fv1.png", width=12, height=10, units="in", res=300)
plot(scantwo.imp.42traits[["Oleic_acid"]], main = "Oleic_acid_int_fv1", lower = "fv1") 
dev.off()

png("~/F2/output/scantwo/figure/Oleic_scantwo_add_fv1.png", width=12, height=10, units="in", res=300)
plot(scantwo.imp.42traits[["Oleic_acid"]], main = "Oleic_acid_add_fv1", upper = "add", lower = "fv1") 
dev.off() 

# look for siginificant loci 
summary(scantwo.imp.42traits[["Oleic_acid"]], thresholds = c(12.3, 9.45, 7.95, 7.37, 4.98)) 
summary(scantwo.imp.42traits[["Oleic_acid"]],perm=scantwo_perm_Oleic,alpha=.05,pval=T)

# load map and make Oleic acid as the only phenotype 
load("~/F2/output/scantwo/LG.f2.after.crossover.final.rename.Rdata")
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Oleic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

# object hold QTL of interest 
qtlm.Oleic <- makeqtl(LG.f2.after.crossover,chr=c("A08", "C03"),pos=c(73, 31))  
qtlm.Oleic
qtl.fit.Oleic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Oleic,formula = y ~ Q1 * Q2 ,get.ests=T) 
summary(qtl.fit.Oleic) 

# refine position
qtl.refine.Oleic <- refineqtl(LG.f2.after.crossover,
                        qtl=qtlm.Oleic,
                        formula= y ~ Q1 * Q2)

# scan the genome for additional QTL 
qtl.add.Oleic <- addqtl(LG.f2.after.crossover,qtl=qtl.refine.Oleic)
summary(qtl.add.Oleic) # no additional loci 

# plot result 
png("~/F2/output/scantwo/figure/Oleic_scantwo_result.png", width=20, height=10, units="in", res=300)
par(mfrow=c(1,2)) 
plot(qtl.refine.Oleic)
plotLodProfile(qtl.refine.Oleic) 
dev.off() 
save(qtl.fit.Oleic, file = "~/F2/output/scantwo/qtl.fit.Oleic.Rdata")
# summary: A08@72.3 and C03@31.0 with interaction  
```  

### Palmitoliec_aicd & Heptadecanoic_acid  
```{r}
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/Palmitoliec_Heptadecanoic/"
files <- list.files(path=path)
setwd(path)
scantwo_perm_Palmitoliec_Heptadecanoic <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

# permutation data check 
summary(do.call(c, scantwo_perm_Palmitoliec_Heptadecanoic)[,1])
# Palmitoliec_aicd (160 permutations)
#      full  fv1  int  add  av1  one
# 5%  100.0 98.8 93.2 91.7 90.6 5.11
# 10%  99.5 98.2 92.0 91.4 90.3 4.64 

summary(scantwo.imp.42traits[["Palmitoliec_acid"]], thresholds = c(100.0, 98.8, 93.2, 91.7, 90.6)) # signal for just additive 

# summary: no QTL 

################ Heptadecanoic acid ######################  
summary(do.call(c, scantwo_perm_Palmitoliec_Heptadecanoic)[,2])
# Heptadecanoic_acid (160 permutations)
#     full  fv1  int  add  av1  one
# 5%  38.3 36.7 29.7 23.9 22.2 5.21
# 10% 36.2 34.1 27.4 21.4 20.1 4.53

summary(scantwo.imp.42traits[["Heptadecanoic_acid"]], thresholds = c(38.3, 36.7, 29.7, 23.9, 22.2)) # signal for just additive 
```

### cis_11_Eicosenoic_acid & Linolenic_acid 
```{r}
path <- "/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/F2/output/scantwo/cis_11_Eicosenoic_Linolenic/"
files <- list.files(path=path)
setwd(path)
scantwo_perm_cis_11_Eicosenoic_Linolenic <- sapply(files, function(x) mget(load(x)), simplify = TRUE) 

# permutation data check 
summary(do.call(c, scantwo_perm_cis_11_Eicosenoic_Linolenic)[,1])
# cis_11_Eicosenoic_acid (160 permutations)
#     full  fv1  int  add  av1  one
# 5%  67.1 65.9 48.2 31.8 30.4 4.91
# 10% 52.9 51.9 42.6 28.7 27.6 4.20

summary(scantwo.imp.42traits[["cis_11_Eicosenoic_acid"]], thresholds = c(67.1, 65.9, 48.2, 31.8, 30.4)) # signal for just additive 

# summary: no QTL 

################ Linolenic acid ######################  
scantwo_perm_Linolenic <- do.call(c, scantwo_perm_cis_11_Eicosenoic_Linolenic)[,2]
summary(scantwo_perm_Linolenic)
# Linolenic_acid (160 permutations)
#     full  fv1  int  add  av1  one
# 5%  13.3 10.9 9.00 8.36 5.96 3.96
# 10% 12.5 10.4 8.43 7.53 5.57 3.84

summary(scantwo.imp.42traits[["Linolenic_acid"]],perm=scantwo_perm_Linolenic,alpha=.05,pval=T)

# load map and make Oleic acid as the only phenotype 
load("~/F2/output/scantwo/LG.f2.after.crossover.final.rename.Rdata")
LG.f2.after.crossover$pheno <- as.data.frame(LG.f2.after.crossover$pheno$Linolenic_acid)
LG.f2.after.crossover <- sim.geno(LG.f2.after.crossover,step=1,n.draws=32) # imputation 
LG.f2.after.crossover <- calc.genoprob(LG.f2.after.crossover,step=1) # calculate probability 

# object hold QTL of interest 
qtlm.Linolenic <- makeqtl(LG.f2.after.crossover,chr=c("A01", "C03", "A08", "C03"),pos=c(86, 39, 73, 36))  
qtlm.Linolenic
qtl.fit.Linolenic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Linolenic,formula = y ~ Q1 + Q2 + Q3 + Q4 + Q3:Q4 ,get.ests=T) 
summary(qtl.fit.Linolenic) 

# drop one marker 
qtlm.Linolenic <- dropfromqtl(qtlm.Linolenic,qtl.name="Q2")
qtlm.Linolenic
qtl.fit.Linolenic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Linolenic,formula = y ~ Q1 + Q2 + Q3 + Q2:Q3) 
summary(qtl.fit.Linolenic)

# drop again 
qtlm.Linolenic <- dropfromqtl(qtlm.Linolenic,qtl.name="Q1")
qtlm.Linolenic
qtl.fit.Linolenic <- fitqtl(cross=LG.f2.after.crossover,qtl=qtlm.Linolenic,formula = y ~ Q1 * Q2, get.ests = T) 
summary(qtl.fit.Linolenic)

# refine position 
qtl.refine.Linolenic <- refineqtl(LG.f2.after.crossover,
                        qtl=qtlm.Linolenic,
                        formula= y ~ Q1 * Q2)

# scan the genome for additional QTL 
qtl.add.Linolenic <- addqtl(LG.f2.after.crossover,qtl=qtl.refine.Linolenic)
summary(qtl.add.Linolenic) # no additional loci

# plot result 
png("~/F2/output/scantwo/figure/Linolenic_scantwo_result.png", width=20, height=8, units="in", res=300)
par(mfrow=c(1,2)) 
plot(qtl.refine.Linolenic)
plotLodProfile(qtl.refine.Linolenic) 
dev.off() 
save(qtl.fit.Linolenic, file = "~/F2/output/scantwo/qtl.fit.Linolenic.Rdata")
# summary: A08@72.3 and C03@31.0 with interaction 
```

#### only work on traits with significant loci 




