Title: code that I didn't keep in the working code but might be useful late sometime 
========================================================

```{r}
# 6) filter using vcftool to get only biallelic SNPs
# extract_SNP.sh & extract_biallelic.sh 
# (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/extract_SNP.sh)  
# (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/extract_biallelic.sh) 
# output: 505_SNP_biallelic.gz 

# 7) filter based on MAF and Q 
# filtering_MAF_Q.sh 
# (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/filtering_MAF_Q.sh)
# output: 505_filtered_MAF_Q.gz 

# 8) calcaulate mean read depth across individuals
# calc_depth.sh
# (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/calc_depth.sh) calculate in R
# output: 505_filtered_MAF_Q.ldepth.mean 

# 9) filter based on read depth (after check mean and distribution of read depth, decide to use 5-500 as the threshold)
# filtering_depth.sh
# (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/filtering_depth.sh)
# output: 505_filtered_MAF_Q_depth.gz 

# 10) calculate missing rate (SNPs with less than median of missing rate should be filtered out) 
# calc_missingness.sh 
# (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/calc_missingness.sh)
# output: 505_filtered_MAF_Q_depth.imiss; 505_filtered_MAF_Q_depth.lmiss 

# 11) filter based on missing rate 
# filter_missingness.sh 
# (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/filtering_missingrate.sh)
# output: 505_filtered_MAF_Q_depth_missingness.gz 

# 12) calculate MAF, missing rate, depth, and LD... 
# calc.sh
# (https://github.com/leejimmy93/KIAT_whitney/blob/master/505/calc.sh)
# output: a lot of stats start with 505_filtered_MAF_Q_depth_missingness prefix 
```

################## format for GWAS ###################

# prepare for GWAS (my MAC) # need to use code to detect which computer I am at later on 
```{r}
# 1) check to see whether the genotype & phentype data ID match  
# sowing number as sample ID???
sample_des_a_revised <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/data/batch_a_revised.csv", header=T)
sample_des_b_revised <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/data/batch_b_revised.csv", header=T)

length(unique(c(sample_des_a_revised$No..of.Sowing, sample_des_b_revised$No..of.Sowing)[!is.na(c(sample_des_a_revised$No..of.Sowing, sample_des_b_revised$No..of.Sowing))])) # 94 

genotype.data.sample.ID.leaf <- unique(c(sample_des_a_revised$No..of.Sowing, sample_des_b_revised$No..of.Sowing)[!is.na(c(sample_des_a_revised$No..of.Sowing, sample_des_b_revised$No..of.Sowing))])

# get phentoype data 
load("~/Desktop/Brassica_project/KIAT_RNA_seq/KIAT505-crossing/mongolia_korea_data_means_combined.Rdata") 
head(data.all)

data.all$SN

phenotype.data.sample.ID <- gsub("K", "", data.all$SN)
phenotype.data.sample.ID

length(intersect(phenotype.data.sample.ID, genotype.data.sample.ID.leaf)) # I see, so they are named according to their sowing number... YES, they matched!!! 

# 2) reform genotype & phentoype data for GWAS 
# geno.505 <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/data/gt_data.csv", header = T)
# prepare for GWAS on whitney 
# GWAS_505_leaf.R # output: leaf_505.2.txt (GD file for GWAS)
# scp to my MAC & also do this on whitney 
# scp ruijuanli@whitney.plb.ucdavis.edu:/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/505/output/leaf_505.2.txt . 

```

# prepare for GWAS (whitney)
```{r}
# 1) check to see whether the genotype & phentype data ID match  
# sowing number as sample ID???
sample_des_a_revised <- read.csv("~/505/data/phenotype/batch_a_revised.csv", header=T)
sample_des_b_revised <- read.csv("~/505/data/phenotype/batch_b_revised.csv", header=T)

length(unique(c(sample_des_a_revised$No..of.Sowing, sample_des_b_revised$No..of.Sowing)[!is.na(c(sample_des_a_revised$No..of.Sowing, sample_des_b_revised$No..of.Sowing))])) # 94 

genotype.data.sample.ID.leaf <- unique(c(sample_des_a_revised$No..of.Sowing, sample_des_b_revised$No..of.Sowing)[!is.na(c(sample_des_a_revised$No..of.Sowing, sample_des_b_revised$No..of.Sowing))])

# get phentoype data 
load("~/KIAT505-crossing/mongolia_korea_data_means_combined.Rdata") 
head(data.all)

data.all$SN

phenotype.data.sample.ID <- gsub("K", "", data.all$SN)
phenotype.data.sample.ID

length(intersect(phenotype.data.sample.ID, genotype.data.sample.ID.leaf)) # I see, so they are named according to their sowing number... YES, they matched!!!  
``` 

### import into GAPIT for GWAS 
```{r}
# source("http://www.bioconductor.org/biocLite.R")
# biocLite("multtest")
# install.packages("gplots")
# install.packages("LDheatmap")
# install.packages("genetics")
# install.packages("EMMREML") 
# install.packages("scatterplot3d")  

library(multtest) 
library(gplots)
library(LDheatmap)
library(genetics)
library(EMMREML)
library(compiler) #this library is already installed in R
library("scatterplot3d")
library(dplyr)

# install GAPIT package 
source("http://zzlab.net/GAPIT/gapit_functions.txt")
# intall EMMA package 
source("http://zzlab.net/GAPIT/emma.txt")

# set working directory and start analysis 
setwd("~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/myGAPIT")
setwd("~/505/output/myGAPIT/")

# genotype data  
# modify genotype data a little to make sure it looks like the right format for GAPIT 

#### GM file  
# cat leaf_505.2.txt | sed 's/"//g' > leaf_505.2.modified.txt 
# manually add "taxa" to the header (ready for GWAS should)
# load table takes very long time, so I decide to work with Rdata, much faster 

load("~/Desktop/Brassica_project/KIAT_RNA_seq/505/data/leaf_505.2.Rdata")
leaf_505.2 <- t(leaf_505.2.t) 
leaf_505.2 <- sample_n(leaf_505.2, size = 40000, replace = F)


class(leaf_505.2.t)  # matrix
leaf_505.2.t[1:10, 1:10]
dim(leaf_505.2.t) # 94 476781 
rownames(leaf_505.2.t) 
# sample 40,000 for test 



leaf_505.2.t.df <- as.data.frame(leaf)

leaf_505.2.t$taxa <- rownames(leaf_505.2.t)
leaf_505_2.t$taxa <- gsub("X", "", rownames(leaf_505.2.t))
leaf_505.2.t$taxa <- gsub("\\.","\\-", rownames(leaf_505.2.t)) 
rownames(leaf_505.2.t)

# still takes very long time, seems impossible, try another way
# cat leaf_505.2.modified.txt | sed 's/ /,/g' > leaf_505.2.modified.csv
# gzip leaf_505.2.modified.csv 

myGD <- read.csv("~/Desktop/Brassica_project/KIAT_RNA_seq/505/data/leaf_505.2.modified.csv.gz", 
                 row.names=1, #this tells R to use the first column as row names
                 na.strings=c("NA")) #this tells R that missing data is denoted as "NA" or "00")

# reform taxa name to sowing number 
sample_des_a_sub2 <- sample_des_a_revised[,c("Sample.ID", "No..of.Sowing", "Name")]
sample_des_a_sub2 <- sample_des_a_sub2[!is.na(sample_des_a_sub2$Sample.ID),]
sample_des_b_sub2 <- sample_des_b_revised[,c("Sample.ID", "No..of.Sowing", "Name")]
sample_des_sub2 <- rbind(sample_des_a_sub2, sample_des_b_sub2)
sample_des_sub2$Sample.ID 


# leaf_505.2.numeric <- data.matrix(as.data.frame(leaf_505.2.t))

SNP.info <- data.frame(matrix(nrow = ncol(leaf_505.2.t), ncol = 3))
colnames(SNP.info) <- c("Name", "Chromosome", "Position")
SNP.info$Name <- colnames(leaf_505.2.t)
SNP.info$Chromosome <- gsub("([[:print:]]+)(_)([[:digit:]]+)", "\\1", colnames(leaf_505.2.t))
SNP.info$Position <- gsub("([[:print:]]+)(_)([[:digit:]]+)", "\\3", colnames(leaf_505.2.t))
head(SNP.info)

# phenotype data 
load("~/Desktop/Brassica_project/KIAT_RNA_seq/505/data/gh.seed.bolt.table.wide.Rdata")
head(gh.seed.bolt.table.wide)
colnames(gh.seed.bolt.table.wide) 

# try GWAS... with data from whitney 
load("~/Desktop/Brassica_project/KIAT_RNA_seq/505/output/for_GAPIT.Rdata")

# replace sample ID with sowing number 
# get right format 
# colnames(geno_505_hmp) <- gsub("X", "", colnames(geno_505_hmp))
# colnames(geno_505_hmp) <- gsub("\\.","\\-", colnames(geno_505_hmp)) 
# merge & replace  
# colnames(geno_505_hmp) 
# geno_505_hmp.t <- data.frame(t(geno_505_hmp))
# dim(geno_505_hmp.t)
# geno_505_hmp.t.header <- geno_505_hmp.t[1:11,]
# geno_505_hmp.t.content <- geno_505_hmp.t[12:105,]
# rownames(geno_505_hmp.t.content)
# geno_505_hmp.t.content[1:10, 1:10]
# geno_505_hmp.t.header[1:10, 1:10]
# geno_505_hmp.t.content$Sample.ID <- rownames(geno_505_hmp.t.content) 
# geno_505_hmp.t.content.merge <- merge(sample_des_sub2, geno_505_hmp.t.content, by = "Sample.ID")
# 
# geno_505_hmp.t.content.merge$No..of.Sowing
# colnames(geno_505_hmp.t.content.merge)[2] <- "taxa"
# geno_505_hmp.t.content.merge$taxa <- paste("K", geno_505_hmp.t.content.merge$taxa, sep = "")
# geno_505_hmp.t.content.merge[1:10, 1:10] 

# only exact lines w/ phenotype data (what if don't worry about this)
# geno_505_hmp.t.content.merge.sub <- merge(myY, geno_505_hmp.t.content.merge, by = "taxa")
# 
# 
# dim(geno_505_hmp.t.content.merge.sub)
# dim(geno_505_hmp.t.header)
# 
# # add header & transpose 
# dim(geno_505_hmp.t.content.merge) 
# dim(geno_505_hmp.t.header)
# 
# rownames(geno_505_hmp.t.content.merge) <- geno_505_hmp.t.content.merge$taxa
# geno_505_hmp.t.content.merge <- geno_505_hmp.t.content.merge[,-c(1:2)]
# 
# geno_505_hmp.t.for.GWAS <- data.frame(rbind(geno_505_hmp.t.header, geno_505_hmp.t.content.merge))
# 
# geno_505_hmp.for.GWAS <- data.frame(t(geno_505_hmp.t.for.GWAS))
# dim(geno_505_hmp.for.GWAS) 
# geno_505_hmp.for.GWAS[1:10, 1:10]
# rownames(geno_505_hmp.for.GWAS) <- c(1:nrow(geno_505_hmp.for.GWAS)) 

################### 

# only keep those that are in the genotype data 
# pheno_data_505 <- gh.seed.bolt.table.wide[which(gh.seed.bolt.table.wide$SN %in% colnames(geno_505_hmp.for.GWAS)[12:105]),]
# 
# identical(
# pheno_data_505$SN[order(pheno_data_505$SN)],
# colnames(geno_505_hmp.for.GWAS)[12:105][order(colnames(geno_505_hmp.for.GWAS)[12:105])]) 

# pheno_data_505 <- pheno_data_505[,c("SN", "FATTY ACID_Erucic acid (C22:1n9)", "FATTY ACID_oil contents (%)", "FATTY ACID_Oleic acid (C18:1n9c)")]

####################

## remove lines with missing data in phenotype data 
# pheno_data_505_nomissing <- pheno_data_505[!is.na(pheno_data_505$Erucic_acid),]
# class(pheno_data_505_nomissing)         
# pheno_data_505_nomissing 
# dim(pheno_data_505_nomissing) # 79 4
# tail(pheno_data_505_nomissing)
# rownames(pheno_data_505_nomissing) <- c(1: nrow(pheno_data_505_nomissing))
# pheno_data_505_nomissing
# 
# ## remove lines with missing data in genotype data
# pheno_data_505
# filter <- pheno_data_505[is.na(pheno_data_505$Erucic_acid),]$Taxa
# filter
# 
# # geno_505_hmp.for.GWAS.nomissing <- geno_505_hmp.for.GWAS[,which(!(colnames(geno_505_hmp.for.GWAS) %in% filter))]
# # 
# # colnames(geno_505_hmp.for.GWAS.nomissing)[12:90]
# # pheno_data_505_nomissing$Taxa
# 
# geno_505_hmp.nomissing <- geno_505_hmp[,which(!(geno_505_hmp[1,] %in% filter))]


# some NA in allele colomn, remove them 
# geno_505_hmp.for.GWAS.nomissing.final <-  geno_505_hmp.for.GWAS.nomissing[!is.na(geno_505_hmp.for.GWAS.nomissing$alleles),] # 463900     90  
# dim(geno_505_hmp.for.GWAS.nomissing.final) # 463900     90 
# geno_505_hmp.for.GWAS.nomissing.final[1:10, 1:13]
# head(pheno_data_505_nomissing)
# rownames(geno_505_hmp.for.GWAS.nomissing.final) <- c(1:nrow(geno_505_hmp.for.GWAS.nomissing.final))
# 
# myY <- pheno_data_505_nomissing
# myG <- geno_505_hmp.for.GWAS.nomissing.final 

# error, check the file to make sure taxa matched  
# identical(
# myY$Taxa[order(myY$Taxa)],
# colnames(myG)[12:90][order(colnames(myG)[12:90])])  

# head(myY)
# myG[1:10, 1:12]
# write.table(myG, file="~/505/output/myG.txt")
# myG2 <- read.table("~/505/output/myG.txt", head=F) 
```


# GWAS on whitney 
```{r}
# myGD <- read.table("~/505/output/leaf_505.2.modified.txt", header = TRUE) # this is a bad idea, doesn't work, takes too long time... 

# cat leaf_505.2.modified.txt | sed 's/ /,/g' > leaf_505.2.modified.csv
# myGD <- read.csv("~/505/output/leaf_505.2.modified.csv.gz", 
#                  row.names=1, #this tells R to use the first column as row names
#                  na.strings=c("NA")) #this tells R that missing data is denoted as "NA" or "00") # doesn't work, takes forever 

geno_505 <- read.table("~/505/output/temp2.txt", header=T)
head(geno_505)

class(geno_505)

rownames(geno_505) <- paste(geno_505$CHROM, geno_505$POS, sep = "_")
head(geno_505)
geno_data_505 <- geno_505[, -c(1:4)]
dim(geno_data_505) # 476781     94

colnames(geno_data_505) <- gsub("X", "", colnames(geno_data_505))
colnames(geno_data_505) <- gsub("\\.","\\-", colnames(geno_data_505)) 

geno_data_505.t <- data.frame(t(geno_data_505))
rownames(geno_data_505.t)
geno_data_505.t$Sample.ID <- rownames(geno_data_505.t)

# reform taxa name to sowing number 
sample_des_a_sub2 <- sample_des_a_revised[,c("Sample.ID", "No..of.Sowing")]
sample_des_a_sub2 <- sample_des_a_sub2[!is.na(sample_des_a_sub2$Sample.ID),]
sample_des_b_sub2 <- sample_des_b_revised[,c("Sample.ID", "No..of.Sowing")]
sample_des_sub2 <- rbind(sample_des_a_sub2, sample_des_b_sub2)
sample_des_sub2

geno_data_505.t.merge <- merge(sample_des_sub2, geno_data_505.t, by = "Sample.ID")
geno_data_505.t.merge$Sample.ID

dim(geno_data_505.t.merge) # 94 476783
geno_data_505.t.merge$No..of.Sowing

rownames(geno_data_505.t.merge)
geno_data_505.t.merge[1:10, 1:10]
dim(geno_data_505.t) # 94 476782
geno_data_505.for.GWAS <- geno_data_505.t.merge[,-1]
colnames(geno_data_505.for.GWAS)[1] <- "taxa"
geno_data_505.for.GWAS$taxa <- paste("K", geno_data_505.for.GWAS$taxa, sep = "")
geno_data_505.for.GWAS[1:10, 1:10]
as.numeric(as.character(geno_data_505.for.GWAS[1,3]))

geno_data_505.for.GWAS.numeric <- data.frame(sapply(geno_data_505.for.GWAS, function(x) as.numeric(as.character(x))))
class(geno_data_505.for.GWAS.numeric[1,1])
geno_data_505.for.GWAS.numeric$taxa <- geno_data_505.t.merge[,2]
geno_data_505.for.GWAS.numeric$taxa <- paste("K", geno_data_505.for.GWAS.numeric$taxa, sep = "")
geno_data_505.for.GWAS.numeric[1:10, 1:10]

### GD file 
SNP.info <- data.frame(matrix(nrow = ncol(geno_data_505.t), ncol = 3))
colnames(SNP.info) <- c("Name", "Chromosome", "Position")
SNP.info$Name <- colnames(geno_data_505.t)
SNP.info$Chromosome <- gsub("([[:print:]]+)(_)([[:digit:]]+)", "\\1", colnames(geno_data_505.t))
SNP.info$Position <- gsub("([[:print:]]+)(_)([[:digit:]]+)", "\\3", colnames(geno_data_505.t))
SNP.info <- SNP.info[1:(nrow(SNP.info)-1),]
dim(SNP.info)
tail(SNP.info)

#### only get phenotype data for which there are genotype data 
load("~/505/data/phenotype/gh.seed.bolt.table.wide.Rdata")
head(gh.seed.bolt.table.wide)
colnames(gh.seed.bolt.table.wide) 
pheno_data_505 <- gh.seed.bolt.table.wide[gh.seed.bolt.table.wide$SN %in% geno_data_505.for.GWAS$taxa,] 
colnames(pheno_data_505_nomissing) <- pheno_data_505[,c("SN", "FATTY ACID_Erucic acid (C22:1n9)", "FATTY ACID_oil contents (%)", "FATTY ACID_Oleic acid (C18:1n9c)")]
          
colnames(pheno_data_505) <- c("taxa", "erucic_acid", "oil_content", "oleic_acid")
head(pheno_data_505)

# remove taxa w/o data 
pheno_data_505_nomissing <- pheno_data_505[!is.na(pheno_data_505$erucic_acid),]
class(pheno_data_505_nomissing)         
pheno_data_505_nomissing 
dim(pheno_data_505_nomissing) # 79 4
tail(pheno_data_505_nomissing)
rownames(pheno_data_505_nomissing) <- c(1: nrow(pheno_data_505_nomissing))

### also need to remove lines w/o phenotype data from the genotype data
geno_data_505.for.GWAS.numeric.final <- geno_data_505.for.GWAS.numeric[geno_data_505.for.GWAS.numeric$taxa %in% pheno_data_505_nomissing$taxa,]

pheno_data_505_nomissing$taxa
geno_data_505.for.GWAS.numeric[1:10, 1:10]
rownames(geno_data_505.for.GWAS.numeric.final)
rownames(geno_data_505.for.GWAS.numeric.final) <- c(1:nrow(geno_data_505.for.GWAS.numeric.final))

dim(geno_data_505.for.GWAS.numeric.final) # 79 476782
dim(pheno_data_505_nomissing) # 79  4 

######## for test 
# subset for one chrosome ChrA01 
SNP.info.chrA01 <- SNP.info[SNP.info$Chromosome == "chrA01",]
dim(SNP.info.chrA01)

filter <- SNP.info.chrA01$Name

geno_data_505.for.GWAS.numeric.final.chrA01 <- 
geno_data_505.for.GWAS.numeric.final[,colnames(geno_data_505.for.GWAS.numeric.final) %in% filter]

geno_data_505.for.GWAS.numeric.final.chrA01[1:10, 1:10]
geno_data_505.for.GWAS.numeric.final.chrA01$taxa <- geno_data_505.for.GWAS.numeric.final$taxa

dim(geno_data_505.for.GWAS.numeric.final.chrA01) 
geno_data_505.for.GWAS.numeric.final.chrA01 <- geno_data_505.for.GWAS.numeric.final.chrA01[,c(21415, 1:21414)] 

############
myGD <- geno_data_505.for.GWAS.numeric.final.chrA01
myGD[is.na(myGD)] <- 0
myGD[1:10, 1:10]
myGM <- SNP.info.chrA01
myY <- pheno_data_505_nomissing

save(myGD, myGM, myY, file="for_GAPIT.Rdata")

###### test for only one chrosome
myGAPIT <- GAPIT( 
Y=myY,
GD=myGD, 
GM=myGM, 
PCA.total=3
)

``` 



